{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cb23330d-0231-4dbe-8c5c-0f88b0af0629",
   "metadata": {},
   "source": [
    "# Variant Calling Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55dbb5e6-f260-4b63-beb1-1aae2bea9b24",
   "metadata": {},
   "source": [
    "## Required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9ebb7618-ea7b-4f34-a458-4e04d259cc65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python 3.9 or higher (.removesuffix() in  SAMtools realign NCBI RefSeq assembly)\n",
    "from tkinter import filedialog\n",
    "import os\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "from datetime import datetime\n",
    "import json\n",
    "import time\n",
    "import shutil\n",
    "import requests\n",
    "import zipfile\n",
    "import subprocess\n",
    "import gzip\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ac92eb-6dd0-4c7a-85c0-8b8aaf49fd1d",
   "metadata": {},
   "source": [
    "## Directory Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fa0994d5-6fdc-4ff5-998a-2d3d4650bf31",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = os.getcwd()\n",
    "\n",
    "main_directory_path = filedialog.askdirectory(title=\"Project\", initialdir=cwd)\n",
    "#main_directory_path = \"/home/student/Variant_Calling/Jupyter/PAO1\"\n",
    "Fastq_directory_path = filedialog.askdirectory(title=\"Fastq files\", initialdir=cwd)\n",
    "#Fastq_directory_path = os.path.join(main_directory_path, \"Fastq\")\n",
    "\n",
    "main_subdirectories = [\"Analysis\"]\n",
    "\n",
    "for directory in  main_subdirectories:\n",
    "    main_subdirectory_path = os.path.join(main_directory_path, directory)\n",
    "    if not os.path.exists(main_subdirectory_path):\n",
    "        os.mkdir(main_subdirectory_path)\n",
    "\n",
    "analysis_directory_path = os.path.join(main_directory_path, \"Analysis\")\n",
    "analysis_subdirectories = [\"fastq\",\"assemblies\",\"variant_calling\",\"samfiles\", \"genomes\"]\n",
    "\n",
    "for directory in analysis_subdirectories:\n",
    "    analysis_subdirectory_path = os.path.join(analysis_directory_path, directory)\n",
    "    if not os.path.exists(analysis_subdirectory_path):\n",
    "        os.mkdir(analysis_subdirectory_path)\n",
    "\n",
    "genomes_directory_path = os.path.join(analysis_directory_path, \"genomes\")\n",
    "genomes_subdirectories = [\"annotated\", \"indexed\"]\n",
    "\n",
    "for directory in genomes_subdirectories:\n",
    "    genomes_subdirectory_path = os.path.join(genomes_directory_path, directory)\n",
    "    if not os.path.exists(genomes_subdirectory_path):\n",
    "        os.mkdir(genomes_subdirectory_path)\n",
    "\n",
    "variant_calling_directory_path = os.path.join(analysis_directory_path, \"variant_calling\")\n",
    "variant_calling_subdirectories = [\"annotated\",\"curated\",\"indels\",\"intermediate\",\"raw\",\"snps\",\"totalpileup\",\"vcf\",\"results\"]\n",
    "\n",
    "for directory in variant_calling_subdirectories:\n",
    "    vc_subdirectory_path = os.path.join(variant_calling_directory_path, directory)\n",
    "    if not os.path.exists(vc_subdirectory_path):\n",
    "        os.mkdir(vc_subdirectory_path)\n",
    "\n",
    "# Analysis subdirectories path\n",
    "fastq_directory_path = os.path.join(analysis_directory_path, \"fastq\")\n",
    "samfiles_directory_path = os.path.join(analysis_directory_path, \"samfiles\")\n",
    "assemblies_directory_path = os.path.join(analysis_directory_path, \"assemblies\")\n",
    "\n",
    "## Variant Calling subdirectories path\n",
    "totalpileup_directory_path = os.path.join(variant_calling_directory_path, \"totalpileup\")\n",
    "raw_directory_path = os.path.join(variant_calling_directory_path, \"raw\")\n",
    "intermediate_directory_path = os.path.join(variant_calling_directory_path, \"intermediate\")\n",
    "vcf_directory_path = os.path.join(variant_calling_directory_path, \"vcf\")\n",
    "indels_directory_path = os.path.join(variant_calling_directory_path, \"indels\")\n",
    "snps_directory_path = os.path.join(variant_calling_directory_path, \"snps\")\n",
    "curated_directory_path = os.path.join(variant_calling_directory_path, \"curated\")\n",
    "annotated_directory_path = os.path.join(variant_calling_directory_path, \"annotated\")\n",
    "results_directory_path = os.path.join(variant_calling_directory_path, \"results\")\n",
    "\n",
    "## Genomes subdirectories path\n",
    "indexed_directory_path = os.path.join(genomes_directory_path, \"indexed\")\n",
    "annotated_genomes_directory_path = os.path.join(genomes_directory_path, \"annotated\")\n",
    "\n",
    "# Softwares\n",
    "picard_jar = \"/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar\"\n",
    "samtools = \"/home/student/miniconda3/envs/VariantCalling/bin/samtools\"\n",
    "bowtie2_build = \"/home/student/miniconda3/envs/VariantCalling/bin/bowtie2-build\"\n",
    "bowtie2 = \"/home/student/miniconda3/envs/VariantCalling/bin/bowtie2\"\n",
    "GATK_jar = \"/home/student/Programs/GenomeAnalysisTK-3.4-46/GenomeAnalysisTK.jar\"\n",
    "java_8 = \"/home/student/javav/jdk1.8.0_131/bin/java\"\n",
    "java_21 = \"/home/student/javav/jdk-21.0.6/bin/java\"\n",
    "snpEff_jar = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.jar\"\n",
    "perl = \"/home/student/miniconda3/envs/VariantCalling/bin/perl\"\n",
    "convert2annovar = \"/home/student/annovar/convert2annovar.pl\"\n",
    "vcfEffOnePerLine = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/scripts/vcfEffOnePerLine.pl\"\n",
    "snpSift_jar = \"/home/student/miniconda3/envs/VariantCalling/share/snpsift-5.2-0/SnpSift.jar\"\n",
    "spades = \"/home/student/miniconda3/envs/VariantCalling/bin/spades.py\"\n",
    "buildDbNcbi = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/scripts/buildDbNcbi.sh\"\n",
    "snpeff_dir_path = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eecbf7b-7de5-47e4-bb8f-a02ddc558221",
   "metadata": {},
   "source": [
    "## Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0a79baa-7262-4098-bc1a-4bd1917b09f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1aa94eda3ba841538c3e5c400fa735d9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(VBox(children=(Checkbox(value=False, description='Configuration file'), VBox(layout=Layout(dispâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load configuration file widget\n",
    "configuration_widget = widgets.Checkbox(description='Configuration file', value=False)\n",
    "\n",
    "configuration_inp_box = widgets.VBox()\n",
    "configuration_section = widgets.VBox([configuration_widget, configuration_inp_box])\n",
    "\n",
    "def update_configuration_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        upload_file = widgets.FileUpload(accept='.json', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        items.append(upload_file)\n",
    "    configuration_inp_box.children = items\n",
    "    configuration_inp_box.layout.display = 'block' if num > 0 else 'none'\n",
    "\n",
    "def on_configuration_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_configuration_inputs(change['new'])\n",
    "\n",
    "configuration_widget.observe(on_configuration_widget_change, names='value')\n",
    "update_configuration_inputs(configuration_widget.value)\n",
    "\n",
    "# Merge widget\n",
    "merge_widget = widgets.Checkbox(description='Merge .fastq files', value=False)\n",
    "\n",
    "# Download reference genome/s widget\n",
    "download_widget = widgets.BoundedIntText(description='Dowload reference genome', value=0,\n",
    "                                         min=0, max=10, step=1,\n",
    "                                        style={'description_width':'200px'})\n",
    "\n",
    "download_inp_box = widgets.VBox()\n",
    "download_section = widgets.VBox([download_widget, download_inp_box])\n",
    "\n",
    "def update_download_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                               placeholder='GCF_000006765.1',\n",
    "                               style={'description_width':'150px'},\n",
    "                               layout=widgets.Layout(width='300px'))\n",
    "        items.append(textbox)\n",
    "    download_inp_box.children = items\n",
    "    download_inp_box.layout.display = 'block' if num > 0 else 'none'\n",
    "\n",
    "def on_download_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_download_inputs(change['new'])\n",
    "\n",
    "download_widget.observe(on_download_widget_change, names='value')\n",
    "update_download_inputs(download_widget.value)\n",
    "\n",
    "# Genome indexing widget\n",
    "index_widget = widgets.BoundedIntText(description='Genome indexing', value=0,\n",
    "                                     min=0, max=10, step=1,\n",
    "                                     style={'description_width':'150px'})\n",
    "\n",
    "index_inp_box = widgets.VBox()\n",
    "index_section = widgets.VBox([index_widget, index_inp_box])\n",
    "\n",
    "def update_index_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                               placeholder='GCF_000006765.1',\n",
    "                               style={'description_width':'150px'},\n",
    "                               layout=widgets.Layout(width='300px'))\n",
    "        items.append(textbox)\n",
    "    index_inp_box.children = items\n",
    "    index_inp_box.layout.display = 'block' if num > 0 else 'none'\n",
    "\n",
    "def on_index_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_index_inputs(change['new'])\n",
    "\n",
    "index_widget.observe(on_index_widget_change, names='value')\n",
    "update_index_inputs(index_widget.value)\n",
    "\n",
    "# Bowtie2 widget\n",
    "bowtie2_widget = widgets.Checkbox(description='Bowtie2 alignment', value=False)\n",
    "\n",
    "alignment_widget = widgets.BoundedIntText(description='Strains', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "alignment_inp_box = widgets.VBox()\n",
    "alignment_section = widgets.VBox([alignment_widget, alignment_inp_box])\n",
    "\n",
    "def update_alignment_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                              placeholder='GCF_000006765.1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    alignment_inp_box.children = items\n",
    "\n",
    "def on_alignment_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_alignment_inputs(change['new'])\n",
    "\n",
    "alignment_widget.observe(on_alignment_widget_change, names='value')\n",
    "update_alignment_inputs(alignment_widget.value)\n",
    "\n",
    "def toggle_alignment_section(change):\n",
    "    alignment_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "bowtie2_widget.observe(toggle_alignment_section, names='value')\n",
    "alignment_section.layout.display = 'block' if bowtie2_widget.value else 'none'\n",
    "\n",
    "# SAMtools realign assembly widget\n",
    "realign_widget = widgets.Checkbox(description='SAMtools realign assembly', value=False)\n",
    "\n",
    "realignment_widget = widgets.BoundedIntText(description='Strains', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "realignment_inp_box = widgets.VBox()\n",
    "realignment_section = widgets.VBox([realignment_widget, realignment_inp_box])\n",
    "\n",
    "def update_realignment_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                              placeholder='GCF_000006765.1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    realignment_inp_box.children = items\n",
    "\n",
    "def on_realignment_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_realignment_inputs(change['new'])\n",
    "\n",
    "realignment_widget.observe(on_realignment_widget_change, names='value')\n",
    "update_realignment_inputs(realignment_widget.value)\n",
    "\n",
    "def toggle_realignment_section(change):\n",
    "    realignment_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "realign_widget.observe(toggle_realignment_section, names='value')\n",
    "realignment_section.layout.display = 'block' if realign_widget.value else 'none'\n",
    "\n",
    "# Find SNPs\n",
    "snps_widget = widgets.Checkbox(description='Find SNPs', value=False)\n",
    "\n",
    "# Find indels\n",
    "indels_widget = widgets.Checkbox(description='Find indels', value=False)\n",
    "\n",
    "# Build genome database (snpEff) widget\n",
    "build_widget = widgets.Checkbox(description='Build genome database', value=False)\n",
    "\n",
    "database_widget = widgets.BoundedIntText(description='Databases', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "database_inp_box = widgets.VBox()\n",
    "database_section = widgets.VBox([database_widget, database_inp_box])\n",
    "\n",
    "def update_database_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'NCBI RefSeq {i+1}',\n",
    "                               placeholder='NC_002516.2',\n",
    "                               style={'description_width':'150px'},\n",
    "                               layout=widgets.Layout(width='300px'))\n",
    "        items.append(textbox)\n",
    "    database_inp_box.children = items\n",
    "    database_inp_box.layout.display = 'block' if build_widget.value else 'none'\n",
    "\n",
    "def on_database_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_database_inputs(change['new'])\n",
    "\n",
    "database_widget.observe(on_database_widget_change, names='value')\n",
    "update_database_inputs(database_widget.value)\n",
    "\n",
    "def toggle_database_section(change):\n",
    "    database_section.layout.display = 'block' if change['new'] else 'none'\n",
    "    if change['new']:\n",
    "        update_database_inputs(database_widget.value)\n",
    "\n",
    "build_widget.observe(toggle_database_section, names='value')\n",
    "database_section.layout.display = 'block' if build_widget.value else 'none'\n",
    "\n",
    "# snpEff annotation\n",
    "snpeff_widget = widgets.Checkbox(description='SNP/indel annotation', value=False)\n",
    "\n",
    "ncbi_refseq_widget = widgets.BoundedIntText(description='NCBI RefSeq', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "ncbi_inp_box = widgets.VBox()\n",
    "ncbi_section = widgets.VBox([ncbi_refseq_widget, ncbi_inp_box])\n",
    "\n",
    "def update_ncbi_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'NCBI RefSeq {i+1}',\n",
    "                              placeholder='NC_002516.2',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    ncbi_inp_box.children = items\n",
    "\n",
    "def on_ncbi_refseq_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_ncbi_inputs(change['new'])\n",
    "\n",
    "ncbi_refseq_widget.observe(on_ncbi_refseq_widget_change, names='value')\n",
    "update_ncbi_inputs(ncbi_refseq_widget.value)\n",
    "\n",
    "def toggle_ncbi_section(change):\n",
    "    ncbi_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "snpeff_widget.observe(toggle_ncbi_section, names='value')\n",
    "ncbi_section.layout.display = 'block' if snpeff_widget.value else 'none'\n",
    "\n",
    "# Filter SNP/indel by parental\n",
    "filter_widget = widgets.Checkbox(description='Filter SNP/indel by parental', value=False)\n",
    "\n",
    "parental_widget = widgets.BoundedIntText(description='Parental', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "parental_inp_box = widgets.VBox()\n",
    "parental_section = widgets.VBox([parental_widget, parental_inp_box])\n",
    "\n",
    "def update_parental_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Parental {i+1}',\n",
    "                              placeholder='PAO1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    parental_inp_box.children = items\n",
    "\n",
    "def on_parental_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_parental_inputs(change['new'])\n",
    "\n",
    "parental_widget.observe(on_parental_widget_change, names='value')\n",
    "update_parental_inputs(parental_widget.value)\n",
    "\n",
    "def toggle_parental_section(change):\n",
    "    parental_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "filter_widget.observe(toggle_parental_section, names='value')\n",
    "parental_section.layout.display = 'block' if filter_widget.value else 'none'\n",
    "\n",
    "# Create sample/mutation table widget\n",
    "table_widget = widgets.Checkbox(description='Sample/mutation table', value=False)\n",
    "\n",
    "threshold_widget = widgets.BoundedIntText(description='Threshold', value=1,\n",
    "                                         min=1, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "strains_widget = widgets.BoundedIntText(description='Strains', value=1,\n",
    "                                        min=1, max=10, step=1,\n",
    "                                        style={'description_width':'150px'})\n",
    "\n",
    "strains_inp_box = widgets.VBox()\n",
    "strains_section = widgets.VBox([strains_widget, strains_inp_box])\n",
    "\n",
    "def update_strains_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Strain {i+1}',\n",
    "                              placeholder='PAO1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    strains_inp_box.children = items\n",
    "\n",
    "def on_strains_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_strains_inputs(change['new'])\n",
    "\n",
    "strains_widget.observe(on_strains_widget_change, names='value')\n",
    "update_strains_inputs(strains_widget.value)\n",
    "\n",
    "table_children_box = widgets.VBox([threshold_widget, strains_section])\n",
    "table_section = widgets.VBox([table_widget, table_children_box])\n",
    "table_children_box.layout.display = 'block' if table_widget.value else 'none'\n",
    "\n",
    "def toggle_table_section(change):\n",
    "    table_children_box.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "table_widget.observe(toggle_table_section, names='value')\n",
    "\n",
    "# De novo genome assembly (SPAdes) widget\n",
    "denovo_widget = widgets.Checkbox(description='De novo genome assembly', value=False)\n",
    "\n",
    "# Compress large files widget\n",
    "compress_widget = widgets.Checkbox(description='Compress large files', value=False,\n",
    "                           layout=widgets.Layout(width='300px'))\n",
    "\n",
    "# Comfirm selected options widget\n",
    "confirm_options = widgets.Checkbox(description='Confirm', value=False, indent=False)\n",
    "\n",
    "# configuration_widget behavior\n",
    "def set_disabled_recursive_configuration(widget, disabled=True):\n",
    "    if isinstance(widget, widgets.Box):\n",
    "        new_children = []\n",
    "        for child in widget.children:\n",
    "            if isinstance(child, widgets.FileUpload):\n",
    "                # Create a new FileUpload widget\n",
    "                new_widget = widgets.FileUpload(\n",
    "                    accept=child.accept,\n",
    "                    multiple=child.multiple,\n",
    "                    layout=child.layout\n",
    "                )\n",
    "                new_children.append(new_widget)\n",
    "            else:\n",
    "                set_disabled_recursive_configuration(child, disabled)\n",
    "                new_children.append(child)\n",
    "        widget.children = tuple(new_children)\n",
    "    else:\n",
    "        if hasattr(widget, 'disabled'):\n",
    "            widget.disabled = disabled\n",
    "        if isinstance(widget, widgets.Checkbox):\n",
    "            widget.value = False\n",
    "        elif isinstance(widget, widgets.BoundedIntText):\n",
    "            widget.value = 0\n",
    "\n",
    "# confirm_widget behavior\n",
    "def set_disabled_recursive_confirm(widget, disabled=True):\n",
    "    widget.disabled = disabled\n",
    "    if hasattr(widget, 'children'):\n",
    "        for child in widget.children:\n",
    "            set_disabled_recursive_confirm(child, disabled)\n",
    "\n",
    "def on_configuration_widget(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_configuration_inputs(change['new'])\n",
    "        state = change['new']\n",
    "\n",
    "        merge_widget.disabled = state\n",
    "        merge_widget.value = False\n",
    "\n",
    "        download_widget.disabled = state\n",
    "        download_widget.value = False\n",
    "        set_disabled_recursive_configuration(download_section, state)\n",
    "\n",
    "        index_widget.disabled = state\n",
    "        index_widget.value = False\n",
    "        set_disabled_recursive_configuration(index_section, state)\n",
    "        \n",
    "        bowtie2_widget.disabled = state\n",
    "        bowtie2_widget.value = False\n",
    "        set_disabled_recursive_configuration(alignment_section, state)\n",
    "        \n",
    "        realign_widget.disabled = state\n",
    "        realign_widget.value = False\n",
    "        set_disabled_recursive_configuration(realignment_section, state)\n",
    "        \n",
    "        snps_widget.disabled = state\n",
    "        snps_widget.value = False\n",
    "        \n",
    "        indels_widget.disabled = state\n",
    "        indels_widget.value = False\n",
    "        \n",
    "        build_widget.disabled = state\n",
    "        build_widget.value = False\n",
    "        set_disabled_recursive_configuration(database_section, state)\n",
    "        \n",
    "        snpeff_widget.disabled = state\n",
    "        snpeff_widget.value = False\n",
    "        set_disabled_recursive_configuration(ncbi_section, state)\n",
    "        \n",
    "        filter_widget.disabled = state\n",
    "        filter_widget.value = False\n",
    "        set_disabled_recursive_configuration(parental_section, state)\n",
    "\n",
    "        table_widget.disabled = state\n",
    "        table_widget.value = False\n",
    "        set_disabled_recursive_configuration(strains_section, state)\n",
    "        \n",
    "        denovo_widget.disabled = state\n",
    "        denovo_widget.value = False\n",
    "\n",
    "        compress_widget.disabled = state\n",
    "        compress_widget.value = False\n",
    "\n",
    "def on_confirm_options_widget(change):\n",
    "    if change['type'] == 'change' and change['name'] == 'value':\n",
    "        state = change['new']\n",
    "\n",
    "        set_disabled_recursive_confirm(configuration_section, state)\n",
    "        merge_widget.disabled = state\n",
    "        set_disabled_recursive_confirm(download_section, state)\n",
    "        set_disabled_recursive_confirm(index_section, state)\n",
    "        set_disabled_recursive_confirm(bowtie2_widget, state)\n",
    "        set_disabled_recursive_confirm(alignment_section, state)\n",
    "        set_disabled_recursive_confirm(realign_widget, state)\n",
    "        set_disabled_recursive_confirm(realignment_section, state)\n",
    "        snps_widget.disabled = state\n",
    "        indels_widget.disabled = state\n",
    "        set_disabled_recursive_confirm(build_widget, state)\n",
    "        set_disabled_recursive_confirm(database_section, state)\n",
    "        set_disabled_recursive_confirm(snpeff_widget, state)\n",
    "        set_disabled_recursive_confirm(ncbi_section, state)\n",
    "        set_disabled_recursive_confirm(filter_widget, state)\n",
    "        set_disabled_recursive_confirm(parental_section, state)\n",
    "        set_disabled_recursive_confirm(table_section, state)\n",
    "        denovo_widget.disabled = state\n",
    "        compress_widget.disabled = state\n",
    "\n",
    "configuration_widget.observe(on_configuration_widget, names='value')\n",
    "confirm_options.observe(on_confirm_options_widget, names='value')\n",
    "\n",
    "column = widgets.VBox([\n",
    "    configuration_section,\n",
    "    merge_widget,\n",
    "    download_section,\n",
    "    index_widget,\n",
    "    index_inp_box,\n",
    "    bowtie2_widget,\n",
    "    alignment_section,\n",
    "    realign_widget,\n",
    "    realignment_section,\n",
    "    snps_widget,\n",
    "    indels_widget,\n",
    "    build_widget,\n",
    "    database_section,\n",
    "    snpeff_widget,\n",
    "    ncbi_section,\n",
    "    filter_widget,\n",
    "    parental_section,\n",
    "    table_section,\n",
    "    denovo_widget,\n",
    "    compress_widget,\n",
    "    confirm_options\n",
    "])\n",
    "\n",
    "display(column)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f37eda25-9eab-41a9-9d2c-f725e8bfce5e",
   "metadata": {},
   "source": [
    "### Configuration file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d60684fc-2a13-4e73-a892-cfc4fcf19712",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Configuration file uploaded successfully.\n"
     ]
    }
   ],
   "source": [
    "if configuration_widget.value:\n",
    "    # Load configuration from json file\n",
    "    try:\n",
    "        configuration_file = configuration_inp_box.children[0].value[0][\"content\"]\n",
    "        configuration = json.loads(bytes(configuration_file).decode(\"utf-8\"))\n",
    "        print(\"Configuration file uploaded successfully.\")\n",
    "        \n",
    "    except IndexError:\n",
    "        print(\"No file uploaded.\")\n",
    "    except KeyError:\n",
    "        print(\"Failed to load configuration file.\")\n",
    "    \n",
    "else:\n",
    "    # Load parental SNP/indel files\n",
    "    if filter_widget.value:\n",
    "        parental_files = []\n",
    "        for i in range(0, parental_widget.value):\n",
    "            strain = parental_inp_box.children[i].children[0].value\n",
    "            snp_file = filedialog.askopenfile(title=f\"{strain} SNPs\", initialdir=\"/home/student/Variant_Calling/Jupyter\")\n",
    "            indels_file = filedialog.askopenfile(title=f\"{strain} indels\", initialdir=\"/home/student/Variant_Calling/Jupyter\")\n",
    "\n",
    "            parental_files.append((snp_file, indels_file))\n",
    "\n",
    "    # Create configuration\n",
    "    configuration = {\n",
    "        \"merge\": merge_widget.value,\n",
    "        \"download\": {\n",
    "            \"n\": download_widget.value,\n",
    "            \"assemblies\": [box.value for box in download_inp_box.children]\n",
    "        },\n",
    "        \"indexing\": {\n",
    "            \"n\": index_widget.value,\n",
    "            \"assemblies\": [box.value for box in index_inp_box.children]\n",
    "        },\n",
    "        \"bowtie2\": {\n",
    "            \"enabled\": bowtie2_widget.value,\n",
    "            \"n\": alignment_widget.value,\n",
    "            \"organisms\": [\n",
    "                {\n",
    "                    \"accession\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "            }\n",
    "            for pair in alignment_inp_box.children\n",
    "        ]\n",
    "    },\n",
    "        \"realign\": {\n",
    "            \"enabled\": realign_widget.value,\n",
    "            \"n\": realignment_widget.value,\n",
    "            \"organisms\": [\n",
    "                {\n",
    "                    \"accession\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                    if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                    else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "                }\n",
    "                for pair in realignment_inp_box.children\n",
    "            ]\n",
    "        },\n",
    "        \"snps\": snps_widget.value,\n",
    "        \"indels\": indels_widget.value,\n",
    "        \"build\": {\n",
    "            \"enabled\": build_widget.value,\n",
    "            \"n\": database_widget.value,\n",
    "            \"databases\": [box.value for box in database_inp_box.children]\n",
    "        },\n",
    "        \"SNP_indel_annotation\": {\n",
    "            \"enabled\": snpeff_widget.value,\n",
    "            \"n\": ncbi_refseq_widget.value,\n",
    "            \"Reference_Sequences\": [\n",
    "                {\n",
    "                    \"accession\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "            }\n",
    "            for pair in ncbi_inp_box.children\n",
    "        ]\n",
    "    },\n",
    "        \"filter_by_parental\": {\n",
    "            \"enabled\": filter_widget.value,\n",
    "            \"n\": parental_widget.value,\n",
    "            \"Parental\": [\n",
    "                {\n",
    "                    \"strain\": pair.children[0].value,\n",
    "                    \"snps_path\": snp.name,\n",
    "                    \"indels_path\": indels.name,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "            }\n",
    "            for pair, (snp, indels) in zip(parental_inp_box.children, parental_files)\n",
    "        ]\n",
    "    },\n",
    "        \"table\": {\n",
    "            \"enabled\": table_widget.value,\n",
    "            \"threshold\": threshold_widget.value,\n",
    "            \"n\": strains_widget.value,\n",
    "            \"strains\": [\n",
    "                {\n",
    "                    \"strain\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "                }\n",
    "                for pair in strains_inp_box.children\n",
    "            ]\n",
    "        },\n",
    "        \"denovo\": denovo_widget.value,\n",
    "        \"compress\": compress_widget.value\n",
    "    }\n",
    "\n",
    "    now = datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "    configuration[\"generated_at\"] = now\n",
    "\n",
    "    config_file_path = os.path.join(main_directory_path, \"config.json\")\n",
    "\n",
    "    # Create configuration file\n",
    "    with open(config_file_path, 'w') as config_file:\n",
    "        json.dump(configuration, config_file, indent=4)\n",
    "\n",
    "    print(\"Configuration file create in\", config_file_path)\n",
    "\n",
    "# Initialize sample dictionaries\n",
    "bowtie2_sample_genome_dict = {}\n",
    "realign_sample_genome_dict = {}\n",
    "annotation_sample_genome_dict = {}\n",
    "filter_sample_strain_dict = {}\n",
    "table_sample_strain_dict = {}\n",
    "\n",
    "for organism in configuration[\"bowtie2\"][\"organisms\"]:\n",
    "    accession = organism[\"accession\"]\n",
    "    samples = organism[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in bowtie2_sample_genome_dict:\n",
    "            bowtie2_sample_genome_dict[sample] = accession\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for organism in configuration[\"realign\"][\"organisms\"]:\n",
    "    accession = organism[\"accession\"]\n",
    "    samples = organism[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in realign_sample_genome_dict:\n",
    "            realign_sample_genome_dict[sample] = accession\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for refseq in configuration[\"SNP_indel_annotation\"][\"Reference_Sequences\"]:\n",
    "    accession = refseq[\"accession\"]\n",
    "    samples = refseq[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in annotation_sample_genome_dict:\n",
    "            annotation_sample_genome_dict[sample] = accession\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for strain in configuration[\"filter_by_parental\"][\"Parental\"]:\n",
    "    parental = strain[\"strain\"]\n",
    "    samples = strain[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in filter_sample_strain_dict:\n",
    "            filter_sample_strain_dict[sample] = parental\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for strain in configuration[\"table\"][\"strains\"]:\n",
    "    strain_name = strain[\"strain\"]\n",
    "    for sample in strain[\"samples\"]:\n",
    "        if sample not in table_sample_strain_dict:\n",
    "            table_sample_strain_dict[sample] = strain_name\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8e10dd3-bf19-44c1-813a-35a300af87f6",
   "metadata": {},
   "source": [
    "## Merge fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "16766547-6862-47c2-9117-5b10d3918101",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These samples are going to be processed: \n",
      "  - PAO1wt12P\n",
      "  - PA14wt0253G\n",
      "  - PA14wt0251G\n",
      "\n",
      "PAO1wt12P processed in 9.74 seconds.\n",
      "PA14wt0253G processed in 20.59 seconds.\n",
      "PA14wt0251G processed in 8.75 seconds.\n",
      "\n",
      "Merge finished!\n"
     ]
    }
   ],
   "source": [
    "# Merge fastq files\n",
    "if configuration[\"merge\"]:\n",
    "\n",
    "    files = os.listdir(Fastq_directory_path)\n",
    "\n",
    "    samples = {file.split(\"_\")[0] for file in files}\n",
    "\n",
    "    print(\"These samples are going to be processed: \")\n",
    "    for sample in samples:\n",
    "        print(f\"  - {sample}\")\n",
    "    \n",
    "    print()\n",
    "\n",
    "    for sample in samples:\n",
    "        start_time = time.time()\n",
    "    \n",
    "        L1_directory = [file for file in files if file.startswith(f\"{sample}_L1\")]\n",
    "        L1_directory_path = os.path.join(Fastq_directory_path, \"\".join(L1_directory))\n",
    "\n",
    "        L2_directory = [file for file in files if file.startswith(f\"{sample}_L2\")]\n",
    "        L2_directory_path = os.path.join(Fastq_directory_path, \"\".join(L2_directory))\n",
    "\n",
    "        L1_files = os.listdir(L1_directory_path)\n",
    "        L2_files = os.listdir(L2_directory_path)\n",
    "    \n",
    "        L1_R1_file = [file for file in L1_files if \"_R1_\" in file]\n",
    "        L1_R1_file_path = os.path.join(L1_directory_path, \"\".join(L1_R1_file))\n",
    "        L2_R1_file = [file for file in L2_files if \"_R1_\" in file]\n",
    "        L2_R1_file_path = os.path.join(L2_directory_path, \"\".join(L2_R1_file))\n",
    "\n",
    "        R1_files_path = [L1_R1_file_path, L2_R1_file_path]\n",
    "\n",
    "        L1_R2_file = [file for file in L1_files if \"_R2_\" in file]\n",
    "        L1_R2_file_path = os.path.join(L1_directory_path, \"\".join(L1_R2_file))\n",
    "        L2_R2_file = [file for file in L2_files if \"_R2_\" in file]\n",
    "        L2_R2_file_path = os.path.join(L2_directory_path, \"\".join(L2_R2_file))\n",
    "\n",
    "        R2_files_path = [L1_R2_file_path, L2_R2_file_path]\n",
    "\n",
    "        ## Merge R1 files\n",
    "        merged_R1_file = f\"{sample}_R1.fastq.gz\"\n",
    "        merged_R1_file_path = os.path.join(fastq_directory_path, merged_R1_file)\n",
    "\n",
    "        with open(merged_R1_file_path, 'wb') as R1_file:\n",
    "            for file in R1_files_path:\n",
    "                with open(file, 'rb') as my_file:\n",
    "                    shutil.copyfileobj(my_file, R1_file)\n",
    "\n",
    "        ## Merge R2 files\n",
    "        merged_R2_file = f\"{sample}_R2.fastq.gz\"\n",
    "        merged_R2_file_path = os.path.join(fastq_directory_path, merged_R2_file)\n",
    "\n",
    "        with open(merged_R2_file_path, 'wb') as R2_file:\n",
    "            for file in R2_files_path:\n",
    "                with open(file, 'rb') as my_file:\n",
    "                    shutil.copyfileobj(my_file, R2_file)\n",
    "    \n",
    "        end_time = time.time()\n",
    "\n",
    "        execution_time = end_time - start_time\n",
    "        print(f\"{sample} processed in {execution_time:.2f} seconds.\")\n",
    "\n",
    "    print()\n",
    "    print(\"Merge finished!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe0af47a-cac8-42f9-b9fe-3c530265b261",
   "metadata": {},
   "source": [
    "## Download reference genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c28cf1f-73e8-4d85-9b0b-db3254cf4bee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCBI RefSeq assembly GCF_000006765.1 downloaded successfully.\n",
      "NCBI RefSeq assembly GCF_000006765.1 unzipped successfully.\n",
      "\n",
      "NCBI RefSeq assembly GCF_000014625.1 downloaded successfully.\n",
      "NCBI RefSeq assembly GCF_000014625.1 unzipped successfully.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"download\"][\"n\"]:\n",
    "\n",
    "    assemblies = configuration[\"download\"][\"assemblies\"]\n",
    "\n",
    "    # Download reference genome .fastq file from NCBI\n",
    "    for assembly_id in assemblies:\n",
    "        url = f\"https://api.ncbi.nlm.nih.gov/datasets/v2/genome/accession/{assembly_id}/download?include_annotation_type=GENOME_FASTA&hydrated=FULLY_HYDRATED\"\n",
    "\n",
    "        try:\n",
    "            response = requests.get(url, stream=True, timeout=60)\n",
    "            response.raise_for_status()\n",
    "        except requests.exceptions.RequestException as download_error:\n",
    "            print(f\"Error downloading the assembly {assembly_id}: {download_error}\")\n",
    "            continue\n",
    "\n",
    "        if response.status_code == 200:\n",
    "            ref_genome_file = f\"{assembly_id}.zip\"\n",
    "            ref_genome_file_path = os.path.join(genomes_directory_path, ref_genome_file)\n",
    "            with open(ref_genome_file_path, 'wb') as genome_file:\n",
    "                for chunk in response.iter_content(chunk_size=8192):\n",
    "                    if chunk:\n",
    "                        genome_file.write(chunk)\n",
    "            print(f\"NCBI RefSeq assembly {assembly_id} downloaded successfully.\")\n",
    "\n",
    "            assembly_id_directory = os.path.join(genomes_directory_path, assembly_id)\n",
    "            if not os.path.exists(assembly_id_directory):\n",
    "                os.mkdir(assembly_id_directory)\n",
    "            \n",
    "            with zipfile.ZipFile(ref_genome_file_path, 'r') as zip_ref:\n",
    "                zip_ref.extractall(path=assembly_id_directory)\n",
    "            print(f\"NCBI RefSeq assembly {assembly_id} unzipped successfully.\")\n",
    "            print()\n",
    "\n",
    "            os.remove(ref_genome_file_path)\n",
    "        \n",
    "        else:\n",
    "            print(f\"Error downloading the NCBI RefSeq assembly {assembly_id}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64bb1ae5-cd5e-4423-a67a-4225fc791a88",
   "metadata": {},
   "source": [
    "## Genome indexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bada94c6-af67-4fb9-bb53-1d80308c2019",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexing GCF_000006765.1 genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO\t2025-06-04 09:04:51\tNormalizeFasta\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    NormalizeFasta -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/GCF_000006765.1/ncbi_dataset/data/GCF_000006765.1/GCF_000006765.1_ASM676v1_genomic.fna -O /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta\n",
      "**********\n",
      "\n",
      "\n",
      "09:04:51.613 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 09:04:51 CEST 2025] NormalizeFasta INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/GCF_000006765.1/ncbi_dataset/data/GCF_000006765.1/GCF_000006765.1_ASM676v1_genomic.fna OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta    LINE_LENGTH=100 TRUNCATE_SEQUENCE_NAMES_AT_WHITESPACE=false VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 09:04:52 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "[Wed Jun 04 09:04:52 CEST 2025] picard.reference.NormalizeFasta done. Elapsed time: 0.02 minutes.\n",
      "Runtime.totalMemory()=168296448\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000006765.1.dict already exists.\n",
      "Settings:\n",
      "  Output files: \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta\n",
      "Reading reference sizes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1566101\n",
      "Using parameters --bmax 1174576 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1174576 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 5; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 783050 (target: 1174575)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (1174576) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 640778 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 640779 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (1174576) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 934176 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 934177 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (1174576) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 682832 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 682833 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (1174576) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 908487 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 908488 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (1174576) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 321897 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 321898 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (1174576) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 940198 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 940199 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (1174576) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 977225 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 977226 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (1174576) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 858804 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 858805 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1056134\n",
      "fchr[G]: 3158821\n",
      "fchr[T]: 5225454\n",
      "fchr[$]: 6264404\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6282714 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.1.bt2.tmp\n",
      "Wrote 1566108 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6264404\n",
      "    bwtLen: 6264405\n",
      "    sz: 1566101\n",
      "    bwtSz: 1566102\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 391526\n",
      "    offsSz: 1566104\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 32628\n",
      "    numLines: 32628\n",
      "    ebwtTotLen: 2088192\n",
      "    ebwtTotSz: 2088192\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:06\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1566101\n",
      "Using parameters --bmax 1174576 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1174576 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:01\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 894914 (target: 1174575)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (1174576) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 960633 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 960634 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (1174576) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 806333 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 806334 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (1174576) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 464337 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 464338 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (1174576) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1050739 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1050740 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (1174576) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1008209 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1008210 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (1174576) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1058957 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1058958 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (1174576) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 915190 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 915191 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1056134\n",
      "fchr[G]: 3158821\n",
      "fchr[T]: 5225454\n",
      "fchr[$]: 6264404\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6282714 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.1.bt2.tmp\n",
      "Wrote 1566108 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6264404\n",
      "    bwtLen: 6264405\n",
      "    sz: 1566101\n",
      "    bwtSz: 1566102\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 391526\n",
      "    offsSz: 1566104\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 32628\n",
      "    numLines: 32628\n",
      "    ebwtTotLen: 2088192\n",
      "    ebwtTotSz: 2088192\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:05\n",
      "GCF_000006765.1 indexed successfully!\n",
      "Indexing GCF_000014625.1 genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.3.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.3.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.4.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.4.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.1.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.2.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.2.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.1.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.2.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.2.bt2\n",
      "INFO\t2025-06-04 09:05:05\tNormalizeFasta\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    NormalizeFasta -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/GCF_000014625.1/ncbi_dataset/data/GCF_000014625.1/GCF_000014625.1_ASM1462v1_genomic.fna -O /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta\n",
      "**********\n",
      "\n",
      "\n",
      "09:05:05.797 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 09:05:05 CEST 2025] NormalizeFasta INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/GCF_000014625.1/ncbi_dataset/data/GCF_000014625.1/GCF_000014625.1_ASM1462v1_genomic.fna OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta    LINE_LENGTH=100 TRUNCATE_SEQUENCE_NAMES_AT_WHITESPACE=false VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 09:05:06 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "[Wed Jun 04 09:05:07 CEST 2025] picard.reference.NormalizeFasta done. Elapsed time: 0.02 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "Building a SMALL index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000014625.1.dict already exists.\n",
      "Settings:\n",
      "  Output files: \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1634409\n",
      "Using parameters --bmax 1225807 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1225807 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 7; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 817204 (target: 1225806)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (1225807) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 913362 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 913363 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (1225807) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 504196 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 504197 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (1225807) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 901912 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 901913 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (1225807) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 463480 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 463481 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (1225807) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1225739 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1225740 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (1225807) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1123799 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1123800 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (1225807) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 269747 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 269748 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (1225807) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 1135395 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1135396 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1102316\n",
      "fchr[G]: 3269680\n",
      "fchr[T]: 5436267\n",
      "fchr[$]: 6537637\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6373902 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.1.bt2.tmp\n",
      "Wrote 1634416 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6537637\n",
      "    bwtLen: 6537638\n",
      "    sz: 1634410\n",
      "    bwtSz: 1634410\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 408603\n",
      "    offsSz: 1634412\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 34051\n",
      "    numLines: 34051\n",
      "    ebwtTotLen: 2179264\n",
      "    ebwtTotSz: 2179264\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:05\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1634409\n",
      "Using parameters --bmax 1225807 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1225807 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 726403 (target: 1225806)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 9\n",
      "  Reserving size (1225807) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 125530 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 125531 for bucket 1\n",
      "Getting block 2 of 9\n",
      "  Reserving size (1225807) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1151245 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1151246 for bucket 2\n",
      "Getting block 3 of 9\n",
      "  Reserving size (1225807) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 855141 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 855142 for bucket 3\n",
      "Getting block 4 of 9\n",
      "  Reserving size (1225807) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1133498 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1133499 for bucket 4\n",
      "Getting block 5 of 9\n",
      "  Reserving size (1225807) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 288814 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 288815 for bucket 5\n",
      "Getting block 6 of 9\n",
      "  Reserving size (1225807) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 974385 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 974386 for bucket 6\n",
      "Getting block 7 of 9\n",
      "  Reserving size (1225807) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 303803 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 303804 for bucket 7\n",
      "Getting block 8 of 9\n",
      "  Reserving size (1225807) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 1088162 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1088163 for bucket 8\n",
      "Getting block 9 of 9\n",
      "  Reserving size (1225807) for bucket 9\n",
      "  Calculating Z arrays for bucket 9\n",
      "  Entering block accumulator loop for bucket 9:\n",
      "  bucket 9: 10%\n",
      "  bucket 9: 20%\n",
      "  bucket 9: 30%\n",
      "  bucket 9: 40%\n",
      "  bucket 9: 50%\n",
      "  bucket 9: 60%\n",
      "  bucket 9: 70%\n",
      "  bucket 9: 80%\n",
      "  bucket 9: 90%\n",
      "  bucket 9: 100%\n",
      "  Sorting block of length 617051 for bucket 9\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 617052 for bucket 9\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1102316\n",
      "fchr[G]: 3269680\n",
      "fchr[T]: 5436267\n",
      "fchr[$]: 6537637\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6373902 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.1.bt2.tmp\n",
      "Wrote 1634416 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6537637\n",
      "    bwtLen: 6537638\n",
      "    sz: 1634410\n",
      "    bwtSz: 1634410\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 408603\n",
      "    offsSz: 1634412\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 34051\n",
      "    numLines: 34051\n",
      "    ebwtTotLen: 2179264\n",
      "    ebwtTotSz: 2179264\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:05\n",
      "GCF_000014625.1 indexed successfully!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.3.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.3.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.4.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.4.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.1.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.2.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.2.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.1.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.2.bt2.tmp to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.2.bt2\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"indexing\"][\"n\"]:\n",
    "\n",
    "    assemblies = configuration[\"indexing\"][\"assemblies\"]\n",
    "\n",
    "    for assembly_id in assemblies:\n",
    "        print(f\"Indexing {assembly_id} genome\")\n",
    "        assembly_fasta_file_directory_path = os.path.join(genomes_directory_path, assembly_id,\n",
    "                                                          \"ncbi_dataset/data\", assembly_id)\n",
    "    \n",
    "        fasta_file = os.listdir(assembly_fasta_file_directory_path)\n",
    "        fasta_file_path = os.path.join(assembly_fasta_file_directory_path, \"\".join(fasta_file))\n",
    "\n",
    "        indexed_assembly_directory_path = os.path.join(genomes_directory_path, \"indexed\", assembly_id)\n",
    "        if not os.path.exists(indexed_assembly_directory_path):\n",
    "            os.mkdir(indexed_assembly_directory_path)\n",
    "        \n",
    "        N_fasta_file_path = os.path.join(indexed_assembly_directory_path, f\"N_{assembly_id}.fasta\")\n",
    "\n",
    "        # Normalize Fasta\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"NormalizeFasta\",\n",
    "            \"I=\" + fasta_file_path,\n",
    "            \"O=\" + N_fasta_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        # Creating .dict file\n",
    "        reference = N_fasta_file_path\n",
    "        dict_file_path = os.path.join(indexed_assembly_directory_path, f\"N_{assembly_id}.dict\")\n",
    "\n",
    "        if os.path.exists(dict_file_path):\n",
    "            print(f\"{assembly_id}.dict already exists.\")\n",
    "        else:\n",
    "            command = [\n",
    "                \"java\", \"-jar\", picard_jar, \"CreateSequenceDictionary\",\n",
    "                \"R=\" + reference,\n",
    "                \"O=\" + dict_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, check=True)\n",
    "\n",
    "        # Creating .fasta.fai files\n",
    "        command = [\n",
    "            samtools, \"faidx\", reference\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        # Building bowtie2 index\n",
    "        index_directory_path = os.path.join(indexed_assembly_directory_path, \"index\")\n",
    "\n",
    "        if not os.path.exists(index_directory_path):\n",
    "            os.mkdir(index_directory_path)\n",
    "\n",
    "        index_file_path = os.path.join(index_directory_path, f\"{assembly_id}_index\")\n",
    "    \n",
    "        command = [\n",
    "            bowtie2_build, reference, index_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "        print(f\"{assembly_id} indexed successfully!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47df1025-481c-4caf-a0a4-145e2ec7021a",
   "metadata": {},
   "source": [
    "## Bowtie2 alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dc79896f-66a4-4628-bfb9-4032d2b4aeee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 samples will be aligned with GCF_000006765.1 if possible...\n",
      "\n",
      "Aligning PAO1wt12P...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3281626 reads; of these:\n",
      "  3281626 (100.00%) were paired; of these:\n",
      "    73435 (2.24%) aligned concordantly 0 times\n",
      "    3151065 (96.02%) aligned concordantly exactly 1 time\n",
      "    57126 (1.74%) aligned concordantly >1 times\n",
      "    ----\n",
      "    73435 pairs aligned concordantly 0 times; of these:\n",
      "      12472 (16.98%) aligned discordantly 1 time\n",
      "    ----\n",
      "    60963 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      121926 mates make up the pairs; of these:\n",
      "        80327 (65.88%) aligned 0 times\n",
      "        40207 (32.98%) aligned exactly 1 time\n",
      "        1392 (1.14%) aligned >1 times\n",
      "98.78% overall alignment rate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PAO1wt12P aligned. Execution time 1980.75 s\n",
      "\n",
      "Aligned 1/1 with GCF_000006765.1\n",
      "\n",
      "2 samples will be aligned with GCF_000014625.1 if possible...\n",
      "\n",
      "Aligning PA14wt0253G...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6236138 reads; of these:\n",
      "  6236138 (100.00%) were paired; of these:\n",
      "    122207 (1.96%) aligned concordantly 0 times\n",
      "    5984387 (95.96%) aligned concordantly exactly 1 time\n",
      "    129544 (2.08%) aligned concordantly >1 times\n",
      "    ----\n",
      "    122207 pairs aligned concordantly 0 times; of these:\n",
      "      39330 (32.18%) aligned discordantly 1 time\n",
      "    ----\n",
      "    82877 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      165754 mates make up the pairs; of these:\n",
      "        107625 (64.93%) aligned 0 times\n",
      "        55538 (33.51%) aligned exactly 1 time\n",
      "        2591 (1.56%) aligned >1 times\n",
      "99.14% overall alignment rate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PA14wt0253G aligned. Execution time 3687.32 s\n",
      "\n",
      "Aligned 1/2 with GCF_000014625.1\n",
      "Aligning PA14wt0251G...\n",
      "\n",
      "PA14wt0251G aligned. Execution time 1734.67 s\n",
      "\n",
      "Aligned 2/2 with GCF_000014625.1\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2717352 reads; of these:\n",
      "  2717352 (100.00%) were paired; of these:\n",
      "    39751 (1.46%) aligned concordantly 0 times\n",
      "    2625130 (96.61%) aligned concordantly exactly 1 time\n",
      "    52471 (1.93%) aligned concordantly >1 times\n",
      "    ----\n",
      "    39751 pairs aligned concordantly 0 times; of these:\n",
      "      9881 (24.86%) aligned discordantly 1 time\n",
      "    ----\n",
      "    29870 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      59740 mates make up the pairs; of these:\n",
      "        35788 (59.91%) aligned 0 times\n",
      "        23034 (38.56%) aligned exactly 1 time\n",
      "        918 (1.54%) aligned >1 times\n",
      "99.34% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"bowtie2\"][\"enabled\"]:\n",
    "    for organism in configuration[\"bowtie2\"][\"organisms\"]:\n",
    "        genome = organism[\"accession\"]\n",
    "        samples = organism[\"samples\"]\n",
    "            \n",
    "        print(f\"{len(samples)} samples will be aligned with {genome} if possible...\")\n",
    "\n",
    "        index_assembly_directory_path = os.path.join(genomes_directory_path, \"indexed\", genome, \"index\", f\"{genome}_index\")\n",
    "\n",
    "        aligned_count = 0\n",
    "\n",
    "        print()\n",
    "    \n",
    "        for sample in samples:\n",
    "            aligned_count += 1        \n",
    "            start_time = time.time()\n",
    "\n",
    "            # Decompress R1 and R2 files\n",
    "            compressed_R1_file_path = os.path.join(fastq_directory_path, f\"{sample}_R1.fastq.gz\")\n",
    "            compressed_R2_file_path = os.path.join(fastq_directory_path, f\"{sample}_R2.fastq.gz\")\n",
    "\n",
    "            decompressed_R1_file_path = os.path.join(fastq_directory_path, f\"{sample}_R1.fastq\")\n",
    "            decompressed_R2_file_path = os.path.join(fastq_directory_path, f\"{sample}_R2.fastq\")\n",
    "            \n",
    "            with gzip.open(compressed_R1_file_path, 'rb') as compressed_R1_file:\n",
    "                with open(decompressed_R1_file_path, 'wb') as decompressed_R1_file:\n",
    "                    shutil.copyfileobj(compressed_R1_file, decompressed_R1_file)\n",
    "\n",
    "            with gzip.open(compressed_R2_file_path, 'rb') as compressed_R2_file:\n",
    "                with open(decompressed_R2_file_path, 'wb') as decompressed_R2_file:\n",
    "                    shutil.copyfileobj(compressed_R2_file, decompressed_R2_file)\n",
    "            \n",
    "            reference_genome_index = index_assembly_directory_path\n",
    "            \n",
    "            # Bowtie2 alignment\n",
    "            print(f\"Aligning {sample}...\")\n",
    "            \n",
    "            R1_file = decompressed_R1_file_path\n",
    "            R2_file = decompressed_R2_file_path\n",
    "\n",
    "            samfiles_file_path = os.path.join(samfiles_directory_path, f\"{sample}_map{genome}.sam\")\n",
    "    \n",
    "            command = [\n",
    "                bowtie2, \"--phred33\", \"-x\", reference_genome_index,\n",
    "                \"-q\", \"-1\", R1_file, \"-2\", R2_file,\n",
    "                \"-X\", \"1000\", \"-S\", samfiles_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, check=True)\n",
    "            \n",
    "            end_time = time.time()\n",
    "            execution_time = end_time - start_time\n",
    "            \n",
    "            print()\n",
    "            print(f\"{sample} aligned. Execution time {execution_time:.2f} s\")\n",
    "            print()\n",
    "            print(f\"Aligned {aligned_count}/{len(samples)} with {genome}\")\n",
    "            \n",
    "            if configuration[\"compress\"]:\n",
    "\n",
    "                print()\n",
    "                print(f\"Compressing .sam {sample} file\")\n",
    "                \n",
    "                # Compress resulting .sam file and remove .sam file\n",
    "                compressed_sam_file_path = os.path.join(samfiles_directory_path, f\"{samfiles_file_path}.gz\")\n",
    "\n",
    "                with open(samfiles_file_path, 'rb') as sam_file:\n",
    "                    with gzip.open(compressed_sam_file_path, 'wb') as compressed_sam_file:\n",
    "                        shutil.copyfileobj(sam_file, compressed_sam_file)\n",
    "\n",
    "                os.remove(samfiles_file_path)\n",
    "            \n",
    "            # Remove decompressed .fastq files\n",
    "            os.remove(decompressed_R1_file_path)\n",
    "            os.remove(decompressed_R2_file_path)\n",
    "\n",
    "        \n",
    "        print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b61aa96e-818a-4c03-86ce-05f0df1e0063",
   "metadata": {},
   "source": [
    "## SAMtools realign NCBI RefSeq assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "585b30c9-e654-46de-9757-51e63a6bf2ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing PA14wt0251G...\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[samopen] SAM header is present: 1 sequences.\n",
      "INFO\t2025-06-04 11:10:14\tSortSam\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    SortSam -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted.bam -SORT_ORDER coordinate\n",
      "**********\n",
      "\n",
      "\n",
      "11:10:14.800 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:10:14 CEST 2025] SortSam INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted.bam SORT_ORDER=coordinate    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:10:15 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:11:00\tSortSam\tFinished reading inputs, merging and writing to output now.\n",
      "[Wed Jun 04 11:11:42 CEST 2025] picard.sam.SortSam done. Elapsed time: 1.46 minutes.\n",
      "Runtime.totalMemory()=1585971200\n",
      "INFO\t2025-06-04 11:11:44\tAddOrReplaceReadGroups\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    AddOrReplaceReadGroups -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg.bam -LB /media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS -PL Illumina -PU /media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS -SM /media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS\n",
      "**********\n",
      "\n",
      "\n",
      "11:11:44.397 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:11:44 CEST 2025] AddOrReplaceReadGroups INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg.bam RGLB=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS RGPL=Illumina RGPU=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS RGSM=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS    RGID=1 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:11:44 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:11:44\tAddOrReplaceReadGroups\tCreated read-group ID=1 PL=Illumina LB=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS SM=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0251G_WGS\n",
      "\n",
      "INFO\t2025-06-04 11:11:56\tAddOrReplaceReadGroups\tProcessed     1,000,000 records.  Elapsed time: 00:00:11s.  Time for last 1,000,000:   11s.  Last read position: NC_008463.1:1,224,992\n",
      "INFO\t2025-06-04 11:12:06\tAddOrReplaceReadGroups\tProcessed     2,000,000 records.  Elapsed time: 00:00:21s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:2,435,852\n",
      "INFO\t2025-06-04 11:12:16\tAddOrReplaceReadGroups\tProcessed     3,000,000 records.  Elapsed time: 00:00:31s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:3,708,923\n",
      "INFO\t2025-06-04 11:12:25\tAddOrReplaceReadGroups\tProcessed     4,000,000 records.  Elapsed time: 00:00:40s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:4,890,146\n",
      "INFO\t2025-06-04 11:12:34\tAddOrReplaceReadGroups\tProcessed     5,000,000 records.  Elapsed time: 00:00:50s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:6,048,687\n",
      "[Wed Jun 04 11:12:39 CEST 2025] picard.sam.AddOrReplaceReadGroups done. Elapsed time: 0.92 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO\t2025-06-04 11:12:41\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bam -METRICS_FILE /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_metrics.txt\n",
      "**********\n",
      "\n",
      "\n",
      "11:12:41.531 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:12:41 CEST 2025] MarkDuplicates INPUT=[/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg.bam] OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bam METRICS_FILE=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_DUP_STRATEGY=FLOW_QUALITY_SUM_STRATEGY FLOW_USE_END_IN_UNPAIRED_READS=false FLOW_USE_UNPAIRED_CLIPPED_END=false FLOW_UNPAIRED_END_UNCERTAINTY=0 FLOW_UNPAIRED_START_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true REMOVE_DUPLICATES=false ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:12:41 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:12:42\tMarkDuplicates\tStart of doWork freeMemory: 158622760; totalMemory: 168361984; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:12:42\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2025-06-04 11:12:42\tMarkDuplicates\tWill retain up to 9688405 data points before spilling to disk.\n",
      "INFO\t2025-06-04 11:12:48\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:06s.  Time for last 1,000,000:    6s.  Last read position: NC_008463.1:1,224,992\n",
      "INFO\t2025-06-04 11:12:48\tMarkDuplicates\tTracking 178 as yet unmatched pairs. 178 records in RAM.\n",
      "INFO\t2025-06-04 11:12:51\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:09s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:2,435,852\n",
      "INFO\t2025-06-04 11:12:51\tMarkDuplicates\tTracking 194 as yet unmatched pairs. 194 records in RAM.\n",
      "INFO\t2025-06-04 11:12:55\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:13s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:3,708,923\n",
      "INFO\t2025-06-04 11:12:55\tMarkDuplicates\tTracking 210 as yet unmatched pairs. 210 records in RAM.\n",
      "INFO\t2025-06-04 11:12:58\tMarkDuplicates\tRead     4,000,000 records.  Elapsed time: 00:00:16s.  Time for last 1,000,000:    2s.  Last read position: NC_008463.1:4,890,146\n",
      "INFO\t2025-06-04 11:12:58\tMarkDuplicates\tTracking 174 as yet unmatched pairs. 174 records in RAM.\n",
      "INFO\t2025-06-04 11:13:01\tMarkDuplicates\tRead     5,000,000 records.  Elapsed time: 00:00:19s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:6,048,687\n",
      "INFO\t2025-06-04 11:13:01\tMarkDuplicates\tTracking 110 as yet unmatched pairs. 110 records in RAM.\n",
      "INFO\t2025-06-04 11:13:02\tMarkDuplicates\tRead 5422468 records. 0 pairs never matched.\n",
      "INFO\t2025-06-04 11:13:03\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 938901560; totalMemory: 1616814080; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:13:03\tMarkDuplicates\tWill retain up to 83562496 duplicate indices before spilling to disk.\n",
      "INFO\t2025-06-04 11:13:04\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2025-06-04 11:13:07\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2025-06-04 11:13:09\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2025-06-04 11:13:10\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1894346008; totalMemory: 2586193920; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:13:10\tMarkDuplicates\tMarking 501807 records as duplicates.\n",
      "INFO\t2025-06-04 11:13:10\tMarkDuplicates\tFound 34633 optical duplicate clusters.\n",
      "INFO\t2025-06-04 11:13:10\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2025-06-04 11:14:07\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2025-06-04 11:14:07\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2025-06-04 11:14:07\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2025-06-04 11:14:08\tMarkDuplicates\tBefore output close freeMemory: 2562543520; totalMemory: 2586193920; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:14:08\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2025-06-04 11:14:08\tMarkDuplicates\tAfter output close freeMemory: 2322094984; totalMemory: 2344411136; maxMemory: 2673999872\n",
      "[Wed Jun 04 11:14:08 CEST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 1.44 minutes.\n",
      "Runtime.totalMemory()=2344411136\n",
      "INFO\t2025-06-04 11:14:09\tBuildBamIndex\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    BuildBamIndex -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bai\n",
      "**********\n",
      "\n",
      "\n",
      "11:14:09.955 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:14:10 CEST 2025] BuildBamIndex INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bai    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:14:10 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:14:21\tBuildBamIndex\tSuccessfully wrote bam index file /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bai\n",
      "[Wed Jun 04 11:14:21 CEST 2025] picard.sam.BuildBamIndex done. Elapsed time: 0.20 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO  11:14:26,115 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:14:26,125 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  11:14:26,126 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  11:14:26,126 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  11:14:26,130 HelpFormatter - Program Args: -T RealignerTargetCreator -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta -o /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1.realigned.intervals \n",
      "INFO  11:14:26,197 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  11:14:26,198 HelpFormatter - Date/Time: 2025/06/04 11:14:26 \n",
      "INFO  11:14:26,199 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:14:26,199 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:14:27,072 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  11:14:27,313 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  11:14:27,327 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  11:14:27,375 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.05 \n",
      "INFO  11:14:27,546 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  11:14:27,682 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  11:14:27,687 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  11:14:27,688 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  11:14:27,689 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  11:14:57,700 ProgressMeter - NC_008463.1:1577665   1572864.0    30.0 s      19.0 s       24.1%     2.1 m      94.0 s \n",
      "INFO  11:15:27,704 ProgressMeter - NC_008463.1:4557053   4554752.0    60.0 s      13.0 s       69.7%    86.0 s      26.0 s \n",
      "INFO  11:15:50,327 ProgressMeter -            done   6537648.0    82.0 s      12.0 s      100.0%    82.0 s       0.0 s \n",
      "INFO  11:15:50,328 ProgressMeter - Total runtime 82.64 secs, 1.38 min, 0.02 hours \n",
      "INFO  11:15:50,331 MicroScheduler - 531814 reads were filtered out during the traversal out of approximately 5469232 total reads (9.72%) \n",
      "INFO  11:15:50,334 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  11:15:50,337 MicroScheduler -   -> 0 reads (0.00% of total) failing BadMateFilter \n",
      "INFO  11:15:50,337 MicroScheduler -   -> 505691 reads (9.25% of total) failing DuplicateReadFilter \n",
      "INFO  11:15:50,337 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  11:15:50,338 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  11:15:50,338 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  11:15:50,340 MicroScheduler -   -> 26123 reads (0.48% of total) failing MappingQualityZeroFilter \n",
      "INFO  11:15:50,340 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  11:15:50,340 MicroScheduler -   -> 0 reads (0.00% of total) failing Platform454Filter \n",
      "INFO  11:15:50,340 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "INFO  11:15:52,184 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "INFO  11:15:55,688 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:15:55,696 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  11:15:55,697 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  11:15:55,697 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  11:15:55,701 HelpFormatter - Program Args: -T IndelRealigner -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta --maxConsensuses 60 --maxReadsForConsensuses 240 --maxReadsForRealignment 6000 --targetIntervals /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1.realigned.intervals -o /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0251G_mapGCF_000014625.1.realigned.bam \n",
      "INFO  11:15:55,721 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  11:15:55,722 HelpFormatter - Date/Time: 2025/06/04 11:15:55 \n",
      "INFO  11:15:55,723 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:15:55,723 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:15:56,893 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  11:15:56,988 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  11:15:57,005 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  11:15:57,037 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.03 \n",
      "INFO  11:15:57,194 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  11:15:57,199 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  11:15:57,206 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  11:15:57,206 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  11:15:57,208 ProgressMeter -        Location |     reads | elapsed |     reads | completed | runtime |   runtime \n",
      "INFO  11:15:57,332 ReadShardBalancer$1 - Loading BAM index data \n",
      "INFO  11:15:57,338 ReadShardBalancer$1 - Done loading BAM index data \n",
      "INFO  11:16:27,682 ProgressMeter - NC_008463.1:732943    600017.0    30.0 s      50.0 s       11.2%     4.5 m       4.0 m \n",
      "INFO  11:16:58,569 ProgressMeter - NC_008463.1:1825691   1400049.0    61.0 s      43.0 s       27.9%     3.6 m       2.6 m \n",
      "INFO  11:17:29,822 ProgressMeter - NC_008463.1:2930605   2300072.0    92.0 s      40.0 s       44.8%     3.4 m     113.0 s \n",
      "INFO  11:17:59,823 ProgressMeter - NC_008463.1:4130684   3300094.0     2.0 m      37.0 s       63.2%     3.2 m      71.0 s \n",
      "INFO  11:18:29,825 ProgressMeter - NC_008463.1:5140743   4200117.0     2.5 m      36.0 s       78.6%     3.2 m      41.0 s \n",
      "INFO  11:19:01,094 ProgressMeter - NC_008463.1:6218415   5100137.0     3.1 m      36.0 s       95.1%     3.2 m       9.0 s \n",
      "INFO  11:19:09,771 ProgressMeter -            done   5434704.0     3.2 m      35.0 s      100.0%     3.2 m       0.0 s \n",
      "INFO  11:19:09,772 ProgressMeter - Total runtime 192.57 secs, 3.21 min, 0.05 hours \n",
      "INFO  11:19:09,774 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 5434704 total reads (0.00%) \n",
      "INFO  11:19:09,775 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  11:19:09,775 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  11:19:11,431 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "[bam_sort_core] merging from 5 files...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PA14wt0251G processed!!\n",
      "\n",
      "Processing PA14wt0253G...\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[samopen] SAM header is present: 1 sequences.\n",
      "INFO\t2025-06-04 11:39:53\tSortSam\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    SortSam -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam -SORT_ORDER coordinate\n",
      "**********\n",
      "\n",
      "\n",
      "11:39:53.301 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:39:53 CEST 2025] SortSam INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam SORT_ORDER=coordinate    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:39:53 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:41:25\tSortSam\tRead    10,000,000 records.  Elapsed time: 00:01:31s.  Time for last 10,000,000:   91s.  Last read position: NC_008463.1:5,857,452.  Last read name: ADX00106:131:HTWKWDRX5:2:2217:25870:16861\n",
      "INFO\t2025-06-04 11:41:45\tSortSam\tFinished reading inputs, merging and writing to output now.\n",
      "INFO\t2025-06-04 11:42:57\tSortSam\tWrote    10,000,000 records from a sorting collection.  Elapsed time: 00:03:03s.  Time for last 10,000,000:   70s.  Last read position: NC_008463.1:5,392,762\n",
      "[Wed Jun 04 11:43:15 CEST 2025] picard.sam.SortSam done. Elapsed time: 3.38 minutes.\n",
      "Runtime.totalMemory()=929165312\n",
      "INFO\t2025-06-04 11:43:18\tAddOrReplaceReadGroups\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    AddOrReplaceReadGroups -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam -LB /media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS -PL Illumina -PU /media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS -SM /media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS\n",
      "**********\n",
      "\n",
      "\n",
      "11:43:18.550 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:43:18 CEST 2025] AddOrReplaceReadGroups INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam RGLB=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS RGPL=Illumina RGPU=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS RGSM=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS    RGID=1 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:43:19 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:43:19\tAddOrReplaceReadGroups\tCreated read-group ID=1 PL=Illumina LB=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS SM=/media/student/EXTERNAL_USB/Variant_Calling/PA14wt0253G_WGS\n",
      "\n",
      "INFO\t2025-06-04 11:43:31\tAddOrReplaceReadGroups\tProcessed     1,000,000 records.  Elapsed time: 00:00:11s.  Time for last 1,000,000:   11s.  Last read position: NC_008463.1:484,706\n",
      "INFO\t2025-06-04 11:43:40\tAddOrReplaceReadGroups\tProcessed     2,000,000 records.  Elapsed time: 00:00:21s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:989,179\n",
      "INFO\t2025-06-04 11:43:50\tAddOrReplaceReadGroups\tProcessed     3,000,000 records.  Elapsed time: 00:00:30s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:1,489,827\n",
      "INFO\t2025-06-04 11:44:00\tAddOrReplaceReadGroups\tProcessed     4,000,000 records.  Elapsed time: 00:00:40s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:2,027,382\n",
      "INFO\t2025-06-04 11:44:09\tAddOrReplaceReadGroups\tProcessed     5,000,000 records.  Elapsed time: 00:00:50s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:2,589,297\n",
      "INFO\t2025-06-04 11:44:19\tAddOrReplaceReadGroups\tProcessed     6,000,000 records.  Elapsed time: 00:00:59s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:3,205,044\n",
      "INFO\t2025-06-04 11:44:28\tAddOrReplaceReadGroups\tProcessed     7,000,000 records.  Elapsed time: 00:01:09s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:3,835,047\n",
      "INFO\t2025-06-04 11:44:38\tAddOrReplaceReadGroups\tProcessed     8,000,000 records.  Elapsed time: 00:01:18s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:4,410,079\n",
      "INFO\t2025-06-04 11:44:47\tAddOrReplaceReadGroups\tProcessed     9,000,000 records.  Elapsed time: 00:01:28s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:4,916,471\n",
      "INFO\t2025-06-04 11:44:56\tAddOrReplaceReadGroups\tProcessed    10,000,000 records.  Elapsed time: 00:01:37s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:5,392,762\n",
      "INFO\t2025-06-04 11:45:06\tAddOrReplaceReadGroups\tProcessed    11,000,000 records.  Elapsed time: 00:01:46s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:5,883,424\n",
      "INFO\t2025-06-04 11:45:15\tAddOrReplaceReadGroups\tProcessed    12,000,000 records.  Elapsed time: 00:01:56s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:6,348,978\n",
      "[Wed Jun 04 11:45:21 CEST 2025] picard.sam.AddOrReplaceReadGroups done. Elapsed time: 2.05 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO\t2025-06-04 11:45:23\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -METRICS_FILE /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_metrics.txt\n",
      "**********\n",
      "\n",
      "\n",
      "11:45:23.727 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:45:23 CEST 2025] MarkDuplicates INPUT=[/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam] OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam METRICS_FILE=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_DUP_STRATEGY=FLOW_QUALITY_SUM_STRATEGY FLOW_USE_END_IN_UNPAIRED_READS=false FLOW_USE_UNPAIRED_CLIPPED_END=false FLOW_UNPAIRED_END_UNCERTAINTY=0 FLOW_UNPAIRED_START_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true REMOVE_DUPLICATES=false ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:45:24 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:45:24\tMarkDuplicates\tStart of doWork freeMemory: 158625304; totalMemory: 168361984; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:45:24\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2025-06-04 11:45:24\tMarkDuplicates\tWill retain up to 9688405 data points before spilling to disk.\n",
      "INFO\t2025-06-04 11:45:30\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:06s.  Time for last 1,000,000:    6s.  Last read position: NC_008463.1:484,706\n",
      "INFO\t2025-06-04 11:45:30\tMarkDuplicates\tTracking 152 as yet unmatched pairs. 152 records in RAM.\n",
      "INFO\t2025-06-04 11:45:34\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:10s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:989,179\n",
      "INFO\t2025-06-04 11:45:34\tMarkDuplicates\tTracking 156 as yet unmatched pairs. 156 records in RAM.\n",
      "INFO\t2025-06-04 11:45:38\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:13s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:1,489,827\n",
      "INFO\t2025-06-04 11:45:38\tMarkDuplicates\tTracking 184 as yet unmatched pairs. 184 records in RAM.\n",
      "INFO\t2025-06-04 11:45:41\tMarkDuplicates\tRead     4,000,000 records.  Elapsed time: 00:00:16s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:2,027,382\n",
      "INFO\t2025-06-04 11:45:41\tMarkDuplicates\tTracking 272 as yet unmatched pairs. 272 records in RAM.\n",
      "INFO\t2025-06-04 11:45:45\tMarkDuplicates\tRead     5,000,000 records.  Elapsed time: 00:00:20s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:2,589,297\n",
      "INFO\t2025-06-04 11:45:45\tMarkDuplicates\tTracking 266 as yet unmatched pairs. 266 records in RAM.\n",
      "INFO\t2025-06-04 11:45:48\tMarkDuplicates\tRead     6,000,000 records.  Elapsed time: 00:00:24s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:3,205,044\n",
      "INFO\t2025-06-04 11:45:48\tMarkDuplicates\tTracking 228 as yet unmatched pairs. 228 records in RAM.\n",
      "INFO\t2025-06-04 11:45:51\tMarkDuplicates\tRead     7,000,000 records.  Elapsed time: 00:00:27s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:3,835,047\n",
      "INFO\t2025-06-04 11:45:51\tMarkDuplicates\tTracking 196 as yet unmatched pairs. 196 records in RAM.\n",
      "INFO\t2025-06-04 11:45:54\tMarkDuplicates\tRead     8,000,000 records.  Elapsed time: 00:00:30s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:4,410,079\n",
      "INFO\t2025-06-04 11:45:54\tMarkDuplicates\tTracking 198 as yet unmatched pairs. 198 records in RAM.\n",
      "INFO\t2025-06-04 11:45:59\tMarkDuplicates\tRead     9,000,000 records.  Elapsed time: 00:00:35s.  Time for last 1,000,000:    4s.  Last read position: NC_008463.1:4,916,471\n",
      "INFO\t2025-06-04 11:45:59\tMarkDuplicates\tTracking 152 as yet unmatched pairs. 152 records in RAM.\n",
      "INFO\t2025-06-04 11:46:08\tMarkDuplicates\tRead    10,000,000 records.  Elapsed time: 00:00:44s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:5,392,762\n",
      "INFO\t2025-06-04 11:46:08\tMarkDuplicates\tTracking 224 as yet unmatched pairs. 224 records in RAM.\n",
      "INFO\t2025-06-04 11:46:12\tMarkDuplicates\tRead    11,000,000 records.  Elapsed time: 00:00:47s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:5,883,424\n",
      "INFO\t2025-06-04 11:46:12\tMarkDuplicates\tTracking 170 as yet unmatched pairs. 170 records in RAM.\n",
      "INFO\t2025-06-04 11:46:15\tMarkDuplicates\tRead    12,000,000 records.  Elapsed time: 00:00:50s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:6,348,978\n",
      "INFO\t2025-06-04 11:46:15\tMarkDuplicates\tTracking 132 as yet unmatched pairs. 132 records in RAM.\n",
      "INFO\t2025-06-04 11:46:16\tMarkDuplicates\tRead 12421314 records. 0 pairs never matched.\n",
      "INFO\t2025-06-04 11:46:19\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 1730031496; totalMemory: 2312515584; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:46:19\tMarkDuplicates\tWill retain up to 83562496 duplicate indices before spilling to disk.\n",
      "INFO\t2025-06-04 11:46:19\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2025-06-04 11:46:22\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2025-06-04 11:46:28\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2025-06-04 11:46:29\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1622062352; totalMemory: 2312515584; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:46:29\tMarkDuplicates\tMarking 1663515 records as duplicates.\n",
      "INFO\t2025-06-04 11:46:29\tMarkDuplicates\tFound 83754 optical duplicate clusters.\n",
      "INFO\t2025-06-04 11:46:29\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2025-06-04 11:48:05\tMarkDuplicates\tWritten    10,000,000 records.  Elapsed time: 00:01:35s.  Time for last 10,000,000:   95s.  Last read position: NC_008463.1:5,392,762\n",
      "INFO\t2025-06-04 11:48:29\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2025-06-04 11:48:29\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2025-06-04 11:48:29\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2025-06-04 11:48:29\tMarkDuplicates\tBefore output close freeMemory: 2290258456; totalMemory: 2312515584; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 11:48:29\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2025-06-04 11:48:29\tMarkDuplicates\tAfter output close freeMemory: 2076978312; totalMemory: 2098061312; maxMemory: 2673999872\n",
      "[Wed Jun 04 11:48:29 CEST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 3.10 minutes.\n",
      "Runtime.totalMemory()=2098061312\n",
      "INFO\t2025-06-04 11:48:32\tBuildBamIndex\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    BuildBamIndex -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bai\n",
      "**********\n",
      "\n",
      "\n",
      "11:48:32.337 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 11:48:32 CEST 2025] BuildBamIndex INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bai    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 11:48:32 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 11:48:52\tBuildBamIndex\tSuccessfully wrote bam index file /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bai\n",
      "[Wed Jun 04 11:48:52 CEST 2025] picard.sam.BuildBamIndex done. Elapsed time: 0.34 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO  11:48:57,528 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:48:57,535 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  11:48:57,536 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  11:48:57,537 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  11:48:57,541 HelpFormatter - Program Args: -T RealignerTargetCreator -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta -o /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.realigned.intervals \n",
      "INFO  11:48:57,606 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  11:48:57,607 HelpFormatter - Date/Time: 2025/06/04 11:48:57 \n",
      "INFO  11:48:57,608 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:48:57,609 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:48:58,644 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  11:48:58,842 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  11:48:58,874 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  11:48:58,964 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.08 \n",
      "INFO  11:48:59,251 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  11:48:59,387 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  11:48:59,388 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  11:48:59,388 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  11:48:59,395 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  11:49:29,405 ProgressMeter - NC_008463.1:810317    802816.0    30.0 s      37.0 s       12.4%     4.0 m       3.5 m \n",
      "INFO  11:49:59,409 ProgressMeter - NC_008463.1:2148105   2146304.0    60.0 s      27.0 s       32.9%     3.0 m       2.0 m \n",
      "INFO  11:50:29,411 ProgressMeter - NC_008463.1:3677417   3670016.0    90.0 s      24.0 s       56.2%     2.7 m      70.0 s \n",
      "INFO  11:50:59,412 ProgressMeter - NC_008463.1:5049573   5046272.0   120.0 s      23.0 s       77.2%     2.6 m      35.0 s \n",
      "INFO  11:51:29,414 ProgressMeter - NC_008463.1:6238021   6225920.0     2.5 m      24.0 s       95.4%     2.6 m       7.0 s \n",
      "INFO  11:51:38,891 ProgressMeter -            done   6537648.0     2.7 m      24.0 s      100.0%     2.7 m       0.0 s \n",
      "INFO  11:51:38,892 ProgressMeter - Total runtime 159.50 secs, 2.66 min, 0.04 hours \n",
      "INFO  11:51:38,894 MicroScheduler - 1738403 reads were filtered out during the traversal out of approximately 12526629 total reads (13.88%) \n",
      "INFO  11:51:38,897 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  11:51:38,897 MicroScheduler -   -> 0 reads (0.00% of total) failing BadMateFilter \n",
      "INFO  11:51:38,898 MicroScheduler -   -> 1675996 reads (13.38% of total) failing DuplicateReadFilter \n",
      "INFO  11:51:38,899 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  11:51:38,904 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  11:51:38,904 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  11:51:38,905 MicroScheduler -   -> 62407 reads (0.50% of total) failing MappingQualityZeroFilter \n",
      "INFO  11:51:38,907 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  11:51:38,907 MicroScheduler -   -> 0 reads (0.00% of total) failing Platform454Filter \n",
      "INFO  11:51:38,908 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "INFO  11:51:40,731 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "INFO  11:51:44,468 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:51:44,476 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  11:51:44,477 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  11:51:44,477 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  11:51:44,480 HelpFormatter - Program Args: -T IndelRealigner -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta --maxConsensuses 60 --maxReadsForConsensuses 240 --maxReadsForRealignment 6000 --targetIntervals /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.realigned.intervals -o /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.realigned.bam \n",
      "INFO  11:51:44,514 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  11:51:44,517 HelpFormatter - Date/Time: 2025/06/04 11:51:44 \n",
      "INFO  11:51:44,518 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:51:44,519 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  11:51:45,749 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  11:51:45,965 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  11:51:45,981 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  11:51:46,049 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.07 \n",
      "INFO  11:51:46,347 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  11:51:46,354 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  11:51:46,373 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  11:51:46,377 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  11:51:46,378 ProgressMeter -        Location |     reads | elapsed |     reads | completed | runtime |   runtime \n",
      "INFO  11:51:46,542 ReadShardBalancer$1 - Loading BAM index data \n",
      "INFO  11:51:46,548 ReadShardBalancer$1 - Done loading BAM index data \n",
      "INFO  11:52:16,799 ProgressMeter - NC_008463.1:235583    400022.0    30.0 s      76.0 s        3.6%    13.9 m      13.4 m \n",
      "INFO  11:52:46,801 ProgressMeter - NC_008463.1:576780   1100064.0    60.0 s      54.0 s        8.8%    11.3 m      10.3 m \n",
      "INFO  11:53:16,802 ProgressMeter - NC_008463.1:920352   1800092.0    90.0 s      50.0 s       14.1%    10.7 m       9.2 m \n",
      "INFO  11:53:47,507 ProgressMeter - NC_008463.1:1246365   2400120.0     2.0 m      50.0 s       19.1%    10.6 m       8.6 m \n",
      "INFO  11:54:17,509 ProgressMeter - NC_008463.1:1637525   3200152.0     2.5 m      47.0 s       25.0%    10.0 m       7.5 m \n",
      "INFO  11:54:47,815 ProgressMeter - NC_008463.1:2017811   3900174.0     3.0 m      46.0 s       30.9%     9.8 m       6.8 m \n",
      "INFO  11:55:18,187 ProgressMeter - NC_008463.1:2426326   4700203.0     3.5 m      45.0 s       37.1%     9.5 m       6.0 m \n",
      "INFO  11:55:48,906 ProgressMeter - NC_008463.1:2886061   5500237.0     4.0 m      44.0 s       44.1%     9.1 m       5.1 m \n",
      "INFO  11:56:19,223 ProgressMeter - NC_008463.1:3414283   6300277.0     4.5 m      43.0 s       52.2%     8.7 m       4.1 m \n",
      "INFO  11:56:49,226 ProgressMeter - NC_008463.1:3888687   7000301.0     5.0 m      43.0 s       59.5%     8.5 m       3.4 m \n",
      "INFO  11:57:20,829 ProgressMeter - NC_008463.1:4336490   7800332.0     5.6 m      42.0 s       66.3%     8.4 m       2.8 m \n",
      "INFO  11:57:52,865 ProgressMeter - NC_008463.1:4732258   8600386.0     6.1 m      42.0 s       72.4%     8.4 m       2.3 m \n",
      "INFO  11:58:23,945 ProgressMeter - NC_008463.1:5117585   9300436.0     6.6 m      42.0 s       78.3%     8.5 m     110.0 s \n",
      "INFO  11:58:57,227 ProgressMeter - NC_008463.1:5477215   1.0100493E7     7.2 m      42.0 s       83.8%     8.6 m      83.0 s \n",
      "INFO  11:59:27,278 ProgressMeter - NC_008463.1:5809684   1.0800557E7     7.7 m      42.0 s       88.9%     8.6 m      57.0 s \n",
      "INFO  11:59:58,598 ProgressMeter - NC_008463.1:6168356   1.1600602E7     8.2 m      42.0 s       94.4%     8.7 m      29.0 s \n",
      "INFO  12:00:28,599 ProgressMeter - NC_008463.1:6507290   1.230062E7     8.7 m      42.0 s       99.5%     8.7 m       2.0 s \n",
      "INFO  12:00:34,931 ProgressMeter -            done   1.2472276E7     8.8 m      42.0 s      100.0%     8.8 m       0.0 s \n",
      "INFO  12:00:34,931 ProgressMeter - Total runtime 528.56 secs, 8.81 min, 0.15 hours \n",
      "INFO  12:00:34,934 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 12472276 total reads (0.00%) \n",
      "INFO  12:00:34,934 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  12:00:34,934 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  12:00:36,829 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "[bam_sort_core] merging from 10 files...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PA14wt0253G processed!!\n",
      "\n",
      "Processing PAO1wt12P...\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[samopen] SAM header is present: 1 sequences.\n",
      "INFO\t2025-06-04 12:41:32\tSortSam\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    SortSam -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam -SORT_ORDER coordinate\n",
      "**********\n",
      "\n",
      "\n",
      "12:41:32.459 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 12:41:32 CEST 2025] SortSam INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam SORT_ORDER=coordinate    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 12:41:32 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 12:42:30\tSortSam\tFinished reading inputs, merging and writing to output now.\n",
      "[Wed Jun 04 12:43:20 CEST 2025] picard.sam.SortSam done. Elapsed time: 1.81 minutes.\n",
      "Runtime.totalMemory()=1190883328\n",
      "INFO\t2025-06-04 12:43:22\tAddOrReplaceReadGroups\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    AddOrReplaceReadGroups -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam -LB /media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS -PL Illumina -PU /media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS -SM /media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS\n",
      "**********\n",
      "\n",
      "\n",
      "12:43:22.929 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 12:43:22 CEST 2025] AddOrReplaceReadGroups INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam RGLB=/media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS RGPL=Illumina RGPU=/media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS RGSM=/media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS    RGID=1 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 12:43:23 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 12:43:23\tAddOrReplaceReadGroups\tCreated read-group ID=1 PL=Illumina LB=/media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS SM=/media/student/EXTERNAL_USB/Variant_Calling/PAO1wt12P_WGS\n",
      "\n",
      "INFO\t2025-06-04 12:43:36\tAddOrReplaceReadGroups\tProcessed     1,000,000 records.  Elapsed time: 00:00:12s.  Time for last 1,000,000:   12s.  Last read position: NC_002516.2:830,564\n",
      "INFO\t2025-06-04 12:43:46\tAddOrReplaceReadGroups\tProcessed     2,000,000 records.  Elapsed time: 00:00:23s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:1,870,533\n",
      "INFO\t2025-06-04 12:43:56\tAddOrReplaceReadGroups\tProcessed     3,000,000 records.  Elapsed time: 00:00:33s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:3,045,985\n",
      "INFO\t2025-06-04 12:44:07\tAddOrReplaceReadGroups\tProcessed     4,000,000 records.  Elapsed time: 00:00:43s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:4,077,540\n",
      "INFO\t2025-06-04 12:44:17\tAddOrReplaceReadGroups\tProcessed     5,000,000 records.  Elapsed time: 00:00:53s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:4,995,202\n",
      "INFO\t2025-06-04 12:44:27\tAddOrReplaceReadGroups\tProcessed     6,000,000 records.  Elapsed time: 00:01:03s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:5,859,008\n",
      "[Wed Jun 04 12:44:34 CEST 2025] picard.sam.AddOrReplaceReadGroups done. Elapsed time: 1.19 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO\t2025-06-04 12:44:36\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -METRICS_FILE /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_metrics.txt\n",
      "**********\n",
      "\n",
      "\n",
      "12:44:36.393 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 12:44:36 CEST 2025] MarkDuplicates INPUT=[/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam] OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam METRICS_FILE=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_DUP_STRATEGY=FLOW_QUALITY_SUM_STRATEGY FLOW_USE_END_IN_UNPAIRED_READS=false FLOW_USE_UNPAIRED_CLIPPED_END=false FLOW_UNPAIRED_END_UNCERTAINTY=0 FLOW_UNPAIRED_START_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true REMOVE_DUPLICATES=false ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 12:44:36 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 12:44:36\tMarkDuplicates\tStart of doWork freeMemory: 158622128; totalMemory: 168361984; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 12:44:36\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2025-06-04 12:44:36\tMarkDuplicates\tWill retain up to 9688405 data points before spilling to disk.\n",
      "INFO\t2025-06-04 12:44:43\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:06s.  Time for last 1,000,000:    6s.  Last read position: NC_002516.2:830,564\n",
      "INFO\t2025-06-04 12:44:43\tMarkDuplicates\tTracking 290 as yet unmatched pairs. 290 records in RAM.\n",
      "INFO\t2025-06-04 12:44:47\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:10s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:1,870,533\n",
      "INFO\t2025-06-04 12:44:47\tMarkDuplicates\tTracking 366 as yet unmatched pairs. 366 records in RAM.\n",
      "INFO\t2025-06-04 12:44:51\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:14s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:3,045,985\n",
      "INFO\t2025-06-04 12:44:51\tMarkDuplicates\tTracking 382 as yet unmatched pairs. 382 records in RAM.\n",
      "INFO\t2025-06-04 12:44:54\tMarkDuplicates\tRead     4,000,000 records.  Elapsed time: 00:00:17s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:4,077,540\n",
      "INFO\t2025-06-04 12:44:54\tMarkDuplicates\tTracking 400 as yet unmatched pairs. 400 records in RAM.\n",
      "INFO\t2025-06-04 12:44:58\tMarkDuplicates\tRead     5,000,000 records.  Elapsed time: 00:00:21s.  Time for last 1,000,000:    4s.  Last read position: NC_002516.2:4,995,202\n",
      "INFO\t2025-06-04 12:44:58\tMarkDuplicates\tTracking 320 as yet unmatched pairs. 320 records in RAM.\n",
      "INFO\t2025-06-04 12:45:02\tMarkDuplicates\tRead     6,000,000 records.  Elapsed time: 00:00:25s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:5,859,008\n",
      "INFO\t2025-06-04 12:45:02\tMarkDuplicates\tTracking 162 as yet unmatched pairs. 162 records in RAM.\n",
      "INFO\t2025-06-04 12:45:04\tMarkDuplicates\tRead 6524016 records. 0 pairs never matched.\n",
      "INFO\t2025-06-04 12:45:05\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 1102420040; totalMemory: 1898385408; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 12:45:05\tMarkDuplicates\tWill retain up to 83562496 duplicate indices before spilling to disk.\n",
      "INFO\t2025-06-04 12:45:06\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2025-06-04 12:45:10\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2025-06-04 12:45:11\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2025-06-04 12:45:12\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1981650712; totalMemory: 2673999872; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 12:45:12\tMarkDuplicates\tMarking 575241 records as duplicates.\n",
      "INFO\t2025-06-04 12:45:12\tMarkDuplicates\tFound 37777 optical duplicate clusters.\n",
      "INFO\t2025-06-04 12:45:12\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2025-06-04 12:46:23\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2025-06-04 12:46:23\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2025-06-04 12:46:23\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2025-06-04 12:46:23\tMarkDuplicates\tBefore output close freeMemory: 2649865144; totalMemory: 2673999872; maxMemory: 2673999872\n",
      "INFO\t2025-06-04 12:46:23\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2025-06-04 12:46:23\tMarkDuplicates\tAfter output close freeMemory: 2400720152; totalMemory: 2423472128; maxMemory: 2673999872\n",
      "[Wed Jun 04 12:46:23 CEST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 1.79 minutes.\n",
      "Runtime.totalMemory()=2423472128\n",
      "INFO\t2025-06-04 12:46:25\tBuildBamIndex\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    BuildBamIndex -INPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -OUTPUT /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bai\n",
      "**********\n",
      "\n",
      "\n",
      "12:46:25.905 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Wed Jun 04 12:46:25 CEST 2025] BuildBamIndex INPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam OUTPUT=/media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bai    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Wed Jun 04 12:46:26 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-06-04 12:46:39\tBuildBamIndex\tSuccessfully wrote bam index file /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bai\n",
      "[Wed Jun 04 12:46:39 CEST 2025] picard.sam.BuildBamIndex done. Elapsed time: 0.22 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO  12:46:43,033 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  12:46:43,042 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  12:46:43,042 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  12:46:43,042 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  12:46:43,047 HelpFormatter - Program Args: -T RealignerTargetCreator -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta -o /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.realigned.intervals \n",
      "INFO  12:46:43,076 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  12:46:43,077 HelpFormatter - Date/Time: 2025/06/04 12:46:43 \n",
      "INFO  12:46:43,077 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  12:46:43,078 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  12:46:44,182 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  12:46:44,687 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  12:46:44,698 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  12:46:44,740 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  12:46:44,907 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  12:46:45,048 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  12:46:45,050 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  12:46:45,050 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  12:46:45,051 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  12:47:15,061 ProgressMeter - NC_002516.2:1664385   1654784.0    30.0 s      18.0 s       26.6%   112.0 s      82.0 s \n",
      "INFO  12:47:45,062 ProgressMeter - NC_002516.2:4084917   4079616.0    60.0 s      14.0 s       65.2%    92.0 s      32.0 s \n",
      "INFO  12:48:15,064 ProgressMeter - NC_002516.2:6039313   6029312.0    90.0 s      14.0 s       96.4%    93.0 s       3.0 s \n",
      "INFO  12:48:20,489 ProgressMeter -            done   6264404.0    95.0 s      15.0 s      100.0%    95.0 s       0.0 s \n",
      "INFO  12:48:20,490 ProgressMeter - Total runtime 95.44 secs, 1.59 min, 0.03 hours \n",
      "INFO  12:48:20,495 MicroScheduler - 624941 reads were filtered out during the traversal out of approximately 6581326 total reads (9.50%) \n",
      "INFO  12:48:20,499 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  12:48:20,500 MicroScheduler -   -> 0 reads (0.00% of total) failing BadMateFilter \n",
      "INFO  12:48:20,501 MicroScheduler -   -> 579127 reads (8.80% of total) failing DuplicateReadFilter \n",
      "INFO  12:48:20,501 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  12:48:20,502 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  12:48:20,502 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  12:48:20,502 MicroScheduler -   -> 45814 reads (0.70% of total) failing MappingQualityZeroFilter \n",
      "INFO  12:48:20,502 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  12:48:20,503 MicroScheduler -   -> 0 reads (0.00% of total) failing Platform454Filter \n",
      "INFO  12:48:20,503 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "INFO  12:48:22,291 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "INFO  12:48:25,743 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  12:48:25,748 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  12:48:25,749 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  12:48:25,749 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  12:48:25,752 HelpFormatter - Program Args: -T IndelRealigner -I /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Variant_Calling/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta --maxConsensuses 60 --maxReadsForConsensuses 240 --maxReadsForRealignment 6000 --targetIntervals /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.realigned.intervals -o /media/student/EXTERNAL_USB/Variant_Calling/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.realigned.bam \n",
      "INFO  12:48:25,776 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  12:48:25,777 HelpFormatter - Date/Time: 2025/06/04 12:48:25 \n",
      "INFO  12:48:25,779 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  12:48:25,779 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  12:48:26,900 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  12:48:26,973 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  12:48:26,983 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  12:48:27,025 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  12:48:27,257 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  12:48:27,267 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  12:48:27,277 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  12:48:27,278 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  12:48:27,280 ProgressMeter -        Location |     reads | elapsed |     reads | completed | runtime |   runtime \n",
      "INFO  12:48:27,416 ReadShardBalancer$1 - Loading BAM index data \n",
      "INFO  12:48:27,422 ReadShardBalancer$1 - Done loading BAM index data \n",
      "INFO  12:48:57,440 ProgressMeter - NC_002516.2:406875    400018.0    30.0 s      75.0 s        6.5%     7.7 m       7.2 m \n",
      "INFO  12:49:30,712 ProgressMeter - NC_002516.2:1095758   1200045.0    63.0 s      52.0 s       17.5%     6.0 m       5.0 m \n",
      "INFO  12:50:01,377 ProgressMeter - NC_002516.2:1980176   2000061.0    94.0 s      47.0 s       31.6%     5.0 m       3.4 m \n",
      "INFO  12:50:31,866 ProgressMeter - NC_002516.2:2940180   2900070.0     2.1 m      42.0 s       46.9%     4.4 m       2.3 m \n",
      "INFO  12:51:01,867 ProgressMeter - NC_002516.2:3832229   3700098.0     2.6 m      41.0 s       61.2%     4.2 m      97.0 s \n",
      "INFO  12:51:31,912 ProgressMeter - NC_002516.2:4533976   4400115.0     3.1 m      41.0 s       72.4%     4.2 m      70.0 s \n",
      "INFO  12:52:02,235 ProgressMeter - NC_002516.2:5268021   5200139.0     3.6 m      41.0 s       84.1%     4.2 m      40.0 s \n",
      "INFO  12:52:33,038 ProgressMeter - NC_002516.2:5938505   6000161.0     4.1 m      40.0 s       94.8%     4.3 m      13.0 s \n",
      "INFO  12:52:48,192 ProgressMeter -            done   6563252.0     4.3 m      39.0 s      100.0%     4.3 m       0.0 s \n",
      "INFO  12:52:48,194 ProgressMeter - Total runtime 260.92 secs, 4.35 min, 0.07 hours \n",
      "INFO  12:52:48,199 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 6563252 total reads (0.00%) \n",
      "INFO  12:52:48,200 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  12:52:48,201 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  12:52:50,211 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "[bam_sort_core] merging from 5 files...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PAO1wt12P processed!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"realign\"][\"enabled\"]:\n",
    "    \n",
    "    sam_files = os.listdir(samfiles_directory_path)\n",
    "\n",
    "    for sam_file in sam_files:\n",
    "        sample = sam_file.split(\"_\")[0]\n",
    "        \n",
    "        print(f\"Processing {sample}...\")\n",
    "        print()\n",
    "    \n",
    "        sam_file_path = os.path.join(samfiles_directory_path, sam_file)\n",
    "\n",
    "        if sam_file.endswith(\".sam.gz\"):\n",
    "            bam_file = sam_file.removesuffix(\".sam.gz\") + \".bam\"\n",
    "            #bam_file = sam_file[:-7]+\"bam\"\n",
    "        else:\n",
    "            bam_file = sam_file.removesuffix(\".sam\") + \".bam\"\n",
    "            #bam_file = sam_file[:-4]+\"bam\"\n",
    "    \n",
    "        bam_file_path = os.path.join(intermediate_directory_path, bam_file)\n",
    "\n",
    "        # Calling samtools view\n",
    "        command = [\n",
    "            samtools, \"view\", \"-b\", \"-S\", sam_file_path\n",
    "        ]\n",
    "    \n",
    "        with open(bam_file_path, 'wb') as bam_out_file:\n",
    "            subprocess.run(command, stdout=bam_out_file, check=True)\n",
    "    \n",
    "        # Sort\n",
    "        sorted_bam_file_path = bam_file_path.removesuffix(\".bam\") + \"_sorted.bam\"\n",
    "        #sorted_bam_file_path = bam_file_path[:-4] + \"_sorted.bam\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"SortSam\",\n",
    "            \"INPUT=\" + bam_file_path, \"OUTPUT=\" + sorted_bam_file_path,\n",
    "            \"SORT_ORDER=coordinate\"\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # Add readgroup information\n",
    "        addrg_bam_file_path = sorted_bam_file_path.removesuffix(\".bam\") + \"_addrg.bam\"\n",
    "        #addrg_bam_file_path = bam_file_path[:-4] + \"_addrg.bam\"\n",
    "\n",
    "        read_group_library = f\"{main_directory_path}/{sample}_WGS\"\n",
    "        read_group_platform = \"Illumina\"\n",
    "        read_group_platform_unit = f\"{main_directory_path}/{sample}_WGS\"\n",
    "        read_group_sample_name = f\"{main_directory_path}/{sample}_WGS\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"AddOrReplaceReadGroups\",\n",
    "            \"INPUT=\" + sorted_bam_file_path,\n",
    "            \"OUTPUT=\" + addrg_bam_file_path,\n",
    "            \"LB=\" + read_group_library,\n",
    "            \"PL=\" + read_group_platform,\n",
    "            \"PU=\" + read_group_platform_unit,\n",
    "            \"SM=\" + read_group_sample_name\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # Mark duplicates\n",
    "        dedup_bam_file_path = addrg_bam_file_path.removesuffix(\".bam\") + \"_dedup.bam\"\n",
    "        #dedup_sorted_bam_file_path = sorted_bam_file_path[:-4] + \"_dedup.bam\"\n",
    "        metrics_file_path = bam_file_path.removesuffix(\".bam\") + \"_metrics.txt\"\n",
    "        #metrics_file_path = bam_file_path[:-4] + \"_metrics.txt\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"MarkDuplicates\",\n",
    "            \"INPUT=\" + addrg_bam_file_path, \"OUTPUT=\" + dedup_bam_file_path,\n",
    "            \"METRICS_FILE=\" + metrics_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        #Index .bam\n",
    "        bai_file_path = dedup_bam_file_path.removesuffix(\"bam\") + \"bai\"\n",
    "        #bai_file_path = bam_file_path[:-3] + \"bai\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"BuildBamIndex\",\n",
    "            \"INPUT=\" + dedup_bam_file_path,\n",
    "            \"OUTPUT=\" + bai_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        #Calling GATK\n",
    "        reference_genome = realign_sample_genome_dict[sample]\n",
    "    \n",
    "        reference_genome_fasta_file = f\"N_{reference_genome}.fasta\"\n",
    "        reference_genome_fasta_file_path = os.path.join(indexed_directory_path, reference_genome, reference_genome_fasta_file)\n",
    "\n",
    "        realigned_intervals_file_path = bam_file_path.removesuffix(\".bam\") + \".realigned.intervals\"\n",
    "        #realigned_intervals_file_path = bam_file_path[:-4] + \".realigned.intervals\"\n",
    "\n",
    "        command = [\n",
    "            java_8, \"-jar\", GATK_jar, \"-T\", \"RealignerTargetCreator\",\n",
    "            \"-I\", dedup_bam_file_path, \"-R\", reference_genome_fasta_file_path,\n",
    "            \"-o\", realigned_intervals_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        realigned_bam_file_path = bam_file_path.removesuffix(\".bam\") + \".realigned.bam\"\n",
    "        #realigned_bam_file_path = bam_file_path[:-4] + \".realigned.bam\"\n",
    "\n",
    "        command = [\n",
    "            java_8, \"-jar\", GATK_jar, \"-T\", \"IndelRealigner\",\n",
    "            \"-I\", dedup_bam_file_path, \"-R\", reference_genome_fasta_file_path,\n",
    "            \"--maxConsensuses\", \"60\", \"--maxReadsForConsensuses\", \"240\",\"--maxReadsForRealignment\", \"6000\",\n",
    "            \"--targetIntervals\", realigned_intervals_file_path,\n",
    "            \"-o\", realigned_bam_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # Calling samtools sort\n",
    "        realigned_sorted_file_path = realigned_bam_file_path.removesuffix(\".bam\") + \".sorted\"\n",
    "        #realigned_sorted_file_path = realigned_bam_file_path[:-4] + \".sorted\"\n",
    "\n",
    "        command = [\n",
    "            samtools, \"sort\", realigned_bam_file_path,\n",
    "            realigned_sorted_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # totalpileup files are used for indel calling\n",
    "        realigned_sorted_bam_file_path = realigned_sorted_file_path + \".bam\"\n",
    "        realigned_totalpileup_file = bam_file.removesuffix(\".bam\") + \".realigned.totalpileup\"\n",
    "        #realigned_totalpileup_file = bam_file[:-4] + \".realigned.totalpileup\"\n",
    "        realigned_totalpileup_file_path = os.path.join(totalpileup_directory_path, realigned_totalpileup_file)\n",
    "\n",
    "        with open(realigned_totalpileup_file_path, 'w') as totalpileup_file:\n",
    "            command = [\n",
    "                samtools, \"pileup\", \"-c\", \"-f\", reference_genome_fasta_file_path,\n",
    "                realigned_sorted_bam_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=totalpileup_file, check=True)\n",
    "    \n",
    "        # raw files are used for SNP calling\n",
    "        realigned_raw_file = bam_file.removesuffix(\".bam\") + \".realigned.raw\"\n",
    "        #realigned_raw_file = bam_file[:-] + \".realigned.raw\"\n",
    "        realigned_raw_file_path = os.path.join(raw_directory_path, realigned_raw_file)\n",
    "\n",
    "        with open(realigned_raw_file_path, 'w') as raw_file:\n",
    "            command = [\n",
    "                samtools, \"pileup\", \"-vc\", \"-f\", reference_genome_fasta_file_path,\n",
    "                realigned_sorted_bam_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=raw_file, check=True)\n",
    "\n",
    "        print()\n",
    "        print(f\"{sample} processed!!\")\n",
    "        print()\n",
    "\n",
    "    shutil.rmtree(intermediate_directory_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c2cbe06-d177-4ebe-a36a-042c7df7b61d",
   "metadata": {},
   "source": [
    "## Find indels without common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0df02208-de11-49d2-b3a9-3e7fa1da739a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if configuration[\"indels\"]:\n",
    "\n",
    "    totalpileup_files = os.listdir(totalpileup_directory_path)\n",
    "\n",
    "    for file in totalpileup_files:\n",
    "        totalpileup_file_path = os.path.join(totalpileup_directory_path, file)\n",
    "    \n",
    "        indels_withoutcommon_file = file.removesuffix(\".realigned.totalpileup\") + \".withoutcommon.indels\"\n",
    "        indels_withoutcommon_file_path = os.path.join(indels_directory_path, indels_withoutcommon_file)\n",
    "\n",
    "        with open(indels_withoutcommon_file_path, 'w') as indels_wc_file:\n",
    "            with open(totalpileup_file_path, 'r') as totalpileup_file:\n",
    "        \n",
    "                indels = []\n",
    "        \n",
    "                for line in totalpileup_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if fields[2] == \"*\":\n",
    "                        if int(fields[5]) >= 500 and int(fields[6]) >= 25:\n",
    "                            if (\n",
    "                                (fields[8] == \"*\" and int(fields[11])*5 > int(fields[7])) or\n",
    "                                (fields[9] == \"*\" and int(fields[10])*5 >= int(fields[7])) or\n",
    "                                (fields[8] != \"*\" and fields[9] != \"*\" and (int(fields[11])*5 >= int(fields[7]) or\n",
    "                                 int(fields[10])*5 >= int(fields[7])))\n",
    "                            ):\n",
    "                                indels.append(line)\n",
    "\n",
    "            indels_wc_file.writelines(indels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e382ac4c-d456-4a50-bf98-89142add422d",
   "metadata": {},
   "source": [
    "## Find SNPs without common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "894886c6-3d2d-4cd6-a00a-d61ada961e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if configuration[\"snps\"]:\n",
    "    raw_files = os.listdir(raw_directory_path)\n",
    "\n",
    "    for file in raw_files:\n",
    "        raw_file_path = os.path.join(raw_directory_path, file)\n",
    "    \n",
    "        snps_withoutcommon_file = file.removesuffix(\".realigned.raw\") + \".snps.withoutcommon\"\n",
    "        snps_withoutcommon_file_path = os.path.join(snps_directory_path, snps_withoutcommon_file)\n",
    "\n",
    "        with open(snps_withoutcommon_file_path, 'w') as snps_wc_file:\n",
    "            with open(raw_file_path, 'r') as raw_file:\n",
    "        \n",
    "                snps = []\n",
    "        \n",
    "                for line in raw_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if (int(fields[5]) >= 50 and int(fields[6]) >= 25 and int(fields[7]) >= 3):\n",
    "                        if fields[3] not in [\"M\",\"R\",\"W\",\"S\",\"Y\",\"K\"]:\n",
    "                            snps.append(line)\n",
    "\n",
    "            snps_wc_file.writelines(snps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf83f75-0771-4d1c-840d-5e9e3268bf13",
   "metadata": {},
   "source": [
    "## Convert .withoutcommon files to .vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8bff7d84-24f0-4da3-acfa-4ec8a88606ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NOTICE: the default --format argument is set as 'pileup'\n",
      "NOTICE: the default --snpqual argument for pileup format is set as 20\n",
      "NOTICE: Column 6-9 in output are heterozygosity status, SNP quality, total reads, reads with mutation\n",
      "NOTICE: Read 271 lines and wrote 275 different variants at 271 genomic positions (127 SNPs and 144 indels)\n",
      "NOTICE: Among 275 different variants at 271 positions, 71 are heterozygotes, 204 are homozygotes\n",
      "NOTICE: Among 127 SNPs, 80 are transitions, 47 are transversions\n",
      "NOTICE: the default --format argument is set as 'pileup'\n",
      "NOTICE: the default --snpqual argument for pileup format is set as 20\n",
      "NOTICE: Column 6-9 in output are heterozygosity status, SNP quality, total reads, reads with mutation\n",
      "NOTICE: Read 307 lines and wrote 309 different variants at 307 genomic positions (126 SNPs and 181 indels)\n",
      "NOTICE: Among 309 different variants at 307 positions, 101 are heterozygotes, 208 are homozygotes\n",
      "NOTICE: Among 126 SNPs, 78 are transitions, 48 are transversions\n",
      "NOTICE: the default --format argument is set as 'pileup'\n",
      "NOTICE: the default --snpqual argument for pileup format is set as 20\n",
      "NOTICE: Column 6-9 in output are heterozygosity status, SNP quality, total reads, reads with mutation\n",
      "NOTICE: Read 118 lines and wrote 119 different variants at 118 genomic positions (16 SNPs and 102 indels)\n",
      "NOTICE: Among 119 different variants at 118 positions, 55 are heterozygotes, 64 are homozygotes\n",
      "NOTICE: Among 16 SNPs, 6 are transitions, 10 are transversions\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"SNP_indel_annotation\"][\"enabled\"]:\n",
    "    snps_files = os.listdir(snps_directory_path)\n",
    "\n",
    "    for file in snps_files:\n",
    "        snps_file_path = os.path.join(snps_directory_path, file)\n",
    "        vcf_file = file.removesuffix(\".raw\") + \".vcf\"\n",
    "        vcf_file_path = os.path.join(vcf_directory_path, vcf_file)\n",
    "\n",
    "        with open(vcf_file_path, 'w') as vcf_out_file:\n",
    "            command = [\n",
    "                perl, convert2annovar, snps_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=vcf_out_file, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da8e7c2a-e5bc-4d20-8b70-9b5b2d6a2afc",
   "metadata": {},
   "source": [
    "## Build genome database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ad7a8965-fefd-41ed-a597-6d8afcb75321",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading genome NC_002516.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100 14.0M    0 14.0M    0     0   904k      0 --:--:--  0:00:15 --:--:-- 1834k\n",
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'build'\n",
      "00:00:00 Building database for 'NC_002516.2'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_002516.2'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:01 done\n",
      "00:00:04 Chromosome: 'NC_002516.2'\tlength: 6264404\n",
      "00:00:04 Create exons from CDS (if needed): \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:04 Exons created for 5572 transcripts.\n",
      "00:00:04 Deleting redundant exons (if needed): \n",
      "00:00:04 \tTotal transcripts with deleted exons: 0\n",
      "00:00:04 Collapsing zero length introns (if needed): \n",
      "00:00:04 \tTotal collapsed transcripts: 0\n",
      "00:00:04 \t\tAdding genomic sequences to genes: \n",
      "00:00:05 \tDone (4779 sequences added).\n",
      "00:00:05 \t\tAdding genomic sequences to exons: \n",
      "00:00:05 \tDone (5573 sequences added, 0 ignored).\n",
      "00:00:05 Finishing up genome\n",
      "00:00:05 Adjusting transcripts: \n",
      "00:00:05 Adjusting genes: \n",
      "00:00:05 Adjusting chromosomes lengths: \n",
      "00:00:05 Ranking exons: \n",
      "00:00:05 Create UTRs from CDS (if needed): \n",
      "00:00:05 Remove empty chromosomes: \n",
      "00:00:05 Marking as 'coding' from CDS information: \n",
      "00:00:05 Done: 0 transcripts marked\n",
      "00:00:05 \n",
      "00:00:05 Caracterizing exons by splicing (stage 1) : \n",
      ".....\n",
      "00:00:05 Caracterizing exons by splicing (stage 2) : \n",
      ".....00:00:05 done.\n",
      "00:00:05 [Optional] Rare amino acid annotations\n",
      "WARNING_FILE_NOT_FOUND: Rare Amino Acid analysis: Cannot read protein sequence file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/protein.fa', nothing done.\n",
      "00:00:05 CDS check: GenBank file format, skipping\n",
      "\n",
      "00:00:05 Protein check file: '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/genes.gbk'\n",
      "\n",
      "00:00:05 Checking database using protein sequences\n",
      "00:00:05 Comparing Proteins...\n",
      "\tLabels:\n",
      "\t\t'+' : OK\n",
      "\t\t'.' : Missing\n",
      "\t\t'*' : Error\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t\t\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\tProtein check:\tNC_002516.2\tOK: 5572\tNot found: 0\tErrors: 0\tError percentage: 0.0%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:06 \n",
      "\n",
      "00:00:06 Saving database\n",
      "00:00:08 Saving sequences for chromosome 'NC_002516.2' to file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/sequence.NC_002516.2.bin'\n",
      "00:00:10 [Optional] Reading regulation elements: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot read optional regulation file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/regulation.gff', nothing done.\n",
      "00:00:10 [Optional] Reading regulation elements: BED \n",
      "00:00:10 Cannot find optional regulation dir '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/regulation.bed/', nothing done.\n",
      "00:00:10 [Optional] Reading motifs: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot open PWMs file /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/pwms.bin. Nothing done\n",
      "00:00:10 Done\n",
      "00:00:10 Logging\n",
      "00:00:11 Checking for updates...\n",
      "00:00:13 Done.\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading genome NC_008463.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100 15.7M    0 15.7M    0     0   785k      0 --:--:--  0:00:20 --:--:-- 1222k\n",
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'build'\n",
      "00:00:00 Building database for 'NC_008463.1'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_008463.1'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:01 done\n",
      "00:00:04 Chromosome: 'NC_008463.1'\tlength: 6537648\n",
      "00:00:04 Create exons from CDS (if needed): \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:04 Exons created for 5961 transcripts.\n",
      "00:00:04 Deleting redundant exons (if needed): \n",
      "00:00:04 \tTotal transcripts with deleted exons: 0\n",
      "00:00:04 Collapsing zero length introns (if needed): \n",
      "00:00:04 \tTotal collapsed transcripts: 0\n",
      "00:00:04 \t\tAdding genomic sequences to genes: \n",
      "00:00:05 \tDone (4989 sequences added).\n",
      "00:00:05 \t\tAdding genomic sequences to exons: \n",
      "00:00:05 \tDone (5966 sequences added, 0 ignored).\n",
      "00:00:05 Finishing up genome\n",
      "00:00:05 Adjusting transcripts: \n",
      "00:00:05 Adjusting genes: \n",
      "00:00:05 Adjusting chromosomes lengths: \n",
      "00:00:05 Ranking exons: \n",
      "00:00:05 Create UTRs from CDS (if needed): \n",
      "00:00:05 Remove empty chromosomes: \n",
      "00:00:05 Marking as 'coding' from CDS information: \n",
      "00:00:05 Done: 70 transcripts marked\n",
      "00:00:05 \n",
      "00:00:05 Caracterizing exons by splicing (stage 1) : \n",
      ".....\n",
      "00:00:05 Caracterizing exons by splicing (stage 2) : \n",
      ".....00:00:05 done.\n",
      "00:00:05 [Optional] Rare amino acid annotations\n",
      "WARNING_FILE_NOT_FOUND: Rare Amino Acid analysis: Cannot read protein sequence file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/protein.fa', nothing done.\n",
      "00:00:05 CDS check: GenBank file format, skipping\n",
      "\n",
      "00:00:05 Protein check file: '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/genes.gbk'\n",
      "\n",
      "00:00:05 Checking database using protein sequences\n",
      "00:00:05 Comparing Proteins...\n",
      "\tLabels:\n",
      "\t\t'+' : OK\n",
      "\t\t'.' : Missing\n",
      "\t\t'*' : Error\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t\t\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++\n",
      "\t+.++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++.++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++.++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++.+++++++++++++++++++.+++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++.++\n",
      "\t+++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++.+++++++++++++++++++++++++.+++++++++++.+++++++++++++++++++++++++++++++++++++.++++++++++++++\n",
      "\t++++++++++++.++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++\n",
      "\t.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++.+++++++++++\n",
      "\t+++++++++++.+++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++\n",
      "\t++++++++++++++++++++++.++++++++++++++++++++++++.+++++++.++++++++++++++++.++++++++++++++++++++.++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+..\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++\n",
      "\t+++.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++.++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++.+++++++++++++..++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++.++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++.+++++++++++++++++++++++++++++++++++++++++.+++++++++++.+++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++.+++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++.+++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++.+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++\tProtein check:\tNC_008463.1\tOK: 5891\tNot found: 70\tErrors: 0\tError percentage: 0.0%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:07 \n",
      "\n",
      "00:00:07 Saving database\n",
      "00:00:09 Saving sequences for chromosome 'NC_008463.1' to file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/sequence.NC_008463.1.bin'\n",
      "00:00:10 [Optional] Reading regulation elements: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot read optional regulation file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/regulation.gff', nothing done.\n",
      "00:00:10 [Optional] Reading regulation elements: BED \n",
      "00:00:10 Cannot find optional regulation dir '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/regulation.bed/', nothing done.\n",
      "00:00:10 [Optional] Reading motifs: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot open PWMs file /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/pwms.bin. Nothing done\n",
      "00:00:10 Done\n",
      "00:00:10 Logging\n",
      "00:00:11 Checking for updates...\n",
      "00:00:13 Done.\n"
     ]
    }
   ],
   "source": [
    "# https://pcingola.github.io/SnpEff/snpeff/build_db/#step-2-option-2-building-a-database-from-genbank-files\n",
    "# https://pcingola.github.io/SnpEff/snpeff/faq/#how-to-building-an-ncbi-genome-genbank-file\n",
    "if configuration[\"build\"][\"enabled\"]:\n",
    "    for database in configuration[\"build\"][\"databases\"]:\n",
    "        command = [\n",
    "            buildDbNcbi, database\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, cwd = snpeff_dir_path, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b34c9070-48e8-47f6-b50a-a4e2bccd2fbb",
   "metadata": {},
   "source": [
    "## SNPs and indels annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "173d1fa0-cd0a-4596-bb01-b4ddb59fd025",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'ann'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_008463.1'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:01 done\n",
      "00:00:01 Reading database for genome version 'NC_008463.1' from file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/snpEffectPredictor.bin' (this might take a while)\n",
      "00:00:03 done\n",
      "00:00:03 Loading Motifs and PWMs\n",
      "00:00:03 Building interval forest\n",
      "00:00:03 done.\n",
      "00:00:03 Genome stats :\n",
      "#-----------------------------------------------\n",
      "# Genome name                : 'NC_008463.1'\n",
      "# Genome version             : 'NC_008463.1'\n",
      "# Genome ID                  : 'NC_008463.1[0]'\n",
      "# Has protein coding info    : true\n",
      "# Has Tr. Support Level info : true\n",
      "# Genes                      : 6041\n",
      "# Protein coding genes       : 5961\n",
      "#-----------------------------------------------\n",
      "# Transcripts                : 5961\n",
      "# Avg. transcripts per gene  : 0.99\n",
      "# TSL transcripts            : 0\n",
      "#-----------------------------------------------\n",
      "# Checked transcripts        : \n",
      "#               AA sequences :   5891 ( 98.83% )\n",
      "#              DNA sequences :      0 ( 0.00% )\n",
      "#-----------------------------------------------\n",
      "# Protein coding transcripts : 5961\n",
      "#              Length errors :     28 ( 0.47% )\n",
      "#  STOP codons in CDS errors :     16 ( 0.27% )\n",
      "#         START codon errors :    531 ( 8.91% )\n",
      "#        STOP codon warnings :     17 ( 0.29% )\n",
      "#              UTR sequences :      0 ( 0.00% )\n",
      "#               Total Errors :    550 ( 9.23% )\n",
      "# WARNING                    : No protein coding transcript has UTR\n",
      "#-----------------------------------------------\n",
      "# Cds                        : 5966\n",
      "# Exons                      : 5966\n",
      "# Exons with sequence        : 5966\n",
      "# Exons without sequence     : 0\n",
      "# Avg. exons per transcript  : 1.00\n",
      "#-----------------------------------------------\n",
      "# Number of chromosomes      : 1\n",
      "# Chromosomes                : Format 'chromo_name size codon_table'\n",
      "#\t\t'NC_008463.1'\t6537648\tStandard\n",
      "#-----------------------------------------------\n",
      "\n",
      "00:00:06 Predicting variants\n",
      "\n",
      "WARNINGS: Some warning were detected\n",
      "Warning type\tNumber of warnings\n",
      "WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS\t14\n",
      "WARNING_TRANSCRIPT_NO_START_CODON\t93\n",
      "WARNING_TRANSCRIPT_NO_STOP_CODON\t1\n",
      "\n",
      "\n",
      "00:00:06 Creating summary file: snpEff_summary.html\n",
      "00:00:08 Creating genes file: snpEff_genes.txt\n",
      "00:00:08 done.\n",
      "00:00:08 Logging\n",
      "00:00:09 Checking for updates...\n",
      "00:00:11 Done.\n",
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'ann'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_008463.1'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:02 done\n",
      "00:00:02 Reading database for genome version 'NC_008463.1' from file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/snpEffectPredictor.bin' (this might take a while)\n",
      "00:00:03 done\n",
      "00:00:03 Loading Motifs and PWMs\n",
      "00:00:03 Building interval forest\n",
      "00:00:04 done.\n",
      "00:00:04 Genome stats :\n",
      "#-----------------------------------------------\n",
      "# Genome name                : 'NC_008463.1'\n",
      "# Genome version             : 'NC_008463.1'\n",
      "# Genome ID                  : 'NC_008463.1[0]'\n",
      "# Has protein coding info    : true\n",
      "# Has Tr. Support Level info : true\n",
      "# Genes                      : 6041\n",
      "# Protein coding genes       : 5961\n",
      "#-----------------------------------------------\n",
      "# Transcripts                : 5961\n",
      "# Avg. transcripts per gene  : 0.99\n",
      "# TSL transcripts            : 0\n",
      "#-----------------------------------------------\n",
      "# Checked transcripts        : \n",
      "#               AA sequences :   5891 ( 98.83% )\n",
      "#              DNA sequences :      0 ( 0.00% )\n",
      "#-----------------------------------------------\n",
      "# Protein coding transcripts : 5961\n",
      "#              Length errors :     28 ( 0.47% )\n",
      "#  STOP codons in CDS errors :     16 ( 0.27% )\n",
      "#         START codon errors :    531 ( 8.91% )\n",
      "#        STOP codon warnings :     17 ( 0.29% )\n",
      "#              UTR sequences :      0 ( 0.00% )\n",
      "#               Total Errors :    550 ( 9.23% )\n",
      "# WARNING                    : No protein coding transcript has UTR\n",
      "#-----------------------------------------------\n",
      "# Cds                        : 5966\n",
      "# Exons                      : 5966\n",
      "# Exons with sequence        : 5966\n",
      "# Exons without sequence     : 0\n",
      "# Avg. exons per transcript  : 1.00\n",
      "#-----------------------------------------------\n",
      "# Number of chromosomes      : 1\n",
      "# Chromosomes                : Format 'chromo_name size codon_table'\n",
      "#\t\t'NC_008463.1'\t6537648\tStandard\n",
      "#-----------------------------------------------\n",
      "\n",
      "00:00:06 Predicting variants\n",
      "\n",
      "WARNINGS: Some warning were detected\n",
      "Warning type\tNumber of warnings\n",
      "WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS\t15\n",
      "WARNING_TRANSCRIPT_NO_START_CODON\t93\n",
      "WARNING_TRANSCRIPT_NO_STOP_CODON\t1\n",
      "\n",
      "\n",
      "00:00:07 Creating summary file: snpEff_summary.html\n",
      "00:00:08 Creating genes file: snpEff_genes.txt\n",
      "00:00:08 done.\n",
      "00:00:08 Logging\n",
      "00:00:09 Checking for updates...\n",
      "00:00:11 Done.\n",
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'ann'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_002516.2'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:01 done\n",
      "00:00:01 Reading database for genome version 'NC_002516.2' from file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/snpEffectPredictor.bin' (this might take a while)\n",
      "00:00:03 done\n",
      "00:00:03 Loading Motifs and PWMs\n",
      "00:00:03 Building interval forest\n",
      "00:00:03 done.\n",
      "00:00:03 Genome stats :\n",
      "#-----------------------------------------------\n",
      "# Genome name                : 'NC_002516.2'\n",
      "# Genome version             : 'NC_002516.2'\n",
      "# Genome ID                  : 'NC_002516.2[0]'\n",
      "# Has protein coding info    : true\n",
      "# Has Tr. Support Level info : true\n",
      "# Genes                      : 5697\n",
      "# Protein coding genes       : 5572\n",
      "#-----------------------------------------------\n",
      "# Transcripts                : 5572\n",
      "# Avg. transcripts per gene  : 0.98\n",
      "# TSL transcripts            : 0\n",
      "#-----------------------------------------------\n",
      "# Checked transcripts        : \n",
      "#               AA sequences :   5572 ( 100.00% )\n",
      "#              DNA sequences :      0 ( 0.00% )\n",
      "#-----------------------------------------------\n",
      "# Protein coding transcripts : 5572\n",
      "#              Length errors :      0 ( 0.00% )\n",
      "#  STOP codons in CDS errors :      0 ( 0.00% )\n",
      "#         START codon errors :    524 ( 9.40% )\n",
      "#        STOP codon warnings :      0 ( 0.00% )\n",
      "#              UTR sequences :      0 ( 0.00% )\n",
      "#               Total Errors :    524 ( 9.40% )\n",
      "# WARNING                    : No protein coding transcript has UTR\n",
      "#-----------------------------------------------\n",
      "# Cds                        : 5573\n",
      "# Exons                      : 5573\n",
      "# Exons with sequence        : 5573\n",
      "# Exons without sequence     : 0\n",
      "# Avg. exons per transcript  : 1.00\n",
      "#-----------------------------------------------\n",
      "# Number of chromosomes      : 1\n",
      "# Chromosomes                : Format 'chromo_name size codon_table'\n",
      "#\t\t'NC_002516.2'\t6264404\tStandard\n",
      "#-----------------------------------------------\n",
      "\n",
      "00:00:04 Predicting variants\n",
      "\n",
      "WARNINGS: Some warning were detected\n",
      "Warning type\tNumber of warnings\n",
      "WARNING_TRANSCRIPT_NO_START_CODON\t16\n",
      "\n",
      "\n",
      "00:00:05 Creating summary file: snpEff_summary.html\n",
      "00:00:06 Creating genes file: snpEff_genes.txt\n",
      "00:00:06 done.\n",
      "00:00:06 Logging\n",
      "00:00:07 Checking for updates...\n",
      "00:00:09 Done.\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"SNP_indel_annotation\"][\"enabled\"]:\n",
    "    vcf_files = [file for file in os.listdir(vcf_directory_path) if file.endswith(\".snps.withoutcommon.vcf\")]\n",
    "\n",
    "    for file in vcf_files:\n",
    "        vcf_file_path = os.path.join(vcf_directory_path, file)\n",
    "    \n",
    "        vcf4snpEff_file_path = vcf_file_path + \"4snpEff\"\n",
    "\n",
    "        with open(vcf4snpEff_file_path, 'w') as vcf4snpEff_file:\n",
    "            with open(vcf_file_path, 'r') as vcf_file:\n",
    "                lines = []\n",
    "\n",
    "                for line in vcf_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if fields[3] != \"-\" and fields[4] != \"-\":\n",
    "                        lines.append(line)\n",
    "\n",
    "            vcf4snpEff_file.writelines(lines)\n",
    "\n",
    "    vcf4snpEff_files = [file for file in os.listdir(vcf_directory_path) if file.endswith(\"vcf4snpEff\")]\n",
    "\n",
    "    for file in vcf4snpEff_files:\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        genome = annotation_sample_genome_dict[sample]\n",
    "    \n",
    "        snps_wc_annotated_file = file.removesuffix(\".vcf4snpEff\") + \".annotated\"\n",
    "        snps_wc_annotated_file_path = os.path.join(annotated_directory_path, snps_wc_annotated_file)\n",
    "\n",
    "        vcf4snpEff_file_path = os.path.join(vcf_directory_path, file)\n",
    "\n",
    "        with open(snps_wc_annotated_file_path, 'w') as snps_annotated_file:\n",
    "    \n",
    "            command = [\n",
    "                java_21, \"-jar\", snpEff_jar,\n",
    "                \"-v\", genome, vcf4snpEff_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=snps_annotated_file, cwd=snpeff_dir_path, check=True)\n",
    "        snps_annotated_file.close()\n",
    "    \n",
    "        temp_directory_path = os.path.join(variant_calling_directory_path, \"temp\")\n",
    "    \n",
    "        if not os.path.exists(temp_directory_path):\n",
    "            os.mkdir(temp_directory_path)\n",
    "\n",
    "        temp1_file_path = os.path.join(temp_directory_path, \"temp1\")\n",
    "    \n",
    "        with open(temp1_file_path, 'w') as temp1_file:\n",
    "            with open(snps_wc_annotated_file_path, 'r') as snps_annotated_file:\n",
    "                command = [\n",
    "                    perl, vcfEffOnePerLine\n",
    "                ]\n",
    "\n",
    "                subprocess.run(command, stdin=snps_annotated_file, stdout=temp1_file, check=True)\n",
    "\n",
    "        temp2_file_path = os.path.join(temp_directory_path, \"temp2\")\n",
    "    \n",
    "        with open(temp2_file_path, 'w') as temp2_file:\n",
    "\n",
    "            command = [\n",
    "                java_21, \"-jar\", snpSift_jar, \"extractFields\",\n",
    "                temp1_file_path, \"POS\", \"REF\", \"ALT\", \"ANN[*].EFFECT\",\n",
    "                \"ANN[*].IMPACT\", \"ANN[*].GENE\", \"ANN[*].GENEID\",\n",
    "                \"ANN[*].HGVS_C\", \"ANN[*].HGVS_P\", \"ANN[*].CDNA_POS\",\n",
    "                \"ANN[*].CDNA_LEN\", \"ANN[*].AA_POS\", \"ANN[*].AA_LEN\"\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=temp2_file, check=True)\n",
    "        temp2_file.close()\n",
    "\n",
    "        semicurated_file = file.removesuffix(\".vcf4snpEff\") + \".semicurated\"\n",
    "        semicurated_file_path = os.path.join(curated_directory_path, semicurated_file)\n",
    "    \n",
    "        with open(semicurated_file_path, 'w') as semicurated_file:\n",
    "            with open(temp2_file_path, 'r') as temp2_file:\n",
    "\n",
    "                content = temp2_file.readlines()\n",
    "            semicurated_file.writelines(content)\n",
    "        \n",
    "        curated_file = file.removesuffix(\".vcf4snpEff\") + \".curated\"\n",
    "        curated_file_path = os.path.join(curated_directory_path, curated_file)\n",
    "    \n",
    "        with open(curated_file_path, 'w') as curated_file:\n",
    "            with open(semicurated_file_path, 'r') as semicurated_file:\n",
    "            \n",
    "                curated_lines = []\n",
    "            \n",
    "                for line in semicurated_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if fields[4] == \"MODERATE\" or fields[4] == \"HIGH\":\n",
    "                        curated_lines.append(line)\n",
    "\n",
    "            curated_file.writelines(curated_lines)\n",
    "\n",
    "        shutil.rmtree(temp_directory_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab2d510d-a585-421b-9622-deaf4efd8cff",
   "metadata": {},
   "source": [
    "## Filter by parental"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7173032-cb65-4d49-ae30-4c6e62c2fe3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if configuration[\"filter_by_parental\"][\"enabled\"]:\n",
    "    \n",
    "    # Filter SNPs\n",
    "    curated_files = [file for file in os.listdir(curated_directory_path) if file.endswith(\".curated\")]\n",
    "    \n",
    "    for file in curated_files:    \n",
    "        sample = file.split(\"_\")[0]\n",
    "        \n",
    "        curated_file_path = os.path.join(curated_directory_path, file)\n",
    "        filtered_file_path = os.path.join(curated_directory_path, file + \".filtered\")\n",
    "            \n",
    "        with open(curated_file_path, 'r') as curated_file:\n",
    "            snp_lines = curated_file.readlines()\n",
    "\n",
    "        processed_lines = []\n",
    "        for line in snp_lines:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            if len(fields) >= 9:\n",
    "                processed_lines.append(\"\\t\".join([fields[0], fields[5], fields[8]]))\n",
    "\n",
    "        processed_lines.sort()\n",
    "\n",
    "        for parental in configuration[\"filter_by_parental\"][\"Parental\"]:\n",
    "            if sample in parental[\"samples\"]:\n",
    "                snp_path = parental[\"snps_path\"]\n",
    "\n",
    "        with open(snp_path, 'r') as parental_file:\n",
    "            parental_lines = []\n",
    "            for line in parental_file:\n",
    "                fields = line.strip().split(\"\\t\")\n",
    "                if len(fields) >= 9:\n",
    "                    processed_line = \"\\t\".join([fields[0], fields[5], fields[8]])\n",
    "                    parental_lines.append(processed_line)\n",
    "\n",
    "        filtered_lines = []\n",
    "        for line in snp_lines:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            compare = \"\\t\".join([fields[0], fields[5], fields[8]])\n",
    "            if compare not in parental_lines:\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "        with open(filtered_file_path, 'w') as filtered_file:\n",
    "            for line in filtered_lines:\n",
    "                filtered_file.write(line + \"\\n\")\n",
    "\n",
    "    # Filter indels\n",
    "    indels_files = [file for file in os.listdir(indels_directory_path) if file.endswith(\".indels\")]\n",
    "        \n",
    "    for file in indels_files:\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "\n",
    "        indels_file_path = os.path.join(indels_directory_path, file)\n",
    "        filtered_file_path = os.path.join(indels_directory_path, file + \".filtered\")\n",
    "        \n",
    "        with open(indels_file_path, 'r') as indels_file:\n",
    "            sample_indels = indels_file.readlines()\n",
    "\n",
    "        for parental in configuration[\"filter_by_parental\"][\"Parental\"]:\n",
    "            if sample in parental[\"samples\"]:\n",
    "                ind_file_path = parental[\"indels_path\"]\n",
    "\n",
    "        with open(ind_file_path, 'r') as parental_file:\n",
    "            parental_lines = parental_file.readlines()\n",
    "\n",
    "        processed_lines = []\n",
    "        for line in parental_lines:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            processed_lines.append(\"\\t\".join(fields[1:4]))\n",
    "\n",
    "        filtered_lines = []\n",
    "\n",
    "        for line in sample_indels:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            compare = \"\\t\".join(fields[1:4])\n",
    "            if compare not in processed_lines:\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "        with open(filtered_file_path, 'w') as filtered_file:\n",
    "            for line in filtered_lines:\n",
    "                filtered_file.write(line + \"\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b786d2a-0e7a-40c2-92f6-c61e2924d8e0",
   "metadata": {},
   "source": [
    "## Create mutations table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "52207479-a10a-45f1-a495-90b0a9fabeb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sample PA14wt0253G does not contain any SNPs.\n",
      "The sample PAO1wt12P does not contain any SNPs.\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"table\"][\"enabled\"]:\n",
    "    \n",
    "    filtered_files = [file for file in os.listdir(curated_directory_path) if file.endswith(\".filtered\")]\n",
    "    if len(filtered_files) > 0:\n",
    "        snp_files = filtered_files\n",
    "    else:\n",
    "        snp_files = [file for file in os.listdir(curated_directory_path) if file.endswith(\".curated\")]\n",
    "    \n",
    "    # Add sample column\n",
    "    for file in snp_files:\n",
    "        snps_file_path = os.path.join(curated_directory_path, file)\n",
    "        try:\n",
    "            df = pd.read_csv(snps_file_path, sep=\"\\t\", header=0)\n",
    "            df[file] = file\n",
    "            df.to_csv(snps_file_path, index=False, sep=\"\\t\")\n",
    "        except pd.errors.EmptyDataError:\n",
    "            sample = file.strip().split(\"_\")[0]\n",
    "            print(f\"The sample {sample} does not contain any SNPs.\")\n",
    "\n",
    "    filtered_files = [file for file in os.listdir(indels_directory_path) if file.endswith(\".filtered\")]\n",
    "    if len(filtered_files) > 0:\n",
    "        ind_files = filtered_files\n",
    "    else:\n",
    "        ind_files = [file for file in os.listdir(indels_directory_path) if file.endswith(\".indels\")]\n",
    "    \n",
    "    for file in ind_files:\n",
    "        ind_file_path = os.path.join(indels_directory_path, file)\n",
    "        try:\n",
    "            df = pd.read_csv(ind_file_path, sep=\"\\t\", header=0)\n",
    "            df[file] = file\n",
    "            df.to_csv(ind_file_path, index=False, sep=\"\\t\")\n",
    "        except EmptyDataError:\n",
    "            sample = file.strip().split(\"_\")[0]\n",
    "            print(f\"The sample {sample} does not contain any indel.\") \n",
    "\n",
    "    # SNP identification\n",
    "    information_directory_path = os.path.join(curated_directory_path, \"information\")\n",
    "\n",
    "    if not os.path.exists(information_directory_path):\n",
    "        os.mkdir(information_directory_path)\n",
    "\n",
    "    ## Obtain old_locus_tag for Pseudomonas aeruginosa UCBPP-PA14 ##\n",
    "    # ------------------------------------------------------------ #\n",
    "    \n",
    "    gff3 = os.path.join(annotated_genomes_directory_path, \"PA14.gff3\")\n",
    "    locus_file_path = os.path.join(information_directory_path, \"PA14_locus_tags.txt\")\n",
    "\n",
    "    PA14_locus_dict = {}\n",
    "\n",
    "    with open(gff3, 'r') as gff3_file:\n",
    "        for line in gff3_file:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "        \n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            if len(fields) < 9:\n",
    "                continue\n",
    "            \n",
    "            if fields[1] == \"RefSeq\" and fields[2] == \"gene\":\n",
    "                attributes = fields[8]\n",
    "                attributes_dict = {}\n",
    "\n",
    "                for attribute in attributes.split(\";\"):\n",
    "                    if \"=\" in attribute:\n",
    "                        key, value = attribute.split(\"=\", 1)\n",
    "                        attributes_dict[key] = value\n",
    "                    if \"locus_tag\" in attributes_dict:\n",
    "                        locus = attributes_dict[\"locus_tag\"]\n",
    "                        old_locus = attributes_dict.get(\"old_locus_tag\", locus)\n",
    "                        PA14_locus_dict[locus] = old_locus\n",
    "\n",
    "    with open(locus_file_path, 'w') as locus_file:\n",
    "        for locus_tag, old_locus_tag in PA14_locus_dict.items():\n",
    "            locus_file.write(f\"{locus_tag}\\t{old_locus_tag}\\n\")\n",
    "\n",
    "    # -------------------------------------------------------------------------- #\n",
    "    ##############################################################################\n",
    "\n",
    "    ## Filter SNPs\n",
    "    snps_dict = {}\n",
    "\n",
    "    for file in snp_files:\n",
    "        file_path = os.path.join(curated_directory_path, file)\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "\n",
    "        with open(file_path, 'r') as snp_file:\n",
    "            snp_file_reader = csv.reader(snp_file, delimiter=\"\\t\")\n",
    "\n",
    "            for row in snp_file_reader:\n",
    "                gene = row[5]\n",
    "                locus = row[6]\n",
    "                aa = row[8]\n",
    "\n",
    "                #########################################\n",
    "                # Modified for PA14 strain locus change #\n",
    "                # ------------------------------------- #\n",
    "                # snp = \",\".join([locus, gene, aa])     #\n",
    "                #########################################\n",
    "                \n",
    "                if table_sample_strain_dict[sample] == \"PA14\":\n",
    "                    old_locus = PA14_locus_dict[locus]\n",
    "                    snp = \",\".join([old_locus, gene, aa])\n",
    "                else:\n",
    "                    snp = \",\".join([locus, gene, aa])\n",
    "\n",
    "                if snp not in snps_dict:\n",
    "                    snps_dict[snp] = 1\n",
    "                else:\n",
    "                    snps_dict[snp] += 1\n",
    "\n",
    "    snps_info_file_path = os.path.join(information_directory_path, \"snps_info.txt\")\n",
    "    \n",
    "    with open(snps_info_file_path, 'w') as snps_info_file:\n",
    "\n",
    "        for snp, count in snps_dict.items():\n",
    "            snps_info_file.write(f\"{count},{snp}\\n\")\n",
    "\n",
    "    filtered_snps_file_path = os.path.join(information_directory_path, \"filtered_snps.txt\")\n",
    "    with open(filtered_snps_file_path, 'w') as filtered_snps_file:\n",
    "        \n",
    "        for snp, count in snps_dict.items():\n",
    "            if count <= configuration[\"table\"][\"threshold\"]:\n",
    "                filtered_snps_file.write(f\"{snp}\\n\")\n",
    "    \n",
    "    ## SNP tabulation\n",
    "    snp_tab_file_path = os.path.join(information_directory_path, \"snp_tabulation.txt\")\n",
    "    with open(snp_tab_file_path, 'w') as snp_tab_file:\n",
    "\n",
    "        for file in snp_files:\n",
    "            file_path = os.path.join(curated_directory_path, file)\n",
    "\n",
    "            with open(file_path, 'r') as snp_file:\n",
    "                snp_file_reader = csv.reader(snp_file, delimiter=\"\\t\")\n",
    "\n",
    "                for row in snp_file_reader:\n",
    "                    gene = row[5]\n",
    "                    locus = row[6]\n",
    "                    aa = row[8]\n",
    "                    sample_name = row[13].split(\"_\")[0]\n",
    "\n",
    "                    #########################################\n",
    "                    # Modified for PA14 strain locus change #\n",
    "                    # ------------------------------------- #\n",
    "                    # snp = \",\".join([locus, gene, aa])     #\n",
    "                    #########################################\n",
    "                    \n",
    "                    if table_sample_strain_dict[sample_name] == \"PA14\":\n",
    "                        old_locus = PA14_locus_dict[locus]\n",
    "                        snp = \",\".join([old_locus, gene, aa])\n",
    "                    else:\n",
    "                        snp = \",\".join([locus, gene, aa])\n",
    "\n",
    "                    if snp in snps_dict and snps_dict[snp] <= configuration[\"table\"][\"threshold\"]:\n",
    "                        snp_tab_file.write(\",\".join([snp, sample_name]) + \"\\n\")\n",
    "\n",
    "    ## SNPs table\n",
    "    snp_table_file_path = os.path.join(results_directory_path, \"snp_table.csv\")\n",
    "    snps_aa_dict = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "    with open(snp_tab_file_path, 'r') as snp_tab_file:\n",
    "        lines = snp_tab_file.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            row = line.strip(\"\\n\").split(\",\")\n",
    "\n",
    "            locus = row[0]\n",
    "            gene = row[1]\n",
    "            aa = row[2]\n",
    "            sample = row[3]\n",
    "\n",
    "            column_name = f\"{locus} ({gene})\"\n",
    "\n",
    "            snps_aa_dict[sample][column_name].append(aa)\n",
    "\n",
    "        with open(snp_table_file_path, 'w', newline='') as snp_table_file:\n",
    "            snp_table_file_writer = csv.writer(snp_table_file)\n",
    "\n",
    "            all_columns = set()\n",
    "\n",
    "            for sample in snps_aa_dict:\n",
    "                for column_name in snps_aa_dict[sample]:\n",
    "                    all_columns.add(column_name)\n",
    "\n",
    "            header = [\"\"] + sorted(list(all_columns))\n",
    "            snp_table_file_writer.writerow(header)\n",
    "\n",
    "            for sample, snps in snps_aa_dict.items():\n",
    "                row = [sample]\n",
    "                for snp in sorted(list(all_columns)):\n",
    "                    aa_mutations = \",\".join(snps.get(snp, []))\n",
    "                    row.append(aa_mutations)\n",
    "                snp_table_file_writer.writerow(row)\n",
    "\n",
    "    # Indels identification\n",
    "    information_directory_path = os.path.join(indels_directory_path, \"information\")\n",
    "    annotated_ind_directory_path = os.path.join(indels_directory_path, \"annotated\")\n",
    "\n",
    "    if not os.path.exists(information_directory_path):\n",
    "        os.mkdir(information_directory_path)\n",
    "    if not os.path.exists(annotated_ind_directory_path):\n",
    "        os.mkdir(annotated_ind_directory_path)\n",
    "\n",
    "    ## Obtain indels position\n",
    "    pos_strain_dict = {}\n",
    "\n",
    "    for file in ind_files:\n",
    "        file_path = os.path.join(indels_directory_path, file)\n",
    "        \n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        strain = table_sample_strain_dict[sample]\n",
    "\n",
    "        if strain not in pos_strain_dict:\n",
    "            pos_strain_dict[strain] = {\n",
    "                \"Positions\": []\n",
    "            }\n",
    "\n",
    "        with open(file_path, 'r') as ind_file:\n",
    "            ind_file_reader = csv.reader(ind_file, delimiter='\\t')\n",
    "\n",
    "            for row in ind_file_reader:\n",
    "                pos = int(row[1])\n",
    "                pos_strain_dict[strain][\"Positions\"].append(pos)\n",
    "\n",
    "    ## Annotate indels\n",
    "    annotated_pos_dict = {}\n",
    "\n",
    "    for file in ind_files:\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        genome = table_sample_strain_dict[sample]\n",
    "    \n",
    "        annotated_genome = table_sample_strain_dict[sample] + \".csv\"\n",
    "        annotated_genome_file_path = os.path.join(annotated_genomes_directory_path, annotated_genome)\n",
    "\n",
    "        with open(annotated_genome_file_path, 'r') as annotated_genome_file:\n",
    "            annotated_genome_reader = csv.reader(annotated_genome_file, delimiter=',')\n",
    "\n",
    "            next(annotated_genome_reader)\n",
    "\n",
    "            if genome not in annotated_pos_dict:\n",
    "                annotated_pos_dict[genome] = {}\n",
    "\n",
    "            for row in annotated_genome_reader:\n",
    "                locus = row[5]\n",
    "                start = int(row[7])\n",
    "                end = int(row[8])\n",
    "                gene = row[10]\n",
    "\n",
    "                for position in pos_strain_dict[genome][\"Positions\"]:\n",
    "                    if position >= start and position <= end:\n",
    "                        annotated_pos_dict[genome][position] = {\n",
    "                            'locus': locus,\n",
    "                            'gene': gene\n",
    "                        }\n",
    "                for position in pos_strain_dict[genome][\"Positions\"]:\n",
    "                    if position not in annotated_pos_dict[genome]:\n",
    "                        annotated_pos_dict[genome][position] = {\n",
    "                            'locus': 'intron',\n",
    "                            'gene': 'intron'\n",
    "                        }\n",
    "\n",
    "    for file in ind_files:\n",
    "        file_path = os.path.join(indels_directory_path, file)\n",
    "\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        genome = table_sample_strain_dict[sample]\n",
    "\n",
    "        with open(file_path, 'r') as ind_file:\n",
    "            ind_file_reader = csv.reader(ind_file, delimiter='\\t')\n",
    "\n",
    "            modified_rows = []\n",
    "\n",
    "            for row in ind_file_reader:\n",
    "                pos = int(row[1])\n",
    "\n",
    "                annotation = annotated_pos_dict.get(genome, {}).get(pos, {\n",
    "                    \"refseq\": \"NA\", \"locus\": \"NA\", \"gene\":\"NA\"\n",
    "                })\n",
    "\n",
    "                locus = annotation[\"locus\"]\n",
    "                gene = annotation[\"gene\"]\n",
    "\n",
    "                if gene != \"\":\n",
    "                    row.extend([locus, gene])\n",
    "                else:\n",
    "                    row.extend([locus, locus])\n",
    "                \n",
    "                del row[2]\n",
    "\n",
    "                modified_rows.append(row)\n",
    "\n",
    "        annotated_ind_file_path = os.path.join(annotated_ind_directory_path, file)\n",
    "\n",
    "        with open(annotated_ind_file_path, 'w', newline='') as annotated_ind_file:\n",
    "            annotated_ind_file_writer = csv.writer(annotated_ind_file)\n",
    "            annotated_ind_file_writer.writerows(modified_rows)\n",
    "\n",
    "\n",
    "    ## Filter indels\n",
    "    inds_dict = {}\n",
    "\n",
    "    for file in ind_files:\n",
    "        file_path = os.path.join(indels_directory_path, file)\n",
    "\n",
    "        with open(file_path, 'r') as ind_file:\n",
    "            ind_file_reader = csv.reader(ind_file, delimiter='\\t')\n",
    "\n",
    "            for row in ind_file_reader:\n",
    "                refseq = row[0]\n",
    "                pos = row[1]\n",
    "                mut = row[3]\n",
    "\n",
    "                indel = \",\".join([refseq, pos, mut])\n",
    "\n",
    "                if indel not in inds_dict:\n",
    "                    inds_dict[indel] = 1\n",
    "                else:\n",
    "                    inds_dict[indel] += 1\n",
    "\n",
    "    inds_info_file_path = os.path.join(information_directory_path, \"indels_info.txt\")\n",
    "    with open(inds_info_file_path, 'w') as inds_info_file:\n",
    "\n",
    "        for indel, count in inds_dict.items():\n",
    "            inds_info_file.write(f\"{count}, {indel}\\n\")\n",
    "\n",
    "    filtered_inds_file_path = os.path.join(information_directory_path, \"filtered_indels.txt\")\n",
    "    with open(filtered_inds_file_path, 'w') as filtered_inds_file:\n",
    "\n",
    "        for indel, count in inds_dict.items():\n",
    "            if count <= configuration[\"table\"][\"threshold\"]:\n",
    "                filtered_inds_file.write(f\"{indel}\\n\")\n",
    "\n",
    "    ## Indels tabulation\n",
    "    ind_tab_file_path = os.path.join(information_directory_path, \"indel_tabulation.txt\")\n",
    "\n",
    "    annot_files = [file for file in os.listdir(annotated_ind_directory_path)]\n",
    "    \n",
    "    with open(ind_tab_file_path, 'w') as ind_tab_file:\n",
    "        for file in annot_files:\n",
    "            file_path = os.path.join(annotated_ind_directory_path, file)\n",
    "\n",
    "            with open(file_path, 'r') as ind_file:\n",
    "                ind_file_reader = csv.reader(ind_file, delimiter=\",\")\n",
    "\n",
    "                for row in ind_file_reader:\n",
    "                    refseq = row[0]\n",
    "                    pos = row[1]\n",
    "                    mut = row[2]\n",
    "                    sample = row[14].split(\"_\")[0]\n",
    "                    locus = row[15]\n",
    "                    gene = row[16]\n",
    "\n",
    "                    indel = \",\".join([refseq, pos, mut])\n",
    "\n",
    "                    if indel in inds_dict and inds_dict[indel] <= configuration[\"table\"][\"threshold\"]:\n",
    "                        ind_mut = f\"{pos} {mut}\"\n",
    "                        ind_tab_file.write(\",\".join([locus, gene, ind_mut, sample]) + \"\\n\")\n",
    "\n",
    "    # Indels table\n",
    "    ind_table_file_path = os.path.join(results_directory_path, \"indels_table.csv\")\n",
    "    \n",
    "    final_inds_dict = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "    with open(ind_tab_file_path, 'r') as ind_tab_file:\n",
    "        lines = ind_tab_file.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            row = line.strip(\"\\n\").split(\",\")\n",
    "            \n",
    "            locus = row[0]\n",
    "            gene = row[1]\n",
    "            indel = row[2]\n",
    "            sample = row[3]\n",
    "\n",
    "            if gene != \"\":\n",
    "                column_name = f\"{locus} ({gene})\"\n",
    "            else:\n",
    "                column_name = f\"{locus} ({locus})\"\n",
    "\n",
    "            final_inds_dict[sample][column_name].append(indel)\n",
    "\n",
    "        with open(ind_table_file_path, 'w', newline='') as ind_table_file:\n",
    "            ind_table_file_writer = csv.writer(ind_table_file)\n",
    "\n",
    "            all_columns = set()\n",
    "\n",
    "            for sample in final_inds_dict:\n",
    "                for column_name in final_inds_dict[sample]:\n",
    "                    all_columns.add(column_name)\n",
    "\n",
    "            header = [\"\"] + sorted(list(all_columns))\n",
    "            ind_table_file_writer.writerow(header)\n",
    "\n",
    "            for sample, inds in final_inds_dict.items():\n",
    "                row = [sample]\n",
    "                for indel in sorted(list(all_columns)):\n",
    "                    indels = \",\".join(inds.get(indel, []))\n",
    "                    row.append(indels)\n",
    "                ind_table_file_writer.writerow(row)\n",
    "\n",
    "    # SNP/indel table\n",
    "    mutations_file_path = os.path.join(results_directory_path, \"mutations.txt\")\n",
    "\n",
    "    with open(mutations_file_path, 'w', newline='') as mutations_file:\n",
    "        mutations_file_writer = csv.writer(mutations_file)\n",
    "\n",
    "        with(open(snp_tab_file_path, 'r') as snp_tab_file,\n",
    "             open(ind_tab_file_path, 'r') as ind_tab_file):\n",
    "\n",
    "            snp_tab_file_reader = csv.reader(snp_tab_file, delimiter=',')\n",
    "            ind_tab_file_reader = csv.reader(ind_tab_file, delimiter=',')\n",
    "\n",
    "            for row in snp_tab_file_reader:\n",
    "                mutations_file_writer.writerow(row)\n",
    "\n",
    "            for row in ind_tab_file_reader:\n",
    "                mutations_file_writer.writerow(row)\n",
    "\n",
    "    ## mutations table\n",
    "    combined_table_file_path = os.path.join(results_directory_path, \"mutations.csv\")\n",
    "\n",
    "    mut_dict = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "    with open(mutations_file_path, 'r') as mut_file:\n",
    "        lines = mut_file.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            row = line.strip().split(\",\")\n",
    "\n",
    "            locus = row[0]\n",
    "            gene = row[1]\n",
    "            mut = row[2]\n",
    "            sample = row[3]\n",
    "\n",
    "            if gene != \"\":\n",
    "                column_name = f\"{locus} ({gene})\"\n",
    "            else:\n",
    "                column_name = f\"{locus} ({locus})\"\n",
    "\n",
    "            mut_dict[sample][column_name].append(mut)\n",
    "\n",
    "    with open(combined_table_file_path, 'w', newline='') as combined_table_file:\n",
    "        combined_table_file_writer = csv.writer(combined_table_file)\n",
    "\n",
    "        all_columns = set()\n",
    "\n",
    "        for sample in mut_dict:\n",
    "            for column_name in mut_dict[sample]:\n",
    "                all_columns.add(column_name)\n",
    "\n",
    "        header = [\"\"] + sorted(list(all_columns))\n",
    "        combined_table_file_writer.writerow(header)\n",
    "\n",
    "        for sample, muts in mut_dict.items():\n",
    "            row = [sample]\n",
    "            for column in sorted(list(all_columns)):\n",
    "                mutations = \", \".join(muts.get(column, []))\n",
    "                row.append(mutations)\n",
    "            combined_table_file_writer.writerow(row)\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be46e20-247c-45ca-beae-b5efa27d854a",
   "metadata": {},
   "source": [
    "## *De novo* genome assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2282ab3f-4569-402d-9dff-78f6e2c88cdf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command line: /home/student/miniconda3/envs/VariantCalling/bin/spades.py\t-o\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes\t-1\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0253G_R1.fastq.gz\t-2\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0253G_R2.fastq.gz\t--isolate\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 4.1.0\n",
      "  Python version: 3.11.0\n",
      "  OS: Linux-5.15.0-139-generic-x86_64-with-glibc2.31\n",
      "\n",
      "Output dir: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes\n",
      "Mode: ONLY assembling (without read error correction)\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Isolate mode\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0253G_R1.fastq.gz']\n",
      "      right reads: ['/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0253G_R2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "  Assembly graph output will use GFA v1.2 format\n",
      "Other parameters:\n",
      "  Dir for temp files: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 10\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/spades.log\n",
      "\n",
      "/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0253G_R1.fastq.gz: max reads length: 151\n",
      "/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0253G_R2.fastq.gz: max reads length: 151\n",
      "\n",
      "Reads length: 151\n",
      "\n",
      "Default k-mer sizes were set to [21, 33, 55, 77] because estimated read length (151) is equal to or greater than 150\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Assembling started. \n",
      "\n",
      "\n",
      "===== K21 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=21\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.004     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.005     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.007     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (read_converter.cpp        :  78)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (read_converter.cpp        :  99)   Converting paired reads\n",
      "  0:00:00.150    34M / 40M   INFO    General                 (binary_converter.cpp      : 127)   16384 reads processed\n",
      "  0:00:00.273    39M / 45M   INFO    General                 (binary_converter.cpp      : 127)   32768 reads processed\n",
      "  0:00:00.715    49M / 54M   INFO    General                 (binary_converter.cpp      : 127)   65536 reads processed\n",
      "  0:00:01.405    59M / 64M   INFO    General                 (binary_converter.cpp      : 127)   131072 reads processed\n",
      "  0:00:02.949    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   262144 reads processed\n",
      "  0:00:05.837    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   524288 reads processed\n",
      "  0:00:12.266    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   1048576 reads processed\n",
      "  0:00:24.272    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   2097152 reads processed\n",
      "  0:00:51.089    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   4194304 reads processed\n",
      "  0:01:15.122    59M / 65M   INFO    General                 (binary_converter.cpp      : 143)   6236138 reads written\n",
      "  0:01:15.128     2M / 65M   INFO    General                 (read_converter.cpp        : 113)   Converting single reads\n",
      "  0:01:15.130    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:01:15.132     2M / 65M   INFO    General                 (read_converter.cpp        : 119)   Converting merged reads\n",
      "  0:01:15.135    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:01:15.136     1M / 65M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:01:15.138     1M / 65M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:01:15.138     1M / 65M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:01:15.139     1M / 65M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:01:15.139     1M / 65M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:01:15.140     1M / 65M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:01:15.140     1M / 65M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:01:15.141     1M / 65M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:01:25.861   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1097460 reads\n",
      "  0:01:34.979   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2195287 reads\n",
      "  0:01:44.529   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3293976 reads\n",
      "  0:01:54.128   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4393151 reads\n",
      "  0:02:03.433   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5493475 reads\n",
      "  0:02:12.569   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6592500 reads\n",
      "  0:02:21.788   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 7691770 reads\n",
      "  0:02:31.486   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8791536 reads\n",
      "  0:02:40.852   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9891207 reads\n",
      "  0:03:47.550   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 17586170 reads\n",
      "  0:04:50.898     1M / 628M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24944552 reads\n",
      "  0:04:50.898     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:05:41.326     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 62772672 kmers in total.\n",
      "  0:05:41.327     1M / 628M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:05:41.329     1M / 628M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:05:41.330     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:05:41.330     1M / 628M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:05:41.330     1M / 628M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:05:41.330     1M / 628M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:05:52.574   601M / 808M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 30214005 kmers\n",
      "  0:06:03.931   601M / 1089M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 60420700 kmers\n",
      "  0:06:04.904   601M / 1089M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 62772679 kmers\n",
      "  0:06:04.904   601M / 1089M INFO    General                 (kmer_splitters.hpp        : 202)   Used 62772679 kmers.\n",
      "  0:06:04.957     1M / 1089M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:06:12.161     1M / 1089M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 61399365 kmers in total.\n",
      "  0:06:12.162     1M / 1089M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:06:24.180    44M / 1089M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 61399365 kmers, 44353808 bytes occupied (5.77906 bits per kmer).\n",
      "  0:06:24.181    44M / 1089M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:06:27.990   104M / 1089M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:07:21.594   104M / 1089M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:07:21.634   104M / 1089M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:07:21.635   104M / 1089M INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:07:21.635   104M / 1089M INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:08:29.365   184M / 1089M INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 811377\n",
      "  0:08:30.210   184M / 1089M INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 823635\n",
      "  0:08:30.234   104M / 1089M INFO   Early tip clipping       (early_simplification.hpp  :  50)   15802391 22-mers were removed by early tip clipper\n",
      "  0:08:30.242   104M / 1089M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:08:30.545   104M / 1089M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:09:36.092   246M / 1089M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 3871487 sequences extracted\n",
      "  0:09:55.282   246M / 1089M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:10:06.544   246M / 1089M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:10:06.637   246M / 1089M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:10:08.740   246M / 1089M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:10:08.741   246M / 1089M INFO    General                 (debruijn_graph_constructor: 516)   Total 7742974 edges to create\n",
      "  0:10:08.742   547M / 1089M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:10:12.096   667M / 1089M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:10:14.722   667M / 1089M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:10:14.771   699M / 1089M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:10:15.623   699M / 1089M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:10:15.623   699M / 1089M INFO    General                 (debruijn_graph_constructor: 542)   Total 2498180 vertices to create\n",
      "  0:10:15.624   931M / 1089M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:10:17.276   720M / 1089M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:10:17.277   720M / 1089M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:10:27.124   763M / 1089M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 62772672 kmers, 45345752 bytes occupied (5.77904 bits per kmer).\n",
      "  0:10:27.239  1003M / 1089M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:19:09.590  1003M / 1089M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:19:15.758  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:19:22.024  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:19:28.466  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:19:34.932  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4000000 edges\n",
      "  0:19:42.228  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5000000 edges\n",
      "  0:19:49.997  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5963249 edges\n",
      "  0:19:55.568  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 6863249 edges\n",
      "  0:20:00.877  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 7704787 edges\n",
      "  0:20:01.060  1003M / 1089M INFO    General                 (coverage_filling.hpp      :  83)   Processed 7742700 edges\n",
      "  0:20:01.514   618M / 1089M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:20:01.530   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 18\n",
      "  0:20:01.533   618M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 255\n",
      "  0:20:01.534   618M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 255. Coverage mad: 65.2344\n",
      "  0:20:01.534   618M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:20:01.984   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:20:02.699   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:20:04.588   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:20:10.944   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:20:14.544   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 221.584. Fitted coverage std. dev: 76.4291\n",
      "  0:20:14.558   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0693853\n",
      "  0:20:14.558   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 21\n",
      "  0:20:14.558   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 21\n",
      "  0:20:14.559   619M / 1089M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6401196\n",
      "  0:20:14.559   618M / 1089M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 221.584\n",
      "  0:20:14.559   618M / 1089M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 21\n",
      "  0:20:14.559   618M / 1089M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:20:14.559   618M / 1089M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:20:14.563   618M / 1089M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:20:14.563   618M / 1089M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:20:14.564   618M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:20:15.387   618M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:20:15.388   618M / 1089M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:20:15.388   618M / 1089M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:20:15.389   618M / 1089M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:20:15.389   618M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:20:15.389   618M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:17.417   618M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 16397 times\n",
      "  0:20:17.418   618M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:31.056   673M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1106293 times\n",
      "  0:21:31.061   673M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:35.068   668M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 255861 times\n",
      "  0:21:35.093   668M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:21:35.094   668M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:35.462   662M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 13813 times\n",
      "  0:21:35.472   662M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.128   655M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 27156 times\n",
      "  0:21:37.136   655M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.254   655M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8758 times\n",
      "  0:21:37.257   655M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:21:37.257   655M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.280   655M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 244 times\n",
      "  0:21:37.280   655M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.528   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2005 times\n",
      "  0:21:37.529   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.585   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2068 times\n",
      "  0:21:37.586   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:21:37.586   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.595   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 43 times\n",
      "  0:21:37.598   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.678   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 393 times\n",
      "  0:21:37.681   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.702   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 913 times\n",
      "  0:21:37.703   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:21:37.704   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.705   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 9 times\n",
      "  0:21:37.705   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.768   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 105 times\n",
      "  0:21:37.771   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.781   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 528 times\n",
      "  0:21:37.781   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:21:37.781   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.782   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 6 times\n",
      "  0:21:37.782   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.795   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 36 times\n",
      "  0:21:37.796   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.799   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 284 times\n",
      "  0:21:37.800   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:21:37.800   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.800   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:21:37.802   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.814   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 18 times\n",
      "  0:21:37.815   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.818   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 207 times\n",
      "  0:21:37.818   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:21:37.818   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.819   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:21:37.819   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.825   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 12 times\n",
      "  0:21:37.826   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.828   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 127 times\n",
      "  0:21:37.829   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:21:37.829   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.829   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:21:37.830   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.836   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 8 times\n",
      "  0:21:37.837   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.838   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 99 times\n",
      "  0:21:37.838   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:21:37.839   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.839   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:37.839   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.844   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:21:37.845   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.846   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 94 times\n",
      "  0:21:37.847   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:21:37.847   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.877   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 37 times\n",
      "  0:21:37.878   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.937   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 15 times\n",
      "  0:21:37.937   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.967   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:21:37.967   653M / 1089M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:21:37.968   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:37.974   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:37.976   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:37.976   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:37.976   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:37.977   653M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:21:38.012   539M / 1089M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:21:38.013   539M / 1089M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:21:38.013   539M / 1089M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:21:38.013   539M / 1089M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:21:38.013   539M / 1089M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:21:38.014   539M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:38.038   539M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:38.039   539M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:38.079   540M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:38.080   540M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:38.103   540M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:38.104   540M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:38.140   540M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:38.141   540M / 1089M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:21:38.165   540M / 1089M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:21:38.362   539M / 1089M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 35758 times\n",
      "  0:21:38.387   538M / 1089M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:21:38.390   538M / 1089M INFO    General                 (simplification.cpp        : 509)     Average coverage = 229.856\n",
      "  0:21:38.420   538M / 1089M INFO    General                 (simplification.cpp        : 510)     Total length = 6488573\n",
      "  0:21:38.436   538M / 1089M INFO    General                 (simplification.cpp        : 511)     Median edge length: 2384\n",
      "  0:21:38.437   538M / 1089M INFO    General                 (simplification.cpp        : 512)     Edges: 22936\n",
      "  0:21:38.437   538M / 1089M INFO    General                 (simplification.cpp        : 513)     Vertices: 14822\n",
      "  0:21:38.437   538M / 1089M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:21:38.726   538M / 1089M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K21/simplified_contigs\"\n",
      "  0:21:38.782   542M / 1089M INFO    General                 (binary_converter.cpp      : 143)   11491 reads written\n",
      "  0:21:38.787   538M / 1089M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:21:38.861     1M / 1089M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 21 minutes 38 seconds\n",
      "\n",
      "===== K21 finished. \n",
      "\n",
      "\n",
      "===== K33 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=33\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.000     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K21/simplified_contigs\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:06.957   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 608983 reads\n",
      "  0:00:14.093   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1216623 reads\n",
      "  0:00:21.046   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1825881 reads\n",
      "  0:00:28.936   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2434386 reads\n",
      "  0:00:35.592   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3043283 reads\n",
      "  0:00:41.992   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3653144 reads\n",
      "  0:00:48.751   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4262379 reads\n",
      "  0:00:57.250   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4871742 reads\n",
      "  0:01:43.005   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8528968 reads\n",
      "  0:03:12.107   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 17057563 reads\n",
      "  0:04:38.764     1M / 571M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24967534 reads\n",
      "  0:04:38.765     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:08:42.387     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 74281939 kmers in total.\n",
      "  0:08:42.387     1M / 571M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:08:42.419     1M / 571M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:08:42.419     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:08:42.419     1M / 571M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:08:42.420     1M / 571M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:08:42.420     1M / 571M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:08:49.347   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 10150977 kmers\n",
      "  0:08:58.080   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 21181589 kmers\n",
      "  0:09:07.858   601M / 692M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 31250866 kmers\n",
      "  0:09:16.140   601M / 789M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 42935127 kmers\n",
      "  0:09:23.362   601M / 893M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 52996970 kmers\n",
      "  0:09:36.627   601M / 965M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 64378399 kmers\n",
      "  0:09:46.788   601M / 1071M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 74281972 kmers\n",
      "  0:09:46.789   601M / 1071M INFO    General                 (kmer_splitters.hpp        : 202)   Used 74281972 kmers.\n",
      "  0:09:46.849     1M / 1071M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:11:20.164     1M / 1071M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 73597170 kmers in total.\n",
      "  0:11:20.165     1M / 1071M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:11:35.484    54M / 1071M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 73597170 kmers, 53163784 bytes occupied (5.77889 bits per kmer).\n",
      "  0:11:35.484    54M / 1071M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:11:44.118   126M / 1071M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:13:06.966   126M / 1071M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:13:06.979   126M / 1071M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:13:06.980   126M / 1071M INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:13:06.981   126M / 1071M INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:14:25.001   206M / 1071M INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 977386\n",
      "  0:14:26.583   206M / 1071M INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 997454\n",
      "  0:14:26.651   126M / 1071M INFO   Early tip clipping       (early_simplification.hpp  :  50)   26010238 34-mers were removed by early tip clipper\n",
      "  0:14:26.653   126M / 1071M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:14:26.656   126M / 1071M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:15:43.026   236M / 1071M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2346402 sequences extracted\n",
      "  0:16:02.738   236M / 1071M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:16:18.371   236M / 1071M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:16:18.433   236M / 1071M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:16:19.889   236M / 1071M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:16:19.890   236M / 1071M INFO    General                 (debruijn_graph_constructor: 516)   Total 4692804 edges to create\n",
      "  0:16:19.898   417M / 1071M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:16:22.592   489M / 1071M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:16:24.453   489M / 1071M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:16:24.475   505M / 1071M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:16:25.135   505M / 1071M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:16:25.136   505M / 1071M INFO    General                 (debruijn_graph_constructor: 542)   Total 1661633 vertices to create\n",
      "  0:16:25.136   661M / 1071M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:16:26.094   537M / 1071M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:16:26.095   537M / 1071M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:16:39.025   591M / 1071M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 74281939 kmers, 53658448 bytes occupied (5.7789 bits per kmer).\n",
      "  0:16:39.201   875M / 1071M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:24:13.957   875M / 1071M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:24:27.634   875M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:24:39.605   875M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1966165 edges\n",
      "  0:24:47.392   875M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2866165 edges\n",
      "  0:24:56.013   875M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3758667 edges\n",
      "  0:25:02.800   875M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4558667 edges\n",
      "  0:25:04.180   875M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4692773 edges\n",
      "  0:25:04.870   413M / 1071M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:25:04.880   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 17\n",
      "  0:25:04.881   413M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 224\n",
      "  0:25:04.881   413M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 224. Coverage mad: 60.7866\n",
      "  0:25:04.882   413M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:25:05.094   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:25:05.704   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:25:07.233   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:25:12.324   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:25:14.674   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 197.76. Fitted coverage std. dev: 69.5032\n",
      "  0:25:14.684   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0364839\n",
      "  0:25:14.684   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 17\n",
      "  0:25:14.684   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 17\n",
      "  0:25:14.686   414M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6411108\n",
      "  0:25:14.689   413M / 1071M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 197.76\n",
      "  0:25:14.689   413M / 1071M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 17\n",
      "  0:25:14.690   413M / 1071M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:25:14.690   413M / 1071M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:25:14.692   413M / 1071M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:25:14.697   413M / 1071M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:25:14.698   413M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:25:15.129   413M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 2 times\n",
      "  0:25:15.136   413M / 1071M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:25:15.137   413M / 1071M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:25:15.137   413M / 1071M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:25:15.137   413M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:25:15.138   413M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:16.210   413M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 33015 times\n",
      "  0:25:16.210   413M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:47.794   430M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 684216 times\n",
      "  0:25:47.804   430M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:49.020   428M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 68206 times\n",
      "  0:25:49.032   428M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:25:49.032   428M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:49.448   424M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 14463 times\n",
      "  0:25:49.451   424M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.116   420M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 14390 times\n",
      "  0:25:50.119   420M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.221   420M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 4216 times\n",
      "  0:25:50.222   420M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:25:50.223   420M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.244   420M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 531 times\n",
      "  0:25:50.245   420M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.386   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1264 times\n",
      "  0:25:50.387   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.409   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 547 times\n",
      "  0:25:50.409   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:25:50.409   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.412   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 47 times\n",
      "  0:25:50.413   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.448   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 171 times\n",
      "  0:25:50.449   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.456   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 157 times\n",
      "  0:25:50.456   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:25:50.456   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.457   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 11 times\n",
      "  0:25:50.457   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.468   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 24 times\n",
      "  0:25:50.473   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.480   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 73 times\n",
      "  0:25:50.481   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:25:50.481   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.481   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 6 times\n",
      "  0:25:50.481   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.489   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 10 times\n",
      "  0:25:50.489   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.492   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 61 times\n",
      "  0:25:50.492   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:25:50.492   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.493   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 4 times\n",
      "  0:25:50.493   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.496   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 4 times\n",
      "  0:25:50.496   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.497   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 37 times\n",
      "  0:25:50.498   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:25:50.498   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.498   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:25:50.498   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.502   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:25:50.502   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.504   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 36 times\n",
      "  0:25:50.504   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:25:50.505   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.505   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:25:50.505   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.508   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:25:50.509   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.510   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 24 times\n",
      "  0:25:50.510   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:25:50.511   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.511   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:25:50.511   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.513   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:25:50.514   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.514   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 13 times\n",
      "  0:25:50.515   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:25:50.515   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.543   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 12 times\n",
      "  0:25:50.544   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.570   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:25:50.571   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.602   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:50.604   419M / 1071M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:25:50.605   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.605   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:50.605   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.605   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:50.606   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:50.606   419M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:50.647   345M / 1071M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:25:50.647   345M / 1071M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:25:50.647   345M / 1071M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:25:50.648   345M / 1071M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:25:50.648   345M / 1071M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:25:50.648   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.671   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:50.671   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.695   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:50.696   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:50.720   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:50.720   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:50.749   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:50.750   345M / 1071M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:25:50.775   345M / 1071M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:25:50.940   343M / 1071M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 89158 times\n",
      "  0:25:50.953   341M / 1071M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:25:50.954   341M / 1071M INFO    General                 (simplification.cpp        : 509)     Average coverage = 203.895\n",
      "  0:25:50.963   341M / 1071M INFO    General                 (simplification.cpp        : 510)     Total length = 6513203\n",
      "  0:25:50.970   341M / 1071M INFO    General                 (simplification.cpp        : 511)     Median edge length: 22424\n",
      "  0:25:50.970   341M / 1071M INFO    General                 (simplification.cpp        : 512)     Edges: 4982\n",
      "  0:25:50.971   341M / 1071M INFO    General                 (simplification.cpp        : 513)     Vertices: 4046\n",
      "  0:25:50.971   341M / 1071M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:25:50.973   341M / 1071M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K33/simplified_contigs\"\n",
      "  0:25:50.999   345M / 1071M INFO    General                 (binary_converter.cpp      : 143)   2493 reads written\n",
      "  0:25:51.004   341M / 1071M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:25:51.053     1M / 1071M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 25 minutes 51 seconds\n",
      "\n",
      "===== K33 finished. \n",
      "\n",
      "\n",
      "===== K55 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=55\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.000     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.007     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K33/simplified_contigs\n",
      "  0:00:00.052     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.052     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:00:00.052     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.052     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.054     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.054     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.054     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:06.936   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 758726 reads\n",
      "  0:00:13.314   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1516834 reads\n",
      "  0:00:19.702   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2274976 reads\n",
      "  0:00:26.095   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3033866 reads\n",
      "  0:00:33.502   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3793509 reads\n",
      "  0:00:40.273   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4552653 reads\n",
      "  0:00:46.537   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5313017 reads\n",
      "  0:00:52.822   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6073528 reads\n",
      "  0:01:19.057   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9111813 reads\n",
      "  0:02:31.523   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 17466388 reads\n",
      "  0:03:38.229     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24949538 reads\n",
      "  0:03:38.229     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:06:46.585     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 80768556 kmers in total.\n",
      "  0:06:46.587     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:06:46.591     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:06:46.591     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:06:46.593     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:06:46.593     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:06:46.594     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:06:56.937   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 9634626 kmers\n",
      "  0:07:04.475   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 21363502 kmers\n",
      "  0:07:16.722   601M / 710M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 31698974 kmers\n",
      "  0:07:22.847   601M / 811M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 41312786 kmers\n",
      "  0:07:30.521   601M / 937M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 52181187 kmers\n",
      "  0:07:41.915   601M / 967M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 63344733 kmers\n",
      "  0:07:51.199   601M / 1068M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 72950204 kmers\n",
      "  0:07:56.284   601M / 1099M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 80768593 kmers\n",
      "  0:07:56.286   601M / 1099M INFO    General                 (kmer_splitters.hpp        : 202)   Used 80768593 kmers.\n",
      "  0:07:56.326     1M / 1099M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:09:32.858     1M / 1099M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 80765942 kmers in total.\n",
      "  0:09:32.862     1M / 1099M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:10:00.049    58M / 1099M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 80765942 kmers, 58341664 bytes occupied (5.77884 bits per kmer).\n",
      "  0:10:00.049    58M / 1099M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:10:10.307   138M / 1099M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:11:35.089   138M / 1099M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:11:35.100   138M / 1099M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:11:35.106   138M / 1099M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:13:30.529   294M / 1099M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 3123409 sequences extracted\n",
      "  0:14:04.299   294M / 1099M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:14:23.469   294M / 1099M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:14:23.570   294M / 1099M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:14:25.634   294M / 1099M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:14:25.635   294M / 1099M INFO    General                 (debruijn_graph_constructor: 516)   Total 6246818 edges to create\n",
      "  0:14:25.638   535M / 1099M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:14:29.344   631M / 1099M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:14:31.750   631M / 1099M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:14:31.813   663M / 1099M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:14:33.055   663M / 1099M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:14:33.056   663M / 1099M INFO    General                 (debruijn_graph_constructor: 542)   Total 3120795 vertices to create\n",
      "  0:14:33.062   952M / 1099M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:14:35.024   776M / 1099M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:14:35.027   776M / 1099M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:14:52.375   834M / 1099M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 80768556 kmers, 58343544 bytes occupied (5.77884 bits per kmer).\n",
      "  0:14:52.588  1146M / 1157M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:21:18.090  1146M / 1157M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:21:32.485  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:21:47.663  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:22:02.419  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:22:09.749  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3682742 edges\n",
      "  0:22:15.467  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4282742 edges\n",
      "  0:22:20.427  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4814020 edges\n",
      "  0:22:25.093  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5314020 edges\n",
      "  0:22:29.474  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5788719 edges\n",
      "  0:22:32.894  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 6151002 edges\n",
      "  0:22:33.882  1146M / 1157M INFO    General                 (coverage_filling.hpp      :  83)   Processed 6246815 edges\n",
      "  0:22:34.377   639M / 1157M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:22:34.382   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 14\n",
      "  0:22:34.383   639M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 174\n",
      "  0:22:34.383   639M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 174. Coverage mad: 50.4084\n",
      "  0:22:34.383   639M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:22:34.617   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:22:35.443   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:22:36.782   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:22:41.472   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:22:43.845   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 156.217. Fitted coverage std. dev: 56.7705\n",
      "  0:22:43.853   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.034713\n",
      "  0:22:43.853   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 15\n",
      "  0:22:43.853   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 15\n",
      "  0:22:43.853   640M / 1157M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6378969\n",
      "  0:22:43.854   639M / 1157M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 156.217\n",
      "  0:22:43.854   639M / 1157M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 15\n",
      "  0:22:43.854   639M / 1157M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:22:43.854   639M / 1157M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:22:43.861   653M / 1157M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:22:45.444   689M / 1157M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1663524 out of 6246815, length: 82189699\n",
      "  0:22:45.731   711M / 1157M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 6278055)\n",
      "  0:22:48.867   711M / 1157M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:23:41.149   753M / 1157M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 60230801 kmers, 43503272 bytes occupied (5.77821 bits per kmer).\n",
      "  0:23:41.447  1213M / 1329M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:24:15.245  1191M / 1329M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:24:38.876  1191M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:25:02.840  1192M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:25:26.939  1192M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:25:50.978  1193M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:26:15.059  1193M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:26:38.631  1194M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:28:36.616  1197M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:32:29.935  1202M / 1329M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:36:26.197  1224M / 1329M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:36:26.696   735M / 1329M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:36:26.699   735M / 1329M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:36:28.598   735M / 1329M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 6778 tips collected, total 7282 connection candidates\n",
      "  0:36:28.629   735M / 1329M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 118 gaps after checking 7153 candidates\n",
      "  0:36:28.777   735M / 1329M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:36:28.876   639M / 1329M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:36:28.883   639M / 1329M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:36:28.884   639M / 1329M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:36:28.884   639M / 1329M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:36:29.508   639M / 1329M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:36:29.509   639M / 1329M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:36:29.509   639M / 1329M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:36:29.510   639M / 1329M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:36:29.510   639M / 1329M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:36:29.510   639M / 1329M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:38.133   604M / 1329M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1161706 times\n",
      "  0:36:38.135   604M / 1329M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:32.738  1836M / 1884M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 226261 times\n",
      "  0:37:32.740  1836M / 1884M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:32.961  1837M / 1886M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8793 times\n",
      "  0:37:32.970  1837M / 1886M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:37:32.971  1837M / 1886M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:33.285  1835M / 1886M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 7943 times\n",
      "  0:37:33.289  1835M / 1886M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.116  1848M / 1897M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2524 times\n",
      "  0:37:34.122  1848M / 1897M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.190  1848M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 756 times\n",
      "  0:37:34.192  1848M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:37:34.193  1848M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.217  1848M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 228 times\n",
      "  0:37:34.219  1848M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.318  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 164 times\n",
      "  0:37:34.320  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.359  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 75 times\n",
      "  0:37:34.361  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:37:34.362  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.365  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 12 times\n",
      "  0:37:34.371  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.397  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 17 times\n",
      "  0:37:34.399  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.405  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 18 times\n",
      "  0:37:34.406  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:37:34.407  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.410  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 4 times\n",
      "  0:37:34.411  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.412  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:34.413  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.416  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 12 times\n",
      "  0:37:34.417  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:37:34.418  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.419  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:34.420  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.425  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:37:34.426  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.429  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6 times\n",
      "  0:37:34.430  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:37:34.431  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.432  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:34.433  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.441  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:34.442  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.444  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:37:34.445  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:37:34.446  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.447  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:34.448  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.449  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:37:34.450  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.452  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:37:34.453  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:37:34.454  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.455  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:34.456  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.457  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:34.458  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.459  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:37:34.460  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:37:34.461  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.462  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:34.463  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.464  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:34.465  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.466  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:37:34.467  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:37:34.468  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.548  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 35 times\n",
      "  0:37:34.549  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.641  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:37:34.642  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.693  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:37:34.695  1849M / 1900M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:37:34.695  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:34.696  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:34.697  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:34.698  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:34.699  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:34.700  1849M / 1900M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:37:34.738  1795M / 1900M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:37:34.743  1809M / 1900M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:37:34.890  1825M / 1900M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 472438 out of 474291, length: 48768889\n",
      "  0:37:34.941  1829M / 1902M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 6278055)\n",
      "  0:37:35.980  1829M / 1902M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:37:51.902  1848M / 1941M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 25649236 kmers, 18526256 bytes occupied (5.77834 bits per kmer).\n",
      "  0:37:52.010  2044M / 2127M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:38:02.830  2040M / 2127M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:38:24.653  2040M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:38:46.140  2040M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:39:07.691  2041M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:39:29.227  2041M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:39:50.694  2041M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:40:12.088  2041M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:41:58.197  2043M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:45:30.011  2045M / 2127M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:49:10.199  2048M / 2127M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:49:10.332  1842M / 2132M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:49:10.339  1842M / 2132M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:49:10.502  1842M / 2132M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 449 tips collected, total 788 connection candidates\n",
      "  0:49:10.511  1842M / 2132M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 788 candidates\n",
      "  0:49:10.567  1842M / 2132M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:49:10.596  1795M / 2132M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:49:10.600  1795M / 2132M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:49:10.601  1795M / 2132M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:49:10.602  1795M / 2132M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:49:10.603  1795M / 2132M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:49:10.604  1795M / 2132M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:49:10.702  1795M / 2132M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:49:10.703  1795M / 2132M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:49:10.829  1796M / 2132M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:49:10.832  1796M / 2132M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:49:10.888  1796M / 2132M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:49:10.890  1796M / 2132M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:49:10.979  1796M / 2132M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:49:10.981  1796M / 2132M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:49:11.068  1796M / 2132M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:49:11.465  1792M / 2132M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 235840 times\n",
      "  0:49:11.489  1785M / 2132M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:49:11.490  1785M / 2132M INFO    General                 (simplification.cpp        : 509)     Average coverage = 158.432\n",
      "  0:49:11.500  1785M / 2132M INFO    General                 (simplification.cpp        : 510)     Total length = 6530004\n",
      "  0:49:11.515  1785M / 2132M INFO    General                 (simplification.cpp        : 511)     Median edge length: 52635\n",
      "  0:49:11.516  1785M / 2132M INFO    General                 (simplification.cpp        : 512)     Edges: 2612\n",
      "  0:49:11.517  1785M / 2132M INFO    General                 (simplification.cpp        : 513)     Vertices: 2742\n",
      "  0:49:11.518  1785M / 2132M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:49:14.411  1785M / 2132M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K55/simplified_contigs\"\n",
      "  0:49:14.446  1789M / 2132M INFO    General                 (binary_converter.cpp      : 143)   1306 reads written\n",
      "  0:49:14.450  1785M / 2132M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:49:19.569     1M / 2132M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 49 minutes 19 seconds\n",
      "\n",
      "===== K55 finished. \n",
      "\n",
      "\n",
      "===== K77 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=77\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.003     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.003     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.004     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.005     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K55/simplified_contigs\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:00:05.378   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 670453 reads\n",
      "  0:00:10.696   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1340357 reads\n",
      "  0:00:16.052   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2010788 reads\n",
      "  0:00:21.301   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2681456 reads\n",
      "  0:00:28.024   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3352592 reads\n",
      "  0:00:33.712   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4024031 reads\n",
      "  0:00:39.063   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4695551 reads\n",
      "  0:00:44.704   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5368155 reads\n",
      "  0:01:12.381   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8728117 reads\n",
      "  0:02:17.641   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 16788912 reads\n",
      "  0:03:40.314     1M / 571M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24947164 reads\n",
      "  0:03:40.315     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:09:06.128     1M / 680M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 75620543 kmers in total.\n",
      "  0:09:06.128     1M / 680M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:09:06.138     1M / 680M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:09:06.141     1M / 680M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:09:06.142     1M / 680M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:09:06.143     1M / 680M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:09:06.143     1M / 680M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:09:11.555   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 5792426 kmers\n",
      "  0:09:18.601   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 12550071 kmers\n",
      "  0:09:28.308   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 19751620 kmers\n",
      "  0:09:40.864   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 27144535 kmers\n",
      "  0:09:48.460   601M / 716M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 34631252 kmers\n",
      "  0:10:00.468   601M / 778M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 42161625 kmers\n",
      "  0:10:14.175   601M / 802M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 49708820 kmers\n",
      "  0:10:21.172   601M / 824M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 57273014 kmers\n",
      "  0:10:37.366   601M / 888M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 64831267 kmers\n",
      "  0:10:46.123   601M / 947M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 72399278 kmers\n",
      "  0:10:52.583   601M / 947M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 75620598 kmers\n",
      "  0:10:52.587   601M / 947M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 75620598 kmers.\n",
      "  0:10:52.643     1M / 947M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:13:25.578     1M / 947M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 76046457 kmers in total.\n",
      "  0:13:25.582     1M / 947M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:13:48.929    55M / 947M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 76046457 kmers, 54932952 bytes occupied (5.77888 bits per kmer).\n",
      "  0:13:48.930    55M / 947M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:14:03.735   131M / 947M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:15:58.447   131M / 947M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:15:58.453   131M / 947M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:15:58.455   131M / 947M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:18:04.772   268M / 1031M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2575578 sequences extracted\n",
      "  0:18:45.584   268M / 1070M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:19:04.130   268M / 1070M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:19:04.324   268M / 1070M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:19:06.251   268M / 1070M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:19:06.252   268M / 1070M INFO    General                 (debruijn_graph_constructor: 516)   Total 5151156 edges to create\n",
      "  0:19:06.255   469M / 1070M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:19:09.813   549M / 1070M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:19:12.037   549M / 1070M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:19:12.079   581M / 1070M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:19:13.282   581M / 1070M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:19:13.282   581M / 1070M INFO    General                 (debruijn_graph_constructor: 542)   Total 3001492 vertices to create\n",
      "  0:19:13.283   862M / 1070M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:19:15.161   710M / 1070M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:19:15.162   710M / 1070M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:19:40.376   763M / 1070M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 75620543 kmers, 54625336 bytes occupied (5.77889 bits per kmer).\n",
      "  0:19:40.883  1055M / 1071M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:25:08.907  1055M / 1071M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:25:24.778  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:25:40.486  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:25:56.949  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:26:02.177  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3512737 edges\n",
      "  0:26:06.280  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3938006 edges\n",
      "  0:26:10.085  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4338006 edges\n",
      "  0:26:13.866  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4736484 edges\n",
      "  0:26:16.569  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5020558 edges\n",
      "  0:26:17.784  1055M / 1071M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5151156 edges\n",
      "  0:26:18.753   580M / 1071M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:26:18.757   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 13\n",
      "  0:26:18.757   580M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 127\n",
      "  0:26:18.758   580M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 127. Coverage mad: 41.5128\n",
      "  0:26:18.758   580M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:26:19.087   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:26:19.631   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:26:20.779   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:26:24.804   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:26:28.465   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 116.71. Fitted coverage std. dev: 44.0833\n",
      "  0:26:28.471   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0263696\n",
      "  0:26:28.472   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 13\n",
      "  0:26:28.472   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 13\n",
      "  0:26:28.472   581M / 1071M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6327560\n",
      "  0:26:28.474   580M / 1071M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 116.71\n",
      "  0:26:28.475   580M / 1071M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 13\n",
      "  0:26:28.475   580M / 1071M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:26:28.475   580M / 1071M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:26:28.481   593M / 1071M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:26:29.725   661M / 1071M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1993397 out of 5151156, length: 98634764\n",
      "  0:26:30.158   685M / 1071M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 5176914)\n",
      "  0:26:33.817   685M / 1071M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:27:35.011   732M / 1071M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 67023893 kmers, 48409688 bytes occupied (5.7782 bits per kmer).\n",
      "  0:27:35.339  1244M / 1331M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:28:15.659  1220M / 1331M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:28:34.332  1221M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:28:53.094  1221M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:29:11.988  1222M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:29:30.880  1223M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:29:49.461  1223M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:30:08.104  1224M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:31:39.239  1227M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:34:43.178  1251M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:37:46.564  1258M / 1331M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:37:47.182   718M / 1331M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:37:47.182   718M / 1331M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:37:48.853   718M / 1331M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 7358 tips collected, total 7822 connection candidates\n",
      "  0:37:48.897   718M / 1331M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 243 gaps after checking 7552 candidates\n",
      "  0:37:49.045   718M / 1331M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:37:49.187   580M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:37:49.210   580M / 1331M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:37:49.216   580M / 1331M INFO    General                 (simplification.cpp        :  63)   Most init cleaning disabled on main iteration\n",
      "  0:37:49.217   580M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:37:49.700   580M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:37:49.701   580M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:37:49.701   580M / 1331M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:37:49.701   580M / 1331M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:37:49.701   580M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:37:49.702   580M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:57.972   543M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1077894 times\n",
      "  0:37:57.974   543M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.423   744M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 18331 times\n",
      "  0:38:02.427   744M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.532   744M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 858 times\n",
      "  0:38:02.533   744M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:38:02.533   744M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.596   744M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1059 times\n",
      "  0:38:02.597   744M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.706   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 158 times\n",
      "  0:38:02.709   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.760   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 158 times\n",
      "  0:38:02.761   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:38:02.761   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.777   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 52 times\n",
      "  0:38:02.777   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.804   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 20 times\n",
      "  0:38:02.805   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.813   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 25 times\n",
      "  0:38:02.814   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:38:02.814   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.822   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 7 times\n",
      "  0:38:02.824   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.826   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:38:02.826   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.828   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8 times\n",
      "  0:38:02.828   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:38:02.829   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.829   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:38:02.829   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.830   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:02.830   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.837   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 5 times\n",
      "  0:38:02.838   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:38:02.838   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.839   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 3 times\n",
      "  0:38:02.841   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.841   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:02.842   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.842   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:38:02.842   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:38:02.843   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.843   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:38:02.843   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.843   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:02.844   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.844   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:38:02.845   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:38:02.845   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.845   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:38:02.846   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.846   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:02.846   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.846   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:38:02.847   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:38:02.847   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.847   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:38:02.848   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.853   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:02.853   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.854   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:38:02.854   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:38:02.854   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:02.855   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:38:02.855   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:02.855   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:02.856   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:02.856   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:38:02.857   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:38:02.857   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:03.047   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 49 times\n",
      "  0:38:03.048   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:03.128   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:03.128   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:03.205   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:38:03.206   745M / 1331M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:38:03.206   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:38:03.206   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:38:03.207   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:38:03.207   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:38:03.207   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:38:03.208   745M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:38:03.268   708M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:38:03.292   721M / 1331M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:38:03.504   738M / 1331M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 890872 out of 892036, length: 69164747\n",
      "  0:38:03.605   745M / 1331M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 5176914)\n",
      "  0:38:05.372   745M / 1331M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:38:35.230   771M / 1331M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 35984151 kmers, 25992262 bytes occupied (5.7786 bits per kmer).\n",
      "  0:38:35.369  1047M / 1331M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:38:55.510  1040M / 1331M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:39:10.134  1041M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:39:22.069  1042M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:39:35.770  1042M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:39:47.422  1043M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:39:59.090  1043M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:40:10.689  1044M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:41:06.832  1047M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:43:00.559  1053M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:44:55.544  1076M / 1331M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:44:56.131   784M / 1331M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:44:56.132   784M / 1331M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:44:56.379   784M / 1331M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 1135 tips collected, total 2039 connection candidates\n",
      "  0:44:56.394   784M / 1331M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 2039 candidates\n",
      "  0:44:56.456   784M / 1331M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:44:56.487   708M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:44:56.493   708M / 1331M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:44:56.494   708M / 1331M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:44:56.494   708M / 1331M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:44:56.494   708M / 1331M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:44:56.495   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:44:56.574   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:44:56.574   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:44:56.696   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:44:56.697   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:44:56.776   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:44:56.776   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:44:56.854   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:44:56.854   708M / 1331M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:44:57.025   708M / 1331M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:44:57.748   697M / 1331M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 445215 times\n",
      "  0:44:57.758   688M / 1331M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:44:57.758   688M / 1331M INFO    General                 (simplification.cpp        : 509)     Average coverage = 116.51\n",
      "  0:44:57.766   688M / 1331M INFO    General                 (simplification.cpp        : 510)     Total length = 6505227\n",
      "  0:44:57.773   688M / 1331M INFO    General                 (simplification.cpp        : 511)     Median edge length: 67843\n",
      "  0:44:57.773   688M / 1331M INFO    General                 (simplification.cpp        : 512)     Edges: 1606\n",
      "  0:44:57.773   688M / 1331M INFO    General                 (simplification.cpp        : 513)     Vertices: 1652\n",
      "  0:44:57.774   688M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Mismatch Correction (id: mismatch_correction)\n",
      "  0:44:57.774   688M / 1331M INFO    General                 (graph_pack_helpers.cpp    :  44)   Index refill\n",
      "  0:44:57.774   688M / 1331M INFO    General                 (edge_index.hpp            : 175)   Using small index (max_id = 5176914)\n",
      "  0:44:58.121   688M / 1331M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:45:00.359   693M / 1331M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 6505227 kmers, 4701962 bytes occupied (5.78238 bits per kmer).\n",
      "  0:45:00.383   745M / 1331M INFO    General                 (edge_index_builders.hpp   : 253)   Collecting edge information from graph, this takes a while.\n",
      "  0:45:02.901   745M / 1331M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 2901750 kmers to process\n",
      "  0:45:04.260   748M / 1331M INFO    General                 (kmer_mapper.hpp           :  87)   Total 9902 kmers with non-trivial roots\n",
      "  0:45:04.288   745M / 1331M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:45:04.288   745M / 1331M INFO    General                 (mismatch_correction.cpp   : 392)   Collect potential mismatches\n",
      "  0:45:09.747   745M / 1331M INFO    General                 (mismatch_correction.cpp   : 193)   Total 908 edges (out of 1606) with 51629 potential mismatch positions (56.8601 positions per edge)\n",
      "  0:45:09.748   745M / 1331M INFO    General                 (mismatch_correction.cpp   : 394)   Potential mismatches collected\n",
      "  0:45:09.749   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:45:11.976   748M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:45:14.176   751M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:45:16.407   752M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:45:18.590   752M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:45:20.803   752M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:45:22.958   753M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:45:31.920   753M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:45:49.078   753M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:46:26.444   753M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 8400000 reads\n",
      "  0:47:37.417   753M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 16800000 reads\n",
      "  0:48:49.989   753M / 1331M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 24944552 reads processed\n",
      "  0:48:50.046   749M / 1331M INFO    General                 (mismatch_correction.cpp   : 387)   All edges processed\n",
      "  0:48:50.047   745M / 1331M INFO    General                 (mismatch_correction.cpp   : 442)   Corrected 0 nucleotides\n",
      "  0:48:50.047   745M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:48:50.047   745M / 1331M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/before_rr.fasta\"\n",
      "  0:48:50.556   745M / 1331M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_after_simplification.gfa\"\n",
      "  0:48:50.667   745M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Paired Information Counting (id: late_pair_info_count)\n",
      "  0:48:50.684   745M / 1331M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 2901750 kmers to process\n",
      "  0:48:50.686   745M / 1331M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:48:50.694   745M / 1331M INFO    General                 (pair_info_count.cpp       : 157)   Min edge length for estimation: 67843\n",
      "  0:48:50.694   745M / 1331M INFO    General                 (pair_info_count.cpp       : 168)   Estimating insert size for library #0\n",
      "  0:48:50.695   745M / 1331M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:48:50.695   745M / 1331M INFO    General                 (paired_info_utils.cpp     :  87)   Estimating insert size (takes a while)\n",
      "  0:48:50.705   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:48:54.336   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:48:58.101   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:49:01.776   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:49:05.322   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:49:08.933   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:49:12.480   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:49:29.256   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:50:01.144   777M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:50:35.860   777M / 1331M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:50:36.225   777M / 1331M INFO    General                 (paired_info_utils.cpp     : 104)   Edge pairs: 8496\n",
      "  0:50:36.227   777M / 1331M INFO    General                 (paired_info_utils.cpp     : 106)   2940158 paired reads (47.1471% of all) aligned to long edges\n",
      "  0:50:36.228   745M / 1331M INFO    General                 (pair_info_count.cpp       : 188)     Insert size = 213.149, deviation = 72.7554, left quantile = 128, right quantile = 310, read length = 151\n",
      "  0:50:36.229   745M / 1331M INFO    General                 (pair_info_count.cpp       : 203)   Filtering data for library #0\n",
      "  0:50:36.230   745M / 1331M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:50:36.230   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:50:39.900   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:50:43.545   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:50:47.168   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:50:50.663   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:50:54.196   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:50:57.803   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:51:14.183   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:51:45.302   745M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:52:20.371   745M / 1331M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:52:20.372   745M / 1331M INFO    General                 (pair_info_count.cpp       : 207)   Mapping library #0\n",
      "  0:52:20.372   745M / 1331M INFO    General                 (pair_info_count.cpp       : 209)   Mapping paired reads (takes a while)\n",
      "  0:52:20.373   745M / 1331M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:52:20.373   745M / 1331M INFO    General                 (paired_info_utils.cpp     : 138)   Left insert size quantile 128, right insert size quantile 310, filtering threshold 2, rounding threshold 0\n",
      "  0:52:20.377   758M / 1331M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:52:24.402   759M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:52:28.334   759M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:52:32.285   759M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:52:36.070   759M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:52:39.875   760M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:52:43.595   760M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:53:02.472   760M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:53:36.796   761M / 1331M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:54:14.514   762M / 1331M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:54:14.520   748M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Distance Estimation (id: distance_estimation)\n",
      "  0:54:14.520   748M / 1331M INFO    General                 (distance_estimation.cpp   :  42)   Processing library #0\n",
      "  0:54:14.523   748M / 1331M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:54:14.525   748M / 1331M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SIMPLE distance estimator\n",
      "  0:54:14.561   749M / 1331M INFO    General                 (distance_estimation_utils.: 131)   Refining clustered pair information\n",
      "  0:54:14.562   749M / 1331M INFO    General                 (distance_estimation_utils.: 133)   The refining of clustered pair information has been finished\n",
      "  0:54:14.562   749M / 1331M INFO    General                 (distance_estimation_utils.: 135)   Improving paired information\n",
      "  0:54:14.623   749M / 1331M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 34; contradictional = 54\n",
      "  0:54:14.673   749M / 1331M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 2; contradictional = 4\n",
      "  0:54:14.674   749M / 1331M INFO    General                 (distance_estimation_utils.:  80)   Filling scaffolding index\n",
      "  0:54:14.675   749M / 1331M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:54:14.675   749M / 1331M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SMOOTHING distance estimator\n",
      "  0:54:14.841   749M / 1331M INFO    General                 (distance_estimation.cpp   :  51)   Clearing raw paired index\n",
      "  0:54:14.844   745M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Repeat Resolving (id: repeat_resolving)\n",
      "  0:54:14.846   745M / 1331M INFO    General                 (repeat_resolving.cpp      :  88)   Using Path-Extend repeat resolving\n",
      "  0:54:14.847   745M / 1331M INFO    General                 (launcher.cpp              : 603)   ExSPAnder repeat resolving tool started\n",
      "  0:54:14.863   746M / 1331M INFO    General                 (launcher.cpp              : 420)   Creating main extenders, unique edge length = 2000\n",
      "  0:54:14.864   746M / 1331M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 116.51\n",
      "  0:54:14.865   746M / 1331M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5104\n",
      "  0:54:14.865   746M / 1331M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 116.51\n",
      "  0:54:14.865   746M / 1331M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5104\n",
      "  0:54:14.881   746M / 1331M INFO    General                 (extenders_logic.cpp       : 550)   Using 1 paired-end library\n",
      "  0:54:14.881   746M / 1331M INFO    General                 (extenders_logic.cpp       : 551)   Using 1 paired-end scaffolding library\n",
      "  0:54:14.882   746M / 1331M INFO    General                 (extenders_logic.cpp       : 552)   Using 0 single read libraries\n",
      "  0:54:14.882   746M / 1331M INFO    General                 (launcher.cpp              : 449)   Total number of extenders is 3\n",
      "  0:54:14.882   746M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 0 paths from 780 (0%)\n",
      "  0:54:14.915   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 79 paths from 780 (10%)\n",
      "  0:54:14.918   747M / 1331M INFO    General                 (path_extenders.cpp        :  34)   Processed 128 paths from 780 (16%)\n",
      "  0:54:14.928   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 158 paths from 780 (20%)\n",
      "  0:54:14.931   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 237 paths from 780 (30%)\n",
      "  0:54:14.932   747M / 1331M INFO    General                 (path_extenders.cpp        :  34)   Processed 256 paths from 780 (32%)\n",
      "  0:54:14.933   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 316 paths from 780 (40%)\n",
      "  0:54:14.933   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 395 paths from 780 (50%)\n",
      "  0:54:14.935   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 474 paths from 780 (60%)\n",
      "  0:54:14.935   747M / 1331M INFO    General                 (path_extenders.cpp        :  34)   Processed 512 paths from 780 (65%)\n",
      "  0:54:14.936   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 553 paths from 780 (70%)\n",
      "  0:54:14.942   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 632 paths from 780 (81%)\n",
      "  0:54:14.946   747M / 1331M INFO    General                 (path_extenders.cpp        :  36)   Processed 711 paths from 780 (91%)\n",
      "  0:54:14.947   747M / 1331M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:54:14.947   747M / 1331M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:54:14.948   747M / 1331M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:54:14.948   747M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:54:14.948   747M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:54:14.949   747M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:54:14.949   747M / 1331M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:54:14.950   747M / 1331M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:54:14.951   747M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:54:14.951   747M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:54:14.952   747M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:54:14.968   747M / 1331M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:54:14.968   747M / 1331M INFO    General                 (launcher.cpp              : 456)   Closing gaps in paths\n",
      "  0:54:14.974   747M / 1331M INFO    General                 (launcher.cpp              : 486)   Gap closing completed\n",
      "  0:54:14.977   745M / 1331M INFO    General                 (launcher.cpp              : 304)   Traversing tandem repeats\n",
      "  0:54:15.001   745M / 1331M INFO    General                 (launcher.cpp              : 314)   Traversed 3 loops\n",
      "  0:54:15.002   745M / 1331M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:54:15.002   745M / 1331M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:54:15.003   745M / 1331M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:54:15.003   745M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:54:15.009   745M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:54:15.010   745M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:54:15.010   745M / 1331M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:54:15.011   745M / 1331M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:54:15.013   745M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:54:15.014   745M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:54:15.015   745M / 1331M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:54:15.028   745M / 1331M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:54:15.031   745M / 1331M INFO    General                 (launcher.cpp              : 666)   ExSPAnder repeat resolving tool finished\n",
      "  0:54:15.032   745M / 1331M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:54:15.032   745M / 1331M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/before_rr.fasta\"\n",
      "  0:54:15.230   745M / 1331M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_with_scaffolds.gfa\"\n",
      "  0:54:15.355   745M / 1331M INFO    General                 (contig_output_stage.cpp   : 169)   Outputting FastG graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph.fastg\"\n",
      "  0:54:15.632   745M / 1331M INFO    General                 (contig_output_stage.cpp   : 200)   Breaking scaffolds\n",
      "  0:54:15.650   752M / 1331M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.fasta\n",
      "  0:54:15.755   752M / 1331M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.paths\n",
      "  0:54:15.774   752M / 1331M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.fasta\n",
      "  0:54:15.904   752M / 1331M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.paths\n",
      "  0:54:15.908   752M / 1331M INFO    General                 (contig_output_stage.cpp   : 114)   Populating GFA with scaffold paths\n",
      "  0:54:15.913   745M / 1331M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:54:16.380     1M / 1331M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 54 minutes 16 seconds\n",
      "\n",
      "===== K77 finished. \n",
      "\n",
      "\n",
      "===== Copy files started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/copy_files.py /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/before_rr.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/before_rr.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/strain_graph.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/strain_graph.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph.fastg /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph.fastg /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.paths\n",
      "\n",
      "\n",
      "===== Copy files finished. \n",
      "\n",
      "\n",
      "===== Assembling finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.fasta --misc_dir /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Assembled contigs are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.fasta\n",
      " * Assembled scaffolds are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.fasta\n",
      " * Paths in the assembly graph corresponding to the contigs are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.paths\n",
      " * Assembly graph is in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph_with_scaffolds.gfa\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0253G_SPAdes/spades.log\n",
      "\n",
      "Thank you for using SPAdes! If you use it in your research, please cite:\n",
      "\n",
      "  Prjibelski, A., Antipov, D., Meleshko, D., Lapidus, A. and Korobeynikov, A., 2020. Using SPAdes de novo assembler. Current protocols in bioinformatics, 70(1), p.e102.\n",
      "  doi.org/10.1002/cpbi.102\n",
      "\n",
      "Command line: /home/student/miniconda3/envs/VariantCalling/bin/spades.py\t-o\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes\t-1\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PAO1wt12P_R1.fastq.gz\t-2\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PAO1wt12P_R2.fastq.gz\t--isolate\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 4.1.0\n",
      "  Python version: 3.11.0\n",
      "  OS: Linux-5.15.0-139-generic-x86_64-with-glibc2.31\n",
      "\n",
      "Output dir: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes\n",
      "Mode: ONLY assembling (without read error correction)\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Isolate mode\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PAO1wt12P_R1.fastq.gz']\n",
      "      right reads: ['/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PAO1wt12P_R2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "  Assembly graph output will use GFA v1.2 format\n",
      "Other parameters:\n",
      "  Dir for temp files: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 10\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/spades.log\n",
      "\n",
      "/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PAO1wt12P_R1.fastq.gz: max reads length: 151\n",
      "/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PAO1wt12P_R2.fastq.gz: max reads length: 151\n",
      "\n",
      "Reads length: 151\n",
      "\n",
      "Default k-mer sizes were set to [21, 33, 55, 77] because estimated read length (151) is equal to or greater than 150\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Assembling started. \n",
      "\n",
      "\n",
      "===== K21 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=21\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  78)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  99)   Converting paired reads\n",
      "  0:00:00.196    34M / 40M   INFO    General                 (binary_converter.cpp      : 127)   16384 reads processed\n",
      "  0:00:00.323    39M / 45M   INFO    General                 (binary_converter.cpp      : 127)   32768 reads processed\n",
      "  0:00:00.809    49M / 55M   INFO    General                 (binary_converter.cpp      : 127)   65536 reads processed\n",
      "  0:00:01.571    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   131072 reads processed\n",
      "  0:00:03.372    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   262144 reads processed\n",
      "  0:00:06.702    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   524288 reads processed\n",
      "  0:00:12.883    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   1048576 reads processed\n",
      "  0:00:29.050    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   2097152 reads processed\n",
      "  0:00:44.758    59M / 65M   INFO    General                 (binary_converter.cpp      : 143)   3281626 reads written\n",
      "  0:00:44.765     2M / 65M   INFO    General                 (read_converter.cpp        : 113)   Converting single reads\n",
      "  0:00:44.769    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:00:44.771     2M / 65M   INFO    General                 (read_converter.cpp        : 119)   Converting merged reads\n",
      "  0:00:44.773    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:00:44.774     1M / 65M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:44.775     1M / 65M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:44.776     1M / 65M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:44.776     1M / 65M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:44.776     1M / 65M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:44.776     1M / 65M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:44.777     1M / 65M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:44.777     1M / 65M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:00:55.092   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1070676 reads\n",
      "  0:01:05.344   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2142854 reads\n",
      "  0:01:15.039   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3215498 reads\n",
      "  0:01:24.643   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4287837 reads\n",
      "  0:01:34.698   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5360694 reads\n",
      "  0:01:44.452   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6432611 reads\n",
      "  0:01:54.139   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 7505758 reads\n",
      "  0:02:03.889   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8576863 reads\n",
      "  0:02:14.018   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9649276 reads\n",
      "  0:02:46.289     1M / 628M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13126504 reads\n",
      "  0:02:46.290     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:02:59.254     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 48436930 kmers in total.\n",
      "  0:02:59.255     1M / 628M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:02:59.257     1M / 628M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:02:59.257     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:02:59.258     1M / 628M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:02:59.258     1M / 628M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:02:59.258     1M / 628M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:03:18.605   601M / 851M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 32076319 kmers\n",
      "  0:03:25.004   601M / 928M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 48436934 kmers\n",
      "  0:03:25.004   601M / 928M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 48436934 kmers.\n",
      "  0:03:25.042     1M / 928M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:29.731     1M / 928M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 47505179 kmers in total.\n",
      "  0:03:29.732     1M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:03:38.149    37M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 47505179 kmers, 34318528 bytes occupied (5.77933 bits per kmer).\n",
      "  0:03:38.150    37M / 928M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:03:41.156    85M / 928M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:04:21.012    85M / 928M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:04:21.036    85M / 928M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:04:21.037    85M / 928M  INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:04:21.037    85M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:05:07.954   125M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 589429\n",
      "  0:05:08.514   125M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 596916\n",
      "  0:05:08.531    85M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  50)   12821123 22-mers were removed by early tip clipper\n",
      "  0:05:08.533    85M / 928M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:05:08.535    85M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:05:54.792   187M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2712000 sequences extracted\n",
      "  0:06:08.236   187M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:06:14.722   187M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 1 loops collected\n",
      "  0:06:14.823   187M / 928M  INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:06:16.128   187M / 928M  INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:06:16.129   187M / 928M  INFO    General                 (debruijn_graph_constructor: 516)   Total 5424002 edges to create\n",
      "  0:06:16.129   395M / 928M  INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:06:18.401   479M / 928M  INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:06:20.152   479M / 928M  INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:06:20.172   495M / 928M  INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:06:20.706   495M / 928M  INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:06:20.707   495M / 928M  INFO    General                 (debruijn_graph_constructor: 542)   Total 1780250 vertices to create\n",
      "  0:06:20.707   660M / 928M  INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:06:21.740   516M / 928M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:06:21.741   516M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:06:28.785   552M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 48436930 kmers, 34991576 bytes occupied (5.77932 bits per kmer).\n",
      "  0:06:28.862   740M / 928M  INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:10:47.156   740M / 928M  INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:10:52.548   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:10:58.384   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:11:08.995   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:11:22.797   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 3989215 edges\n",
      "  0:11:31.294   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 4889215 edges\n",
      "  0:11:36.292   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 5423789 edges\n",
      "  0:11:36.614   433M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:11:36.619   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 15\n",
      "  0:11:36.620   433M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 119\n",
      "  0:11:36.625   433M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 124. Coverage mad: 31.1346\n",
      "  0:11:36.627   433M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:11:37.161   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:11:38.064   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:11:40.072   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:11:46.639   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:11:47.549   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 123.373. Fitted coverage std. dev: 32.3346\n",
      "  0:11:47.566   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0888995\n",
      "  0:11:47.566   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 18\n",
      "  0:11:47.571   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 18\n",
      "  0:11:47.572   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6171935\n",
      "  0:11:47.573   433M / 928M  INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 123.373\n",
      "  0:11:47.573   433M / 928M  INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 18\n",
      "  0:11:47.574   433M / 928M  INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:11:47.574   433M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:11:47.575   433M / 928M  INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:11:47.575   433M / 928M  INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:11:47.575   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:11:48.619   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:11:48.620   433M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:11:48.620   433M / 928M  INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:11:48.621   433M / 928M  INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:11:48.621   433M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:11:48.621   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:50.896   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 9936 times\n",
      "  0:11:50.900   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:31.765   466M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 810733 times\n",
      "  0:12:31.773   466M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:33.960   465M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 120365 times\n",
      "  0:12:33.975   465M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:12:33.976   465M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:34.143   463M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 8015 times\n",
      "  0:12:34.145   463M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.033   457M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 14404 times\n",
      "  0:12:35.037   457M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.205   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 9153 times\n",
      "  0:12:35.208   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:12:35.208   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.217   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 183 times\n",
      "  0:12:35.218   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.372   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1081 times\n",
      "  0:12:35.373   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.393   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1138 times\n",
      "  0:12:35.394   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:12:35.394   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.396   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 23 times\n",
      "  0:12:35.396   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.426   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 160 times\n",
      "  0:12:35.427   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.432   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 418 times\n",
      "  0:12:35.433   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:12:35.433   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.434   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 8 times\n",
      "  0:12:35.434   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.451   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 58 times\n",
      "  0:12:35.452   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.455   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 230 times\n",
      "  0:12:35.455   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:12:35.455   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.456   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:12:35.456   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.464   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 20 times\n",
      "  0:12:35.464   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.466   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 111 times\n",
      "  0:12:35.466   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:12:35.467   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.467   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:12:35.467   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.474   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 14 times\n",
      "  0:12:35.474   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.475   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 64 times\n",
      "  0:12:35.476   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:12:35.476   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.476   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:12:35.476   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.478   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:12:35.478   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.479   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 55 times\n",
      "  0:12:35.479   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:12:35.480   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.480   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:12:35.480   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.486   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 7 times\n",
      "  0:12:35.486   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.487   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 31 times\n",
      "  0:12:35.487   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:12:35.488   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.488   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:12:35.488   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.489   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:12:35.490   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.490   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:12:35.491   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:12:35.491   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.509   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 17 times\n",
      "  0:12:35.510   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.548   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 12 times\n",
      "  0:12:35.549   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.567   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:12:35.567   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:12:35.567   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.569   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:12:35.569   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.569   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:12:35.569   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:12:35.569   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:12:35.602   378M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:12:35.603   378M / 928M  INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:12:35.603   378M / 928M  INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:12:35.604   378M / 928M  INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:12:35.604   378M / 928M  INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:12:35.604   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.642   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:12:35.643   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.701   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:12:35.702   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:12:35.720   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:12:35.720   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:12:35.751   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:12:35.751   378M / 928M  INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:12:35.768   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:12:35.843   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 30034 times\n",
      "  0:12:35.868   377M / 928M  INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:12:35.868   377M / 928M  INFO    General                 (simplification.cpp        : 509)     Average coverage = 128.483\n",
      "  0:12:35.881   377M / 928M  INFO    General                 (simplification.cpp        : 510)     Total length = 6215872\n",
      "  0:12:35.904   377M / 928M  INFO    General                 (simplification.cpp        : 511)     Median edge length: 2364\n",
      "  0:12:35.905   377M / 928M  INFO    General                 (simplification.cpp        : 512)     Edges: 20281\n",
      "  0:12:35.905   377M / 928M  INFO    General                 (simplification.cpp        : 513)     Vertices: 12924\n",
      "  0:12:35.905   377M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:12:35.908   377M / 928M  INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K21/simplified_contigs\"\n",
      "  0:12:35.956   380M / 928M  INFO    General                 (binary_converter.cpp      : 143)   10159 reads written\n",
      "  0:12:35.960   377M / 928M  INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:12:36.005     1M / 928M  INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 12 minutes 36 seconds\n",
      "\n",
      "===== K21 finished. \n",
      "\n",
      "\n",
      "===== K33 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=33\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K21/simplified_contigs\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:06.488   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 592528 reads\n",
      "  0:00:12.814   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1184368 reads\n",
      "  0:00:19.385   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1777339 reads\n",
      "  0:00:25.587   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2370360 reads\n",
      "  0:00:32.057   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2964046 reads\n",
      "  0:00:38.800   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3556781 reads\n",
      "  0:00:45.079   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4149044 reads\n",
      "  0:00:51.394   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4742126 reads\n",
      "  0:01:34.656   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8892369 reads\n",
      "  0:02:21.929     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13146822 reads\n",
      "  0:02:21.929     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:59.411     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 56063755 kmers in total.\n",
      "  0:03:59.412     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:03:59.445     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:03:59.446     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:03:59.446     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:03:59.447     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:03:59.447     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:04:06.781   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 12497616 kmers\n",
      "  0:04:15.718   601M / 616M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 24991680 kmers\n",
      "  0:04:25.680   601M / 708M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 37496564 kmers\n",
      "  0:04:33.884   601M / 725M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 49952961 kmers\n",
      "  0:04:42.887   601M / 735M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 56063775 kmers\n",
      "  0:04:42.890   601M / 735M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 56063775 kmers.\n",
      "  0:04:42.927     1M / 735M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:05:04.250     1M / 735M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 55622538 kmers in total.\n",
      "  0:05:04.251     1M / 735M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:05:16.546    41M / 735M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 55622538 kmers, 40181504 bytes occupied (5.77917 bits per kmer).\n",
      "  0:05:16.547    41M / 735M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:05:23.195    97M / 735M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:06:39.529    97M / 735M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:06:39.538    97M / 735M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:06:39.539    97M / 735M  INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:06:39.540    97M / 735M  INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:07:33.388   177M / 735M  INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 686919\n",
      "  0:07:34.281   177M / 735M  INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 698361\n",
      "  0:07:34.290    97M / 735M  INFO   Early tip clipping       (early_simplification.hpp  :  50)   19565956 34-mers were removed by early tip clipper\n",
      "  0:07:34.294    97M / 735M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:07:34.296    97M / 735M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:08:30.088   172M / 1055M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1608188 sequences extracted\n",
      "  0:08:50.237   172M / 1080M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:09:04.311   172M / 1080M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:09:04.376   172M / 1080M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:09:05.722   172M / 1080M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:09:05.726   172M / 1080M INFO    General                 (debruijn_graph_constructor: 516)   Total 3216376 edges to create\n",
      "  0:09:05.728   297M / 1080M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:09:08.053   349M / 1080M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:09:09.413   349M / 1080M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:09:09.424   365M / 1080M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:09:10.014   365M / 1080M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:09:10.015   365M / 1080M INFO    General                 (debruijn_graph_constructor: 542)   Total 1166971 vertices to create\n",
      "  0:09:10.016   473M / 1080M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:09:10.697   380M / 1080M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:09:10.698   380M / 1080M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:09:22.692   420M / 1080M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 56063755 kmers, 40500264 bytes occupied (5.77917 bits per kmer).\n",
      "  0:09:22.849   636M / 1080M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:14:40.959   636M / 1080M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:14:53.852   636M / 1080M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:15:04.483   636M / 1080M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1916697 edges\n",
      "  0:15:13.675   636M / 1080M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2748032 edges\n",
      "  0:15:18.579   636M / 1080M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3216358 edges\n",
      "  0:15:19.176   285M / 1080M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:15:19.189   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 13\n",
      "  0:15:19.190   285M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 105\n",
      "  0:15:19.191   285M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 110. Coverage mad: 29.652\n",
      "  0:15:19.191   285M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:15:19.354   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:15:20.276   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:15:21.556   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:15:25.039   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:15:26.067   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 109.759. Fitted coverage std. dev: 29.6043\n",
      "  0:15:26.072   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0689098\n",
      "  0:15:26.072   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 15\n",
      "  0:15:26.073   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 15\n",
      "  0:15:26.073   286M / 1080M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6183950\n",
      "  0:15:26.074   285M / 1080M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 109.759\n",
      "  0:15:26.074   285M / 1080M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 15\n",
      "  0:15:26.075   285M / 1080M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:15:26.075   285M / 1080M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:15:26.077   285M / 1080M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:15:26.082   285M / 1080M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:15:26.083   285M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:15:26.451   285M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 2 times\n",
      "  0:15:26.452   285M / 1080M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:15:26.453   285M / 1080M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:15:26.453   285M / 1080M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:15:26.454   285M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:15:26.454   285M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:27.311   285M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 18250 times\n",
      "  0:15:27.312   285M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:49.857   298M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 479712 times\n",
      "  0:15:49.858   298M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:50.664   297M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 28417 times\n",
      "  0:15:50.671   297M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:15:50.671   297M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:50.828   296M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 7377 times\n",
      "  0:15:50.831   296M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.185   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 5931 times\n",
      "  0:15:51.188   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.338   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3084 times\n",
      "  0:15:51.340   294M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:15:51.340   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.356   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 189 times\n",
      "  0:15:51.357   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.426   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 349 times\n",
      "  0:15:51.428   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.491   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 296 times\n",
      "  0:15:51.492   294M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:15:51.492   294M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.495   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 17 times\n",
      "  0:15:51.495   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.523   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 45 times\n",
      "  0:15:51.526   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.539   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 95 times\n",
      "  0:15:51.539   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:15:51.540   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.540   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:15:51.540   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.547   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 16 times\n",
      "  0:15:51.547   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.552   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:15:51.552   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:15:51.552   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.553   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:15:51.553   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.556   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:15:51.557   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.559   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:15:51.559   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:15:51.559   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.560   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:51.560   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.560   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:15:51.561   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.564   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:15:51.564   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:15:51.564   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.565   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:15:51.565   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.565   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:15:51.565   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.566   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6 times\n",
      "  0:15:51.566   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:15:51.566   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.567   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:51.567   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.567   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:51.567   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.568   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 10 times\n",
      "  0:15:51.569   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:15:51.569   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.569   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:15:51.569   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.570   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:51.570   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.571   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8 times\n",
      "  0:15:51.571   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:15:51.571   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.592   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 8 times\n",
      "  0:15:51.592   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.622   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:15:51.622   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.660   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:51.667   293M / 1080M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:15:51.667   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.667   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:51.668   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.668   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:51.668   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:51.668   293M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:51.692   239M / 1080M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:15:51.704   239M / 1080M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:15:51.705   239M / 1080M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:15:51.706   239M / 1080M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:15:51.707   239M / 1080M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:15:51.707   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.744   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:51.745   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.791   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:51.791   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:51.822   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:51.823   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:51.844   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:51.845   239M / 1080M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:15:51.877   239M / 1080M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:15:52.058   238M / 1080M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 77310 times\n",
      "  0:15:52.065   236M / 1080M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:15:52.065   236M / 1080M INFO    General                 (simplification.cpp        : 509)     Average coverage = 113.738\n",
      "  0:15:52.070   236M / 1080M INFO    General                 (simplification.cpp        : 510)     Total length = 6235779\n",
      "  0:15:52.077   236M / 1080M INFO    General                 (simplification.cpp        : 511)     Median edge length: 29936\n",
      "  0:15:52.077   236M / 1080M INFO    General                 (simplification.cpp        : 512)     Edges: 4221\n",
      "  0:15:52.083   236M / 1080M INFO    General                 (simplification.cpp        : 513)     Vertices: 3240\n",
      "  0:15:52.084   236M / 1080M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:15:52.376   236M / 1080M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K33/simplified_contigs\"\n",
      "  0:15:52.428   240M / 1080M INFO    General                 (binary_converter.cpp      : 143)   2112 reads written\n",
      "  0:15:52.432   236M / 1080M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:15:52.467     1M / 1080M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 15 minutes 52 seconds\n",
      "\n",
      "===== K33 finished. \n",
      "\n",
      "\n",
      "===== K55 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/config.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/isolate_mode.info\"\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=55\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.004     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.004     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.005     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.006     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K33/simplified_contigs\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.011     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:00.011     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.012     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.012     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.017     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.017     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:06.720   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 732839 reads\n",
      "  0:00:13.788   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1466116 reads\n",
      "  0:00:20.575   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2200621 reads\n",
      "  0:00:27.074   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2934328 reads\n",
      "  0:00:33.687   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3667680 reads\n",
      "  0:00:40.064   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4401944 reads\n",
      "  0:01:05.387   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5137066 reads\n",
      "  0:01:12.792   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5870399 reads\n",
      "  0:01:38.576   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8806720 reads\n",
      "  0:02:19.049     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13130728 reads\n",
      "  0:02:19.050     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:42.827     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 59773652 kmers in total.\n",
      "  0:03:42.829     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:03:43.006     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:03:43.008     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:03:43.011     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:03:43.012     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:03:43.014     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:03:51.945   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 12211803 kmers\n",
      "  0:04:01.025   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 24434436 kmers\n",
      "  0:04:10.766   601M / 692M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 36636966 kmers\n",
      "  0:04:20.280   601M / 782M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 48837641 kmers\n",
      "  0:04:32.854   601M / 847M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 59773672 kmers\n",
      "  0:04:32.855   601M / 847M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 59773672 kmers.\n",
      "  0:04:32.916     1M / 847M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:05:20.184     1M / 847M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 59807304 kmers in total.\n",
      "  0:05:20.185     1M / 847M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:05:33.511    44M / 847M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 59807304 kmers, 43203896 bytes occupied (5.77908 bits per kmer).\n",
      "  0:05:33.511    44M / 847M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:05:40.307   104M / 847M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:06:59.510   104M / 847M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:06:59.523   104M / 847M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:06:59.526   104M / 847M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:08:29.507   215M / 1167M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2141534 sequences extracted\n",
      "  0:09:01.922   215M / 1200M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:09:17.589   215M / 1200M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:09:17.735   215M / 1200M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:09:19.614   215M / 1200M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:09:19.616   215M / 1200M INFO    General                 (debruijn_graph_constructor: 516)   Total 4283068 edges to create\n",
      "  0:09:19.619   383M / 1200M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:09:22.983   451M / 1200M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:09:24.895   451M / 1200M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:09:24.943   483M / 1200M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:09:25.950   483M / 1200M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:09:25.951   483M / 1200M INFO    General                 (debruijn_graph_constructor: 542)   Total 2175186 vertices to create\n",
      "  0:09:25.951   688M / 1200M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:09:27.230   552M / 1200M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:09:27.233   552M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:09:39.244   596M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 59773652 kmers, 43179632 bytes occupied (5.77909 bits per kmer).\n",
      "  0:09:39.373   828M / 1200M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:14:17.114   828M / 1200M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:14:34.162   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:14:51.742   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:15:00.541   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2682776 edges\n",
      "  0:15:06.593   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3248828 edges\n",
      "  0:15:12.020   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3748828 edges\n",
      "  0:15:16.753   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4190744 edges\n",
      "  0:15:17.840   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4283068 edges\n",
      "  0:15:18.286   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:15:18.290   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 10\n",
      "  0:15:18.290   449M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 84\n",
      "  0:15:18.291   449M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 87. Coverage mad: 25.2042\n",
      "  0:15:18.291   449M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:15:18.421   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:15:19.201   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:15:20.218   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:15:23.009   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:15:23.407   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 86.4573. Fitted coverage std. dev: 24.6508\n",
      "  0:15:23.415   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0955892\n",
      "  0:15:23.415   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 13\n",
      "  0:15:23.415   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 13\n",
      "  0:15:23.416   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6179273\n",
      "  0:15:23.427   449M / 1200M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 86.4573\n",
      "  0:15:23.428   449M / 1200M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 13\n",
      "  0:15:23.428   449M / 1200M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:15:23.428   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:15:23.443   462M / 1200M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:15:24.621   498M / 1200M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1173697 out of 4283068, length: 62361050\n",
      "  0:15:24.795   513M / 1200M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 4304486)\n",
      "  0:15:27.265   513M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:16:11.368   545M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 44394808 kmers, 32065504 bytes occupied (5.77824 bits per kmer).\n",
      "  0:16:11.582   885M / 1200M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:16:39.625   870M / 1200M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:17:07.919   871M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:17:36.335   871M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:18:05.080   872M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:18:33.025   873M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:19:00.780   873M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:19:28.818   874M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:21:46.153   877M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:24:13.744   881M / 1200M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:24:14.136   525M / 1200M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:24:14.155   525M / 1200M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:24:15.539   525M / 1200M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 3555 tips collected, total 3760 connection candidates\n",
      "  0:24:15.566   525M / 1200M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 84 gaps after checking 3668 candidates\n",
      "  0:24:15.862   525M / 1200M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:24:15.969   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:24:15.984   449M / 1200M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:24:15.986   449M / 1200M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:24:15.986   449M / 1200M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:24:16.478   449M / 1200M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:24:16.479   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:24:16.480   449M / 1200M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:24:16.480   449M / 1200M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:24:16.480   449M / 1200M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:24:16.481   449M / 1200M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:24:23.943   433M / 1200M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 779125 times\n",
      "  0:24:23.944   433M / 1200M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:07.231  1231M / 1264M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 149716 times\n",
      "  0:25:07.233  1231M / 1264M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:07.407  1231M / 1264M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 4463 times\n",
      "  0:25:07.410  1231M / 1264M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:25:07.410  1231M / 1264M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:07.585  1230M / 1265M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 3604 times\n",
      "  0:25:07.598  1230M / 1265M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.152  1234M / 1268M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 757 times\n",
      "  0:25:08.155  1234M / 1268M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.286  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 965 times\n",
      "  0:25:08.287  1234M / 1269M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:25:08.288  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.301  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 70 times\n",
      "  0:25:08.302  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.326  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 40 times\n",
      "  0:25:08.328  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.353  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 67 times\n",
      "  0:25:08.354  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:25:08.355  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.362  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:08.363  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.378  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 4 times\n",
      "  0:25:08.380  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.405  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:25:08.409  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:25:08.410  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.412  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:08.413  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.414  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:25:08.422  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.425  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:25:08.426  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:25:08.427  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.427  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:25:08.428  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.433  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:25:08.437  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.439  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:25:08.439  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:25:08.441  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.441  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:08.442  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.442  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:08.443  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.445  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:08.445  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:25:08.446  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.450  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:08.451  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.452  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:08.452  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.454  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:25:08.457  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:25:08.457  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.458  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:08.458  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.462  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:08.462  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.463  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:08.463  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:25:08.464  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.465  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:08.465  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.473  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:08.474  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.475  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:25:08.476  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:25:08.478  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.558  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 10 times\n",
      "  0:25:08.559  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.653  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:08.655  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.736  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:08.737  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:25:08.738  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:08.739  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:08.739  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:08.740  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:08.741  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:08.741  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:08.767  1208M / 1270M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:25:08.777  1221M / 1270M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:25:08.892  1230M / 1283M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 379843 out of 381554, length: 38553946\n",
      "  0:25:08.949  1233M / 1284M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 4304486)\n",
      "  0:25:09.906  1233M / 1284M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:25:24.577  1247M / 1317M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 20240601 kmers, 14619792 bytes occupied (5.7784 bits per kmer).\n",
      "  0:25:24.683  1403M / 1471M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:25:34.416  1400M / 1471M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:26:10.322  1401M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:26:45.791  1401M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:27:21.140  1401M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:27:56.370  1402M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:28:32.089  1402M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:29:07.242  1402M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:32:00.762  1404M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:35:11.147  1405M / 1471M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:35:11.249  1240M / 1471M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:35:11.252  1240M / 1471M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:35:11.339  1240M / 1471M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 168 tips collected, total 313 connection candidates\n",
      "  0:35:11.341  1240M / 1471M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 313 candidates\n",
      "  0:35:11.401  1240M / 1471M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:35:11.423  1208M / 1471M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:35:11.429  1208M / 1471M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:35:11.430  1208M / 1471M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:35:11.431  1208M / 1471M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:35:11.437  1208M / 1471M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:35:11.438  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:35:11.517  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:35:11.522  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:35:11.647  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:35:11.648  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:35:11.726  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:35:11.730  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:35:11.824  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:35:11.826  1208M / 1471M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:35:11.939  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:35:12.380  1204M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 189682 times\n",
      "  0:35:12.398  1199M / 1471M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:35:12.399  1199M / 1471M INFO    General                 (simplification.cpp        : 509)     Average coverage = 88.2127\n",
      "  0:35:12.410  1199M / 1471M INFO    General                 (simplification.cpp        : 510)     Total length = 6251864\n",
      "  0:35:12.427  1199M / 1471M INFO    General                 (simplification.cpp        : 511)     Median edge length: 76224\n",
      "  0:35:12.428  1199M / 1471M INFO    General                 (simplification.cpp        : 512)     Edges: 2190\n",
      "  0:35:12.429  1199M / 1471M INFO    General                 (simplification.cpp        : 513)     Vertices: 2090\n",
      "  0:35:12.430  1199M / 1471M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:35:12.718  1199M / 1471M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K55/simplified_contigs\"\n",
      "  0:35:12.753  1203M / 1471M INFO    General                 (binary_converter.cpp      : 143)   1095 reads written\n",
      "  0:35:12.758  1199M / 1471M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:35:16.486     1M / 1471M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 35 minutes 16 seconds\n",
      "\n",
      "===== K55 finished. \n",
      "\n",
      "\n",
      "===== K77 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/config.info\"\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/isolate_mode.info\"\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=77\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.011     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.011     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.011     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.012     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.012     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.012     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.024     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.025     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.025     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K55/simplified_contigs\n",
      "  0:00:00.031     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.031     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:00.031     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.031     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.032     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.032     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.032     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:00:06.089   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 641445 reads\n",
      "  0:00:11.759   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1283786 reads\n",
      "  0:00:17.993   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1927256 reads\n",
      "  0:00:24.174   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2570681 reads\n",
      "  0:00:30.375   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3215168 reads\n",
      "  0:00:36.239   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3857538 reads\n",
      "  0:00:42.012   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4501434 reads\n",
      "  0:00:47.930   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5144971 reads\n",
      "  0:01:22.916   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9003256 reads\n",
      "  0:02:03.001     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13128694 reads\n",
      "  0:02:03.002     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:05:24.384     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 55473231 kmers in total.\n",
      "  0:05:24.385     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:05:24.388     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:05:24.388     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:05:24.389     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:05:24.389     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:05:24.390     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:05:30.288   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 6508312 kmers\n",
      "  0:05:46.495   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 14004744 kmers\n",
      "  0:05:55.685   601M / 705M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 21305028 kmers\n",
      "  0:06:12.128   601M / 808M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 27792127 kmers\n",
      "  0:06:18.380   601M / 899M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 34360547 kmers\n",
      "  0:06:26.249   601M / 969M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 41990532 kmers\n",
      "  0:06:39.538   601M / 1057M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 49175349 kmers\n",
      "  0:06:48.851   601M / 1130M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 55473269 kmers\n",
      "  0:06:48.851   601M / 1130M INFO    General                 (kmer_splitters.hpp        : 202)   Used 55473269 kmers.\n",
      "  0:06:48.899     1M / 1130M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:08:33.327     1M / 1130M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 55798725 kmers in total.\n",
      "  0:08:33.328     1M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:09:04.786    41M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 55798725 kmers, 40308768 bytes occupied (5.77917 bits per kmer).\n",
      "  0:09:04.787    41M / 1130M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:09:16.626    97M / 1130M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:10:56.622    97M / 1130M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:10:56.638    97M / 1130M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:10:56.644    97M / 1130M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:12:30.760   191M / 1130M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1748112 sequences extracted\n",
      "  0:13:02.081   191M / 1130M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:13:16.091   191M / 1130M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:13:16.169   191M / 1130M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:13:17.506   191M / 1130M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:13:17.507   191M / 1130M INFO    General                 (debruijn_graph_constructor: 516)   Total 3496224 edges to create\n",
      "  0:13:17.511   327M / 1130M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:13:20.621   383M / 1130M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:13:22.150   383M / 1130M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:13:22.208   399M / 1130M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:13:23.229   399M / 1130M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:13:23.230   399M / 1130M INFO    General                 (debruijn_graph_constructor: 542)   Total 2073606 vertices to create\n",
      "  0:13:23.231   592M / 1130M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:13:24.625   493M / 1130M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:13:24.630   493M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:13:37.024   533M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 55473231 kmers, 40073624 bytes occupied (5.77917 bits per kmer).\n",
      "  0:13:37.162   745M / 1130M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:17:21.162   745M / 1130M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:17:39.624   745M / 1130M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:17:58.469   745M / 1130M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:18:05.310   745M / 1130M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2528789 edges\n",
      "  0:18:09.651   745M / 1130M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2928789 edges\n",
      "  0:18:13.785   745M / 1130M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3306614 edges\n",
      "  0:18:15.997   745M / 1130M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3496224 edges\n",
      "  0:18:16.936   397M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:18:16.941   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 7\n",
      "  0:18:16.942   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 61\n",
      "  0:18:16.942   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 64. Coverage mad: 20.7564\n",
      "  0:18:16.942   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:18:17.163   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:18:18.023   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:18:19.094   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:18:21.423   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:18:22.236   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 64.4987. Fitted coverage std. dev: 19.6675\n",
      "  0:18:22.240   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.253612\n",
      "  0:18:22.240   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 11\n",
      "  0:18:22.240   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 11\n",
      "  0:18:22.241   397M / 1130M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6167580\n",
      "  0:18:22.241   397M / 1130M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 64.4987\n",
      "  0:18:22.241   397M / 1130M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 11\n",
      "  0:18:22.241   397M / 1130M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:18:22.242   397M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:18:22.247   410M / 1130M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:18:23.195   446M / 1130M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1393903 out of 3496224, length: 72584154\n",
      "  0:18:23.390   463M / 1130M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 3513708)\n",
      "  0:18:26.259   463M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:19:16.913   498M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 48309169 kmers, 34892696 bytes occupied (5.77823 bits per kmer).\n",
      "  0:19:17.170   870M / 1130M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:19:47.390   853M / 1130M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:20:09.120   854M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:20:30.875   855M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:20:52.588   856M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:21:14.210   856M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:21:35.773   857M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:21:57.546   858M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:23:34.064   862M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:25:12.226   884M / 1130M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:25:12.577   487M / 1130M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:25:12.588   487M / 1130M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:25:13.627   487M / 1130M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 3527 tips collected, total 3718 connection candidates\n",
      "  0:25:13.645   487M / 1130M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 166 gaps after checking 3526 candidates\n",
      "  0:25:13.745   487M / 1130M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:25:13.789   397M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:25:13.798   397M / 1130M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:25:13.799   397M / 1130M INFO    General                 (simplification.cpp        :  63)   Most init cleaning disabled on main iteration\n",
      "  0:25:13.800   397M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:25:14.113   397M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:25:14.114   397M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:25:14.114   397M / 1130M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:25:14.114   397M / 1130M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:25:14.115   397M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:25:14.115   397M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:19.557   365M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 713838 times\n",
      "  0:25:19.559   365M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.563   448M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 7702 times\n",
      "  0:25:21.565   448M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.665   448M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 442 times\n",
      "  0:25:21.666   448M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:25:21.667   448M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.690   448M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 351 times\n",
      "  0:25:21.691   448M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.704   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 28 times\n",
      "  0:25:21.705   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.838   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 341 times\n",
      "  0:25:21.841   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:25:21.841   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.843   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 12 times\n",
      "  0:25:21.850   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.857   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 7 times\n",
      "  0:25:21.857   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.873   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 22 times\n",
      "  0:25:21.881   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:25:21.882   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.883   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:25:21.883   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.883   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:21.884   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.885   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 5 times\n",
      "  0:25:21.890   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:25:21.890   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.890   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:21.890   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.891   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:25:21.891   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.892   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:25:21.893   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:25:21.893   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.893   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:21.893   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.894   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:21.894   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.895   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:25:21.895   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:25:21.896   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.896   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:21.896   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.896   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:21.898   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.898   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:21.898   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:25:21.898   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.899   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:21.899   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.902   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:21.903   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.903   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:25:21.903   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:25:21.903   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.904   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:21.904   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.904   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:21.904   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.904   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:21.905   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:25:21.905   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:21.905   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:21.905   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:21.906   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:21.906   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:21.906   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:21.906   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:25:21.906   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:22.042   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 19 times\n",
      "  0:25:22.043   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:22.170   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:22.171   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:22.267   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:22.274   449M / 1130M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:25:22.274   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:22.275   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:25:22.275   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:25:22.275   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:25:22.276   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:25:22.281   449M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:25:22.352   431M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:25:22.366   444M / 1130M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:25:22.630   461M / 1130M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 668151 out of 669302, length: 51836680\n",
      "  0:25:22.703   467M / 1130M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 3513708)\n",
      "  0:25:23.973   467M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:25:43.721   486M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 27152757 kmers, 19613528 bytes occupied (5.77872 bits per kmer).\n",
      "  0:25:43.831   694M / 1130M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:25:57.005   688M / 1130M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:26:13.815   689M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:26:30.526   689M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:26:47.124   690M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:27:04.048   691M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:27:20.679   691M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:27:37.173   692M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:28:59.471   695M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:30:28.820   698M / 1130M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:30:28.979   482M / 1130M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:30:28.979   482M / 1130M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:30:29.195   482M / 1130M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 374 tips collected, total 699 connection candidates\n",
      "  0:30:29.205   482M / 1130M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 699 candidates\n",
      "  0:30:29.276   482M / 1130M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:30:29.309   431M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:30:29.311   431M / 1130M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:30:29.311   431M / 1130M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:30:29.312   431M / 1130M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:30:29.312   431M / 1130M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:30:29.312   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:30:29.431   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:30:29.432   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:30:29.545   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:30:29.554   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:30:29.626   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:30:29.627   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:30:29.747   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:30:29.747   431M / 1130M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:30:29.868   431M / 1130M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:30:30.471   425M / 1130M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 333943 times\n",
      "  0:30:30.478   416M / 1130M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:30:30.478   416M / 1130M INFO    General                 (simplification.cpp        : 509)     Average coverage = 64.978\n",
      "  0:30:30.483   416M / 1130M INFO    General                 (simplification.cpp        : 510)     Total length = 6237203\n",
      "  0:30:30.492   416M / 1130M INFO    General                 (simplification.cpp        : 511)     Median edge length: 94276\n",
      "  0:30:30.493   416M / 1130M INFO    General                 (simplification.cpp        : 512)     Edges: 1416\n",
      "  0:30:30.493   416M / 1130M INFO    General                 (simplification.cpp        : 513)     Vertices: 1292\n",
      "  0:30:30.493   416M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Mismatch Correction (id: mismatch_correction)\n",
      "  0:30:30.494   416M / 1130M INFO    General                 (graph_pack_helpers.cpp    :  44)   Index refill\n",
      "  0:30:30.494   416M / 1130M INFO    General                 (edge_index.hpp            : 175)   Using small index (max_id = 3513708)\n",
      "  0:30:30.874   416M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:30:33.198   420M / 1130M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 6237203 kmers, 4508168 bytes occupied (5.78229 bits per kmer).\n",
      "  0:30:33.224   468M / 1130M INFO    General                 (edge_index_builders.hpp   : 253)   Collecting edge information from graph, this takes a while.\n",
      "  0:30:35.427   468M / 1130M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 1212192 kmers to process\n",
      "  0:30:36.115   470M / 1130M INFO    General                 (kmer_mapper.hpp           :  87)   Total 3086 kmers with non-trivial roots\n",
      "  0:30:36.119   468M / 1130M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:30:36.120   468M / 1130M INFO    General                 (mismatch_correction.cpp   : 392)   Collect potential mismatches\n",
      "  0:30:38.412   468M / 1130M INFO    General                 (mismatch_correction.cpp   : 193)   Total 572 edges (out of 1416) with 21750 potential mismatch positions (38.0245 positions per edge)\n",
      "  0:30:38.413   468M / 1130M INFO    General                 (mismatch_correction.cpp   : 394)   Potential mismatches collected\n",
      "  0:30:38.414   468M / 1130M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:30:40.721   470M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:30:43.029   471M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:30:45.333   472M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:30:47.598   472M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:30:49.850   472M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:30:52.112   472M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:31:01.844   472M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:31:20.906   472M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:31:59.832   472M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 8400000 reads\n",
      "  0:32:45.175   472M / 1130M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 13126504 reads processed\n",
      "  0:32:45.231   470M / 1130M INFO    General                 (mismatch_correction.cpp   : 387)   All edges processed\n",
      "  0:32:45.233   468M / 1130M INFO    General                 (mismatch_correction.cpp   : 442)   Corrected 0 nucleotides\n",
      "  0:32:45.233   468M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:32:45.233   468M / 1130M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/before_rr.fasta\"\n",
      "  0:32:45.405   468M / 1130M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_after_simplification.gfa\"\n",
      "  0:32:45.514   468M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Paired Information Counting (id: late_pair_info_count)\n",
      "  0:32:45.535   468M / 1130M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 1212192 kmers to process\n",
      "  0:32:45.540   468M / 1130M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:32:45.549   468M / 1130M INFO    General                 (pair_info_count.cpp       : 157)   Min edge length for estimation: 94276\n",
      "  0:32:45.553   468M / 1130M INFO    General                 (pair_info_count.cpp       : 168)   Estimating insert size for library #0\n",
      "  0:32:45.553   468M / 1130M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:32:45.554   468M / 1130M INFO    General                 (paired_info_utils.cpp     :  87)   Estimating insert size (takes a while)\n",
      "  0:32:45.579   500M / 1130M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:32:49.515   500M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:32:53.456   501M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:32:57.352   501M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:33:01.437   501M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:33:05.634   501M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:33:09.622   501M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:33:27.490   501M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:33:47.831   500M / 1130M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:33:48.207   500M / 1130M INFO    General                 (paired_info_utils.cpp     : 104)   Edge pairs: 9274\n",
      "  0:33:48.209   500M / 1130M INFO    General                 (paired_info_utils.cpp     : 106)   1470073 paired reads (44.7971% of all) aligned to long edges\n",
      "  0:33:48.211   468M / 1130M INFO    General                 (pair_info_count.cpp       : 188)     Insert size = 262.599, deviation = 93.6092, left quantile = 153, right quantile = 381, read length = 151\n",
      "  0:33:48.212   468M / 1130M INFO    General                 (pair_info_count.cpp       : 203)   Filtering data for library #0\n",
      "  0:33:48.212   468M / 1130M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:33:48.213   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:33:52.368   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:33:56.266   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:34:00.324   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:34:04.365   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:34:08.560   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:34:12.541   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:34:30.093   469M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:34:50.238   469M / 1130M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:34:50.239   468M / 1130M INFO    General                 (pair_info_count.cpp       : 207)   Mapping library #0\n",
      "  0:34:50.239   468M / 1130M INFO    General                 (pair_info_count.cpp       : 209)   Mapping paired reads (takes a while)\n",
      "  0:34:50.240   468M / 1130M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:34:50.240   468M / 1130M INFO    General                 (paired_info_utils.cpp     : 138)   Left insert size quantile 153, right insert size quantile 381, filtering threshold 2, rounding threshold 0\n",
      "  0:34:50.245   482M / 1130M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:34:54.533   482M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:34:58.731   483M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:35:02.803   483M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:35:06.881   483M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:35:11.189   483M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:35:15.306   483M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:35:33.677   484M / 1130M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:35:54.655   484M / 1130M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:35:54.661   471M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Distance Estimation (id: distance_estimation)\n",
      "  0:35:54.662   471M / 1130M INFO    General                 (distance_estimation.cpp   :  42)   Processing library #0\n",
      "  0:35:54.662   471M / 1130M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:35:54.663   471M / 1130M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SIMPLE distance estimator\n",
      "  0:35:54.699   471M / 1130M INFO    General                 (distance_estimation_utils.: 131)   Refining clustered pair information\n",
      "  0:35:54.700   471M / 1130M INFO    General                 (distance_estimation_utils.: 133)   The refining of clustered pair information has been finished\n",
      "  0:35:54.701   471M / 1130M INFO    General                 (distance_estimation_utils.: 135)   Improving paired information\n",
      "  0:35:54.748   471M / 1130M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 156; contradictional = 96\n",
      "  0:35:54.800   471M / 1130M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 0; contradictional = 0\n",
      "  0:35:54.800   471M / 1130M INFO    General                 (distance_estimation_utils.:  80)   Filling scaffolding index\n",
      "  0:35:54.801   471M / 1130M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:35:54.801   471M / 1130M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SMOOTHING distance estimator\n",
      "  0:35:54.889   471M / 1130M INFO    General                 (distance_estimation.cpp   :  51)   Clearing raw paired index\n",
      "  0:35:54.890   469M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Repeat Resolving (id: repeat_resolving)\n",
      "  0:35:54.892   469M / 1130M INFO    General                 (repeat_resolving.cpp      :  88)   Using Path-Extend repeat resolving\n",
      "  0:35:54.895   469M / 1130M INFO    General                 (launcher.cpp              : 603)   ExSPAnder repeat resolving tool started\n",
      "  0:35:54.903   469M / 1130M INFO    General                 (launcher.cpp              : 420)   Creating main extenders, unique edge length = 2000\n",
      "  0:35:54.906   469M / 1130M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 64.978\n",
      "  0:35:54.910   469M / 1130M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5820\n",
      "  0:35:54.910   469M / 1130M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 64.978\n",
      "  0:35:54.910   469M / 1130M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5820\n",
      "  0:35:54.922   469M / 1130M INFO    General                 (extenders_logic.cpp       : 550)   Using 1 paired-end library\n",
      "  0:35:54.923   469M / 1130M INFO    General                 (extenders_logic.cpp       : 551)   Using 1 paired-end scaffolding library\n",
      "  0:35:54.923   469M / 1130M INFO    General                 (extenders_logic.cpp       : 552)   Using 0 single read libraries\n",
      "  0:35:54.923   469M / 1130M INFO    General                 (launcher.cpp              : 449)   Total number of extenders is 3\n",
      "  0:35:54.924   469M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 0 paths from 681 (0%)\n",
      "  0:35:55.016   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 69 paths from 681 (10%)\n",
      "  0:35:55.021   471M / 1130M INFO    General                 (path_extenders.cpp        :  34)   Processed 128 paths from 681 (18%)\n",
      "  0:35:55.026   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 138 paths from 681 (20%)\n",
      "  0:35:55.027   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 207 paths from 681 (30%)\n",
      "  0:35:55.029   471M / 1130M INFO    General                 (path_extenders.cpp        :  34)   Processed 256 paths from 681 (37%)\n",
      "  0:35:55.030   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 276 paths from 681 (40%)\n",
      "  0:35:55.037   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 345 paths from 681 (50%)\n",
      "  0:35:55.038   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 414 paths from 681 (60%)\n",
      "  0:35:55.043   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 483 paths from 681 (70%)\n",
      "  0:35:55.046   471M / 1130M INFO    General                 (path_extenders.cpp        :  34)   Processed 512 paths from 681 (75%)\n",
      "  0:35:55.047   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 552 paths from 681 (81%)\n",
      "  0:35:55.047   471M / 1130M INFO    General                 (path_extenders.cpp        :  36)   Processed 621 paths from 681 (91%)\n",
      "  0:35:55.049   471M / 1130M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:35:55.049   471M / 1130M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:35:55.050   471M / 1130M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:35:55.060   471M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:35:55.061   471M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:35:55.061   471M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:35:55.062   471M / 1130M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:35:55.064   471M / 1130M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:35:55.065   471M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:35:55.066   471M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:35:55.066   471M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:35:55.086   471M / 1130M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:35:55.091   471M / 1130M INFO    General                 (launcher.cpp              : 456)   Closing gaps in paths\n",
      "  0:35:55.092   471M / 1130M INFO    General                 (launcher.cpp              : 486)   Gap closing completed\n",
      "  0:35:55.098   469M / 1130M INFO    General                 (launcher.cpp              : 304)   Traversing tandem repeats\n",
      "  0:35:55.123   469M / 1130M INFO    General                 (launcher.cpp              : 314)   Traversed 2 loops\n",
      "  0:35:55.124   469M / 1130M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:35:55.124   469M / 1130M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:35:55.124   469M / 1130M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:35:55.125   469M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:35:55.125   469M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:35:55.125   469M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:35:55.125   469M / 1130M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:35:55.126   469M / 1130M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:35:55.128   469M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:35:55.128   469M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:35:55.128   469M / 1130M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:35:55.137   469M / 1130M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:35:55.140   469M / 1130M INFO    General                 (launcher.cpp              : 666)   ExSPAnder repeat resolving tool finished\n",
      "  0:35:55.141   469M / 1130M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:35:55.141   469M / 1130M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/before_rr.fasta\"\n",
      "  0:35:55.338   469M / 1130M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_with_scaffolds.gfa\"\n",
      "  0:35:55.452   469M / 1130M INFO    General                 (contig_output_stage.cpp   : 169)   Outputting FastG graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph.fastg\"\n",
      "  0:35:55.806   469M / 1130M INFO    General                 (contig_output_stage.cpp   : 200)   Breaking scaffolds\n",
      "  0:35:55.842   475M / 1130M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.fasta\n",
      "  0:35:55.962   475M / 1130M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.paths\n",
      "  0:35:55.980   475M / 1130M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.fasta\n",
      "  0:35:56.126   475M / 1130M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.paths\n",
      "  0:35:56.129   475M / 1130M INFO    General                 (contig_output_stage.cpp   : 114)   Populating GFA with scaffold paths\n",
      "  0:35:56.134   469M / 1130M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:35:56.323     1M / 1130M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 35 minutes 56 seconds\n",
      "\n",
      "===== K77 finished. \n",
      "\n",
      "\n",
      "===== Copy files started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/copy_files.py /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/before_rr.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/before_rr.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/strain_graph.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/strain_graph.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph.fastg /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph.fastg /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.paths\n",
      "\n",
      "\n",
      "===== Copy files finished. \n",
      "\n",
      "\n",
      "===== Assembling finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.fasta --misc_dir /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Assembled contigs are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.fasta\n",
      " * Assembled scaffolds are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.fasta\n",
      " * Paths in the assembly graph corresponding to the contigs are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.paths\n",
      " * Assembly graph is in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph_with_scaffolds.gfa\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PAO1wt12P_SPAdes/spades.log\n",
      "\n",
      "Thank you for using SPAdes! If you use it in your research, please cite:\n",
      "\n",
      "  Prjibelski, A., Antipov, D., Meleshko, D., Lapidus, A. and Korobeynikov, A., 2020. Using SPAdes de novo assembler. Current protocols in bioinformatics, 70(1), p.e102.\n",
      "  doi.org/10.1002/cpbi.102\n",
      "\n",
      "Command line: /home/student/miniconda3/envs/VariantCalling/bin/spades.py\t-o\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes\t-1\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0251G_R1.fastq.gz\t-2\t/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0251G_R2.fastq.gz\t--isolate\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 4.1.0\n",
      "  Python version: 3.11.0\n",
      "  OS: Linux-5.15.0-139-generic-x86_64-with-glibc2.31\n",
      "\n",
      "Output dir: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes\n",
      "Mode: ONLY assembling (without read error correction)\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Isolate mode\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0251G_R1.fastq.gz']\n",
      "      right reads: ['/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0251G_R2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "  Assembly graph output will use GFA v1.2 format\n",
      "Other parameters:\n",
      "  Dir for temp files: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 10\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/spades.log\n",
      "\n",
      "/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0251G_R1.fastq.gz: max reads length: 151\n",
      "/media/student/EXTERNAL_USB/Variant_Calling/Analysis/fastq/PA14wt0251G_R2.fastq.gz: max reads length: 151\n",
      "\n",
      "Reads length: 151\n",
      "\n",
      "Default k-mer sizes were set to [21, 33, 55, 77] because estimated read length (151) is equal to or greater than 150\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Assembling started. \n",
      "\n",
      "\n",
      "===== K21 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K21/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K21/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K21/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K21/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/dataset.info\") with K=21\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (read_converter.cpp        :  78)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (read_converter.cpp        :  99)   Converting paired reads\n",
      "  0:00:00.240    34M / 40M   INFO    General                 (binary_converter.cpp      : 127)   16384 reads processed\n",
      "  0:00:00.370    39M / 45M   INFO    General                 (binary_converter.cpp      : 127)   32768 reads processed\n",
      "  0:00:00.927    49M / 55M   INFO    General                 (binary_converter.cpp      : 127)   65536 reads processed\n",
      "  0:00:01.739    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   131072 reads processed\n",
      "  0:00:03.685    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   262144 reads processed\n",
      "  0:00:09.098    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   524288 reads processed\n",
      "  0:00:15.638    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   1048576 reads processed\n",
      "  0:00:29.156    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   2097152 reads processed\n",
      "  0:00:37.267    59M / 65M   INFO    General                 (binary_converter.cpp      : 143)   2717352 reads written\n",
      "  0:00:37.273     2M / 65M   INFO    General                 (read_converter.cpp        : 113)   Converting single reads\n",
      "  0:00:37.276    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:00:37.277     2M / 65M   INFO    General                 (read_converter.cpp        : 119)   Converting merged reads\n",
      "  0:00:37.280    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:00:37.282     1M / 65M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:37.283     1M / 65M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:37.283     1M / 65M   INFO    General                 (construction.cpp          : 156)   Average read length 145.894\n",
      "  0:00:37.284     1M / 65M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:37.284     1M / 65M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:37.284     1M / 65M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:37.284     1M / 65M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:37.285     1M / 65M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:00:47.825   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1073084 reads\n",
      "  0:00:57.725   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2146458 reads\n",
      "  0:01:07.790   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3220697 reads\n",
      "  0:01:18.105   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4294526 reads\n",
      "  0:01:28.008   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5365083 reads\n",
      "  0:01:37.953   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6438137 reads\n",
      "  0:01:48.175   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 7512965 reads\n",
      "  0:01:58.108   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8586690 reads\n",
      "  0:02:08.291   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9660148 reads\n",
      "  0:02:19.758     1M / 628M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 10869408 reads\n",
      "  0:02:19.759     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:02:25.481     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 33351246 kmers in total.\n",
      "  0:02:25.482     1M / 628M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:02:25.485     1M / 628M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:02:25.485     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:02:25.486     1M / 628M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:02:25.486     1M / 628M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:02:25.487     1M / 628M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:02:38.532   601M / 881M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 33351246 kmers\n",
      "  0:02:38.533   601M / 881M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 33351246 kmers.\n",
      "  0:02:38.577     1M / 881M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:02:41.284     1M / 881M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 32721478 kmers in total.\n",
      "  0:02:41.285     1M / 881M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:02:46.611    23M / 881M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 32721478 kmers, 23640928 bytes occupied (5.77992 bits per kmer).\n",
      "  0:02:46.612    23M / 881M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:02:48.648    55M / 881M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:03:18.618    55M / 881M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:03:18.633    55M / 881M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:03:18.634    55M / 881M  INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:03:18.634    55M / 881M  INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:04:02.475    95M / 881M  INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 400374\n",
      "  0:04:03.108    95M / 881M  INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 404172\n",
      "  0:04:03.123    55M / 881M  INFO   Early tip clipping       (early_simplification.hpp  :  50)   8210777 22-mers were removed by early tip clipper\n",
      "  0:04:03.123    55M / 881M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:04:03.125    55M / 881M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:04:35.914   125M / 881M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1843424 sequences extracted\n",
      "  0:04:47.187   125M / 881M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:04:53.683   125M / 881M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:04:53.734   125M / 881M  INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:04:55.033   125M / 881M  INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:04:55.034   125M / 881M  INFO    General                 (debruijn_graph_constructor: 516)   Total 3686848 edges to create\n",
      "  0:04:55.034   269M / 881M  INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:04:58.043   329M / 881M  INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:05:00.324   329M / 881M  INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:05:00.337   345M / 881M  INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:05:00.974   345M / 881M  INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:05:00.975   345M / 881M  INFO    General                 (debruijn_graph_constructor: 542)   Total 1213656 vertices to create\n",
      "  0:05:00.975   457M / 881M  INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:05:01.772   352M / 881M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:05:01.775   352M / 881M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:05:08.562   376M / 881M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 33351246 kmers, 24095768 bytes occupied (5.77988 bits per kmer).\n",
      "  0:05:08.622   504M / 881M  INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:08:40.139   504M / 881M  INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:08:47.520   504M / 881M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:08:53.862   504M / 881M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:09:01.256   504M / 881M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 2994195 edges\n",
      "  0:09:05.532   504M / 881M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 3686697 edges\n",
      "  0:09:05.633   298M / 881M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:09:05.635   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 8\n",
      "  0:09:05.636   298M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 111\n",
      "  0:09:05.636   298M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 111. Coverage mad: 28.1694\n",
      "  0:09:05.637   298M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:09:05.772   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:09:06.342   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:09:07.219   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:09:11.221   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:09:17.777   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 32\n",
      "  0:09:19.760   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 99.6127. Fitted coverage std. dev: 32.6421\n",
      "  0:09:19.766   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.284849\n",
      "  0:09:19.766   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 14\n",
      "  0:09:19.767   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 14\n",
      "  0:09:19.767   299M / 881M  INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6325016\n",
      "  0:09:19.767   298M / 881M  INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 99.6127\n",
      "  0:09:19.767   298M / 881M  INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 14\n",
      "  0:09:19.768   298M / 881M  INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:09:19.768   298M / 881M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:09:19.768   298M / 881M  INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:09:19.768   298M / 881M  INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:09:19.769   298M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:09:20.067   298M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:09:20.067   298M / 881M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:09:20.068   298M / 881M  INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:09:20.068   298M / 881M  INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:09:20.068   298M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:09:20.068   298M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:20.954   298M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 6142 times\n",
      "  0:09:20.955   298M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:44.353   325M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 560788 times\n",
      "  0:09:44.355   325M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:45.622   322M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 59137 times\n",
      "  0:09:45.631   322M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:09:45.632   322M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:45.733   322M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 4909 times\n",
      "  0:09:45.735   322M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:46.632   318M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 10472 times\n",
      "  0:09:46.635   318M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:46.764   317M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6371 times\n",
      "  0:09:46.767   317M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:09:46.767   317M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:46.776   317M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 207 times\n",
      "  0:09:46.778   317M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:46.845   317M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 917 times\n",
      "  0:09:46.855   317M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:46.894   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1168 times\n",
      "  0:09:46.899   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:09:46.899   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:46.901   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 19 times\n",
      "  0:09:46.905   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:46.955   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 206 times\n",
      "  0:09:46.959   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:46.968   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 332 times\n",
      "  0:09:46.969   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:09:46.969   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:46.970   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:09:46.970   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:46.985   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 46 times\n",
      "  0:09:46.988   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:46.995   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 196 times\n",
      "  0:09:46.996   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:09:46.996   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:46.997   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 3 times\n",
      "  0:09:46.998   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.001   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 17 times\n",
      "  0:09:47.005   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:47.009   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 101 times\n",
      "  0:09:47.009   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:09:47.009   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.010   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:09:47.010   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.019   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 8 times\n",
      "  0:09:47.022   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:47.024   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 73 times\n",
      "  0:09:47.027   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:09:47.028   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.029   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:09:47.029   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.046   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 13 times\n",
      "  0:09:47.053   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:47.055   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 42 times\n",
      "  0:09:47.055   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:09:47.056   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.056   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:09:47.056   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.060   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 6 times\n",
      "  0:09:47.060   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:47.061   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 33 times\n",
      "  0:09:47.062   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:09:47.062   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.062   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:09:47.063   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.064   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:09:47.064   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:47.065   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:09:47.065   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:09:47.068   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.101   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 8 times\n",
      "  0:09:47.102   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.166   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 8 times\n",
      "  0:09:47.169   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:47.201   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:09:47.202   316M / 881M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:09:47.203   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.204   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:09:47.204   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.204   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:09:47.204   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:09:47.205   316M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:09:47.220   261M / 881M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:09:47.226   261M / 881M  INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:09:47.226   261M / 881M  INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:09:47.226   261M / 881M  INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:09:47.227   261M / 881M  INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:09:47.227   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.261   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:09:47.262   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.316   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:09:47.327   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:09:47.353   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:09:47.355   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:09:47.409   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:09:47.413   261M / 881M  INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:09:47.440   261M / 881M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:09:47.504   260M / 881M  INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 13811 times\n",
      "  0:09:47.519   260M / 881M  INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:09:47.520   260M / 881M  INFO    General                 (simplification.cpp        : 509)     Average coverage = 103.451\n",
      "  0:09:47.535   260M / 881M  INFO    General                 (simplification.cpp        : 510)     Total length = 6441678\n",
      "  0:09:47.559   260M / 881M  INFO    General                 (simplification.cpp        : 511)     Median edge length: 2359\n",
      "  0:09:47.559   260M / 881M  INFO    General                 (simplification.cpp        : 512)     Edges: 20753\n",
      "  0:09:47.560   260M / 881M  INFO    General                 (simplification.cpp        : 513)     Vertices: 13574\n",
      "  0:09:47.560   260M / 881M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:09:47.560   260M / 881M  INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K21/simplified_contigs\"\n",
      "  0:09:47.628   264M / 881M  INFO    General                 (binary_converter.cpp      : 143)   10399 reads written\n",
      "  0:09:47.634   260M / 881M  INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:09:47.711     1M / 881M  INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 9 minutes 47 seconds\n",
      "\n",
      "===== K21 finished. \n",
      "\n",
      "\n",
      "===== K33 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K33/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K33/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K33/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K33/configs/isolate_mode.info\"\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/dataset.info\") with K=33\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.006     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.006     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.006     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.007     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K21/simplified_contigs\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.894\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:07.385   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 593487 reads\n",
      "  0:00:14.361   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1186849 reads\n",
      "  0:00:21.187   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1780647 reads\n",
      "  0:00:28.068   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2374566 reads\n",
      "  0:00:35.805   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2967266 reads\n",
      "  0:00:42.966   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3560757 reads\n",
      "  0:00:49.751   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4155305 reads\n",
      "  0:00:56.785   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4747567 reads\n",
      "  0:01:43.328   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8900510 reads\n",
      "  0:02:07.079     1M / 571M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 10890206 reads\n",
      "  0:02:07.079     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:02:49.815     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 38671992 kmers in total.\n",
      "  0:02:49.816     1M / 571M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:02:49.819     1M / 571M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:02:49.819     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:02:49.820     1M / 571M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:02:49.820     1M / 571M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:02:49.820     1M / 571M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:02:58.722   601M / 742M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 14316855 kmers\n",
      "  0:03:09.063   601M / 1001M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 28617773 kmers\n",
      "  0:03:16.839   601M / 1104M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 38672002 kmers\n",
      "  0:03:16.840   601M / 1104M INFO    General                 (kmer_splitters.hpp        : 202)   Used 38672002 kmers.\n",
      "  0:03:16.904     1M / 1104M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:25.106     1M / 1104M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 38354427 kmers in total.\n",
      "  0:03:25.107     1M / 1104M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:03:38.020    28M / 1104M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 38354427 kmers, 27709312 bytes occupied (5.77963 bits per kmer).\n",
      "  0:03:38.021    28M / 1104M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:03:43.185    68M / 1104M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:04:25.436    68M / 1104M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:04:25.454    68M / 1104M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:04:25.455    68M / 1104M INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:04:25.455    68M / 1104M INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:05:04.007   108M / 1104M INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 477256\n",
      "  0:05:04.652   108M / 1104M INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 483187\n",
      "  0:05:04.678    68M / 1104M INFO   Early tip clipping       (early_simplification.hpp  :  50)   13180484 34-mers were removed by early tip clipper\n",
      "  0:05:04.683    68M / 1104M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:05:04.685    68M / 1104M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:05:40.077   119M / 1104M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1102211 sequences extracted\n",
      "  0:05:51.082   119M / 1104M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:05:57.694   119M / 1104M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:05:57.738   119M / 1104M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:05:58.512   119M / 1104M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:05:58.513   119M / 1104M INFO    General                 (debruijn_graph_constructor: 516)   Total 2204422 edges to create\n",
      "  0:05:58.514   208M / 1104M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:05:59.639   244M / 1104M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:06:00.441   244M / 1104M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:06:00.452   252M / 1104M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:06:00.663   252M / 1104M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:06:00.667   252M / 1104M INFO    General                 (debruijn_graph_constructor: 542)   Total 784646 vertices to create\n",
      "  0:06:00.668   328M / 1104M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:06:01.248   267M / 1104M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:06:01.248   267M / 1104M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:06:08.820   294M / 1104M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 38671992 kmers, 27938792 bytes occupied (5.77964 bits per kmer).\n",
      "  0:06:08.908   442M / 1104M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:09:48.704   442M / 1104M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:10:00.018   442M / 1104M INFO    General                 (coverage_filling.hpp      :  83)   Processed 978469 edges\n",
      "  0:10:10.645   442M / 1104M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1878469 edges\n",
      "  0:10:14.550   442M / 1104M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2204407 edges\n",
      "  0:10:14.786   200M / 1104M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:10:14.788   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 8\n",
      "  0:10:14.789   200M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 100\n",
      "  0:10:14.789   200M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 100. Coverage mad: 26.6868\n",
      "  0:10:14.790   200M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:10:14.959   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:10:15.707   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:10:16.565   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:10:19.902   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:10:25.241   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 32\n",
      "  0:10:26.224   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 89.166. Fitted coverage std. dev: 29.7332\n",
      "  0:10:26.232   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.18459\n",
      "  0:10:26.233   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 12\n",
      "  0:10:26.233   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 12\n",
      "  0:10:26.234   201M / 1104M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6331912\n",
      "  0:10:26.234   200M / 1104M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 89.166\n",
      "  0:10:26.238   200M / 1104M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 12\n",
      "  0:10:26.238   200M / 1104M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:10:26.239   200M / 1104M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:10:26.239   200M / 1104M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:10:26.240   200M / 1104M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:10:26.241   200M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:10:26.501   200M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 4 times\n",
      "  0:10:26.503   200M / 1104M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:10:26.503   200M / 1104M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:10:26.504   200M / 1104M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:10:26.505   200M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:10:26.505   200M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:27.389   200M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 11575 times\n",
      "  0:10:27.390   200M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:39.605   199M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 334601 times\n",
      "  0:10:39.609   199M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:39.953   198M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14121 times\n",
      "  0:10:39.967   198M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:10:39.968   198M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.147   197M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 4443 times\n",
      "  0:10:40.151   197M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.497   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 4214 times\n",
      "  0:10:40.500   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.632   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2533 times\n",
      "  0:10:40.634   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:10:40.634   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.644   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 201 times\n",
      "  0:10:40.646   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.679   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 327 times\n",
      "  0:10:40.680   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.691   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 190 times\n",
      "  0:10:40.692   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:10:40.693   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.694   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 13 times\n",
      "  0:10:40.694   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.702   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 38 times\n",
      "  0:10:40.703   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.747   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 134 times\n",
      "  0:10:40.748   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:10:40.749   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.749   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 6 times\n",
      "  0:10:40.750   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.754   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 14 times\n",
      "  0:10:40.757   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.768   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 31 times\n",
      "  0:10:40.768   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:10:40.769   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.769   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:10:40.769   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.770   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:10:40.770   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.771   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 19 times\n",
      "  0:10:40.772   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:10:40.772   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.772   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:10:40.773   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.775   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:10:40.780   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.781   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 12 times\n",
      "  0:10:40.781   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:10:40.782   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.782   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:10:40.782   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.784   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:10:40.788   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.789   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 10 times\n",
      "  0:10:40.790   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:10:40.790   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.791   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:10:40.793   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.793   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:10:40.793   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.794   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6 times\n",
      "  0:10:40.795   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:10:40.795   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.795   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:10:40.796   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.798   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:10:40.798   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.800   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:10:40.802   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:10:40.802   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.836   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 7 times\n",
      "  0:10:40.836   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.867   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:10:40.872   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.900   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:10:40.901   196M / 1104M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:10:40.901   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.902   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:10:40.902   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.902   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:10:40.902   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:10:40.903   196M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:10:40.925   169M / 1104M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:10:40.931   169M / 1104M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:10:40.931   169M / 1104M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:10:40.932   169M / 1104M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:10:40.932   169M / 1104M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:10:40.932   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:40.955   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:10:40.958   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:40.982   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:10:40.990   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:41.019   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:10:41.019   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:10:41.050   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:10:41.050   169M / 1104M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:10:41.071   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:10:41.227   169M / 1104M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 39085 times\n",
      "  0:10:41.233   167M / 1104M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:10:41.240   167M / 1104M INFO    General                 (simplification.cpp        : 509)     Average coverage = 91.8134\n",
      "  0:10:41.249   167M / 1104M INFO    General                 (simplification.cpp        : 510)     Total length = 6468970\n",
      "  0:10:41.264   167M / 1104M INFO    General                 (simplification.cpp        : 511)     Median edge length: 17548\n",
      "  0:10:41.265   167M / 1104M INFO    General                 (simplification.cpp        : 512)     Edges: 4714\n",
      "  0:10:41.265   167M / 1104M INFO    General                 (simplification.cpp        : 513)     Vertices: 3964\n",
      "  0:10:41.266   167M / 1104M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:10:41.266   167M / 1104M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K33/simplified_contigs\"\n",
      "  0:10:41.309   171M / 1104M INFO    General                 (binary_converter.cpp      : 143)   2360 reads written\n",
      "  0:10:41.313   167M / 1104M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:10:41.360     1M / 1104M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 10 minutes 41 seconds\n",
      "\n",
      "===== K33 finished. \n",
      "\n",
      "\n",
      "===== K55 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K55/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K55/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K55/configs/config.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K55/configs/isolate_mode.info\"\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/dataset.info\") with K=55\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.007     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.007     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.009     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.010     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K33/simplified_contigs\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.894\n",
      "  0:00:00.016     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.017     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.017     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.018     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.019     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:07.920   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 736010 reads\n",
      "  0:00:14.978   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1470220 reads\n",
      "  0:00:22.329   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2206103 reads\n",
      "  0:00:29.355   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2941574 reads\n",
      "  0:00:36.693   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3677270 reads\n",
      "  0:00:44.915   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4412113 reads\n",
      "  0:00:52.283   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5145979 reads\n",
      "  0:00:59.716   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5879250 reads\n",
      "  0:01:29.111   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8823724 reads\n",
      "  0:01:51.977     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 10874128 reads\n",
      "  0:01:51.978     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:02:28.214     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 41662901 kmers in total.\n",
      "  0:02:28.215     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:02:28.217     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:02:28.217     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:02:28.218     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:02:28.218     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:02:28.218     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:02:37.827   601M / 719M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 14056111 kmers\n",
      "  0:02:48.212   601M / 970M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 28090050 kmers\n",
      "  0:02:58.711   601M / 1186M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 41662912 kmers\n",
      "  0:02:58.713   601M / 1186M INFO    General                 (kmer_splitters.hpp        : 202)   Used 41662912 kmers.\n",
      "  0:02:58.790     1M / 1186M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:13.282     1M / 1186M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 41662228 kmers in total.\n",
      "  0:03:13.283     1M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:03:21.235    30M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 41662228 kmers, 30098392 bytes occupied (5.77951 bits per kmer).\n",
      "  0:03:21.236    30M / 1186M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:03:26.853    70M / 1186M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:04:10.339    70M / 1186M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:04:10.434    70M / 1186M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:04:10.436    70M / 1186M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:05:07.290   145M / 1186M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1499575 sequences extracted\n",
      "  0:05:26.955   145M / 1186M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:05:34.326   145M / 1186M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:05:34.401   145M / 1186M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:05:35.388   145M / 1186M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:05:35.389   145M / 1186M INFO    General                 (debruijn_graph_constructor: 516)   Total 2999150 edges to create\n",
      "  0:05:35.390   261M / 1186M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:05:37.519   309M / 1186M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:05:38.762   309M / 1186M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:05:38.794   325M / 1186M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:05:39.381   325M / 1186M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:05:39.382   325M / 1186M INFO    General                 (debruijn_graph_constructor: 542)   Total 1498902 vertices to create\n",
      "  0:05:39.383   466M / 1186M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:05:40.173   379M / 1186M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:05:40.176   379M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:05:48.375   408M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 41662901 kmers, 30098888 bytes occupied (5.77951 bits per kmer).\n",
      "  0:05:48.489   568M / 1186M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:08:40.031   568M / 1186M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:08:54.188   568M / 1186M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:09:05.034   568M / 1186M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1833475 edges\n",
      "  0:09:11.056   568M / 1186M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2429791 edges\n",
      "  0:09:15.941   568M / 1186M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2913514 edges\n",
      "  0:09:16.850   568M / 1186M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2999149 edges\n",
      "  0:09:17.058   310M / 1186M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:09:17.059   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 8\n",
      "  0:09:17.060   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 79\n",
      "  0:09:17.060   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 79. Coverage mad: 22.239\n",
      "  0:09:17.060   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:09:17.174   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:09:17.845   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:09:18.548   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:09:20.562   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:09:24.328   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 32\n",
      "  0:09:25.899   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 71.0642. Fitted coverage std. dev: 24.3386\n",
      "  0:09:25.906   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.142612\n",
      "  0:09:25.906   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 12\n",
      "  0:09:25.907   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 12\n",
      "  0:09:25.907   310M / 1186M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6265555\n",
      "  0:09:25.907   310M / 1186M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 71.0642\n",
      "  0:09:25.908   310M / 1186M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 12\n",
      "  0:09:25.908   310M / 1186M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:09:25.908   310M / 1186M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:09:25.924   323M / 1186M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:09:26.731   340M / 1186M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 778385 out of 2999149, length: 39175113\n",
      "  0:09:26.822   351M / 1186M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 3014148)\n",
      "  0:09:28.606   351M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:09:52.081   372M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 28840938 kmers, 20831496 bytes occupied (5.77831 bits per kmer).\n",
      "  0:09:52.223   596M / 1186M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:10:05.219   585M / 1186M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:10:27.737   585M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:10:49.639   586M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:11:11.934   586M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:11:35.479   587M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:11:58.131   587M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:12:20.240   588M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:14:09.257   590M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:15:05.578   591M / 1186M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 2717352 reads processed\n",
      "  0:15:05.699   356M / 1186M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:15:05.700   356M / 1186M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:15:06.542   356M / 1186M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 2060 tips collected, total 2243 connection candidates\n",
      "  0:15:06.553   356M / 1186M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 56 gaps after checking 2180 candidates\n",
      "  0:15:06.627   356M / 1186M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:15:06.646   310M / 1186M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:15:06.650   310M / 1186M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:15:06.651   310M / 1186M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:15:06.651   310M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:15:06.940   310M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:15:06.947   310M / 1186M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:15:06.947   310M / 1186M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:15:06.948   310M / 1186M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:15:06.948   310M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:15:06.948   310M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:11.198   293M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 550024 times\n",
      "  0:15:11.199   293M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.144   865M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 105981 times\n",
      "  0:15:33.145   865M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.254   865M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2508 times\n",
      "  0:15:33.255   865M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:15:33.256   865M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.323   864M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2098 times\n",
      "  0:15:33.325   864M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.639   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 469 times\n",
      "  0:15:33.646   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.812   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 891 times\n",
      "  0:15:33.814   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:15:33.814   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.819   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 38 times\n",
      "  0:15:33.820   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.831   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 17 times\n",
      "  0:15:33.831   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.848   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 60 times\n",
      "  0:15:33.851   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:15:33.851   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.857   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:15:33.860   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.863   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 4 times\n",
      "  0:15:33.870   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.884   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:15:33.887   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:15:33.887   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.888   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:33.889   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.891   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:33.891   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.895   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 13 times\n",
      "  0:15:33.904   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:15:33.904   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.905   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:33.905   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.909   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:15:33.911   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.912   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 4 times\n",
      "  0:15:33.913   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:15:33.913   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.914   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:33.914   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.918   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:15:33.919   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.919   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:15:33.920   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:15:33.921   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.922   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:33.923   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.923   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:33.923   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.924   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:33.924   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:15:33.925   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.926   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:33.926   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.930   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:33.930   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.931   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:15:33.931   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:15:33.932   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.932   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:33.933   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:33.933   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:33.934   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:33.935   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:15:33.935   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:15:33.941   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:33.998   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 9 times\n",
      "  0:15:34.006   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:34.061   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:15:34.070   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:34.129   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:34.131   867M / 1186M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:15:34.131   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:34.132   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:34.133   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:34.133   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:34.134   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:34.138   867M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:34.165   845M / 1186M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:15:34.178   859M / 1186M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:15:34.296   864M / 1186M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 219127 out of 220930, length: 28302635\n",
      "  0:15:34.322   866M / 1186M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 3014148)\n",
      "  0:15:35.268   866M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:15:43.391   877M / 1186M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 15722425 kmers, 11356488 bytes occupied (5.77849 bits per kmer).\n",
      "  0:15:43.462   997M / 1186M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:15:49.389   995M / 1186M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:16:00.839   995M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:16:12.082   996M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:16:24.126   996M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:16:35.314   997M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:16:46.356   997M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:16:57.555   997M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:17:53.385   999M / 1186M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:18:21.861  1000M / 1186M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 2717352 reads processed\n",
      "  0:18:21.935   874M / 1186M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:18:21.937   874M / 1186M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:18:21.985   874M / 1186M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 474 tips collected, total 761 connection candidates\n",
      "  0:18:21.987   874M / 1186M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 761 candidates\n",
      "  0:18:22.001   874M / 1186M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:18:22.009   845M / 1186M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:18:22.010   845M / 1186M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:18:22.011   845M / 1186M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:18:22.011   845M / 1186M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:18:22.012   845M / 1186M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:18:22.012   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:18:22.036   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:18:22.037   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:18:22.061   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:18:22.062   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:18:22.119   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:18:22.120   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:18:22.172   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:18:22.179   845M / 1186M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:18:22.239   845M / 1186M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:18:22.587   843M / 1186M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 109185 times\n",
      "  0:18:22.597   840M / 1186M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:18:22.598   840M / 1186M INFO    General                 (simplification.cpp        : 509)     Average coverage = 71.7323\n",
      "  0:18:22.603   840M / 1186M INFO    General                 (simplification.cpp        : 510)     Total length = 6477614\n",
      "  0:18:22.609   840M / 1186M INFO    General                 (simplification.cpp        : 511)     Median edge length: 27999\n",
      "  0:18:22.610   840M / 1186M INFO    General                 (simplification.cpp        : 512)     Edges: 2560\n",
      "  0:18:22.610   840M / 1186M INFO    General                 (simplification.cpp        : 513)     Vertices: 2698\n",
      "  0:18:22.611   840M / 1186M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:18:22.612   840M / 1186M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K55/simplified_contigs\"\n",
      "  0:18:22.638   844M / 1186M INFO    General                 (binary_converter.cpp      : 143)   1280 reads written\n",
      "  0:18:22.934   840M / 1186M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:18:24.261     1M / 1186M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 18 minutes 24 seconds\n",
      "\n",
      "===== K55 finished. \n",
      "\n",
      "\n",
      "===== K77 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/configs/config.info /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/dataset.info\") with K=77\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.000     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K55/simplified_contigs\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.894\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:00:06.019   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 645155 reads\n",
      "  0:00:12.159   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1289712 reads\n",
      "  0:00:18.146   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1934396 reads\n",
      "  0:00:24.133   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2579046 reads\n",
      "  0:00:30.083   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3224610 reads\n",
      "  0:00:36.801   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3869941 reads\n",
      "  0:00:43.425   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4514021 reads\n",
      "  0:00:49.445   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5156260 reads\n",
      "  0:01:25.681   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9023894 reads\n",
      "  0:01:44.218     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 10871968 reads\n",
      "  0:01:44.219     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:29.903     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 39233270 kmers in total.\n",
      "  0:03:29.904     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:03:30.085     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:03:30.085     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:03:30.086     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:03:30.087     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:03:30.087     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:03:37.209   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 8126066 kmers\n",
      "  0:03:45.811   601M / 607M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 16258628 kmers\n",
      "  0:03:59.473   601M / 698M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 24391327 kmers\n",
      "  0:04:08.568   601M / 756M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 32542191 kmers\n",
      "  0:04:15.242   601M / 784M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 39233290 kmers\n",
      "  0:04:15.243   601M / 784M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 39233290 kmers.\n",
      "  0:04:15.287     1M / 784M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:05:15.140     1M / 784M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 39433703 kmers in total.\n",
      "  0:05:15.141     1M / 784M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:05:23.962    29M / 784M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 39433703 kmers, 28488752 bytes occupied (5.77957 bits per kmer).\n",
      "  0:05:23.963    29M / 784M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:05:32.143    69M / 784M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:06:15.393    69M / 784M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:06:15.402    69M / 784M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:06:15.409    69M / 784M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:07:03.902   135M / 1064M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1225422 sequences extracted\n",
      "  0:07:21.290   135M / 1082M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:07:28.549   135M / 1082M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:07:28.650   135M / 1082M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:07:29.463   135M / 1082M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:07:29.464   135M / 1082M INFO    General                 (debruijn_graph_constructor: 516)   Total 2450844 edges to create\n",
      "  0:07:29.464   231M / 1082M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:07:31.350   271M / 1082M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:07:32.417   271M / 1082M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:07:32.436   287M / 1082M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:07:32.928   287M / 1082M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:07:32.929   287M / 1082M INFO    General                 (debruijn_graph_constructor: 542)   Total 1425855 vertices to create\n",
      "  0:07:32.930   420M / 1082M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:07:33.599   345M / 1082M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:07:33.602   345M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:07:41.034   372M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 39233270 kmers, 28344056 bytes occupied (5.7796 bits per kmer).\n",
      "  0:07:41.140   524M / 1082M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:10:00.990   524M / 1082M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:10:16.604   524M / 1082M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:10:26.802   524M / 1082M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1729738 edges\n",
      "  0:10:31.412   524M / 1082M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2169076 edges\n",
      "  0:10:34.122   524M / 1082M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2450844 edges\n",
      "  0:10:34.644   277M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:10:34.646   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 7\n",
      "  0:10:34.646   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 56\n",
      "  0:10:34.646   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 56. Coverage mad: 19.2738\n",
      "  0:10:34.647   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:10:34.791   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:10:35.369   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:10:35.896   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:10:36.878   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:10:40.269   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 32\n",
      "  0:10:43.115   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 53.8056. Fitted coverage std. dev: 18.9416\n",
      "  0:10:43.119   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.247386\n",
      "  0:10:43.120   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 12\n",
      "  0:10:43.120   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 12\n",
      "  0:10:43.120   277M / 1082M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6160820\n",
      "  0:10:43.123   277M / 1082M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 53.8056\n",
      "  0:10:43.124   277M / 1082M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 12\n",
      "  0:10:43.124   277M / 1082M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:10:43.124   277M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:10:43.130   290M / 1082M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:10:43.788   325M / 1082M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 932214 out of 2450844, length: 47141828\n",
      "  0:10:43.896   337M / 1082M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 2463101)\n",
      "  0:10:45.879   337M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:11:14.615   360M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 32182538 kmers, 23245032 bytes occupied (5.7783 bits per kmer).\n",
      "  0:11:14.748   608M / 1082M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:11:32.742   596M / 1082M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:11:50.827   597M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:12:10.978   598M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:12:29.509   598M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:12:48.021   599M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:13:05.811   599M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:13:24.306   600M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:14:58.242   603M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:15:46.327   605M / 1082M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 2717352 reads processed\n",
      "  0:15:46.488   345M / 1082M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:15:46.492   345M / 1082M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:15:47.256   345M / 1082M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 2257 tips collected, total 2486 connection candidates\n",
      "  0:15:47.269   345M / 1082M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 117 gaps after checking 2360 candidates\n",
      "  0:15:47.345   345M / 1082M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:15:47.374   277M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:15:47.380   277M / 1082M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:15:47.381   277M / 1082M INFO    General                 (simplification.cpp        :  63)   Most init cleaning disabled on main iteration\n",
      "  0:15:47.381   277M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:15:47.669   277M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:15:47.670   277M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:15:47.671   277M / 1082M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:15:47.671   277M / 1082M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:15:47.671   277M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:15:47.671   277M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:52.229   258M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 510806 times\n",
      "  0:15:52.230   258M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.663   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 5645 times\n",
      "  0:15:53.665   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.802   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 325 times\n",
      "  0:15:53.803   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:15:53.803   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.836   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 194 times\n",
      "  0:15:53.837   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.849   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 8 times\n",
      "  0:15:53.849   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.961   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 219 times\n",
      "  0:15:53.963   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:15:53.963   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.964   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 4 times\n",
      "  0:15:53.964   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.965   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:15:53.965   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.972   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 25 times\n",
      "  0:15:53.981   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:15:53.982   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.982   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:53.983   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.983   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:53.983   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.984   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 5 times\n",
      "  0:15:53.985   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:15:53.985   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.985   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:53.985   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.986   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:53.986   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.986   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:53.987   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:15:53.989   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.990   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:53.990   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.991   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:53.991   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.991   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:15:53.991   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:15:53.992   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.992   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:53.992   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.992   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:53.993   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.993   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:15:53.993   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:15:53.993   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.994   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:53.994   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.994   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:53.994   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.995   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:53.995   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:15:53.995   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.995   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:53.996   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.996   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:53.996   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.996   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:53.997   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:15:53.997   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:53.997   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:53.997   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:53.998   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:53.998   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:53.998   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:53.998   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:15:53.999   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:54.100   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 18 times\n",
      "  0:15:54.101   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:54.197   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:54.198   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:54.284   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:54.285   320M / 1082M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:15:54.285   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:15:54.286   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:15:54.286   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:15:54.286   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:15:54.286   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:15:54.287   320M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:15:54.336   303M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:15:54.360   316M / 1082M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:15:54.550   325M / 1082M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 413496 out of 414654, length: 38350998\n",
      "  0:15:54.613   328M / 1082M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 2463101)\n",
      "  0:15:55.735   328M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:16:10.241   343M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 20446811 kmers, 14769660 bytes occupied (5.77876 bits per kmer).\n",
      "  0:16:10.328   499M / 1082M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:16:20.873   495M / 1082M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:16:29.199   496M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:16:36.957   497M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:16:44.119   498M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:16:51.540   499M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:16:59.072   499M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:17:06.435   500M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:17:42.018   504M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:17:59.760   506M / 1082M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 2717352 reads processed\n",
      "  0:17:59.911   346M / 1082M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:17:59.912   346M / 1082M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:18:00.060   346M / 1082M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 777 tips collected, total 1192 connection candidates\n",
      "  0:18:00.063   346M / 1082M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 1192 candidates\n",
      "  0:18:00.115   346M / 1082M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:18:00.152   303M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:18:00.164   303M / 1082M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:18:00.165   303M / 1082M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:18:00.168   303M / 1082M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:18:00.169   303M / 1082M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:18:00.169   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:18:00.259   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:18:00.260   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:18:00.357   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:18:00.357   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:18:00.447   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:18:00.448   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:18:00.536   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:18:00.540   303M / 1082M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:18:00.675   303M / 1082M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:18:01.236   298M / 1082M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 206406 times\n",
      "  0:18:01.243   293M / 1082M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:18:01.243   293M / 1082M INFO    General                 (simplification.cpp        : 509)     Average coverage = 53.0232\n",
      "  0:18:01.247   293M / 1082M INFO    General                 (simplification.cpp        : 510)     Total length = 6469631\n",
      "  0:18:01.253   293M / 1082M INFO    General                 (simplification.cpp        : 511)     Median edge length: 31730\n",
      "  0:18:01.254   293M / 1082M INFO    General                 (simplification.cpp        : 512)     Edges: 1842\n",
      "  0:18:01.255   293M / 1082M INFO    General                 (simplification.cpp        : 513)     Vertices: 2124\n",
      "  0:18:01.255   293M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Mismatch Correction (id: mismatch_correction)\n",
      "  0:18:01.255   293M / 1082M INFO    General                 (graph_pack_helpers.cpp    :  44)   Index refill\n",
      "  0:18:01.256   293M / 1082M INFO    General                 (edge_index.hpp            : 175)   Using small index (max_id = 2463101)\n",
      "  0:18:01.746   293M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:18:04.710   298M / 1082M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 6469631 kmers, 4674460 bytes occupied (5.78019 bits per kmer).\n",
      "  0:18:04.744   350M / 1082M INFO    General                 (edge_index_builders.hpp   : 253)   Collecting edge information from graph, this takes a while.\n",
      "  0:18:07.513   350M / 1082M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 885304 kmers to process\n",
      "  0:18:08.039   351M / 1082M INFO    General                 (kmer_mapper.hpp           :  87)   Total 1214 kmers with non-trivial roots\n",
      "  0:18:08.044   350M / 1082M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:18:08.045   350M / 1082M INFO    General                 (mismatch_correction.cpp   : 392)   Collect potential mismatches\n",
      "  0:18:10.091   350M / 1082M INFO    General                 (mismatch_correction.cpp   : 193)   Total 904 edges (out of 1842) with 15869 potential mismatch positions (17.5542 positions per edge)\n",
      "  0:18:10.092   350M / 1082M INFO    General                 (mismatch_correction.cpp   : 394)   Potential mismatches collected\n",
      "  0:18:10.096   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:18:12.311   351M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:18:14.592   352M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:18:16.908   352M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:18:19.249   353M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:18:21.431   353M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:18:23.809   353M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:18:33.664   353M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:18:52.493   353M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:19:32.092   353M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 8400000 reads\n",
      "  0:19:53.945   353M / 1082M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 10869408 reads processed\n",
      "  0:19:54.003   351M / 1082M INFO    General                 (mismatch_correction.cpp   : 387)   All edges processed\n",
      "  0:19:54.005   350M / 1082M INFO    General                 (mismatch_correction.cpp   : 442)   Corrected 0 nucleotides\n",
      "  0:19:54.005   350M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:19:54.005   350M / 1082M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/before_rr.fasta\"\n",
      "  0:19:54.564   350M / 1082M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/assembly_graph_after_simplification.gfa\"\n",
      "  0:19:54.662   350M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Paired Information Counting (id: late_pair_info_count)\n",
      "  0:19:54.671   350M / 1082M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 885304 kmers to process\n",
      "  0:19:54.672   350M / 1082M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:19:54.677   350M / 1082M INFO    General                 (pair_info_count.cpp       : 157)   Min edge length for estimation: 31730\n",
      "  0:19:54.677   350M / 1082M INFO    General                 (pair_info_count.cpp       : 168)   Estimating insert size for library #0\n",
      "  0:19:54.679   350M / 1082M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:19:54.679   350M / 1082M INFO    General                 (paired_info_utils.cpp     :  87)   Estimating insert size (takes a while)\n",
      "  0:19:54.693   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:19:58.497   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:20:02.255   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:20:05.989   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:20:10.130   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:20:14.132   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:20:18.262   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:20:35.912   382M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:20:44.999   382M / 1082M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 2717352 reads processed\n",
      "  0:20:45.419   382M / 1082M INFO    General                 (paired_info_utils.cpp     : 104)   Edge pairs: 8862\n",
      "  0:20:45.428   382M / 1082M INFO    General                 (paired_info_utils.cpp     : 106)   1281036 paired reads (47.1428% of all) aligned to long edges\n",
      "  0:20:45.429   350M / 1082M INFO    General                 (pair_info_count.cpp       : 188)     Insert size = 241.549, deviation = 79.0189, left quantile = 147, right quantile = 346, read length = 151\n",
      "  0:20:45.434   350M / 1082M INFO    General                 (pair_info_count.cpp       : 203)   Filtering data for library #0\n",
      "  0:20:45.434   350M / 1082M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:20:45.435   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:20:49.371   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:20:53.237   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:20:56.897   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:21:00.730   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:21:04.502   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:21:08.667   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:21:25.839   350M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:21:34.698   350M / 1082M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 2717352 reads processed\n",
      "  0:21:34.699   350M / 1082M INFO    General                 (pair_info_count.cpp       : 207)   Mapping library #0\n",
      "  0:21:34.699   350M / 1082M INFO    General                 (pair_info_count.cpp       : 209)   Mapping paired reads (takes a while)\n",
      "  0:21:34.699   350M / 1082M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:21:34.700   350M / 1082M INFO    General                 (paired_info_utils.cpp     : 138)   Left insert size quantile 147, right insert size quantile 346, filtering threshold 2, rounding threshold 0\n",
      "  0:21:34.707   364M / 1082M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:21:38.988   365M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:21:42.744   365M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:21:46.553   365M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:21:50.359   365M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:21:54.004   365M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:21:57.810   366M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:22:15.747   366M / 1082M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:22:24.933   366M / 1082M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 2717352 reads processed\n",
      "  0:22:24.939   353M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Distance Estimation (id: distance_estimation)\n",
      "  0:22:24.939   353M / 1082M INFO    General                 (distance_estimation.cpp   :  42)   Processing library #0\n",
      "  0:22:24.942   353M / 1082M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:22:24.943   353M / 1082M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SIMPLE distance estimator\n",
      "  0:22:24.978   353M / 1082M INFO    General                 (distance_estimation_utils.: 131)   Refining clustered pair information\n",
      "  0:22:24.979   353M / 1082M INFO    General                 (distance_estimation_utils.: 133)   The refining of clustered pair information has been finished\n",
      "  0:22:24.980   353M / 1082M INFO    General                 (distance_estimation_utils.: 135)   Improving paired information\n",
      "  0:22:25.016   353M / 1082M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 46; contradictional = 36\n",
      "  0:22:25.058   353M / 1082M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 4; contradictional = 8\n",
      "  0:22:25.058   353M / 1082M INFO    General                 (distance_estimation_utils.:  80)   Filling scaffolding index\n",
      "  0:22:25.059   353M / 1082M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:22:25.059   353M / 1082M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SMOOTHING distance estimator\n",
      "  0:22:25.122   353M / 1082M INFO    General                 (distance_estimation.cpp   :  51)   Clearing raw paired index\n",
      "  0:22:25.123   351M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Repeat Resolving (id: repeat_resolving)\n",
      "  0:22:25.123   351M / 1082M INFO    General                 (repeat_resolving.cpp      :  88)   Using Path-Extend repeat resolving\n",
      "  0:22:25.125   351M / 1082M INFO    General                 (launcher.cpp              : 603)   ExSPAnder repeat resolving tool started\n",
      "  0:22:25.130   351M / 1082M INFO    General                 (launcher.cpp              : 420)   Creating main extenders, unique edge length = 2000\n",
      "  0:22:25.131   351M / 1082M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 53.0232\n",
      "  0:22:25.132   351M / 1082M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5410\n",
      "  0:22:25.133   351M / 1082M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 53.0232\n",
      "  0:22:25.133   351M / 1082M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5410\n",
      "  0:22:25.152   352M / 1082M INFO    General                 (extenders_logic.cpp       : 550)   Using 1 paired-end library\n",
      "  0:22:25.156   352M / 1082M INFO    General                 (extenders_logic.cpp       : 551)   Using 1 paired-end scaffolding library\n",
      "  0:22:25.157   352M / 1082M INFO    General                 (extenders_logic.cpp       : 552)   Using 0 single read libraries\n",
      "  0:22:25.157   352M / 1082M INFO    General                 (launcher.cpp              : 449)   Total number of extenders is 3\n",
      "  0:22:25.157   352M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 0 paths from 899 (0%)\n",
      "  0:22:25.168   352M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 90 paths from 899 (10%)\n",
      "  0:22:25.174   352M / 1082M INFO    General                 (path_extenders.cpp        :  34)   Processed 128 paths from 899 (14%)\n",
      "  0:22:25.182   352M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 180 paths from 899 (20%)\n",
      "  0:22:25.199   352M / 1082M INFO    General                 (path_extenders.cpp        :  34)   Processed 256 paths from 899 (28%)\n",
      "  0:22:25.200   352M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 270 paths from 899 (30%)\n",
      "  0:22:25.208   352M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 360 paths from 899 (40%)\n",
      "  0:22:25.212   353M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 450 paths from 899 (50%)\n",
      "  0:22:25.213   353M / 1082M INFO    General                 (path_extenders.cpp        :  34)   Processed 512 paths from 899 (56%)\n",
      "  0:22:25.220   353M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 540 paths from 899 (60%)\n",
      "  0:22:25.222   353M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 630 paths from 899 (70%)\n",
      "  0:22:25.224   353M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 720 paths from 899 (80%)\n",
      "  0:22:25.226   353M / 1082M INFO    General                 (path_extenders.cpp        :  36)   Processed 810 paths from 899 (90%)\n",
      "  0:22:25.227   352M / 1082M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:22:25.227   352M / 1082M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:22:25.228   352M / 1082M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:22:25.228   352M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:22:25.228   352M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:22:25.229   352M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:22:25.229   352M / 1082M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:22:25.230   352M / 1082M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:22:25.234   352M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:22:25.235   352M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:22:25.236   352M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:22:25.249   352M / 1082M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:22:25.256   352M / 1082M INFO    General                 (launcher.cpp              : 456)   Closing gaps in paths\n",
      "  0:22:25.258   353M / 1082M INFO    General                 (launcher.cpp              : 486)   Gap closing completed\n",
      "  0:22:25.259   351M / 1082M INFO    General                 (launcher.cpp              : 304)   Traversing tandem repeats\n",
      "  0:22:25.285   351M / 1082M INFO    General                 (launcher.cpp              : 314)   Traversed 3 loops\n",
      "  0:22:25.288   351M / 1082M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:22:25.289   351M / 1082M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:22:25.289   351M / 1082M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:22:25.290   351M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:22:25.290   351M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:22:25.290   351M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:22:25.291   351M / 1082M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:22:25.292   351M / 1082M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:22:25.294   351M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:22:25.294   351M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:22:25.295   351M / 1082M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:22:25.299   351M / 1082M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:22:25.306   351M / 1082M INFO    General                 (launcher.cpp              : 666)   ExSPAnder repeat resolving tool finished\n",
      "  0:22:25.306   351M / 1082M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:22:25.306   351M / 1082M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/before_rr.fasta\"\n",
      "  0:22:25.529   351M / 1082M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/assembly_graph_with_scaffolds.gfa\"\n",
      "  0:22:25.669   351M / 1082M INFO    General                 (contig_output_stage.cpp   : 169)   Outputting FastG graph to \"/media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/assembly_graph.fastg\"\n",
      "  0:22:26.069   351M / 1082M INFO    General                 (contig_output_stage.cpp   : 200)   Breaking scaffolds\n",
      "  0:22:26.110   358M / 1082M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/final_contigs.fasta\n",
      "  0:22:26.264   358M / 1082M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/final_contigs.paths\n",
      "  0:22:26.304   358M / 1082M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/scaffolds.fasta\n",
      "  0:22:26.463   358M / 1082M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/scaffolds.paths\n",
      "  0:22:26.466   358M / 1082M INFO    General                 (contig_output_stage.cpp   : 114)   Populating GFA with scaffold paths\n",
      "  0:22:26.469   351M / 1082M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:22:26.668     1M / 1082M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 22 minutes 26 seconds\n",
      "\n",
      "===== K77 finished. \n",
      "\n",
      "\n",
      "===== Copy files started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/copy_files.py /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/before_rr.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/before_rr.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/final_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/strain_graph.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/strain_graph.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/scaffolds.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/scaffolds.fasta /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/scaffolds.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/scaffolds.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/assembly_graph.fastg /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/assembly_graph.fastg /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/K77/final_contigs.paths /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/contigs.paths\n",
      "\n",
      "\n",
      "===== Copy files finished. \n",
      "\n",
      "\n",
      "===== Assembling finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/scaffolds.fasta --misc_dir /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Assembled contigs are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/contigs.fasta\n",
      " * Assembled scaffolds are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/scaffolds.fasta\n",
      " * Paths in the assembly graph corresponding to the contigs are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/scaffolds.paths\n",
      " * Assembly graph is in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/assembly_graph_with_scaffolds.gfa\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /media/student/EXTERNAL_USB/Variant_Calling/Analysis/assemblies/PA14wt0251G_SPAdes/spades.log\n",
      "\n",
      "Thank you for using SPAdes! If you use it in your research, please cite:\n",
      "\n",
      "  Prjibelski, A., Antipov, D., Meleshko, D., Lapidus, A. and Korobeynikov, A., 2020. Using SPAdes de novo assembler. Current protocols in bioinformatics, 70(1), p.e102.\n",
      "  doi.org/10.1002/cpbi.102\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"denovo\"]:\n",
    "\n",
    "    samples = {file.split(\"_\")[0] for file in os.listdir(fastq_directory_path)}\n",
    "\n",
    "    for sample in samples:\n",
    "        files = os.listdir(fastq_directory_path)\n",
    "\n",
    "        R1_file = sample + \"_R1.fastq.gz\"\n",
    "        R1_file_path = os.path.join(fastq_directory_path, R1_file)\n",
    "        R2_file = sample + \"_R2.fastq.gz\"\n",
    "        R2_file_path = os.path.join(fastq_directory_path, R2_file)\n",
    "\n",
    "        denovo_directory = sample + \"_SPAdes\"\n",
    "        denovo_directory_path = os.path.join(assemblies_directory_path, denovo_directory)\n",
    "\n",
    "        if os.path.exists(denovo_directory_path):\n",
    "            shutil.rmtree(denovo_directory_path)\n",
    "\n",
    "        command = [\n",
    "            \"python\", spades, \"-o\", denovo_directory_path,\n",
    "            \"-1\", R1_file_path, \"-2\", R2_file_path,\n",
    "            \"--isolate\"\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        contigs_file_path = os.path.join(denovo_directory_path, \"contigs.fasta\")\n",
    "\n",
    "        denovo_file = sample + \".SPAdes.denovoassembly.fasta\"\n",
    "        denovo_file_path = os.path.join(assemblies_directory_path, denovo_file)\n",
    "        \n",
    "        os.rename(contigs_file_path, denovo_file_path)\n",
    "\n",
    "        shutil.rmtree(denovo_directory_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
