{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cb23330d-0231-4dbe-8c5c-0f88b0af0629",
   "metadata": {},
   "source": [
    "# Variant Calling Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55dbb5e6-f260-4b63-beb1-1aae2bea9b24",
   "metadata": {},
   "source": [
    "## Required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9ebb7618-ea7b-4f34-a458-4e04d259cc65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python 3.9 or higher (.removesuffix() in  SAMtools realign NCBI RefSeq assembly)\n",
    "from tkinter import filedialog\n",
    "import os\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "from datetime import datetime\n",
    "import json\n",
    "import time\n",
    "import shutil\n",
    "import requests\n",
    "import zipfile\n",
    "import subprocess\n",
    "import gzip\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ac92eb-6dd0-4c7a-85c0-8b8aaf49fd1d",
   "metadata": {},
   "source": [
    "## Directory Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fa0994d5-6fdc-4ff5-998a-2d3d4650bf31",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = os.getcwd()\n",
    "\n",
    "main_directory_path = filedialog.askdirectory(title=\"Project\", initialdir=cwd)\n",
    "#main_directory_path = \"/home/student/Variant_Calling/Jupyter/PAO1\"\n",
    "Fastq_directory_path = filedialog.askdirectory(title=\"Fastq files\", initialdir=cwd)\n",
    "#Fastq_directory_path = os.path.join(main_directory_path, \"Fastq\")\n",
    "\n",
    "main_subdirectories = [\"Analysis\"]\n",
    "\n",
    "for directory in  main_subdirectories:\n",
    "    main_subdirectory_path = os.path.join(main_directory_path, directory)\n",
    "    if not os.path.exists(main_subdirectory_path):\n",
    "        os.mkdir(main_subdirectory_path)\n",
    "\n",
    "analysis_directory_path = os.path.join(main_directory_path, \"Analysis\")\n",
    "analysis_subdirectories = [\"fastq\",\"assemblies\",\"variant_calling\",\"samfiles\", \"genomes\"]\n",
    "\n",
    "for directory in analysis_subdirectories:\n",
    "    analysis_subdirectory_path = os.path.join(analysis_directory_path, directory)\n",
    "    if not os.path.exists(analysis_subdirectory_path):\n",
    "        os.mkdir(analysis_subdirectory_path)\n",
    "\n",
    "genomes_directory_path = os.path.join(analysis_directory_path, \"genomes\")\n",
    "genomes_subdirectories = [\"annotated\", \"indexed\"]\n",
    "\n",
    "for directory in genomes_subdirectories:\n",
    "    genomes_subdirectory_path = os.path.join(genomes_directory_path, directory)\n",
    "    if not os.path.exists(genomes_subdirectory_path):\n",
    "        os.mkdir(genomes_subdirectory_path)\n",
    "\n",
    "variant_calling_directory_path = os.path.join(analysis_directory_path, \"variant_calling\")\n",
    "variant_calling_subdirectories = [\"annotated\",\"curated\",\"indels\",\"intermediate\",\"raw\",\"snps\",\"totalpileup\",\"vcf\",\"results\"]\n",
    "\n",
    "for directory in variant_calling_subdirectories:\n",
    "    vc_subdirectory_path = os.path.join(variant_calling_directory_path, directory)\n",
    "    if not os.path.exists(vc_subdirectory_path):\n",
    "        os.mkdir(vc_subdirectory_path)\n",
    "\n",
    "# Analysis subdirectories path\n",
    "fastq_directory_path = os.path.join(analysis_directory_path, \"fastq\")\n",
    "samfiles_directory_path = os.path.join(analysis_directory_path, \"samfiles\")\n",
    "assemblies_directory_path = os.path.join(analysis_directory_path, \"assemblies\")\n",
    "\n",
    "## Variant Calling subdirectories path\n",
    "totalpileup_directory_path = os.path.join(variant_calling_directory_path, \"totalpileup\")\n",
    "raw_directory_path = os.path.join(variant_calling_directory_path, \"raw\")\n",
    "intermediate_directory_path = os.path.join(variant_calling_directory_path, \"intermediate\")\n",
    "vcf_directory_path = os.path.join(variant_calling_directory_path, \"vcf\")\n",
    "indels_directory_path = os.path.join(variant_calling_directory_path, \"indels\")\n",
    "snps_directory_path = os.path.join(variant_calling_directory_path, \"snps\")\n",
    "curated_directory_path = os.path.join(variant_calling_directory_path, \"curated\")\n",
    "annotated_directory_path = os.path.join(variant_calling_directory_path, \"annotated\")\n",
    "results_directory_path = os.path.join(variant_calling_directory_path, \"results\")\n",
    "\n",
    "## Genomes subdirectories path\n",
    "indexed_directory_path = os.path.join(genomes_directory_path, \"indexed\")\n",
    "annotated_genomes_directory_path = os.path.join(genomes_directory_path, \"annotated\")\n",
    "\n",
    "# Softwares\n",
    "picard_jar = \"/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar\"\n",
    "samtools = \"/home/student/miniconda3/envs/VariantCalling/bin/samtools\"\n",
    "bowtie2_build = \"/home/student/miniconda3/envs/VariantCalling/bin/bowtie2-build\"\n",
    "bowtie2 = \"/home/student/miniconda3/envs/VariantCalling/bin/bowtie2\"\n",
    "GATK_jar = \"/home/student/Programs/GenomeAnalysisTK-3.4-46/GenomeAnalysisTK.jar\"\n",
    "java_8 = \"/home/student/javav/jdk1.8.0_131/bin/java\"\n",
    "java_21 = \"/home/student/javav/jdk-21.0.6/bin/java\"\n",
    "snpEff_jar = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.jar\"\n",
    "perl = \"/home/student/miniconda3/envs/VariantCalling/bin/perl\"\n",
    "convert2annovar = \"/home/student/annovar/convert2annovar.pl\"\n",
    "vcfEffOnePerLine = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/scripts/vcfEffOnePerLine.pl\"\n",
    "snpSift_jar = \"/home/student/miniconda3/envs/VariantCalling/share/snpsift-5.2-0/SnpSift.jar\"\n",
    "spades = \"/home/student/miniconda3/envs/VariantCalling/bin/spades.py\"\n",
    "buildDbNcbi = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/scripts/buildDbNcbi.sh\"\n",
    "snpeff_dir_path = \"/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eecbf7b-7de5-47e4-bb8f-a02ddc558221",
   "metadata": {},
   "source": [
    "## Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0a79baa-7262-4098-bc1a-4bd1917b09f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fe3c5c7e64c24a38b63773343cbeeaa7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(VBox(children=(Checkbox(value=False, description='Configuration file'), VBox(layout=Layout(dispâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load configuration file widget\n",
    "configuration_widget = widgets.Checkbox(description='Configuration file', value=False)\n",
    "\n",
    "configuration_inp_box = widgets.VBox()\n",
    "configuration_section = widgets.VBox([configuration_widget, configuration_inp_box])\n",
    "\n",
    "def update_configuration_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        upload_file = widgets.FileUpload(accept='.json', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        items.append(upload_file)\n",
    "    configuration_inp_box.children = items\n",
    "    configuration_inp_box.layout.display = 'block' if num > 0 else 'none'\n",
    "\n",
    "def on_configuration_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_configuration_inputs(change['new'])\n",
    "\n",
    "configuration_widget.observe(on_configuration_widget_change, names='value')\n",
    "update_configuration_inputs(configuration_widget.value)\n",
    "\n",
    "# Merge widget\n",
    "merge_widget = widgets.Checkbox(description='Merge .fastq files', value=False)\n",
    "\n",
    "# Download reference genome/s widget\n",
    "download_widget = widgets.BoundedIntText(description='Dowload reference genome', value=0,\n",
    "                                         min=0, max=10, step=1,\n",
    "                                        style={'description_width':'200px'})\n",
    "\n",
    "download_inp_box = widgets.VBox()\n",
    "download_section = widgets.VBox([download_widget, download_inp_box])\n",
    "\n",
    "def update_download_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                               placeholder='GCF_000006765.1',\n",
    "                               style={'description_width':'150px'},\n",
    "                               layout=widgets.Layout(width='300px'))\n",
    "        items.append(textbox)\n",
    "    download_inp_box.children = items\n",
    "    download_inp_box.layout.display = 'block' if num > 0 else 'none'\n",
    "\n",
    "def on_download_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_download_inputs(change['new'])\n",
    "\n",
    "download_widget.observe(on_download_widget_change, names='value')\n",
    "update_download_inputs(download_widget.value)\n",
    "\n",
    "# Genome indexing widget\n",
    "index_widget = widgets.BoundedIntText(description='Genome indexing', value=0,\n",
    "                                     min=0, max=10, step=1,\n",
    "                                     style={'description_width':'150px'})\n",
    "\n",
    "index_inp_box = widgets.VBox()\n",
    "index_section = widgets.VBox([index_widget, index_inp_box])\n",
    "\n",
    "def update_index_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                               placeholder='GCF_000006765.1',\n",
    "                               style={'description_width':'150px'},\n",
    "                               layout=widgets.Layout(width='300px'))\n",
    "        items.append(textbox)\n",
    "    index_inp_box.children = items\n",
    "    index_inp_box.layout.display = 'block' if num > 0 else 'none'\n",
    "\n",
    "def on_index_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_index_inputs(change['new'])\n",
    "\n",
    "index_widget.observe(on_index_widget_change, names='value')\n",
    "update_index_inputs(index_widget.value)\n",
    "\n",
    "# Bowtie2 widget\n",
    "bowtie2_widget = widgets.Checkbox(description='Bowtie2 alignment', value=False)\n",
    "\n",
    "alignment_widget = widgets.BoundedIntText(description='Strains', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "alignment_inp_box = widgets.VBox()\n",
    "alignment_section = widgets.VBox([alignment_widget, alignment_inp_box])\n",
    "\n",
    "def update_alignment_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                              placeholder='GCF_000006765.1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    alignment_inp_box.children = items\n",
    "\n",
    "def on_alignment_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_alignment_inputs(change['new'])\n",
    "\n",
    "alignment_widget.observe(on_alignment_widget_change, names='value')\n",
    "update_alignment_inputs(alignment_widget.value)\n",
    "\n",
    "def toggle_alignment_section(change):\n",
    "    alignment_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "bowtie2_widget.observe(toggle_alignment_section, names='value')\n",
    "alignment_section.layout.display = 'block' if bowtie2_widget.value else 'none'\n",
    "\n",
    "# SAMtools realign assembly widget\n",
    "realign_widget = widgets.Checkbox(description='SAMtools realign assembly', value=False)\n",
    "\n",
    "realignment_widget = widgets.BoundedIntText(description='Strains', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "realignment_inp_box = widgets.VBox()\n",
    "realignment_section = widgets.VBox([realignment_widget, realignment_inp_box])\n",
    "\n",
    "def update_realignment_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Assembly Accession {i+1}',\n",
    "                              placeholder='GCF_000006765.1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    realignment_inp_box.children = items\n",
    "\n",
    "def on_realignment_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_realignment_inputs(change['new'])\n",
    "\n",
    "realignment_widget.observe(on_realignment_widget_change, names='value')\n",
    "update_realignment_inputs(realignment_widget.value)\n",
    "\n",
    "def toggle_realignment_section(change):\n",
    "    realignment_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "realign_widget.observe(toggle_realignment_section, names='value')\n",
    "realignment_section.layout.display = 'block' if realign_widget.value else 'none'\n",
    "\n",
    "# Find SNPs\n",
    "snps_widget = widgets.Checkbox(description='Find SNPs', value=False)\n",
    "\n",
    "# Find indels\n",
    "indels_widget = widgets.Checkbox(description='Find indels', value=False)\n",
    "\n",
    "# Build genome database (snpEff) widget\n",
    "build_widget = widgets.Checkbox(description='Build genome database', value=False)\n",
    "\n",
    "database_widget = widgets.BoundedIntText(description='Databases', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "database_inp_box = widgets.VBox()\n",
    "database_section = widgets.VBox([database_widget, database_inp_box])\n",
    "\n",
    "def update_database_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'NCBI RefSeq {i+1}',\n",
    "                               placeholder='NC_002516.2',\n",
    "                               style={'description_width':'150px'},\n",
    "                               layout=widgets.Layout(width='300px'))\n",
    "        items.append(textbox)\n",
    "    database_inp_box.children = items\n",
    "    database_inp_box.layout.display = 'block' if build_widget.value else 'none'\n",
    "\n",
    "def on_database_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_database_inputs(change['new'])\n",
    "\n",
    "database_widget.observe(on_database_widget_change, names='value')\n",
    "update_database_inputs(database_widget.value)\n",
    "\n",
    "def toggle_database_section(change):\n",
    "    database_section.layout.display = 'block' if change['new'] else 'none'\n",
    "    if change['new']:\n",
    "        update_database_inputs(database_widget.value)\n",
    "\n",
    "build_widget.observe(toggle_database_section, names='value')\n",
    "database_section.layout.display = 'block' if build_widget.value else 'none'\n",
    "\n",
    "# snpEff annotation\n",
    "snpeff_widget = widgets.Checkbox(description='SNP/indel annotation', value=False)\n",
    "\n",
    "ncbi_refseq_widget = widgets.BoundedIntText(description='NCBI RefSeq', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "ncbi_inp_box = widgets.VBox()\n",
    "ncbi_section = widgets.VBox([ncbi_refseq_widget, ncbi_inp_box])\n",
    "\n",
    "def update_ncbi_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'NCBI RefSeq {i+1}',\n",
    "                              placeholder='NC_002516.2',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    ncbi_inp_box.children = items\n",
    "\n",
    "def on_ncbi_refseq_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_ncbi_inputs(change['new'])\n",
    "\n",
    "ncbi_refseq_widget.observe(on_ncbi_refseq_widget_change, names='value')\n",
    "update_ncbi_inputs(ncbi_refseq_widget.value)\n",
    "\n",
    "def toggle_ncbi_section(change):\n",
    "    ncbi_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "snpeff_widget.observe(toggle_ncbi_section, names='value')\n",
    "ncbi_section.layout.display = 'block' if snpeff_widget.value else 'none'\n",
    "\n",
    "# Filter SNP/indel by parental\n",
    "filter_widget = widgets.Checkbox(description='Filter SNP/indel by parental', value=False)\n",
    "\n",
    "parental_widget = widgets.BoundedIntText(description='Parental', value=1,\n",
    "                                         min=1, max=10, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "parental_inp_box = widgets.VBox()\n",
    "parental_section = widgets.VBox([parental_widget, parental_inp_box])\n",
    "\n",
    "def update_parental_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Parental {i+1}',\n",
    "                              placeholder='PAO1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    parental_inp_box.children = items\n",
    "\n",
    "def on_parental_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_parental_inputs(change['new'])\n",
    "\n",
    "parental_widget.observe(on_parental_widget_change, names='value')\n",
    "update_parental_inputs(parental_widget.value)\n",
    "\n",
    "def toggle_parental_section(change):\n",
    "    parental_section.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "filter_widget.observe(toggle_parental_section, names='value')\n",
    "parental_section.layout.display = 'block' if filter_widget.value else 'none'\n",
    "\n",
    "# Create sample/mutation table widget\n",
    "table_widget = widgets.Checkbox(description='Sample/mutation table', value=False)\n",
    "\n",
    "threshold_widget = widgets.BoundedIntText(description='Threshold', value=1,\n",
    "                                         min=1, step=1,\n",
    "                                         style={'description_width':'150px'})\n",
    "\n",
    "strains_widget = widgets.BoundedIntText(description='Strains', value=1,\n",
    "                                        min=1, max=10, step=1,\n",
    "                                        style={'description_width':'150px'})\n",
    "\n",
    "strains_inp_box = widgets.VBox()\n",
    "strains_section = widgets.VBox([strains_widget, strains_inp_box])\n",
    "\n",
    "def update_strains_inputs(num):\n",
    "    items = []\n",
    "    for i in range(num):\n",
    "        textbox = widgets.Text(description=f'Strain {i+1}',\n",
    "                              placeholder='PAO1',\n",
    "                              style={'description_width':'150px'},\n",
    "                              layout=widgets.Layout(width='300px'))\n",
    "        file_upload = widgets.FileUpload(accept='.txt', multiple=False,\n",
    "                                        layout=widgets.Layout(width='300px'))\n",
    "        input_pair = widgets.HBox([textbox, file_upload])\n",
    "        items.append(input_pair)\n",
    "    strains_inp_box.children = items\n",
    "\n",
    "def on_strains_widget_change(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_strains_inputs(change['new'])\n",
    "\n",
    "strains_widget.observe(on_strains_widget_change, names='value')\n",
    "update_strains_inputs(strains_widget.value)\n",
    "\n",
    "table_children_box = widgets.VBox([threshold_widget, strains_section])\n",
    "table_section = widgets.VBox([table_widget, table_children_box])\n",
    "table_children_box.layout.display = 'block' if table_widget.value else 'none'\n",
    "\n",
    "def toggle_table_section(change):\n",
    "    table_children_box.layout.display = 'block' if change['new'] else 'none'\n",
    "\n",
    "table_widget.observe(toggle_table_section, names='value')\n",
    "\n",
    "# De novo genome assembly (SPAdes) widget\n",
    "denovo_widget = widgets.Checkbox(description='De novo genome assembly', value=False)\n",
    "\n",
    "# Compress large files widget\n",
    "compress_widget = widgets.Checkbox(description='Compress large files', value=False,\n",
    "                           layout=widgets.Layout(width='300px'))\n",
    "\n",
    "# Comfirm selected options widget\n",
    "confirm_options = widgets.Checkbox(description='Confirm', value=False, indent=False)\n",
    "\n",
    "# configuration_widget behavior\n",
    "def set_disabled_recursive_configuration(widget, disabled=True):\n",
    "    if isinstance(widget, widgets.Box):\n",
    "        new_children = []\n",
    "        for child in widget.children:\n",
    "            if isinstance(child, widgets.FileUpload):\n",
    "                # Create a new FileUpload widget\n",
    "                new_widget = widgets.FileUpload(\n",
    "                    accept=child.accept,\n",
    "                    multiple=child.multiple,\n",
    "                    layout=child.layout\n",
    "                )\n",
    "                new_children.append(new_widget)\n",
    "            else:\n",
    "                set_disabled_recursive_configuration(child, disabled)\n",
    "                new_children.append(child)\n",
    "        widget.children = tuple(new_children)\n",
    "    else:\n",
    "        if hasattr(widget, 'disabled'):\n",
    "            widget.disabled = disabled\n",
    "        if isinstance(widget, widgets.Checkbox):\n",
    "            widget.value = False\n",
    "        elif isinstance(widget, widgets.BoundedIntText):\n",
    "            widget.value = 0\n",
    "\n",
    "# confirm_widget behavior\n",
    "def set_disabled_recursive_confirm(widget, disabled=True):\n",
    "    widget.disabled = disabled\n",
    "    if hasattr(widget, 'children'):\n",
    "        for child in widget.children:\n",
    "            set_disabled_recursive_confirm(child, disabled)\n",
    "\n",
    "def on_configuration_widget(change):\n",
    "    if change['name'] == 'value':\n",
    "        update_configuration_inputs(change['new'])\n",
    "        state = change['new']\n",
    "\n",
    "        merge_widget.disabled = state\n",
    "        merge_widget.value = False\n",
    "\n",
    "        download_widget.disabled = state\n",
    "        download_widget.value = False\n",
    "        set_disabled_recursive_configuration(download_section, state)\n",
    "\n",
    "        index_widget.disabled = state\n",
    "        index_widget.value = False\n",
    "        set_disabled_recursive_configuration(index_section, state)\n",
    "        \n",
    "        bowtie2_widget.disabled = state\n",
    "        bowtie2_widget.value = False\n",
    "        set_disabled_recursive_configuration(alignment_section, state)\n",
    "        \n",
    "        realign_widget.disabled = state\n",
    "        realign_widget.value = False\n",
    "        set_disabled_recursive_configuration(realignment_section, state)\n",
    "        \n",
    "        snps_widget.disabled = state\n",
    "        snps_widget.value = False\n",
    "        \n",
    "        indels_widget.disabled = state\n",
    "        indels_widget.value = False\n",
    "        \n",
    "        build_widget.disabled = state\n",
    "        build_widget.value = False\n",
    "        set_disabled_recursive_configuration(database_section, state)\n",
    "        \n",
    "        snpeff_widget.disabled = state\n",
    "        snpeff_widget.value = False\n",
    "        set_disabled_recursive_configuration(ncbi_section, state)\n",
    "        \n",
    "        filter_widget.disabled = state\n",
    "        filter_widget.value = False\n",
    "        set_disabled_recursive_configuration(parental_section, state)\n",
    "\n",
    "        table_widget.disabled = state\n",
    "        table_widget.value = False\n",
    "        set_disabled_recursive_configuration(strains_section, state)\n",
    "        \n",
    "        denovo_widget.disabled = state\n",
    "        denovo_widget.value = False\n",
    "\n",
    "        compress_widget.disabled = state\n",
    "        compress_widget.value = False\n",
    "\n",
    "def on_confirm_options_widget(change):\n",
    "    if change['type'] == 'change' and change['name'] == 'value':\n",
    "        state = change['new']\n",
    "\n",
    "        set_disabled_recursive_confirm(configuration_section, state)\n",
    "        merge_widget.disabled = state\n",
    "        set_disabled_recursive_confirm(download_section, state)\n",
    "        set_disabled_recursive_confirm(index_section, state)\n",
    "        set_disabled_recursive_confirm(bowtie2_widget, state)\n",
    "        set_disabled_recursive_confirm(alignment_section, state)\n",
    "        set_disabled_recursive_confirm(realign_widget, state)\n",
    "        set_disabled_recursive_confirm(realignment_section, state)\n",
    "        snps_widget.disabled = state\n",
    "        indels_widget.disabled = state\n",
    "        set_disabled_recursive_confirm(build_widget, state)\n",
    "        set_disabled_recursive_confirm(database_section, state)\n",
    "        set_disabled_recursive_confirm(snpeff_widget, state)\n",
    "        set_disabled_recursive_confirm(ncbi_section, state)\n",
    "        set_disabled_recursive_confirm(filter_widget, state)\n",
    "        set_disabled_recursive_confirm(parental_section, state)\n",
    "        set_disabled_recursive_confirm(table_section, state)\n",
    "        denovo_widget.disabled = state\n",
    "        compress_widget.disabled = state\n",
    "\n",
    "configuration_widget.observe(on_configuration_widget, names='value')\n",
    "confirm_options.observe(on_confirm_options_widget, names='value')\n",
    "\n",
    "column = widgets.VBox([\n",
    "    configuration_section,\n",
    "    merge_widget,\n",
    "    download_section,\n",
    "    index_widget,\n",
    "    index_inp_box,\n",
    "    bowtie2_widget,\n",
    "    alignment_section,\n",
    "    realign_widget,\n",
    "    realignment_section,\n",
    "    snps_widget,\n",
    "    indels_widget,\n",
    "    build_widget,\n",
    "    database_section,\n",
    "    snpeff_widget,\n",
    "    ncbi_section,\n",
    "    filter_widget,\n",
    "    parental_section,\n",
    "    table_section,\n",
    "    denovo_widget,\n",
    "    compress_widget,\n",
    "    confirm_options\n",
    "])\n",
    "\n",
    "display(column)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f37eda25-9eab-41a9-9d2c-f725e8bfce5e",
   "metadata": {},
   "source": [
    "### Configuration file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d60684fc-2a13-4e73-a892-cfc4fcf19712",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Configuration file create in /media/student/EXTERNAL_USB/Prova/config.json\n"
     ]
    }
   ],
   "source": [
    "if configuration_widget.value:\n",
    "    # Load configuration from json file\n",
    "    try:\n",
    "        configuration_file = configuration_inp_box.children[0].value[0][\"content\"]\n",
    "        configuration = json.loads(bytes(configuration_file).decode(\"utf-8\"))\n",
    "        print(\"Configuration file uploaded successfully.\")\n",
    "        \n",
    "    except IndexError:\n",
    "        print(\"No file uploaded.\")\n",
    "    except KeyError:\n",
    "        print(\"Failed to load configuration file.\")\n",
    "    \n",
    "else:\n",
    "    # Load parental SNP/indel files\n",
    "    if filter_widget.value:\n",
    "        parental_files = []\n",
    "        for i in range(0, parental_widget.value):\n",
    "            strain = parental_inp_box.children[i].children[0].value\n",
    "            snp_file = filedialog.askopenfile(title=f\"{strain} SNPs\", initialdir=\"/home/student/Variant_Calling/Jupyter\")\n",
    "            indels_file = filedialog.askopenfile(title=f\"{strain} indels\", initialdir=\"/home/student/Variant_Calling/Jupyter\")\n",
    "\n",
    "            parental_files.append((snp_file, indels_file))\n",
    "\n",
    "    # Create configuration\n",
    "    configuration = {\n",
    "        \"merge\": merge_widget.value,\n",
    "        \"download\": {\n",
    "            \"n\": download_widget.value,\n",
    "            \"assemblies\": [box.value for box in download_inp_box.children]\n",
    "        },\n",
    "        \"indexing\": {\n",
    "            \"n\": index_widget.value,\n",
    "            \"assemblies\": [box.value for box in index_inp_box.children]\n",
    "        },\n",
    "        \"bowtie2\": {\n",
    "            \"enabled\": bowtie2_widget.value,\n",
    "            \"n\": alignment_widget.value,\n",
    "            \"organisms\": [\n",
    "                {\n",
    "                    \"accession\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "            }\n",
    "            for pair in alignment_inp_box.children\n",
    "        ]\n",
    "    },\n",
    "        \"realign\": {\n",
    "            \"enabled\": realign_widget.value,\n",
    "            \"n\": realignment_widget.value,\n",
    "            \"organisms\": [\n",
    "                {\n",
    "                    \"accession\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                    if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                    else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "                }\n",
    "                for pair in realignment_inp_box.children\n",
    "            ]\n",
    "        },\n",
    "        \"snps\": snps_widget.value,\n",
    "        \"indels\": indels_widget.value,\n",
    "        \"build\": {\n",
    "            \"enabled\": build_widget.value,\n",
    "            \"n\": database_widget.value,\n",
    "            \"databases\": [box.value for box in database_inp_box.children]\n",
    "        },\n",
    "        \"SNP_indel_annotation\": {\n",
    "            \"enabled\": snpeff_widget.value,\n",
    "            \"n\": ncbi_refseq_widget.value,\n",
    "            \"Reference_Sequences\": [\n",
    "                {\n",
    "                    \"accession\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "            }\n",
    "            for pair in ncbi_inp_box.children\n",
    "        ]\n",
    "    },\n",
    "        \"filter_by_parental\": {\n",
    "            \"enabled\": filter_widget.value,\n",
    "            \"n\": parental_widget.value,\n",
    "            \"Parental\": [\n",
    "                {\n",
    "                    \"strain\": pair.children[0].value,\n",
    "                    \"snps_path\": snp.name,\n",
    "                    \"indels_path\": indels.name,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "            }\n",
    "            for pair, (snp, indels) in zip(parental_inp_box.children, parental_files)\n",
    "        ]\n",
    "    },\n",
    "        \"table\": {\n",
    "            \"enabled\": table_widget.value,\n",
    "            \"threshold\": threshold_widget.value,\n",
    "            \"n\": strains_widget.value,\n",
    "            \"strains\": [\n",
    "                {\n",
    "                    \"strain\": pair.children[0].value,\n",
    "                    \"file\": pair.children[1].value[0]['name']\n",
    "                        if pair.children[1].value and len(pair.children[1].value) > 0\n",
    "                        else None,\n",
    "                    \"samples\": [\n",
    "                        line.strip()\n",
    "                        for line in bytes(pair.children[1].value[0]['content']).decode(\"utf-8\").splitlines()\n",
    "                        if line.strip()\n",
    "                    ] if pair.children[1].value and len(pair.children[1].value) > 0 else []\n",
    "                }\n",
    "                for pair in strains_inp_box.children\n",
    "            ]\n",
    "        },\n",
    "        \"denovo\": denovo_widget.value,\n",
    "        \"compress\": compress_widget.value\n",
    "    }\n",
    "\n",
    "    now = datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "    configuration[\"generated_at\"] = now\n",
    "\n",
    "    config_file_path = os.path.join(main_directory_path, \"config.json\")\n",
    "\n",
    "    # Create configuration file\n",
    "    with open(config_file_path, 'w') as config_file:\n",
    "        json.dump(configuration, config_file, indent=4)\n",
    "\n",
    "    print(\"Configuration file create in\", config_file_path)\n",
    "\n",
    "# Initialize sample dictionaries\n",
    "bowtie2_sample_genome_dict = {}\n",
    "realign_sample_genome_dict = {}\n",
    "annotation_sample_genome_dict = {}\n",
    "filter_sample_strain_dict = {}\n",
    "table_sample_strain_dict = {}\n",
    "\n",
    "for organism in configuration[\"bowtie2\"][\"organisms\"]:\n",
    "    accession = organism[\"accession\"]\n",
    "    samples = organism[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in bowtie2_sample_genome_dict:\n",
    "            bowtie2_sample_genome_dict[sample] = accession\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for organism in configuration[\"realign\"][\"organisms\"]:\n",
    "    accession = organism[\"accession\"]\n",
    "    samples = organism[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in realign_sample_genome_dict:\n",
    "            realign_sample_genome_dict[sample] = accession\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for refseq in configuration[\"SNP_indel_annotation\"][\"Reference_Sequences\"]:\n",
    "    accession = refseq[\"accession\"]\n",
    "    samples = refseq[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in annotation_sample_genome_dict:\n",
    "            annotation_sample_genome_dict[sample] = accession\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for strain in configuration[\"filter_by_parental\"][\"Parental\"]:\n",
    "    parental = strain[\"strain\"]\n",
    "    samples = strain[\"samples\"]\n",
    "    for sample in samples:\n",
    "        if sample not in filter_sample_strain_dict:\n",
    "            filter_sample_strain_dict[sample] = parental\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")\n",
    "\n",
    "for strain in configuration[\"table\"][\"strains\"]:\n",
    "    strain_name = strain[\"strain\"]\n",
    "    for sample in strain[\"samples\"]:\n",
    "        if sample not in table_sample_strain_dict:\n",
    "            table_sample_strain_dict[sample] = strain_name\n",
    "        else:\n",
    "            print(f\"{sample} is duplicated.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8e10dd3-bf19-44c1-813a-35a300af87f6",
   "metadata": {},
   "source": [
    "## Merge fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "16766547-6862-47c2-9117-5b10d3918101",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These samples are going to be processed: \n",
      "  - PA14wt0253G\n",
      "  - PAO1wt12P\n",
      "\n",
      "PA14wt0253G processed in 27.54 seconds.\n",
      "PAO1wt12P processed in 18.25 seconds.\n",
      "\n",
      "Merge finished!\n"
     ]
    }
   ],
   "source": [
    "# Merge fastq files\n",
    "if configuration[\"merge\"]:\n",
    "\n",
    "    files = os.listdir(Fastq_directory_path)\n",
    "\n",
    "    samples = {file.split(\"_\")[0] for file in files}\n",
    "\n",
    "    print(\"These samples are going to be processed: \")\n",
    "    for sample in samples:\n",
    "        print(f\"  - {sample}\")\n",
    "    \n",
    "    print()\n",
    "\n",
    "    for sample in samples:\n",
    "        start_time = time.time()\n",
    "    \n",
    "        L1_directory = [file for file in files if file.startswith(f\"{sample}_L1\")]\n",
    "        L1_directory_path = os.path.join(Fastq_directory_path, \"\".join(L1_directory))\n",
    "\n",
    "        L2_directory = [file for file in files if file.startswith(f\"{sample}_L2\")]\n",
    "        L2_directory_path = os.path.join(Fastq_directory_path, \"\".join(L2_directory))\n",
    "\n",
    "        L1_files = os.listdir(L1_directory_path)\n",
    "        L2_files = os.listdir(L2_directory_path)\n",
    "    \n",
    "        L1_R1_file = [file for file in L1_files if \"_R1_\" in file]\n",
    "        L1_R1_file_path = os.path.join(L1_directory_path, \"\".join(L1_R1_file))\n",
    "        L2_R1_file = [file for file in L2_files if \"_R1_\" in file]\n",
    "        L2_R1_file_path = os.path.join(L2_directory_path, \"\".join(L2_R1_file))\n",
    "\n",
    "        R1_files_path = [L1_R1_file_path, L2_R1_file_path]\n",
    "\n",
    "        L1_R2_file = [file for file in L1_files if \"_R2_\" in file]\n",
    "        L1_R2_file_path = os.path.join(L1_directory_path, \"\".join(L1_R2_file))\n",
    "        L2_R2_file = [file for file in L2_files if \"_R2_\" in file]\n",
    "        L2_R2_file_path = os.path.join(L2_directory_path, \"\".join(L2_R2_file))\n",
    "\n",
    "        R2_files_path = [L1_R2_file_path, L2_R2_file_path]\n",
    "\n",
    "        ## Merge R1 files\n",
    "        merged_R1_file = f\"{sample}_R1.fastq.gz\"\n",
    "        merged_R1_file_path = os.path.join(fastq_directory_path, merged_R1_file)\n",
    "\n",
    "        with open(merged_R1_file_path, 'wb') as R1_file:\n",
    "            for file in R1_files_path:\n",
    "                with open(file, 'rb') as my_file:\n",
    "                    shutil.copyfileobj(my_file, R1_file)\n",
    "\n",
    "        ## Merge R2 files\n",
    "        merged_R2_file = f\"{sample}_R2.fastq.gz\"\n",
    "        merged_R2_file_path = os.path.join(fastq_directory_path, merged_R2_file)\n",
    "\n",
    "        with open(merged_R2_file_path, 'wb') as R2_file:\n",
    "            for file in R2_files_path:\n",
    "                with open(file, 'rb') as my_file:\n",
    "                    shutil.copyfileobj(my_file, R2_file)\n",
    "    \n",
    "        end_time = time.time()\n",
    "\n",
    "        execution_time = end_time - start_time\n",
    "        print(f\"{sample} processed in {execution_time:.2f} seconds.\")\n",
    "\n",
    "    print()\n",
    "    print(\"Merge finished!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe0af47a-cac8-42f9-b9fe-3c530265b261",
   "metadata": {},
   "source": [
    "## Download reference genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c28cf1f-73e8-4d85-9b0b-db3254cf4bee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCBI RefSeq assembly GCF_000006765.1 downloaded successfully.\n",
      "NCBI RefSeq assembly GCF_000006765.1 unzipped successfully.\n",
      "\n",
      "NCBI RefSeq assembly GCF_000014625.1 downloaded successfully.\n",
      "NCBI RefSeq assembly GCF_000014625.1 unzipped successfully.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"download\"][\"n\"]:\n",
    "\n",
    "    assemblies = configuration[\"download\"][\"assemblies\"]\n",
    "\n",
    "    # Download reference genome .fastq file from NCBI\n",
    "    for assembly_id in assemblies:\n",
    "        url = f\"https://api.ncbi.nlm.nih.gov/datasets/v2/genome/accession/{assembly_id}/download?include_annotation_type=GENOME_FASTA&hydrated=FULLY_HYDRATED\"\n",
    "\n",
    "        try:\n",
    "            response = requests.get(url, stream=True, timeout=60)\n",
    "            response.raise_for_status()\n",
    "        except requests.exceptions.RequestException as download_error:\n",
    "            print(f\"Error downloading the assembly {assembly_id}: {download_error}\")\n",
    "            continue\n",
    "\n",
    "        if response.status_code == 200:\n",
    "            ref_genome_file = f\"{assembly_id}.zip\"\n",
    "            ref_genome_file_path = os.path.join(genomes_directory_path, ref_genome_file)\n",
    "            with open(ref_genome_file_path, 'wb') as genome_file:\n",
    "                for chunk in response.iter_content(chunk_size=8192):\n",
    "                    if chunk:\n",
    "                        genome_file.write(chunk)\n",
    "            print(f\"NCBI RefSeq assembly {assembly_id} downloaded successfully.\")\n",
    "\n",
    "            assembly_id_directory = os.path.join(genomes_directory_path, assembly_id)\n",
    "            if not os.path.exists(assembly_id_directory):\n",
    "                os.mkdir(assembly_id_directory)\n",
    "            \n",
    "            with zipfile.ZipFile(ref_genome_file_path, 'r') as zip_ref:\n",
    "                zip_ref.extractall(path=assembly_id_directory)\n",
    "            print(f\"NCBI RefSeq assembly {assembly_id} unzipped successfully.\")\n",
    "            print()\n",
    "\n",
    "            os.remove(ref_genome_file_path)\n",
    "        \n",
    "        else:\n",
    "            print(f\"Error downloading the NCBI RefSeq assembly {assembly_id}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64bb1ae5-cd5e-4423-a67a-4225fc791a88",
   "metadata": {},
   "source": [
    "## Genome indexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bada94c6-af67-4fb9-bb53-1d80308c2019",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexing GCF_000006765.1 genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO\t2025-05-12 19:52:33\tNormalizeFasta\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    NormalizeFasta -I /media/student/EXTERNAL_USB/Prova/Analysis/genomes/GCF_000006765.1/ncbi_dataset/data/GCF_000006765.1/GCF_000006765.1_ASM676v1_genomic.fna -O /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta\n",
      "**********\n",
      "\n",
      "\n",
      "19:52:33.280 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 19:52:33 CEST 2025] NormalizeFasta INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/GCF_000006765.1/ncbi_dataset/data/GCF_000006765.1/GCF_000006765.1_ASM676v1_genomic.fna OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta    LINE_LENGTH=100 TRUNCATE_SEQUENCE_NAMES_AT_WHITESPACE=false VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 19:52:34 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "[Mon May 12 19:52:36 CEST 2025] picard.reference.NormalizeFasta done. Elapsed time: 0.06 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO\t2025-05-12 19:52:42\tCreateSequenceDictionary\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    CreateSequenceDictionary -R /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta -O /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.dict\n",
      "**********\n",
      "\n",
      "\n",
      "19:52:42.749 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 19:52:43 CEST 2025] CreateSequenceDictionary OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.dict REFERENCE=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 19:52:43 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "[Mon May 12 19:52:44 CEST 2025] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.03 minutes.\n",
      "Runtime.totalMemory()=168296448\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1566101\n",
      "Using parameters --bmax 1174576 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1174576 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 5; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 783050 (target: 1174575)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (1174576) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 640778 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 640779 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (1174576) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 934176 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 934177 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (1174576) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 682832 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 682833 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (1174576) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 908487 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 908488 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (1174576) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 321897 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 321898 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (1174576) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 940198 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 940199 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (1174576) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 977225 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 977226 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (1174576) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 858804 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 858805 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1056134\n",
      "fchr[G]: 3158821\n",
      "fchr[T]: 5225454\n",
      "fchr[$]: 6264404\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6282714 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.1.bt2.tmp\n",
      "Wrote 1566108 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6264404\n",
      "    bwtLen: 6264405\n",
      "    sz: 1566101\n",
      "    bwtSz: 1566102\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 391526\n",
      "    offsSz: 1566104\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 32628\n",
      "    numLines: 32628\n",
      "    ebwtTotLen: 2088192\n",
      "    ebwtTotSz: 2088192\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:05\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1566101\n",
      "Using parameters --bmax 1174576 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1174576 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 894914 (target: 1174575)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (1174576) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 960633 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 960634 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (1174576) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 806333 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 806334 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (1174576) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 464337 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 464338 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (1174576) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1050739 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1050740 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (1174576) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1008209 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1008210 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (1174576) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1058957 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1058958 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (1174576) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 915190 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 915191 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1056134\n",
      "fchr[G]: 3158821\n",
      "fchr[T]: 5225454\n",
      "fchr[$]: 6264404\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6282714 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.1.bt2.tmp\n",
      "Wrote 1566108 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6264404\n",
      "    bwtLen: 6264405\n",
      "    sz: 1566101\n",
      "    bwtSz: 1566102\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 391526\n",
      "    offsSz: 1566104\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 32628\n",
      "    numLines: 32628\n",
      "    ebwtTotLen: 2088192\n",
      "    ebwtTotSz: 2088192\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:07\n",
      "GCF_000006765.1 indexed successfully!\n",
      "Indexing GCF_000014625.1 genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.3.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.3.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.4.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.4.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.1.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.2.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.2.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.1.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.2.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/index/GCF_000006765.1_index.rev.2.bt2\n",
      "INFO\t2025-05-12 19:53:02\tNormalizeFasta\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    NormalizeFasta -I /media/student/EXTERNAL_USB/Prova/Analysis/genomes/GCF_000014625.1/ncbi_dataset/data/GCF_000014625.1/GCF_000014625.1_ASM1462v1_genomic.fna -O /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta\n",
      "**********\n",
      "\n",
      "\n",
      "19:53:02.522 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 19:53:03 CEST 2025] NormalizeFasta INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/GCF_000014625.1/ncbi_dataset/data/GCF_000014625.1/GCF_000014625.1_ASM1462v1_genomic.fna OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta    LINE_LENGTH=100 TRUNCATE_SEQUENCE_NAMES_AT_WHITESPACE=false VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 19:53:03 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "[Mon May 12 19:53:05 CEST 2025] picard.reference.NormalizeFasta done. Elapsed time: 0.04 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO\t2025-05-12 19:53:10\tCreateSequenceDictionary\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    CreateSequenceDictionary -R /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta -O /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.dict\n",
      "**********\n",
      "\n",
      "\n",
      "19:53:10.442 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 19:53:10 CEST 2025] CreateSequenceDictionary OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.dict REFERENCE=/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 19:53:11 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "[Mon May 12 19:53:12 CEST 2025] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.04 minutes.\n",
      "Runtime.totalMemory()=168296448\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"/media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta\n",
      "Reading reference sizes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1634409\n",
      "Using parameters --bmax 1225807 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1225807 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 7; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 817204 (target: 1225806)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (1225807) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 913362 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 913363 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (1225807) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 504196 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 504197 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (1225807) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 901912 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 901913 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (1225807) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 463480 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 463481 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (1225807) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1225739 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1225740 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (1225807) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1123799 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1123800 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (1225807) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 269747 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 269748 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (1225807) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 1135395 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1135396 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1102316\n",
      "fchr[G]: 3269680\n",
      "fchr[T]: 5436267\n",
      "fchr[$]: 6537637\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6373902 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.1.bt2.tmp\n",
      "Wrote 1634416 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6537637\n",
      "    bwtLen: 6537638\n",
      "    sz: 1634410\n",
      "    bwtSz: 1634410\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 408603\n",
      "    offsSz: 1634412\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 34051\n",
      "    numLines: 34051\n",
      "    ebwtTotLen: 2179264\n",
      "    ebwtTotSz: 2179264\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:12\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 1634409\n",
      "Using parameters --bmax 1225807 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 1225807 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:01\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 726403 (target: 1225806)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 9\n",
      "  Reserving size (1225807) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 125530 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 125531 for bucket 1\n",
      "Getting block 2 of 9\n",
      "  Reserving size (1225807) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1151245 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1151246 for bucket 2\n",
      "Getting block 3 of 9\n",
      "  Reserving size (1225807) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 855141 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 855142 for bucket 3\n",
      "Getting block 4 of 9\n",
      "  Reserving size (1225807) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1133498 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1133499 for bucket 4\n",
      "Getting block 5 of 9\n",
      "  Reserving size (1225807) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 288814 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 288815 for bucket 5\n",
      "Getting block 6 of 9\n",
      "  Reserving size (1225807) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 974385 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 974386 for bucket 6\n",
      "Getting block 7 of 9\n",
      "  Reserving size (1225807) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 303803 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 303804 for bucket 7\n",
      "Getting block 8 of 9\n",
      "  Reserving size (1225807) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 1088162 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1088163 for bucket 8\n",
      "Getting block 9 of 9\n",
      "  Reserving size (1225807) for bucket 9\n",
      "  Calculating Z arrays for bucket 9\n",
      "  Entering block accumulator loop for bucket 9:\n",
      "  bucket 9: 10%\n",
      "  bucket 9: 20%\n",
      "  bucket 9: 30%\n",
      "  bucket 9: 40%\n",
      "  bucket 9: 50%\n",
      "  bucket 9: 60%\n",
      "  bucket 9: 70%\n",
      "  bucket 9: 80%\n",
      "  bucket 9: 90%\n",
      "  bucket 9: 100%\n",
      "  Sorting block of length 617051 for bucket 9\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 617052 for bucket 9\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1102316\n",
      "fchr[G]: 3269680\n",
      "fchr[T]: 5436267\n",
      "fchr[$]: 6537637\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 6373902 bytes to primary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.1.bt2.tmp\n",
      "Wrote 1634416 bytes to secondary EBWT file: /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 6537637\n",
      "    bwtLen: 6537638\n",
      "    sz: 1634410\n",
      "    bwtSz: 1634410\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 408603\n",
      "    offsSz: 1634412\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 34051\n",
      "    numLines: 34051\n",
      "    ebwtTotLen: 2179264\n",
      "    ebwtTotSz: 2179264\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:13\n",
      "GCF_000014625.1 indexed successfully!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.3.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.3.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.4.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.4.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.1.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.2.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.2.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.1.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.1.bt2\n",
      "Renaming /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.2.bt2.tmp to /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/index/GCF_000014625.1_index.rev.2.bt2\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"indexing\"][\"n\"]:\n",
    "\n",
    "    assemblies = configuration[\"indexing\"][\"assemblies\"]\n",
    "\n",
    "    for assembly_id in assemblies:\n",
    "        print(f\"Indexing {assembly_id} genome\")\n",
    "        assembly_fasta_file_directory_path = os.path.join(genomes_directory_path, assembly_id,\n",
    "                                                          \"ncbi_dataset/data\", assembly_id)\n",
    "    \n",
    "        fasta_file = os.listdir(assembly_fasta_file_directory_path)\n",
    "        fasta_file_path = os.path.join(assembly_fasta_file_directory_path, \"\".join(fasta_file))\n",
    "\n",
    "        indexed_assembly_directory_path = os.path.join(genomes_directory_path, \"indexed\", assembly_id)\n",
    "        if not os.path.exists(indexed_assembly_directory_path):\n",
    "            os.mkdir(indexed_assembly_directory_path)\n",
    "        \n",
    "        N_fasta_file_path = os.path.join(indexed_assembly_directory_path, f\"N_{assembly_id}.fasta\")\n",
    "\n",
    "        # Normalize Fasta\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"NormalizeFasta\",\n",
    "            \"I=\" + fasta_file_path,\n",
    "            \"O=\" + N_fasta_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        # Creating .dict file\n",
    "        reference = N_fasta_file_path\n",
    "        dict_file_path = os.path.join(indexed_assembly_directory_path, f\"N_{assembly_id}.dict\")\n",
    "\n",
    "        if os.path.exists(dict_file_path):\n",
    "            print(f\"{assembly_id}.dict already exists.\")\n",
    "        else:\n",
    "            command = [\n",
    "                \"java\", \"-jar\", picard_jar, \"CreateSequenceDictionary\",\n",
    "                \"R=\" + reference,\n",
    "                \"O=\" + dict_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, check=True)\n",
    "\n",
    "        # Creating .fasta.fai files\n",
    "        command = [\n",
    "            samtools, \"faidx\", reference\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        # Building bowtie2 index\n",
    "        index_directory_path = os.path.join(indexed_assembly_directory_path, \"index\")\n",
    "\n",
    "        if not os.path.exists(index_directory_path):\n",
    "            os.mkdir(index_directory_path)\n",
    "\n",
    "        index_file_path = os.path.join(index_directory_path, f\"{assembly_id}_index\")\n",
    "    \n",
    "        command = [\n",
    "            bowtie2_build, reference, index_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "        print(f\"{assembly_id} indexed successfully!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47df1025-481c-4caf-a0a4-145e2ec7021a",
   "metadata": {},
   "source": [
    "## Bowtie2 alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dc79896f-66a4-4628-bfb9-4032d2b4aeee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 samples will be aligned with GCF_000006765.1 if possible...\n",
      "\n",
      "Aligning PAO1wt12P...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3281626 reads; of these:\n",
      "  3281626 (100.00%) were paired; of these:\n",
      "    73435 (2.24%) aligned concordantly 0 times\n",
      "    3151065 (96.02%) aligned concordantly exactly 1 time\n",
      "    57126 (1.74%) aligned concordantly >1 times\n",
      "    ----\n",
      "    73435 pairs aligned concordantly 0 times; of these:\n",
      "      12472 (16.98%) aligned discordantly 1 time\n",
      "    ----\n",
      "    60963 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      121926 mates make up the pairs; of these:\n",
      "        80327 (65.88%) aligned 0 times\n",
      "        40207 (32.98%) aligned exactly 1 time\n",
      "        1392 (1.14%) aligned >1 times\n",
      "98.78% overall alignment rate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PAO1wt12P aligned. Execution time 2143.74 s\n",
      "\n",
      "Aligned 1/1 with GCF_000006765.1\n",
      "\n",
      "1 samples will be aligned with GCF_000014625.1 if possible...\n",
      "\n",
      "Aligning PA14wt0253G...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6236138 reads; of these:\n",
      "  6236138 (100.00%) were paired; of these:\n",
      "    122207 (1.96%) aligned concordantly 0 times\n",
      "    5984387 (95.96%) aligned concordantly exactly 1 time\n",
      "    129544 (2.08%) aligned concordantly >1 times\n",
      "    ----\n",
      "    122207 pairs aligned concordantly 0 times; of these:\n",
      "      39330 (32.18%) aligned discordantly 1 time\n",
      "    ----\n",
      "    82877 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      165754 mates make up the pairs; of these:\n",
      "        107625 (64.93%) aligned 0 times\n",
      "        55538 (33.51%) aligned exactly 1 time\n",
      "        2591 (1.56%) aligned >1 times\n",
      "99.14% overall alignment rate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PA14wt0253G aligned. Execution time 3861.96 s\n",
      "\n",
      "Aligned 1/1 with GCF_000014625.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"bowtie2\"][\"enabled\"]:\n",
    "    for organism in configuration[\"bowtie2\"][\"organisms\"]:\n",
    "        genome = organism[\"accession\"]\n",
    "        samples = organism[\"samples\"]\n",
    "            \n",
    "        print(f\"{len(samples)} samples will be aligned with {genome} if possible...\")\n",
    "\n",
    "        index_assembly_directory_path = os.path.join(genomes_directory_path, \"indexed\", genome, \"index\", f\"{genome}_index\")\n",
    "\n",
    "        aligned_count = 0\n",
    "\n",
    "        print()\n",
    "    \n",
    "        for sample in samples:\n",
    "            aligned_count += 1        \n",
    "            start_time = time.time()\n",
    "\n",
    "            # Decompress R1 and R2 files\n",
    "            compressed_R1_file_path = os.path.join(fastq_directory_path, f\"{sample}_R1.fastq.gz\")\n",
    "            compressed_R2_file_path = os.path.join(fastq_directory_path, f\"{sample}_R2.fastq.gz\")\n",
    "\n",
    "            decompressed_R1_file_path = os.path.join(fastq_directory_path, f\"{sample}_R1.fastq\")\n",
    "            decompressed_R2_file_path = os.path.join(fastq_directory_path, f\"{sample}_R2.fastq\")\n",
    "            \n",
    "            with gzip.open(compressed_R1_file_path, 'rb') as compressed_R1_file:\n",
    "                with open(decompressed_R1_file_path, 'wb') as decompressed_R1_file:\n",
    "                    shutil.copyfileobj(compressed_R1_file, decompressed_R1_file)\n",
    "\n",
    "            with gzip.open(compressed_R2_file_path, 'rb') as compressed_R2_file:\n",
    "                with open(decompressed_R2_file_path, 'wb') as decompressed_R2_file:\n",
    "                    shutil.copyfileobj(compressed_R2_file, decompressed_R2_file)\n",
    "            \n",
    "            reference_genome_index = index_assembly_directory_path\n",
    "            \n",
    "            # Bowtie2 alignment\n",
    "            print(f\"Aligning {sample}...\")\n",
    "            \n",
    "            R1_file = decompressed_R1_file_path\n",
    "            R2_file = decompressed_R2_file_path\n",
    "\n",
    "            samfiles_file_path = os.path.join(samfiles_directory_path, f\"{sample}_map{genome}.sam\")\n",
    "    \n",
    "            command = [\n",
    "                bowtie2, \"--phred33\", \"-x\", reference_genome_index,\n",
    "                \"-q\", \"-1\", R1_file, \"-2\", R2_file,\n",
    "                \"-X\", \"1000\", \"-S\", samfiles_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, check=True)\n",
    "            \n",
    "            end_time = time.time()\n",
    "            execution_time = end_time - start_time\n",
    "            \n",
    "            print()\n",
    "            print(f\"{sample} aligned. Execution time {execution_time:.2f} s\")\n",
    "            print()\n",
    "            print(f\"Aligned {aligned_count}/{len(samples)} with {genome}\")\n",
    "            \n",
    "            if configuration[\"compress\"]:\n",
    "\n",
    "                print()\n",
    "                print(f\"Compressing .sam {sample} file\")\n",
    "                \n",
    "                # Compress resulting .sam file and remove .sam file\n",
    "                compressed_sam_file_path = os.path.join(samfiles_directory_path, f\"{samfiles_file_path}.gz\")\n",
    "\n",
    "                with open(samfiles_file_path, 'rb') as sam_file:\n",
    "                    with gzip.open(compressed_sam_file_path, 'wb') as compressed_sam_file:\n",
    "                        shutil.copyfileobj(sam_file, compressed_sam_file)\n",
    "\n",
    "                os.remove(samfiles_file_path)\n",
    "            \n",
    "            # Remove decompressed .fastq files\n",
    "            os.remove(decompressed_R1_file_path)\n",
    "            os.remove(decompressed_R2_file_path)\n",
    "\n",
    "        \n",
    "        print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b61aa96e-818a-4c03-86ce-05f0df1e0063",
   "metadata": {},
   "source": [
    "## SAMtools realign NCBI RefSeq assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "585b30c9-e654-46de-9757-51e63a6bf2ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing PA14wt0253G...\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[samopen] SAM header is present: 1 sequences.\n",
      "INFO\t2025-05-12 21:37:38\tSortSam\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    SortSam -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam -SORT_ORDER coordinate\n",
      "**********\n",
      "\n",
      "\n",
      "21:37:38.206 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 21:37:38 CEST 2025] SortSam INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.bam OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam SORT_ORDER=coordinate    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 21:37:38 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 21:39:07\tSortSam\tRead    10,000,000 records.  Elapsed time: 00:01:28s.  Time for last 10,000,000:   88s.  Last read position: NC_008463.1:5,857,452.  Last read name: ADX00106:131:HTWKWDRX5:2:2217:25870:16861\n",
      "INFO\t2025-05-12 21:39:23\tSortSam\tFinished reading inputs, merging and writing to output now.\n",
      "INFO\t2025-05-12 21:40:37\tSortSam\tWrote    10,000,000 records from a sorting collection.  Elapsed time: 00:02:58s.  Time for last 10,000,000:   72s.  Last read position: NC_008463.1:5,392,762\n",
      "[Mon May 12 21:40:55 CEST 2025] picard.sam.SortSam done. Elapsed time: 3.29 minutes.\n",
      "Runtime.totalMemory()=872009728\n",
      "INFO\t2025-05-12 21:40:57\tAddOrReplaceReadGroups\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    AddOrReplaceReadGroups -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam -LB /media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS -PL Illumina -PU /media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS -SM /media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS\n",
      "**********\n",
      "\n",
      "\n",
      "21:40:57.731 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 21:40:57 CEST 2025] AddOrReplaceReadGroups INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted.bam OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam RGLB=/media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS RGPL=Illumina RGPU=/media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS RGSM=/media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS    RGID=1 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 21:40:58 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 21:40:58\tAddOrReplaceReadGroups\tCreated read-group ID=1 PL=Illumina LB=/media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS SM=/media/student/EXTERNAL_USB/Prova/PA14wt0253G_WGS\n",
      "\n",
      "INFO\t2025-05-12 21:41:10\tAddOrReplaceReadGroups\tProcessed     1,000,000 records.  Elapsed time: 00:00:11s.  Time for last 1,000,000:   11s.  Last read position: NC_008463.1:484,706\n",
      "INFO\t2025-05-12 21:41:19\tAddOrReplaceReadGroups\tProcessed     2,000,000 records.  Elapsed time: 00:00:21s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:989,179\n",
      "INFO\t2025-05-12 21:41:29\tAddOrReplaceReadGroups\tProcessed     3,000,000 records.  Elapsed time: 00:00:31s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:1,489,827\n",
      "INFO\t2025-05-12 21:41:39\tAddOrReplaceReadGroups\tProcessed     4,000,000 records.  Elapsed time: 00:00:40s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:2,027,382\n",
      "INFO\t2025-05-12 21:41:48\tAddOrReplaceReadGroups\tProcessed     5,000,000 records.  Elapsed time: 00:00:50s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:2,589,297\n",
      "INFO\t2025-05-12 21:41:57\tAddOrReplaceReadGroups\tProcessed     6,000,000 records.  Elapsed time: 00:00:59s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:3,205,044\n",
      "INFO\t2025-05-12 21:42:07\tAddOrReplaceReadGroups\tProcessed     7,000,000 records.  Elapsed time: 00:01:09s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:3,835,047\n",
      "INFO\t2025-05-12 21:42:16\tAddOrReplaceReadGroups\tProcessed     8,000,000 records.  Elapsed time: 00:01:18s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:4,410,079\n",
      "INFO\t2025-05-12 21:42:25\tAddOrReplaceReadGroups\tProcessed     9,000,000 records.  Elapsed time: 00:01:27s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:4,916,471\n",
      "INFO\t2025-05-12 21:42:35\tAddOrReplaceReadGroups\tProcessed    10,000,000 records.  Elapsed time: 00:01:37s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:5,392,762\n",
      "INFO\t2025-05-12 21:42:45\tAddOrReplaceReadGroups\tProcessed    11,000,000 records.  Elapsed time: 00:01:47s.  Time for last 1,000,000:   10s.  Last read position: NC_008463.1:5,883,424\n",
      "INFO\t2025-05-12 21:42:54\tAddOrReplaceReadGroups\tProcessed    12,000,000 records.  Elapsed time: 00:01:56s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:6,348,978\n",
      "[Mon May 12 21:43:00 CEST 2025] picard.sam.AddOrReplaceReadGroups done. Elapsed time: 2.05 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO\t2025-05-12 21:43:02\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -METRICS_FILE /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_metrics.txt\n",
      "**********\n",
      "\n",
      "\n",
      "21:43:02.228 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 21:43:02 CEST 2025] MarkDuplicates INPUT=[/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg.bam] OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam METRICS_FILE=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_DUP_STRATEGY=FLOW_QUALITY_SUM_STRATEGY FLOW_USE_END_IN_UNPAIRED_READS=false FLOW_USE_UNPAIRED_CLIPPED_END=false FLOW_UNPAIRED_END_UNCERTAINTY=0 FLOW_UNPAIRED_START_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true REMOVE_DUPLICATES=false ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 21:43:02 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 21:43:02\tMarkDuplicates\tStart of doWork freeMemory: 158623280; totalMemory: 168361984; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 21:43:02\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2025-05-12 21:43:02\tMarkDuplicates\tWill retain up to 9688405 data points before spilling to disk.\n",
      "INFO\t2025-05-12 21:43:08\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:05s.  Time for last 1,000,000:    5s.  Last read position: NC_008463.1:484,706\n",
      "INFO\t2025-05-12 21:43:08\tMarkDuplicates\tTracking 152 as yet unmatched pairs. 152 records in RAM.\n",
      "INFO\t2025-05-12 21:43:12\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:09s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:989,179\n",
      "INFO\t2025-05-12 21:43:12\tMarkDuplicates\tTracking 156 as yet unmatched pairs. 156 records in RAM.\n",
      "INFO\t2025-05-12 21:43:15\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:12s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:1,489,827\n",
      "INFO\t2025-05-12 21:43:15\tMarkDuplicates\tTracking 184 as yet unmatched pairs. 184 records in RAM.\n",
      "INFO\t2025-05-12 21:43:18\tMarkDuplicates\tRead     4,000,000 records.  Elapsed time: 00:00:15s.  Time for last 1,000,000:    2s.  Last read position: NC_008463.1:2,027,382\n",
      "INFO\t2025-05-12 21:43:18\tMarkDuplicates\tTracking 272 as yet unmatched pairs. 272 records in RAM.\n",
      "INFO\t2025-05-12 21:43:22\tMarkDuplicates\tRead     5,000,000 records.  Elapsed time: 00:00:19s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:2,589,297\n",
      "INFO\t2025-05-12 21:43:22\tMarkDuplicates\tTracking 266 as yet unmatched pairs. 266 records in RAM.\n",
      "INFO\t2025-05-12 21:43:25\tMarkDuplicates\tRead     6,000,000 records.  Elapsed time: 00:00:22s.  Time for last 1,000,000:    3s.  Last read position: NC_008463.1:3,205,044\n",
      "INFO\t2025-05-12 21:43:25\tMarkDuplicates\tTracking 228 as yet unmatched pairs. 228 records in RAM.\n",
      "INFO\t2025-05-12 21:43:28\tMarkDuplicates\tRead     7,000,000 records.  Elapsed time: 00:00:25s.  Time for last 1,000,000:    2s.  Last read position: NC_008463.1:3,835,047\n",
      "INFO\t2025-05-12 21:43:28\tMarkDuplicates\tTracking 196 as yet unmatched pairs. 196 records in RAM.\n",
      "INFO\t2025-05-12 21:43:30\tMarkDuplicates\tRead     8,000,000 records.  Elapsed time: 00:00:28s.  Time for last 1,000,000:    2s.  Last read position: NC_008463.1:4,410,079\n",
      "INFO\t2025-05-12 21:43:30\tMarkDuplicates\tTracking 198 as yet unmatched pairs. 198 records in RAM.\n",
      "INFO\t2025-05-12 21:43:35\tMarkDuplicates\tRead     9,000,000 records.  Elapsed time: 00:00:32s.  Time for last 1,000,000:    4s.  Last read position: NC_008463.1:4,916,471\n",
      "INFO\t2025-05-12 21:43:35\tMarkDuplicates\tTracking 152 as yet unmatched pairs. 152 records in RAM.\n",
      "INFO\t2025-05-12 21:43:44\tMarkDuplicates\tRead    10,000,000 records.  Elapsed time: 00:00:42s.  Time for last 1,000,000:    9s.  Last read position: NC_008463.1:5,392,762\n",
      "INFO\t2025-05-12 21:43:44\tMarkDuplicates\tTracking 224 as yet unmatched pairs. 224 records in RAM.\n",
      "INFO\t2025-05-12 21:43:47\tMarkDuplicates\tRead    11,000,000 records.  Elapsed time: 00:00:44s.  Time for last 1,000,000:    2s.  Last read position: NC_008463.1:5,883,424\n",
      "INFO\t2025-05-12 21:43:47\tMarkDuplicates\tTracking 170 as yet unmatched pairs. 170 records in RAM.\n",
      "INFO\t2025-05-12 21:43:50\tMarkDuplicates\tRead    12,000,000 records.  Elapsed time: 00:00:47s.  Time for last 1,000,000:    2s.  Last read position: NC_008463.1:6,348,978\n",
      "INFO\t2025-05-12 21:43:50\tMarkDuplicates\tTracking 132 as yet unmatched pairs. 132 records in RAM.\n",
      "INFO\t2025-05-12 21:43:51\tMarkDuplicates\tRead 12421314 records. 0 pairs never matched.\n",
      "INFO\t2025-05-12 21:43:54\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 1730008128; totalMemory: 2312515584; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 21:43:54\tMarkDuplicates\tWill retain up to 83562496 duplicate indices before spilling to disk.\n",
      "INFO\t2025-05-12 21:43:54\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2025-05-12 21:43:57\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2025-05-12 21:44:02\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2025-05-12 21:44:02\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1622066384; totalMemory: 2312515584; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 21:44:02\tMarkDuplicates\tMarking 1663515 records as duplicates.\n",
      "INFO\t2025-05-12 21:44:02\tMarkDuplicates\tFound 83754 optical duplicate clusters.\n",
      "INFO\t2025-05-12 21:44:02\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2025-05-12 21:45:42\tMarkDuplicates\tWritten    10,000,000 records.  Elapsed time: 00:01:39s.  Time for last 10,000,000:   99s.  Last read position: NC_008463.1:5,392,762\n",
      "INFO\t2025-05-12 21:46:06\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2025-05-12 21:46:06\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2025-05-12 21:46:06\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2025-05-12 21:46:07\tMarkDuplicates\tBefore output close freeMemory: 2290258176; totalMemory: 2312515584; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 21:46:07\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2025-05-12 21:46:07\tMarkDuplicates\tAfter output close freeMemory: 2076977960; totalMemory: 2098061312; maxMemory: 2673999872\n",
      "[Mon May 12 21:46:07 CEST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 3.09 minutes.\n",
      "Runtime.totalMemory()=2098061312\n",
      "INFO\t2025-05-12 21:46:09\tBuildBamIndex\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    BuildBamIndex -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bai\n",
      "**********\n",
      "\n",
      "\n",
      "21:46:09.269 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 21:46:09 CEST 2025] BuildBamIndex INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bai    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 21:46:09 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 21:46:33\tBuildBamIndex\tSuccessfully wrote bam index file /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bai\n",
      "[Mon May 12 21:46:33 CEST 2025] picard.sam.BuildBamIndex done. Elapsed time: 0.41 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO  21:46:37,430 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  21:46:37,436 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  21:46:37,436 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  21:46:37,436 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  21:46:37,447 HelpFormatter - Program Args: -T RealignerTargetCreator -I /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta -o /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.realigned.intervals \n",
      "INFO  21:46:37,456 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  21:46:37,457 HelpFormatter - Date/Time: 2025/05/12 21:46:37 \n",
      "INFO  21:46:37,462 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  21:46:37,462 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  21:46:38,450 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  21:46:38,906 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  21:46:38,917 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  21:46:38,956 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  21:46:39,140 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  21:46:39,221 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  21:46:39,224 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  21:46:39,224 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  21:46:39,225 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  21:47:09,238 ProgressMeter - NC_008463.1:730297    720896.0    30.0 s      41.0 s       11.2%     4.5 m       4.0 m \n",
      "INFO  21:47:39,240 ProgressMeter - NC_008463.1:1928129   1916928.0    60.0 s      31.0 s       29.5%     3.4 m       2.4 m \n",
      "INFO  21:48:09,241 ProgressMeter - NC_008463.1:3313669   3309568.0    90.0 s      27.0 s       50.7%     3.0 m      87.0 s \n",
      "INFO  21:48:39,242 ProgressMeter - NC_008463.1:4644373   4636672.0   120.0 s      25.0 s       71.0%     2.8 m      48.0 s \n",
      "INFO  21:49:09,244 ProgressMeter - NC_008463.1:5806537   5799936.0     2.5 m      25.0 s       88.8%     2.8 m      18.0 s \n",
      "INFO  21:49:31,303 ProgressMeter -            done   6537648.0     2.9 m      26.0 s      100.0%     2.9 m       0.0 s \n",
      "INFO  21:49:31,304 ProgressMeter - Total runtime 172.08 secs, 2.87 min, 0.05 hours \n",
      "INFO  21:49:31,306 MicroScheduler - 1738403 reads were filtered out during the traversal out of approximately 12526629 total reads (13.88%) \n",
      "INFO  21:49:31,309 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  21:49:31,309 MicroScheduler -   -> 0 reads (0.00% of total) failing BadMateFilter \n",
      "INFO  21:49:31,310 MicroScheduler -   -> 1675996 reads (13.38% of total) failing DuplicateReadFilter \n",
      "INFO  21:49:31,310 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  21:49:31,311 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  21:49:31,311 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  21:49:31,311 MicroScheduler -   -> 62407 reads (0.50% of total) failing MappingQualityZeroFilter \n",
      "INFO  21:49:31,312 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  21:49:31,312 MicroScheduler -   -> 0 reads (0.00% of total) failing Platform454Filter \n",
      "INFO  21:49:31,315 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "INFO  21:49:32,635 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "INFO  21:49:36,005 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  21:49:36,011 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  21:49:36,011 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  21:49:36,012 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  21:49:36,016 HelpFormatter - Program Args: -T IndelRealigner -I /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000014625.1/N_GCF_000014625.1.fasta --maxConsensuses 60 --maxReadsForConsensuses 240 --maxReadsForRealignment 6000 --targetIntervals /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.realigned.intervals -o /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PA14wt0253G_mapGCF_000014625.1.realigned.bam \n",
      "INFO  21:49:36,027 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  21:49:36,030 HelpFormatter - Date/Time: 2025/05/12 21:49:36 \n",
      "INFO  21:49:36,030 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  21:49:36,031 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  21:49:37,132 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  21:49:37,229 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  21:49:37,242 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  21:49:37,276 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.03 \n",
      "INFO  21:49:37,455 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  21:49:37,460 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  21:49:37,463 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  21:49:37,463 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  21:49:37,464 ProgressMeter -        Location |     reads | elapsed |     reads | completed | runtime |   runtime \n",
      "INFO  21:49:37,606 ReadShardBalancer$1 - Loading BAM index data \n",
      "INFO  21:49:37,607 ReadShardBalancer$1 - Done loading BAM index data \n",
      "INFO  21:50:07,473 ProgressMeter - NC_008463.1:221919    400022.0    30.0 s      75.0 s        3.4%    14.7 m      14.2 m \n",
      "INFO  21:50:38,691 ProgressMeter - NC_008463.1:533136   1000061.0    61.0 s      61.0 s        8.2%    12.5 m      11.5 m \n",
      "INFO  21:51:08,851 ProgressMeter - NC_008463.1:895655   1800092.0    91.0 s      50.0 s       13.7%    11.1 m       9.6 m \n",
      "INFO  21:51:41,302 ProgressMeter - NC_008463.1:1246365   2400120.0     2.1 m      51.0 s       19.1%    10.8 m       8.7 m \n",
      "INFO  21:52:12,201 ProgressMeter - NC_008463.1:1624603   3200152.0     2.6 m      48.0 s       24.8%    10.3 m       7.8 m \n",
      "INFO  21:52:44,052 ProgressMeter - NC_008463.1:2027602   3900174.0     3.1 m      47.0 s       31.0%    10.0 m       6.9 m \n",
      "INFO  21:53:15,256 ProgressMeter - NC_008463.1:2416771   4600201.0     3.6 m      47.0 s       37.0%     9.8 m       6.2 m \n",
      "INFO  21:53:46,106 ProgressMeter - NC_008463.1:2851205   5400235.0     4.1 m      46.0 s       43.6%     9.5 m       5.3 m \n",
      "INFO  21:54:16,571 ProgressMeter - NC_008463.1:3350618   6200271.0     4.7 m      45.0 s       51.3%     9.1 m       4.4 m \n",
      "INFO  21:54:46,796 ProgressMeter - NC_008463.1:3835318   6900286.0     5.2 m      44.0 s       58.7%     8.8 m       3.6 m \n",
      "INFO  21:55:16,797 ProgressMeter - NC_008463.1:4224734   7600328.0     5.7 m      44.0 s       64.6%     8.7 m       3.1 m \n",
      "INFO  21:55:46,799 ProgressMeter - NC_008463.1:4616727   8400371.0     6.2 m      43.0 s       70.6%     8.7 m       2.6 m \n",
      "INFO  21:56:17,761 ProgressMeter - NC_008463.1:5007649   9100431.0     6.7 m      43.0 s       76.6%     8.7 m       2.0 m \n",
      "INFO  21:56:48,288 ProgressMeter - NC_008463.1:5315351   9800471.0     7.2 m      43.0 s       81.3%     8.8 m      98.0 s \n",
      "INFO  21:57:18,885 ProgressMeter - NC_008463.1:5623506   1.0400525E7     7.7 m      44.0 s       86.0%     8.9 m      74.0 s \n",
      "INFO  21:57:48,885 ProgressMeter - NC_008463.1:5956830   1.1100575E7     8.2 m      44.0 s       91.1%     9.0 m      47.0 s \n",
      "INFO  21:58:18,887 ProgressMeter - NC_008463.1:6287667   1.1800608E7     8.7 m      44.0 s       96.2%     9.0 m      20.0 s \n",
      "INFO  21:58:43,584 ProgressMeter -            done   1.2472276E7     9.1 m      43.0 s      100.0%     9.1 m       0.0 s \n",
      "INFO  21:58:43,585 ProgressMeter - Total runtime 546.12 secs, 9.10 min, 0.15 hours \n",
      "INFO  21:58:43,591 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 12472276 total reads (0.00%) \n",
      "INFO  21:58:43,592 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  21:58:43,592 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  21:58:44,951 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "[bam_sort_core] merging from 10 files...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PA14wt0253G processed!!\n",
      "\n",
      "Processing PAO1wt12P...\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[samopen] SAM header is present: 1 sequences.\n",
      "INFO\t2025-05-12 22:40:12\tSortSam\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    SortSam -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam -SORT_ORDER coordinate\n",
      "**********\n",
      "\n",
      "\n",
      "22:40:12.546 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 22:40:12 CEST 2025] SortSam INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.bam OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam SORT_ORDER=coordinate    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 22:40:13 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 22:41:11\tSortSam\tFinished reading inputs, merging and writing to output now.\n",
      "[Mon May 12 22:42:02 CEST 2025] picard.sam.SortSam done. Elapsed time: 1.83 minutes.\n",
      "Runtime.totalMemory()=1110773760\n",
      "INFO\t2025-05-12 22:42:04\tAddOrReplaceReadGroups\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    AddOrReplaceReadGroups -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam -LB /media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS -PL Illumina -PU /media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS -SM /media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS\n",
      "**********\n",
      "\n",
      "\n",
      "22:42:04.634 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 22:42:04 CEST 2025] AddOrReplaceReadGroups INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted.bam OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam RGLB=/media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS RGPL=Illumina RGPU=/media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS RGSM=/media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS    RGID=1 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 22:42:05 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 22:42:05\tAddOrReplaceReadGroups\tCreated read-group ID=1 PL=Illumina LB=/media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS SM=/media/student/EXTERNAL_USB/Prova/PAO1wt12P_WGS\n",
      "\n",
      "INFO\t2025-05-12 22:42:17\tAddOrReplaceReadGroups\tProcessed     1,000,000 records.  Elapsed time: 00:00:12s.  Time for last 1,000,000:   12s.  Last read position: NC_002516.2:830,564\n",
      "INFO\t2025-05-12 22:42:28\tAddOrReplaceReadGroups\tProcessed     2,000,000 records.  Elapsed time: 00:00:22s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:1,870,533\n",
      "INFO\t2025-05-12 22:42:38\tAddOrReplaceReadGroups\tProcessed     3,000,000 records.  Elapsed time: 00:00:33s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:3,045,985\n",
      "INFO\t2025-05-12 22:42:49\tAddOrReplaceReadGroups\tProcessed     4,000,000 records.  Elapsed time: 00:00:43s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:4,077,540\n",
      "INFO\t2025-05-12 22:42:59\tAddOrReplaceReadGroups\tProcessed     5,000,000 records.  Elapsed time: 00:00:53s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:4,995,202\n",
      "INFO\t2025-05-12 22:43:09\tAddOrReplaceReadGroups\tProcessed     6,000,000 records.  Elapsed time: 00:01:03s.  Time for last 1,000,000:   10s.  Last read position: NC_002516.2:5,859,008\n",
      "[Mon May 12 22:43:16 CEST 2025] picard.sam.AddOrReplaceReadGroups done. Elapsed time: 1.19 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO\t2025-05-12 22:43:18\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -METRICS_FILE /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_metrics.txt\n",
      "**********\n",
      "\n",
      "\n",
      "22:43:18.393 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 22:43:18 CEST 2025] MarkDuplicates INPUT=[/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg.bam] OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam METRICS_FILE=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_metrics.txt    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_DUP_STRATEGY=FLOW_QUALITY_SUM_STRATEGY FLOW_USE_END_IN_UNPAIRED_READS=false FLOW_USE_UNPAIRED_CLIPPED_END=false FLOW_UNPAIRED_END_UNCERTAINTY=0 FLOW_UNPAIRED_START_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true REMOVE_DUPLICATES=false ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 22:43:18 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 22:43:19\tMarkDuplicates\tStart of doWork freeMemory: 158624656; totalMemory: 168361984; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 22:43:19\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2025-05-12 22:43:19\tMarkDuplicates\tWill retain up to 9688405 data points before spilling to disk.\n",
      "INFO\t2025-05-12 22:43:24\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:05s.  Time for last 1,000,000:    5s.  Last read position: NC_002516.2:830,564\n",
      "INFO\t2025-05-12 22:43:24\tMarkDuplicates\tTracking 290 as yet unmatched pairs. 290 records in RAM.\n",
      "INFO\t2025-05-12 22:43:28\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:09s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:1,870,533\n",
      "INFO\t2025-05-12 22:43:28\tMarkDuplicates\tTracking 366 as yet unmatched pairs. 366 records in RAM.\n",
      "INFO\t2025-05-12 22:43:32\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:13s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:3,045,985\n",
      "INFO\t2025-05-12 22:43:32\tMarkDuplicates\tTracking 382 as yet unmatched pairs. 382 records in RAM.\n",
      "INFO\t2025-05-12 22:43:35\tMarkDuplicates\tRead     4,000,000 records.  Elapsed time: 00:00:16s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:4,077,540\n",
      "INFO\t2025-05-12 22:43:35\tMarkDuplicates\tTracking 400 as yet unmatched pairs. 400 records in RAM.\n",
      "INFO\t2025-05-12 22:43:39\tMarkDuplicates\tRead     5,000,000 records.  Elapsed time: 00:00:20s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:4,995,202\n",
      "INFO\t2025-05-12 22:43:39\tMarkDuplicates\tTracking 320 as yet unmatched pairs. 320 records in RAM.\n",
      "INFO\t2025-05-12 22:43:42\tMarkDuplicates\tRead     6,000,000 records.  Elapsed time: 00:00:23s.  Time for last 1,000,000:    3s.  Last read position: NC_002516.2:5,859,008\n",
      "INFO\t2025-05-12 22:43:42\tMarkDuplicates\tTracking 162 as yet unmatched pairs. 162 records in RAM.\n",
      "INFO\t2025-05-12 22:43:44\tMarkDuplicates\tRead 6524016 records. 0 pairs never matched.\n",
      "INFO\t2025-05-12 22:43:45\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 1102540152; totalMemory: 1898381312; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 22:43:45\tMarkDuplicates\tWill retain up to 83562496 duplicate indices before spilling to disk.\n",
      "INFO\t2025-05-12 22:43:46\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2025-05-12 22:43:50\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2025-05-12 22:43:51\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2025-05-12 22:43:52\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1981822616; totalMemory: 2673999872; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 22:43:52\tMarkDuplicates\tMarking 575241 records as duplicates.\n",
      "INFO\t2025-05-12 22:43:52\tMarkDuplicates\tFound 37777 optical duplicate clusters.\n",
      "INFO\t2025-05-12 22:43:52\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2025-05-12 22:45:02\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2025-05-12 22:45:02\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2025-05-12 22:45:02\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2025-05-12 22:45:02\tMarkDuplicates\tBefore output close freeMemory: 2650009688; totalMemory: 2673999872; maxMemory: 2673999872\n",
      "INFO\t2025-05-12 22:45:02\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2025-05-12 22:45:02\tMarkDuplicates\tAfter output close freeMemory: 2400851248; totalMemory: 2423472128; maxMemory: 2673999872\n",
      "[Mon May 12 22:45:02 CEST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 1.73 minutes.\n",
      "Runtime.totalMemory()=2423472128\n",
      "INFO\t2025-05-12 22:45:04\tBuildBamIndex\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** \n",
      "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    BuildBamIndex -INPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -OUTPUT /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bai\n",
      "**********\n",
      "\n",
      "\n",
      "22:45:04.434 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/student/miniconda3/envs/VariantCalling/share/picard-3.4.0-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon May 12 22:45:04 CEST 2025] BuildBamIndex INPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam OUTPUT=/media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bai    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon May 12 22:45:04 CEST 2025] Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; OpenJDK 64-Bit Server VM 17.0.3-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: 3.4.0\n",
      "INFO\t2025-05-12 22:45:17\tBuildBamIndex\tSuccessfully wrote bam index file /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bai\n",
      "[Mon May 12 22:45:17 CEST 2025] picard.sam.BuildBamIndex done. Elapsed time: 0.23 minutes.\n",
      "Runtime.totalMemory()=168296448\n",
      "INFO  22:45:21,580 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  22:45:21,583 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  22:45:21,584 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  22:45:21,584 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  22:45:21,588 HelpFormatter - Program Args: -T RealignerTargetCreator -I /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta -o /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.realigned.intervals \n",
      "INFO  22:45:21,598 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  22:45:21,599 HelpFormatter - Date/Time: 2025/05/12 22:45:21 \n",
      "INFO  22:45:21,601 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  22:45:21,602 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  22:45:22,667 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  22:45:23,140 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  22:45:23,154 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  22:45:23,193 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  22:45:23,366 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  22:45:23,484 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  22:45:23,486 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  22:45:23,487 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  22:45:23,488 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  22:45:53,499 ProgressMeter - NC_002516.2:1406441   1392640.0    30.0 s      21.0 s       22.5%     2.2 m     103.0 s \n",
      "INFO  22:46:23,500 ProgressMeter - NC_002516.2:4145069   4128768.0    60.0 s      14.0 s       66.2%    90.0 s      30.0 s \n",
      "INFO  22:46:53,502 ProgressMeter - NC_002516.2:6264389   6258688.0    90.0 s      14.0 s      100.0%    90.0 s       0.0 s \n",
      "INFO  22:46:54,038 ProgressMeter -            done   6264404.0    90.0 s      14.0 s      100.0%    90.0 s       0.0 s \n",
      "INFO  22:46:54,038 ProgressMeter - Total runtime 90.55 secs, 1.51 min, 0.03 hours \n",
      "INFO  22:46:54,049 MicroScheduler - 624941 reads were filtered out during the traversal out of approximately 6581326 total reads (9.50%) \n",
      "INFO  22:46:54,054 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  22:46:54,054 MicroScheduler -   -> 0 reads (0.00% of total) failing BadMateFilter \n",
      "INFO  22:46:54,054 MicroScheduler -   -> 579127 reads (8.80% of total) failing DuplicateReadFilter \n",
      "INFO  22:46:54,054 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  22:46:54,055 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  22:46:54,055 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  22:46:54,058 MicroScheduler -   -> 45814 reads (0.70% of total) failing MappingQualityZeroFilter \n",
      "INFO  22:46:54,062 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  22:46:54,062 MicroScheduler -   -> 0 reads (0.00% of total) failing Platform454Filter \n",
      "INFO  22:46:54,062 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "INFO  22:46:55,398 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "INFO  22:46:58,994 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  22:46:59,004 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.4-46-gbc02625, Compiled 2015/07/09 17:38:12 \n",
      "INFO  22:46:59,005 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  22:46:59,005 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  22:46:59,008 HelpFormatter - Program Args: -T IndelRealigner -I /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1_sorted_addrg_dedup.bam -R /media/student/EXTERNAL_USB/Prova/Analysis/genomes/indexed/GCF_000006765.1/N_GCF_000006765.1.fasta --maxConsensuses 60 --maxReadsForConsensuses 240 --maxReadsForRealignment 6000 --targetIntervals /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.realigned.intervals -o /media/student/EXTERNAL_USB/Prova/Analysis/variant_calling/intermediate/PAO1wt12P_mapGCF_000006765.1.realigned.bam \n",
      "INFO  22:46:59,017 HelpFormatter - Executing as student@ubuntum0151 on Linux 5.15.0-139-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_131-b11. \n",
      "INFO  22:46:59,019 HelpFormatter - Date/Time: 2025/05/12 22:46:58 \n",
      "INFO  22:46:59,020 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  22:46:59,022 HelpFormatter - --------------------------------------------------------------------------------- \n",
      "INFO  22:47:00,274 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  22:47:00,444 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  22:47:00,468 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  22:47:00,527 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.06 \n",
      "INFO  22:47:00,796 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  22:47:00,801 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  22:47:00,814 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  22:47:00,816 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  22:47:00,817 ProgressMeter -        Location |     reads | elapsed |     reads | completed | runtime |   runtime \n",
      "INFO  22:47:00,993 ReadShardBalancer$1 - Loading BAM index data \n",
      "INFO  22:47:00,996 ReadShardBalancer$1 - Done loading BAM index data \n",
      "INFO  22:47:30,832 ProgressMeter - NC_002516.2:397546    400018.0    30.0 s      75.0 s        6.3%     7.9 m       7.4 m \n",
      "INFO  22:48:02,118 ProgressMeter - NC_002516.2:1001843   1100044.0    61.0 s      55.0 s       16.0%     6.4 m       5.3 m \n",
      "INFO  22:48:32,515 ProgressMeter - NC_002516.2:1791089   1900060.0    91.0 s      48.0 s       28.6%     5.3 m       3.8 m \n",
      "INFO  22:49:04,202 ProgressMeter - NC_002516.2:2702756   2600067.0     2.1 m      47.0 s       43.1%     4.8 m       2.7 m \n",
      "INFO  22:49:34,704 ProgressMeter - NC_002516.2:3670829   3500091.0     2.6 m      43.0 s       58.6%     4.4 m     108.0 s \n",
      "INFO  22:50:05,174 ProgressMeter - NC_002516.2:4304776   4200111.0     3.1 m      43.0 s       68.7%     4.5 m      83.0 s \n",
      "INFO  22:50:35,175 ProgressMeter - NC_002516.2:4995448   4900133.0     3.6 m      43.0 s       79.7%     4.5 m      54.0 s \n",
      "INFO  22:51:05,555 ProgressMeter - NC_002516.2:5690328   5700150.0     4.1 m      42.0 s       90.8%     4.5 m      24.0 s \n",
      "INFO  22:51:35,699 ProgressMeter -            done   6563252.0     4.6 m      41.0 s      100.0%     4.6 m       0.0 s \n",
      "INFO  22:51:35,701 ProgressMeter - Total runtime 274.89 secs, 4.58 min, 0.08 hours \n",
      "INFO  22:51:35,707 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 6563252 total reads (0.00%) \n",
      "INFO  22:51:35,708 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  22:51:35,708 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  22:51:37,262 GATKRunReport - Uploaded run statistics report to AWS S3 \n",
      "[bam_sort_core] merging from 5 files...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PAO1wt12P processed!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"realign\"][\"enabled\"]:\n",
    "    \n",
    "    sam_files = os.listdir(samfiles_directory_path)\n",
    "\n",
    "    for sam_file in sam_files:\n",
    "        sample = sam_file.split(\"_\")[0]\n",
    "        \n",
    "        print(f\"Processing {sample}...\")\n",
    "        print()\n",
    "    \n",
    "        sam_file_path = os.path.join(samfiles_directory_path, sam_file)\n",
    "\n",
    "        if sam_file.endswith(\".sam.gz\"):\n",
    "            bam_file = sam_file.removesuffix(\".sam.gz\") + \".bam\"\n",
    "            #bam_file = sam_file[:-7]+\"bam\"\n",
    "        else:\n",
    "            bam_file = sam_file.removesuffix(\".sam\") + \".bam\"\n",
    "            #bam_file = sam_file[:-4]+\"bam\"\n",
    "    \n",
    "        bam_file_path = os.path.join(intermediate_directory_path, bam_file)\n",
    "\n",
    "        # Calling samtools view\n",
    "        command = [\n",
    "            samtools, \"view\", \"-b\", \"-S\", sam_file_path\n",
    "        ]\n",
    "    \n",
    "        with open(bam_file_path, 'wb') as bam_out_file:\n",
    "            subprocess.run(command, stdout=bam_out_file, check=True)\n",
    "    \n",
    "        # Sort\n",
    "        sorted_bam_file_path = bam_file_path.removesuffix(\".bam\") + \"_sorted.bam\"\n",
    "        #sorted_bam_file_path = bam_file_path[:-4] + \"_sorted.bam\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"SortSam\",\n",
    "            \"INPUT=\" + bam_file_path, \"OUTPUT=\" + sorted_bam_file_path,\n",
    "            \"SORT_ORDER=coordinate\"\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # Add readgroup information\n",
    "        addrg_bam_file_path = sorted_bam_file_path.removesuffix(\".bam\") + \"_addrg.bam\"\n",
    "        #addrg_bam_file_path = bam_file_path[:-4] + \"_addrg.bam\"\n",
    "\n",
    "        read_group_library = f\"{main_directory_path}/{sample}_WGS\"\n",
    "        read_group_platform = \"Illumina\"\n",
    "        read_group_platform_unit = f\"{main_directory_path}/{sample}_WGS\"\n",
    "        read_group_sample_name = f\"{main_directory_path}/{sample}_WGS\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"AddOrReplaceReadGroups\",\n",
    "            \"INPUT=\" + sorted_bam_file_path,\n",
    "            \"OUTPUT=\" + addrg_bam_file_path,\n",
    "            \"LB=\" + read_group_library,\n",
    "            \"PL=\" + read_group_platform,\n",
    "            \"PU=\" + read_group_platform_unit,\n",
    "            \"SM=\" + read_group_sample_name\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # Mark duplicates\n",
    "        dedup_bam_file_path = addrg_bam_file_path.removesuffix(\".bam\") + \"_dedup.bam\"\n",
    "        #dedup_sorted_bam_file_path = sorted_bam_file_path[:-4] + \"_dedup.bam\"\n",
    "        metrics_file_path = bam_file_path.removesuffix(\".bam\") + \"_metrics.txt\"\n",
    "        #metrics_file_path = bam_file_path[:-4] + \"_metrics.txt\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"MarkDuplicates\",\n",
    "            \"INPUT=\" + addrg_bam_file_path, \"OUTPUT=\" + dedup_bam_file_path,\n",
    "            \"METRICS_FILE=\" + metrics_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        #Index .bam\n",
    "        bai_file_path = dedup_bam_file_path.removesuffix(\"bam\") + \"bai\"\n",
    "        #bai_file_path = bam_file_path[:-3] + \"bai\"\n",
    "\n",
    "        command = [\n",
    "            \"java\", \"-jar\", picard_jar, \"BuildBamIndex\",\n",
    "            \"INPUT=\" + dedup_bam_file_path,\n",
    "            \"OUTPUT=\" + bai_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        #Calling GATK\n",
    "        reference_genome = realign_sample_genome_dict[sample]\n",
    "    \n",
    "        reference_genome_fasta_file = f\"N_{reference_genome}.fasta\"\n",
    "        reference_genome_fasta_file_path = os.path.join(indexed_directory_path, reference_genome, reference_genome_fasta_file)\n",
    "\n",
    "        realigned_intervals_file_path = bam_file_path.removesuffix(\".bam\") + \".realigned.intervals\"\n",
    "        #realigned_intervals_file_path = bam_file_path[:-4] + \".realigned.intervals\"\n",
    "\n",
    "        command = [\n",
    "            java_8, \"-jar\", GATK_jar, \"-T\", \"RealignerTargetCreator\",\n",
    "            \"-I\", dedup_bam_file_path, \"-R\", reference_genome_fasta_file_path,\n",
    "            \"-o\", realigned_intervals_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        realigned_bam_file_path = bam_file_path.removesuffix(\".bam\") + \".realigned.bam\"\n",
    "        #realigned_bam_file_path = bam_file_path[:-4] + \".realigned.bam\"\n",
    "\n",
    "        command = [\n",
    "            java_8, \"-jar\", GATK_jar, \"-T\", \"IndelRealigner\",\n",
    "            \"-I\", dedup_bam_file_path, \"-R\", reference_genome_fasta_file_path,\n",
    "            \"--maxConsensuses\", \"60\", \"--maxReadsForConsensuses\", \"240\",\"--maxReadsForRealignment\", \"6000\",\n",
    "            \"--targetIntervals\", realigned_intervals_file_path,\n",
    "            \"-o\", realigned_bam_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # Calling samtools sort\n",
    "        realigned_sorted_file_path = realigned_bam_file_path.removesuffix(\".bam\") + \".sorted\"\n",
    "        #realigned_sorted_file_path = realigned_bam_file_path[:-4] + \".sorted\"\n",
    "\n",
    "        command = [\n",
    "            samtools, \"sort\", realigned_bam_file_path,\n",
    "            realigned_sorted_file_path\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "    \n",
    "        # totalpileup files are used for indel calling\n",
    "        realigned_sorted_bam_file_path = realigned_sorted_file_path + \".bam\"\n",
    "        realigned_totalpileup_file = bam_file.removesuffix(\".bam\") + \".realigned.totalpileup\"\n",
    "        #realigned_totalpileup_file = bam_file[:-4] + \".realigned.totalpileup\"\n",
    "        realigned_totalpileup_file_path = os.path.join(totalpileup_directory_path, realigned_totalpileup_file)\n",
    "\n",
    "        with open(realigned_totalpileup_file_path, 'w') as totalpileup_file:\n",
    "            command = [\n",
    "                samtools, \"pileup\", \"-c\", \"-f\", reference_genome_fasta_file_path,\n",
    "                realigned_sorted_bam_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=totalpileup_file, check=True)\n",
    "    \n",
    "        # raw files are used for SNP calling\n",
    "        realigned_raw_file = bam_file.removesuffix(\".bam\") + \".realigned.raw\"\n",
    "        #realigned_raw_file = bam_file[:-] + \".realigned.raw\"\n",
    "        realigned_raw_file_path = os.path.join(raw_directory_path, realigned_raw_file)\n",
    "\n",
    "        with open(realigned_raw_file_path, 'w') as raw_file:\n",
    "            command = [\n",
    "                samtools, \"pileup\", \"-vc\", \"-f\", reference_genome_fasta_file_path,\n",
    "                realigned_sorted_bam_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=raw_file, check=True)\n",
    "\n",
    "        print()\n",
    "        print(f\"{sample} processed!!\")\n",
    "        print()\n",
    "\n",
    "    shutil.rmtree(intermediate_directory_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c2cbe06-d177-4ebe-a36a-042c7df7b61d",
   "metadata": {},
   "source": [
    "## Find indels without common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0df02208-de11-49d2-b3a9-3e7fa1da739a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if configuration[\"indels\"]:\n",
    "\n",
    "    totalpileup_files = os.listdir(totalpileup_directory_path)\n",
    "\n",
    "    for file in totalpileup_files:\n",
    "        totalpileup_file_path = os.path.join(totalpileup_directory_path, file)\n",
    "    \n",
    "        indels_withoutcommon_file = file.removesuffix(\".realigned.totalpileup\") + \".withoutcommon.indels\"\n",
    "        indels_withoutcommon_file_path = os.path.join(indels_directory_path, indels_withoutcommon_file)\n",
    "\n",
    "        with open(indels_withoutcommon_file_path, 'w') as indels_wc_file:\n",
    "            with open(totalpileup_file_path, 'r') as totalpileup_file:\n",
    "        \n",
    "                indels = []\n",
    "        \n",
    "                for line in totalpileup_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if fields[2] == \"*\":\n",
    "                        if int(fields[5]) >= 500 and int(fields[6]) >= 25:\n",
    "                            if (\n",
    "                                (fields[8] == \"*\" and int(fields[11])*5 > int(fields[7])) or\n",
    "                                (fields[9] == \"*\" and int(fields[10])*5 >= int(fields[7])) or\n",
    "                                (fields[8] != \"*\" and fields[9] != \"*\" and (int(fields[11])*5 >= int(fields[7]) or\n",
    "                                 int(fields[10])*5 >= int(fields[7])))\n",
    "                            ):\n",
    "                                indels.append(line)\n",
    "\n",
    "            indels_wc_file.writelines(indels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e382ac4c-d456-4a50-bf98-89142add422d",
   "metadata": {},
   "source": [
    "## Find SNPs without common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "894886c6-3d2d-4cd6-a00a-d61ada961e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if configuration[\"snps\"]:\n",
    "    raw_files = os.listdir(raw_directory_path)\n",
    "\n",
    "    for file in raw_files:\n",
    "        raw_file_path = os.path.join(raw_directory_path, file)\n",
    "    \n",
    "        snps_withoutcommon_file = file.removesuffix(\".realigned.raw\") + \".snps.withoutcommon\"\n",
    "        snps_withoutcommon_file_path = os.path.join(snps_directory_path, snps_withoutcommon_file)\n",
    "\n",
    "        with open(snps_withoutcommon_file_path, 'w') as snps_wc_file:\n",
    "            with open(raw_file_path, 'r') as raw_file:\n",
    "        \n",
    "                snps = []\n",
    "        \n",
    "                for line in raw_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if (int(fields[5]) >= 50 and int(fields[6]) >= 25 and int(fields[7]) >= 3):\n",
    "                        if fields[3] not in [\"M\",\"R\",\"W\",\"S\",\"Y\",\"K\"]:\n",
    "                            snps.append(line)\n",
    "\n",
    "            snps_wc_file.writelines(snps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf83f75-0771-4d1c-840d-5e9e3268bf13",
   "metadata": {},
   "source": [
    "## Convert .withoutcommon files to .vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8bff7d84-24f0-4da3-acfa-4ec8a88606ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NOTICE: the default --format argument is set as 'pileup'\n",
      "NOTICE: the default --snpqual argument for pileup format is set as 20\n",
      "NOTICE: Column 6-9 in output are heterozygosity status, SNP quality, total reads, reads with mutation\n",
      "NOTICE: Read 307 lines and wrote 309 different variants at 307 genomic positions (126 SNPs and 181 indels)\n",
      "NOTICE: Among 309 different variants at 307 positions, 101 are heterozygotes, 208 are homozygotes\n",
      "NOTICE: Among 126 SNPs, 78 are transitions, 48 are transversions\n",
      "NOTICE: the default --format argument is set as 'pileup'\n",
      "NOTICE: the default --snpqual argument for pileup format is set as 20\n",
      "NOTICE: Column 6-9 in output are heterozygosity status, SNP quality, total reads, reads with mutation\n",
      "NOTICE: Read 118 lines and wrote 119 different variants at 118 genomic positions (16 SNPs and 102 indels)\n",
      "NOTICE: Among 119 different variants at 118 positions, 55 are heterozygotes, 64 are homozygotes\n",
      "NOTICE: Among 16 SNPs, 6 are transitions, 10 are transversions\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"SNP_indel_annotation\"][\"enabled\"]:\n",
    "    snps_files = os.listdir(snps_directory_path)\n",
    "\n",
    "    for file in snps_files:\n",
    "        snps_file_path = os.path.join(snps_directory_path, file)\n",
    "        vcf_file = file.removesuffix(\".raw\") + \".vcf\"\n",
    "        vcf_file_path = os.path.join(vcf_directory_path, vcf_file)\n",
    "\n",
    "        with open(vcf_file_path, 'w') as vcf_out_file:\n",
    "            command = [\n",
    "                perl, convert2annovar, snps_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=vcf_out_file, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da8e7c2a-e5bc-4d20-8b70-9b5b2d6a2afc",
   "metadata": {},
   "source": [
    "## Build genome database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ad7a8965-fefd-41ed-a597-6d8afcb75321",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading genome NC_002516.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100 14.0M    0 14.0M    0     0   788k      0 --:--:--  0:00:18 --:--:--  950k\n",
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'build'\n",
      "00:00:00 Building database for 'NC_002516.2'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_002516.2'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:01 done\n",
      "00:00:04 Chromosome: 'NC_002516.2'\tlength: 6264404\n",
      "00:00:04 Create exons from CDS (if needed): \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:04 Exons created for 5572 transcripts.\n",
      "00:00:04 Deleting redundant exons (if needed): \n",
      "00:00:04 \tTotal transcripts with deleted exons: 0\n",
      "00:00:04 Collapsing zero length introns (if needed): \n",
      "00:00:04 \tTotal collapsed transcripts: 0\n",
      "00:00:04 \t\tAdding genomic sequences to genes: \n",
      "00:00:04 \tDone (4779 sequences added).\n",
      "00:00:04 \t\tAdding genomic sequences to exons: \n",
      "00:00:05 \tDone (5573 sequences added, 0 ignored).\n",
      "00:00:05 Finishing up genome\n",
      "00:00:05 Adjusting transcripts: \n",
      "00:00:05 Adjusting genes: \n",
      "00:00:05 Adjusting chromosomes lengths: \n",
      "00:00:05 Ranking exons: \n",
      "00:00:05 Create UTRs from CDS (if needed): \n",
      "00:00:05 Remove empty chromosomes: \n",
      "00:00:05 Marking as 'coding' from CDS information: \n",
      "00:00:05 Done: 0 transcripts marked\n",
      "00:00:05 \n",
      "00:00:05 Caracterizing exons by splicing (stage 1) : \n",
      ".....\n",
      "00:00:05 Caracterizing exons by splicing (stage 2) : \n",
      ".....00:00:05 done.\n",
      "00:00:05 [Optional] Rare amino acid annotations\n",
      "WARNING_FILE_NOT_FOUND: Rare Amino Acid analysis: Cannot read protein sequence file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/protein.fa', nothing done.\n",
      "00:00:05 CDS check: GenBank file format, skipping\n",
      "\n",
      "00:00:05 Protein check file: '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/genes.gbk'\n",
      "\n",
      "00:00:05 Checking database using protein sequences\n",
      "00:00:05 Comparing Proteins...\n",
      "\tLabels:\n",
      "\t\t'+' : OK\n",
      "\t\t'.' : Missing\n",
      "\t\t'*' : Error\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t\t\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\tProtein check:\tNC_002516.2\tOK: 5572\tNot found: 0\tErrors: 0\tError percentage: 0.0%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:06 \n",
      "\n",
      "00:00:06 Saving database\n",
      "00:00:08 Saving sequences for chromosome 'NC_002516.2' to file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/sequence.NC_002516.2.bin'\n",
      "00:00:10 [Optional] Reading regulation elements: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot read optional regulation file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/regulation.gff', nothing done.\n",
      "00:00:10 [Optional] Reading regulation elements: BED \n",
      "00:00:10 Cannot find optional regulation dir '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/regulation.bed/', nothing done.\n",
      "00:00:10 [Optional] Reading motifs: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot open PWMs file /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/pwms.bin. Nothing done\n",
      "00:00:10 Done\n",
      "00:00:10 Logging\n",
      "00:00:11 Checking for updates...\n",
      "00:00:13 Done.\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading genome NC_008463.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100 15.7M    0 15.7M    0     0   953k      0 --:--:--  0:00:16 --:--:-- 2022k\n",
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'build'\n",
      "00:00:00 Building database for 'NC_008463.1'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_008463.1'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:01 done\n",
      "00:00:04 Chromosome: 'NC_008463.1'\tlength: 6537648\n",
      "00:00:04 Create exons from CDS (if needed): \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:04 Exons created for 5961 transcripts.\n",
      "00:00:04 Deleting redundant exons (if needed): \n",
      "00:00:04 \tTotal transcripts with deleted exons: 0\n",
      "00:00:04 Collapsing zero length introns (if needed): \n",
      "00:00:04 \tTotal collapsed transcripts: 0\n",
      "00:00:04 \t\tAdding genomic sequences to genes: \n",
      "00:00:05 \tDone (4989 sequences added).\n",
      "00:00:05 \t\tAdding genomic sequences to exons: \n",
      "00:00:05 \tDone (5966 sequences added, 0 ignored).\n",
      "00:00:05 Finishing up genome\n",
      "00:00:05 Adjusting transcripts: \n",
      "00:00:05 Adjusting genes: \n",
      "00:00:05 Adjusting chromosomes lengths: \n",
      "00:00:05 Ranking exons: \n",
      "00:00:05 Create UTRs from CDS (if needed): \n",
      "00:00:05 Remove empty chromosomes: \n",
      "00:00:05 Marking as 'coding' from CDS information: \n",
      "00:00:05 Done: 70 transcripts marked\n",
      "00:00:05 \n",
      "00:00:05 Caracterizing exons by splicing (stage 1) : \n",
      ".."
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t\t\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...\n",
      "00:00:05 Caracterizing exons by splicing (stage 2) : \n",
      ".....00:00:05 done.\n",
      "00:00:05 [Optional] Rare amino acid annotations\n",
      "WARNING_FILE_NOT_FOUND: Rare Amino Acid analysis: Cannot read protein sequence file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/protein.fa', nothing done.\n",
      "00:00:05 CDS check: GenBank file format, skipping\n",
      "\n",
      "00:00:05 Protein check file: '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/genes.gbk'\n",
      "\n",
      "00:00:05 Checking database using protein sequences\n",
      "00:00:05 Comparing Proteins...\n",
      "\tLabels:\n",
      "\t\t'+' : OK\n",
      "\t\t'.' : Missing\n",
      "\t\t'*' : Error\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++\n",
      "\t+.++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++.++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++.++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++.+++++++++++++++++++.+++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++.++\n",
      "\t+++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++.+++++++++++++++++++++++++.+++++++++++.+++++++++++++++++++++++++++++++++++++.++++++++++++++\n",
      "\t++++++++++++.++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++\n",
      "\t.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++.+++++++++++\n",
      "\t+++++++++++.+++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++\n",
      "\t++++++++++++++++++++++.++++++++++++++++++++++++.+++++++.++++++++++++++++.++++++++++++++++++++.++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+..\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++\n",
      "\t+++.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++.++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++.+++++++++++++..++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++.++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++.+++++++++++++++++++++++++++++++++++++++++.+++++++++++.+++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++.+++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++.++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t+++++.+++++++++++++++++++++++++++++++++.++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++.+++++++++++.+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++.+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\t++++++++++++++++++++++++++++++++++++++++++++++++.+++++++++++++\tProtein check:\tNC_008463.1\tOK: 5891\tNot found: 70\tErrors: 0\tError percentage: 0.0%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:07 \n",
      "\n",
      "00:00:07 Saving database\n",
      "00:00:09 Saving sequences for chromosome 'NC_008463.1' to file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/sequence.NC_008463.1.bin'\n",
      "00:00:10 [Optional] Reading regulation elements: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot read optional regulation file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/regulation.gff', nothing done.\n",
      "00:00:10 [Optional] Reading regulation elements: BED \n",
      "00:00:10 Cannot find optional regulation dir '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/regulation.bed/', nothing done.\n",
      "00:00:10 [Optional] Reading motifs: GFF\n",
      "WARNING_FILE_NOT_FOUND: Cannot open PWMs file /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/pwms.bin. Nothing done\n",
      "00:00:10 Done\n",
      "00:00:10 Logging\n",
      "00:00:11 Checking for updates...\n",
      "00:00:13 Done.\n"
     ]
    }
   ],
   "source": [
    "# https://pcingola.github.io/SnpEff/snpeff/build_db/#step-2-option-2-building-a-database-from-genbank-files\n",
    "# https://pcingola.github.io/SnpEff/snpeff/faq/#how-to-building-an-ncbi-genome-genbank-file\n",
    "if configuration[\"build\"][\"enabled\"]:\n",
    "    for database in configuration[\"build\"][\"databases\"]:\n",
    "        command = [\n",
    "            buildDbNcbi, database\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, cwd = snpeff_dir_path, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b34c9070-48e8-47f6-b50a-a4e2bccd2fbb",
   "metadata": {},
   "source": [
    "## SNPs and indels annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "173d1fa0-cd0a-4596-bb01-b4ddb59fd025",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'ann'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_008463.1'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:01 done\n",
      "00:00:01 Reading database for genome version 'NC_008463.1' from file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_008463.1/snpEffectPredictor.bin' (this might take a while)\n",
      "00:00:03 done\n",
      "00:00:03 Loading Motifs and PWMs\n",
      "00:00:03 Building interval forest\n",
      "00:00:04 done.\n",
      "00:00:04 Genome stats :\n",
      "#-----------------------------------------------\n",
      "# Genome name                : 'NC_008463.1'\n",
      "# Genome version             : 'NC_008463.1'\n",
      "# Genome ID                  : 'NC_008463.1[0]'\n",
      "# Has protein coding info    : true\n",
      "# Has Tr. Support Level info : true\n",
      "# Genes                      : 6041\n",
      "# Protein coding genes       : 5961\n",
      "#-----------------------------------------------\n",
      "# Transcripts                : 5961\n",
      "# Avg. transcripts per gene  : 0.99\n",
      "# TSL transcripts            : 0\n",
      "#-----------------------------------------------\n",
      "# Checked transcripts        : \n",
      "#               AA sequences :   5891 ( 98.83% )\n",
      "#              DNA sequences :      0 ( 0.00% )\n",
      "#-----------------------------------------------\n",
      "# Protein coding transcripts : 5961\n",
      "#              Length errors :     28 ( 0.47% )\n",
      "#  STOP codons in CDS errors :     16 ( 0.27% )\n",
      "#         START codon errors :    531 ( 8.91% )\n",
      "#        STOP codon warnings :     17 ( 0.29% )\n",
      "#              UTR sequences :      0 ( 0.00% )\n",
      "#               Total Errors :    550 ( 9.23% )\n",
      "# WARNING                    : No protein coding transcript has UTR\n",
      "#-----------------------------------------------\n",
      "# Cds                        : 5966\n",
      "# Exons                      : 5966\n",
      "# Exons with sequence        : 5966\n",
      "# Exons without sequence     : 0\n",
      "# Avg. exons per transcript  : 1.00\n",
      "#-----------------------------------------------\n",
      "# Number of chromosomes      : 1\n",
      "# Chromosomes                : Format 'chromo_name size codon_table'\n",
      "#\t\t'NC_008463.1'\t6537648\tStandard\n",
      "#-----------------------------------------------\n",
      "\n",
      "00:00:06 Predicting variants\n",
      "\n",
      "WARNINGS: Some warning were detected\n",
      "Warning type\tNumber of warnings\n",
      "WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS\t15\n",
      "WARNING_TRANSCRIPT_NO_START_CODON\t93\n",
      "WARNING_TRANSCRIPT_NO_STOP_CODON\t1\n",
      "\n",
      "\n",
      "00:00:06 Creating summary file: snpEff_summary.html\n",
      "00:00:08 Creating genes file: snpEff_genes.txt\n",
      "00:00:08 done.\n",
      "00:00:08 Logging\n",
      "00:00:09 Checking for updates...\n",
      "00:00:11 Done.\n",
      "00:00:00 SnpEff version SnpEff 5.2 (build 2023-09-29 06:17), by Pablo Cingolani\n",
      "00:00:00 Command: 'ann'\n",
      "00:00:00 Reading configuration file 'snpEff.config'. Genome: 'NC_002516.2'\n",
      "00:00:00 Reading config file: /home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/snpEff.config\n",
      "00:00:02 done\n",
      "00:00:02 Reading database for genome version 'NC_002516.2' from file '/home/student/miniconda3/envs/VariantCalling/share/snpeff-5.2-1/./data/NC_002516.2/snpEffectPredictor.bin' (this might take a while)\n",
      "00:00:03 done\n",
      "00:00:03 Loading Motifs and PWMs\n",
      "00:00:03 Building interval forest\n",
      "00:00:04 done.\n",
      "00:00:04 Genome stats :\n",
      "#-----------------------------------------------\n",
      "# Genome name                : 'NC_002516.2'\n",
      "# Genome version             : 'NC_002516.2'\n",
      "# Genome ID                  : 'NC_002516.2[0]'\n",
      "# Has protein coding info    : true\n",
      "# Has Tr. Support Level info : true\n",
      "# Genes                      : 5697\n",
      "# Protein coding genes       : 5572\n",
      "#-----------------------------------------------\n",
      "# Transcripts                : 5572\n",
      "# Avg. transcripts per gene  : 0.98\n",
      "# TSL transcripts            : 0\n",
      "#-----------------------------------------------\n",
      "# Checked transcripts        : \n",
      "#               AA sequences :   5572 ( 100.00% )\n",
      "#              DNA sequences :      0 ( 0.00% )\n",
      "#-----------------------------------------------\n",
      "# Protein coding transcripts : 5572\n",
      "#              Length errors :      0 ( 0.00% )\n",
      "#  STOP codons in CDS errors :      0 ( 0.00% )\n",
      "#         START codon errors :    524 ( 9.40% )\n",
      "#        STOP codon warnings :      0 ( 0.00% )\n",
      "#              UTR sequences :      0 ( 0.00% )\n",
      "#               Total Errors :    524 ( 9.40% )\n",
      "# WARNING                    : No protein coding transcript has UTR\n",
      "#-----------------------------------------------\n",
      "# Cds                        : 5573\n",
      "# Exons                      : 5573\n",
      "# Exons with sequence        : 5573\n",
      "# Exons without sequence     : 0\n",
      "# Avg. exons per transcript  : 1.00\n",
      "#-----------------------------------------------\n",
      "# Number of chromosomes      : 1\n",
      "# Chromosomes                : Format 'chromo_name size codon_table'\n",
      "#\t\t'NC_002516.2'\t6264404\tStandard\n",
      "#-----------------------------------------------\n",
      "\n",
      "00:00:04 Predicting variants\n",
      "\n",
      "WARNINGS: Some warning were detected\n",
      "Warning type\tNumber of warnings\n",
      "WARNING_TRANSCRIPT_NO_START_CODON\t16\n",
      "\n",
      "\n",
      "00:00:05 Creating summary file: snpEff_summary.html\n",
      "00:00:06 Creating genes file: snpEff_genes.txt\n",
      "00:00:06 done.\n",
      "00:00:06 Logging\n",
      "00:00:07 Checking for updates...\n",
      "00:00:09 Done.\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"SNP_indel_annotation\"][\"enabled\"]:\n",
    "    vcf_files = [file for file in os.listdir(vcf_directory_path) if file.endswith(\".snps.withoutcommon.vcf\")]\n",
    "\n",
    "    for file in vcf_files:\n",
    "        vcf_file_path = os.path.join(vcf_directory_path, file)\n",
    "    \n",
    "        vcf4snpEff_file_path = vcf_file_path + \"4snpEff\"\n",
    "\n",
    "        with open(vcf4snpEff_file_path, 'w') as vcf4snpEff_file:\n",
    "            with open(vcf_file_path, 'r') as vcf_file:\n",
    "                lines = []\n",
    "\n",
    "                for line in vcf_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if fields[3] != \"-\" and fields[4] != \"-\":\n",
    "                        lines.append(line)\n",
    "\n",
    "            vcf4snpEff_file.writelines(lines)\n",
    "\n",
    "    vcf4snpEff_files = [file for file in os.listdir(vcf_directory_path) if file.endswith(\"vcf4snpEff\")]\n",
    "\n",
    "    for file in vcf4snpEff_files:\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        genome = annotation_sample_genome_dict[sample]\n",
    "    \n",
    "        snps_wc_annotated_file = file.removesuffix(\".vcf4snpEff\") + \".annotated\"\n",
    "        snps_wc_annotated_file_path = os.path.join(annotated_directory_path, snps_wc_annotated_file)\n",
    "\n",
    "        vcf4snpEff_file_path = os.path.join(vcf_directory_path, file)\n",
    "\n",
    "        with open(snps_wc_annotated_file_path, 'w') as snps_annotated_file:\n",
    "    \n",
    "            command = [\n",
    "                java_21, \"-jar\", snpEff_jar,\n",
    "                \"-v\", genome, vcf4snpEff_file_path\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=snps_annotated_file, cwd=snpeff_dir_path, check=True)\n",
    "        snps_annotated_file.close()\n",
    "    \n",
    "        temp_directory_path = os.path.join(variant_calling_directory_path, \"temp\")\n",
    "    \n",
    "        if not os.path.exists(temp_directory_path):\n",
    "            os.mkdir(temp_directory_path)\n",
    "\n",
    "        temp1_file_path = os.path.join(temp_directory_path, \"temp1\")\n",
    "    \n",
    "        with open(temp1_file_path, 'w') as temp1_file:\n",
    "            with open(snps_wc_annotated_file_path, 'r') as snps_annotated_file:\n",
    "                command = [\n",
    "                    perl, vcfEffOnePerLine\n",
    "                ]\n",
    "\n",
    "                subprocess.run(command, stdin=snps_annotated_file, stdout=temp1_file, check=True)\n",
    "\n",
    "        temp2_file_path = os.path.join(temp_directory_path, \"temp2\")\n",
    "    \n",
    "        with open(temp2_file_path, 'w') as temp2_file:\n",
    "\n",
    "            command = [\n",
    "                java_21, \"-jar\", snpSift_jar, \"extractFields\",\n",
    "                temp1_file_path, \"POS\", \"REF\", \"ALT\", \"ANN[*].EFFECT\",\n",
    "                \"ANN[*].IMPACT\", \"ANN[*].GENE\", \"ANN[*].GENEID\",\n",
    "                \"ANN[*].HGVS_C\", \"ANN[*].HGVS_P\", \"ANN[*].CDNA_POS\",\n",
    "                \"ANN[*].CDNA_LEN\", \"ANN[*].AA_POS\", \"ANN[*].AA_LEN\"\n",
    "            ]\n",
    "\n",
    "            subprocess.run(command, stdout=temp2_file, check=True)\n",
    "        temp2_file.close()\n",
    "\n",
    "        semicurated_file = file.removesuffix(\".vcf4snpEff\") + \".semicurated\"\n",
    "        semicurated_file_path = os.path.join(curated_directory_path, semicurated_file)\n",
    "    \n",
    "        with open(semicurated_file_path, 'w') as semicurated_file:\n",
    "            with open(temp2_file_path, 'r') as temp2_file:\n",
    "\n",
    "                content = temp2_file.readlines()\n",
    "            semicurated_file.writelines(content)\n",
    "        \n",
    "        curated_file = file.removesuffix(\".vcf4snpEff\") + \".curated\"\n",
    "        curated_file_path = os.path.join(curated_directory_path, curated_file)\n",
    "    \n",
    "        with open(curated_file_path, 'w') as curated_file:\n",
    "            with open(semicurated_file_path, 'r') as semicurated_file:\n",
    "            \n",
    "                curated_lines = []\n",
    "            \n",
    "                for line in semicurated_file:\n",
    "                    fields = line.strip().split()\n",
    "\n",
    "                    if fields[4] == \"MODERATE\" or fields[4] == \"HIGH\":\n",
    "                        curated_lines.append(line)\n",
    "\n",
    "            curated_file.writelines(curated_lines)\n",
    "\n",
    "        shutil.rmtree(temp_directory_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab2d510d-a585-421b-9622-deaf4efd8cff",
   "metadata": {},
   "source": [
    "## Filter by parental"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7173032-cb65-4d49-ae30-4c6e62c2fe3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if configuration[\"filter_by_parental\"][\"enabled\"]:\n",
    "    \n",
    "    # Filter SNPs\n",
    "    curated_files = [file for file in os.listdir(curated_directory_path) if file.endswith(\".curated\")]\n",
    "    \n",
    "    for file in curated_files:    \n",
    "        sample = file.split(\"_\")[0]\n",
    "        \n",
    "        curated_file_path = os.path.join(curated_directory_path, file)\n",
    "        filtered_file_path = os.path.join(curated_directory_path, file + \".filtered\")\n",
    "            \n",
    "        with open(curated_file_path, 'r') as curated_file:\n",
    "            snp_lines = curated_file.readlines()\n",
    "\n",
    "        processed_lines = []\n",
    "        for line in snp_lines:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            if len(fields) >= 9:\n",
    "                processed_lines.append(\"\\t\".join([fields[0], fields[5], fields[8]]))\n",
    "\n",
    "        processed_lines.sort()\n",
    "\n",
    "        for parental in configuration[\"filter_by_parental\"][\"Parental\"]:\n",
    "            if sample in parental[\"samples\"]:\n",
    "                snp_path = parental[\"snps_path\"]\n",
    "\n",
    "        with open(snp_path, 'r') as parental_file:\n",
    "            parental_lines = []\n",
    "            for line in parental_file:\n",
    "                fields = line.strip().split(\"\\t\")\n",
    "                if len(fields) >= 9:\n",
    "                    processed_line = \"\\t\".join([fields[0], fields[5], fields[8]])\n",
    "                    parental_lines.append(processed_line)\n",
    "\n",
    "        filtered_lines = []\n",
    "        for line in snp_lines:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            compare = \"\\t\".join([fields[0], fields[5], fields[8]])\n",
    "            if compare not in parental_lines:\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "        with open(filtered_file_path, 'w') as filtered_file:\n",
    "            for line in filtered_lines:\n",
    "                filtered_file.write(line + \"\\n\")\n",
    "\n",
    "    # Filter indels\n",
    "    indels_files = [file for file in os.listdir(indels_directory_path) if file.endswith(\".indels\")]\n",
    "        \n",
    "    for file in indels_files:\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "\n",
    "        indels_file_path = os.path.join(indels_directory_path, file)\n",
    "        filtered_file_path = os.path.join(indels_directory_path, file + \".filtered\")\n",
    "        \n",
    "        with open(indels_file_path, 'r') as indels_file:\n",
    "            sample_indels = indels_file.readlines()\n",
    "\n",
    "        for parental in configuration[\"filter_by_parental\"][\"Parental\"]:\n",
    "            if sample in parental[\"samples\"]:\n",
    "                ind_file_path = parental[\"indels_path\"]\n",
    "\n",
    "        with open(ind_file_path, 'r') as parental_file:\n",
    "            parental_lines = parental_file.readlines()\n",
    "\n",
    "        processed_lines = []\n",
    "        for line in parental_lines:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            processed_lines.append(\"\\t\".join(fields[1:4]))\n",
    "\n",
    "        filtered_lines = []\n",
    "\n",
    "        for line in sample_indels:\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            compare = \"\\t\".join(fields[1:4])\n",
    "            if compare not in processed_lines:\n",
    "                filtered_lines.append(line)\n",
    "\n",
    "        with open(filtered_file_path, 'w') as filtered_file:\n",
    "            for line in filtered_lines:\n",
    "                filtered_file.write(line + \"\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b786d2a-0e7a-40c2-92f6-c61e2924d8e0",
   "metadata": {},
   "source": [
    "## Create mutations table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "52207479-a10a-45f1-a495-90b0a9fabeb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sample PA14wt0253G does not contain any SNPs.\n",
      "The sample PAO1wt12P does not contain any SNPs.\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"table\"][\"enabled\"]:\n",
    "    \n",
    "    filtered_files = [file for file in os.listdir(curated_directory_path) if file.endswith(\".filtered\")]\n",
    "    if len(filtered_files) > 0:\n",
    "        snp_files = filtered_files\n",
    "    else:\n",
    "        snp_files = [file for file in os.listdir(curated_directory_path) if file.endswith(\".curated\")]\n",
    "    \n",
    "    # Add sample column\n",
    "    for file in snp_files:\n",
    "        snps_file_path = os.path.join(curated_directory_path, file)\n",
    "        try:\n",
    "            df = pd.read_csv(snps_file_path, sep=\"\\t\", header=0)\n",
    "            df[file] = file\n",
    "            df.to_csv(snps_file_path, index=False, sep=\"\\t\")\n",
    "        except pd.errors.EmptyDataError:\n",
    "            sample = file.strip().split(\"_\")[0]\n",
    "            print(f\"The sample {sample} does not contain any SNPs.\")\n",
    "\n",
    "    filtered_files = [file for file in os.listdir(indels_directory_path) if file.endswith(\".filtered\")]\n",
    "    if len(filtered_files) > 0:\n",
    "        ind_files = filtered_files\n",
    "    else:\n",
    "        ind_files = [file for file in os.listdir(indels_directory_path) if file.endswith(\".indels\")]\n",
    "    \n",
    "    for file in ind_files:\n",
    "        ind_file_path = os.path.join(indels_directory_path, file)\n",
    "        try:\n",
    "            df = pd.read_csv(ind_file_path, sep=\"\\t\", header=0)\n",
    "            df[file] = file\n",
    "            df.to_csv(ind_file_path, index=False, sep=\"\\t\")\n",
    "        except EmptyDataError:\n",
    "            sample = file.strip().split(\"_\")[0]\n",
    "            print(f\"The sample {sample} does not contain any indel.\") \n",
    "\n",
    "    # SNP identification\n",
    "    information_directory_path = os.path.join(curated_directory_path, \"information\")\n",
    "\n",
    "    if not os.path.exists(information_directory_path):\n",
    "        os.mkdir(information_directory_path)\n",
    "\n",
    "    ## Obtain old_locus_tag for Pseudomonas aeruginosa UCBPP-PA14 ##\n",
    "    # ------------------------------------------------------------ #\n",
    "    \n",
    "    gff3 = os.path.join(annotated_genomes_directory_path, \"PA14.gff3\")\n",
    "    locus_file_path = os.path.join(information_directory_path, \"PA14_locus_tags.txt\")\n",
    "\n",
    "    PA14_locus_dict = {}\n",
    "\n",
    "    with open(gff3, 'r') as gff3_file:\n",
    "        for line in gff3_file:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "        \n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            if len(fields) < 9:\n",
    "                continue\n",
    "            \n",
    "            if fields[1] == \"RefSeq\" and fields[2] == \"gene\":\n",
    "                attributes = fields[8]\n",
    "                attributes_dict = {}\n",
    "\n",
    "                for attribute in attributes.split(\";\"):\n",
    "                    if \"=\" in attribute:\n",
    "                        key, value = attribute.split(\"=\", 1)\n",
    "                        attributes_dict[key] = value\n",
    "                    if \"locus_tag\" in attributes_dict:\n",
    "                        locus = attributes_dict[\"locus_tag\"]\n",
    "                        old_locus = attributes_dict.get(\"old_locus_tag\", locus)\n",
    "                        PA14_locus_dict[locus] = old_locus\n",
    "\n",
    "    with open(locus_file_path, 'w') as locus_file:\n",
    "        for locus_tag, old_locus_tag in PA14_locus_dict.items():\n",
    "            locus_file.write(f\"{locus_tag}\\t{old_locus_tag}\\n\")\n",
    "\n",
    "    # -------------------------------------------------------------------------- #\n",
    "    ##############################################################################\n",
    "\n",
    "    ## Filter SNPs\n",
    "    snps_dict = {}\n",
    "\n",
    "    for file in snp_files:\n",
    "        file_path = os.path.join(curated_directory_path, file)\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "\n",
    "        with open(file_path, 'r') as snp_file:\n",
    "            snp_file_reader = csv.reader(snp_file, delimiter=\"\\t\")\n",
    "\n",
    "            for row in snp_file_reader:\n",
    "                gene = row[5]\n",
    "                locus = row[6]\n",
    "                aa = row[8]\n",
    "\n",
    "                #########################################\n",
    "                # Modified for PA14 strain locus change #\n",
    "                # ------------------------------------- #\n",
    "                # snp = \",\".join([locus, gene, aa])     #\n",
    "                #########################################\n",
    "                \n",
    "                if table_sample_strain_dict[sample] == \"PA14\":\n",
    "                    old_locus = PA14_locus_dict[locus]\n",
    "                    snp = \",\".join([old_locus, gene, aa])\n",
    "                else:\n",
    "                    snp = \",\".join([locus, gene, aa])\n",
    "\n",
    "                if snp not in snps_dict:\n",
    "                    snps_dict[snp] = 1\n",
    "                else:\n",
    "                    snps_dict[snp] += 1\n",
    "\n",
    "    snps_info_file_path = os.path.join(information_directory_path, \"snps_info.txt\")\n",
    "    \n",
    "    with open(snps_info_file_path, 'w') as snps_info_file:\n",
    "\n",
    "        for snp, count in snps_dict.items():\n",
    "            snps_info_file.write(f\"{count},{snp}\\n\")\n",
    "\n",
    "    filtered_snps_file_path = os.path.join(information_directory_path, \"filtered_snps.txt\")\n",
    "    with open(filtered_snps_file_path, 'w') as filtered_snps_file:\n",
    "        \n",
    "        for snp, count in snps_dict.items():\n",
    "            if count <= configuration[\"table\"][\"threshold\"]:\n",
    "                filtered_snps_file.write(f\"{snp}\\n\")\n",
    "    \n",
    "    ## SNP tabulation\n",
    "    snp_tab_file_path = os.path.join(information_directory_path, \"snp_tabulation.txt\")\n",
    "    with open(snp_tab_file_path, 'w') as snp_tab_file:\n",
    "\n",
    "        for file in snp_files:\n",
    "            file_path = os.path.join(curated_directory_path, file)\n",
    "\n",
    "            with open(file_path, 'r') as snp_file:\n",
    "                snp_file_reader = csv.reader(snp_file, delimiter=\"\\t\")\n",
    "\n",
    "                for row in snp_file_reader:\n",
    "                    gene = row[5]\n",
    "                    locus = row[6]\n",
    "                    aa = row[8]\n",
    "                    sample_name = row[13].split(\"_\")[0]\n",
    "\n",
    "                    #########################################\n",
    "                    # Modified for PA14 strain locus change #\n",
    "                    # ------------------------------------- #\n",
    "                    # snp = \",\".join([locus, gene, aa])     #\n",
    "                    #########################################\n",
    "                    \n",
    "                    if table_sample_strain_dict[sample_name] == \"PA14\":\n",
    "                        old_locus = PA14_locus_dict[locus]\n",
    "                        snp = \",\".join([old_locus, gene, aa])\n",
    "                    else:\n",
    "                        snp = \",\".join([locus, gene, aa])\n",
    "\n",
    "                    if snp in snps_dict and snps_dict[snp] <= configuration[\"table\"][\"threshold\"]:\n",
    "                        snp_tab_file.write(\",\".join([snp, sample_name]) + \"\\n\")\n",
    "\n",
    "    ## SNPs table\n",
    "    snp_table_file_path = os.path.join(results_directory_path, \"snp_table.csv\")\n",
    "    snps_aa_dict = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "    with open(snp_tab_file_path, 'r') as snp_tab_file:\n",
    "        lines = snp_tab_file.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            row = line.strip(\"\\n\").split(\",\")\n",
    "\n",
    "            locus = row[0]\n",
    "            gene = row[1]\n",
    "            aa = row[2]\n",
    "            sample = row[3]\n",
    "\n",
    "            column_name = f\"{locus} ({gene})\"\n",
    "\n",
    "            snps_aa_dict[sample][column_name].append(aa)\n",
    "\n",
    "        with open(snp_table_file_path, 'w', newline='') as snp_table_file:\n",
    "            snp_table_file_writer = csv.writer(snp_table_file)\n",
    "\n",
    "            all_columns = set()\n",
    "\n",
    "            for sample in snps_aa_dict:\n",
    "                for column_name in snps_aa_dict[sample]:\n",
    "                    all_columns.add(column_name)\n",
    "\n",
    "            header = [\"\"] + sorted(list(all_columns))\n",
    "            snp_table_file_writer.writerow(header)\n",
    "\n",
    "            for sample, snps in snps_aa_dict.items():\n",
    "                row = [sample]\n",
    "                for snp in sorted(list(all_columns)):\n",
    "                    aa_mutations = \",\".join(snps.get(snp, []))\n",
    "                    row.append(aa_mutations)\n",
    "                snp_table_file_writer.writerow(row)\n",
    "\n",
    "    # Indels identification\n",
    "    information_directory_path = os.path.join(indels_directory_path, \"information\")\n",
    "    annotated_ind_directory_path = os.path.join(indels_directory_path, \"annotated\")\n",
    "\n",
    "    if not os.path.exists(information_directory_path):\n",
    "        os.mkdir(information_directory_path)\n",
    "    if not os.path.exists(annotated_ind_directory_path):\n",
    "        os.mkdir(annotated_ind_directory_path)\n",
    "\n",
    "    ## Obtain indels position\n",
    "    pos_strain_dict = {}\n",
    "\n",
    "    for file in ind_files:\n",
    "        file_path = os.path.join(indels_directory_path, file)\n",
    "        \n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        strain = table_sample_strain_dict[sample]\n",
    "\n",
    "        if strain not in pos_strain_dict:\n",
    "            pos_strain_dict[strain] = {\n",
    "                \"Positions\": []\n",
    "            }\n",
    "\n",
    "        with open(file_path, 'r') as ind_file:\n",
    "            ind_file_reader = csv.reader(ind_file, delimiter='\\t')\n",
    "\n",
    "            for row in ind_file_reader:\n",
    "                pos = int(row[1])\n",
    "                pos_strain_dict[strain][\"Positions\"].append(pos)\n",
    "\n",
    "    ## Annotate indels\n",
    "    annotated_pos_dict = {}\n",
    "\n",
    "    for file in ind_files:\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        genome = table_sample_strain_dict[sample]\n",
    "    \n",
    "        annotated_genome = table_sample_strain_dict[sample] + \".csv\"\n",
    "        annotated_genome_file_path = os.path.join(annotated_genomes_directory_path, annotated_genome)\n",
    "\n",
    "        with open(annotated_genome_file_path, 'r') as annotated_genome_file:\n",
    "            annotated_genome_reader = csv.reader(annotated_genome_file, delimiter=',')\n",
    "\n",
    "            next(annotated_genome_reader)\n",
    "\n",
    "            if genome not in annotated_pos_dict:\n",
    "                annotated_pos_dict[genome] = {}\n",
    "\n",
    "            for row in annotated_genome_reader:\n",
    "                locus = row[5]\n",
    "                start = int(row[7])\n",
    "                end = int(row[8])\n",
    "                gene = row[10]\n",
    "\n",
    "                for position in pos_strain_dict[genome][\"Positions\"]:\n",
    "                    if position >= start and position <= end:\n",
    "                        annotated_pos_dict[genome][position] = {\n",
    "                            'locus': locus,\n",
    "                            'gene': gene\n",
    "                        }\n",
    "                for position in pos_strain_dict[genome][\"Positions\"]:\n",
    "                    if position not in annotated_pos_dict[genome]:\n",
    "                        annotated_pos_dict[genome][position] = {\n",
    "                            'locus': 'intron',\n",
    "                            'gene': 'intron'\n",
    "                        }\n",
    "\n",
    "    for file in ind_files:\n",
    "        file_path = os.path.join(indels_directory_path, file)\n",
    "\n",
    "        sample = file.strip().split(\"_\")[0]\n",
    "        genome = table_sample_strain_dict[sample]\n",
    "\n",
    "        with open(file_path, 'r') as ind_file:\n",
    "            ind_file_reader = csv.reader(ind_file, delimiter='\\t')\n",
    "\n",
    "            modified_rows = []\n",
    "\n",
    "            for row in ind_file_reader:\n",
    "                pos = int(row[1])\n",
    "\n",
    "                annotation = annotated_pos_dict.get(genome, {}).get(pos, {\n",
    "                    \"refseq\": \"NA\", \"locus\": \"NA\", \"gene\":\"NA\"\n",
    "                })\n",
    "\n",
    "                locus = annotation[\"locus\"]\n",
    "                gene = annotation[\"gene\"]\n",
    "\n",
    "                if gene != \"\":\n",
    "                    row.extend([locus, gene])\n",
    "                else:\n",
    "                    row.extend([locus, locus])\n",
    "                \n",
    "                del row[2]\n",
    "\n",
    "                modified_rows.append(row)\n",
    "\n",
    "        annotated_ind_file_path = os.path.join(annotated_ind_directory_path, file)\n",
    "\n",
    "        with open(annotated_ind_file_path, 'w', newline='') as annotated_ind_file:\n",
    "            annotated_ind_file_writer = csv.writer(annotated_ind_file)\n",
    "            annotated_ind_file_writer.writerows(modified_rows)\n",
    "\n",
    "\n",
    "    ## Filter indels\n",
    "    inds_dict = {}\n",
    "\n",
    "    for file in ind_files:\n",
    "        file_path = os.path.join(indels_directory_path, file)\n",
    "\n",
    "        with open(file_path, 'r') as ind_file:\n",
    "            ind_file_reader = csv.reader(ind_file, delimiter='\\t')\n",
    "\n",
    "            for row in ind_file_reader:\n",
    "                refseq = row[0]\n",
    "                pos = row[1]\n",
    "                mut = row[3]\n",
    "\n",
    "                indel = \",\".join([refseq, pos, mut])\n",
    "\n",
    "                if indel not in inds_dict:\n",
    "                    inds_dict[indel] = 1\n",
    "                else:\n",
    "                    inds_dict[indel] += 1\n",
    "\n",
    "    inds_info_file_path = os.path.join(information_directory_path, \"indels_info.txt\")\n",
    "    with open(inds_info_file_path, 'w') as inds_info_file:\n",
    "\n",
    "        for indel, count in inds_dict.items():\n",
    "            inds_info_file.write(f\"{count}, {indel}\\n\")\n",
    "\n",
    "    filtered_inds_file_path = os.path.join(information_directory_path, \"filtered_indels.txt\")\n",
    "    with open(filtered_inds_file_path, 'w') as filtered_inds_file:\n",
    "\n",
    "        for indel, count in inds_dict.items():\n",
    "            if count <= configuration[\"table\"][\"threshold\"]:\n",
    "                filtered_inds_file.write(f\"{indel}\\n\")\n",
    "\n",
    "    ## Indels tabulation\n",
    "    ind_tab_file_path = os.path.join(information_directory_path, \"indel_tabulation.txt\")\n",
    "\n",
    "    annot_files = [file for file in os.listdir(annotated_ind_directory_path)]\n",
    "    \n",
    "    with open(ind_tab_file_path, 'w') as ind_tab_file:\n",
    "        for file in annot_files:\n",
    "            file_path = os.path.join(annotated_ind_directory_path, file)\n",
    "\n",
    "            with open(file_path, 'r') as ind_file:\n",
    "                ind_file_reader = csv.reader(ind_file, delimiter=\",\")\n",
    "\n",
    "                for row in ind_file_reader:\n",
    "                    refseq = row[0]\n",
    "                    pos = row[1]\n",
    "                    mut = row[2]\n",
    "                    sample = row[14].split(\"_\")[0]\n",
    "                    locus = row[15]\n",
    "                    gene = row[16]\n",
    "\n",
    "                    indel = \",\".join([refseq, pos, mut])\n",
    "\n",
    "                    if indel in inds_dict and inds_dict[indel] <= configuration[\"table\"][\"threshold\"]:\n",
    "                        ind_mut = f\"{pos} {mut}\"\n",
    "                        ind_tab_file.write(\",\".join([locus, gene, ind_mut, sample]) + \"\\n\")\n",
    "\n",
    "    # Indels table\n",
    "    ind_table_file_path = os.path.join(results_directory_path, \"indels_table.csv\")\n",
    "    \n",
    "    final_inds_dict = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "    with open(ind_tab_file_path, 'r') as ind_tab_file:\n",
    "        lines = ind_tab_file.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            row = line.strip(\"\\n\").split(\",\")\n",
    "            \n",
    "            locus = row[0]\n",
    "            gene = row[1]\n",
    "            indel = row[2]\n",
    "            sample = row[3]\n",
    "\n",
    "            if gene != \"\":\n",
    "                column_name = f\"{locus} ({gene})\"\n",
    "            else:\n",
    "                column_name = f\"{locus} ({locus})\"\n",
    "\n",
    "            final_inds_dict[sample][column_name].append(indel)\n",
    "\n",
    "        with open(ind_table_file_path, 'w', newline='') as ind_table_file:\n",
    "            ind_table_file_writer = csv.writer(ind_table_file)\n",
    "\n",
    "            all_columns = set()\n",
    "\n",
    "            for sample in final_inds_dict:\n",
    "                for column_name in final_inds_dict[sample]:\n",
    "                    all_columns.add(column_name)\n",
    "\n",
    "            header = [\"\"] + sorted(list(all_columns))\n",
    "            ind_table_file_writer.writerow(header)\n",
    "\n",
    "            for sample, inds in final_inds_dict.items():\n",
    "                row = [sample]\n",
    "                for indel in sorted(list(all_columns)):\n",
    "                    indels = \",\".join(inds.get(indel, []))\n",
    "                    row.append(indels)\n",
    "                ind_table_file_writer.writerow(row)\n",
    "\n",
    "    # SNP/indel table\n",
    "    mutations_file_path = os.path.join(results_directory_path, \"mutations.txt\")\n",
    "\n",
    "    with open(mutations_file_path, 'w', newline='') as mutations_file:\n",
    "        mutations_file_writer = csv.writer(mutations_file)\n",
    "\n",
    "        with(open(snp_tab_file_path, 'r') as snp_tab_file,\n",
    "             open(ind_tab_file_path, 'r') as ind_tab_file):\n",
    "\n",
    "            snp_tab_file_reader = csv.reader(snp_tab_file, delimiter=',')\n",
    "            ind_tab_file_reader = csv.reader(ind_tab_file, delimiter=',')\n",
    "\n",
    "            for row in snp_tab_file_reader:\n",
    "                mutations_file_writer.writerow(row)\n",
    "\n",
    "            for row in ind_tab_file_reader:\n",
    "                mutations_file_writer.writerow(row)\n",
    "\n",
    "    ## mutations table\n",
    "    combined_table_file_path = os.path.join(results_directory_path, \"mutations.csv\")\n",
    "\n",
    "    mut_dict = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "    with open(mutations_file_path, 'r') as mut_file:\n",
    "        lines = mut_file.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            row = line.strip().split(\",\")\n",
    "\n",
    "            locus = row[0]\n",
    "            gene = row[1]\n",
    "            mut = row[2]\n",
    "            sample = row[3]\n",
    "\n",
    "            if gene != \"\":\n",
    "                column_name = f\"{locus} ({gene})\"\n",
    "            else:\n",
    "                column_name = f\"{locus} ({locus})\"\n",
    "\n",
    "            mut_dict[sample][column_name].append(mut)\n",
    "\n",
    "    with open(combined_table_file_path, 'w', newline='') as combined_table_file:\n",
    "        combined_table_file_writer = csv.writer(combined_table_file)\n",
    "\n",
    "        all_columns = set()\n",
    "\n",
    "        for sample in mut_dict:\n",
    "            for column_name in mut_dict[sample]:\n",
    "                all_columns.add(column_name)\n",
    "\n",
    "        header = [\"\"] + sorted(list(all_columns))\n",
    "        combined_table_file_writer.writerow(header)\n",
    "\n",
    "        for sample, muts in mut_dict.items():\n",
    "            row = [sample]\n",
    "            for column in sorted(list(all_columns)):\n",
    "                mutations = \", \".join(muts.get(column, []))\n",
    "                row.append(mutations)\n",
    "            combined_table_file_writer.writerow(row)\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be46e20-247c-45ca-beae-b5efa27d854a",
   "metadata": {},
   "source": [
    "## *De novo* genome assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2282ab3f-4569-402d-9dff-78f6e2c88cdf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command line: /home/student/miniconda3/envs/VariantCalling/bin/spades.py\t-o\t/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes\t-1\t/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PA14wt0253G_R1.fastq.gz\t-2\t/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PA14wt0253G_R2.fastq.gz\t--isolate\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 4.1.0\n",
      "  Python version: 3.11.0\n",
      "  OS: Linux-5.15.0-139-generic-x86_64-with-glibc2.31\n",
      "\n",
      "Output dir: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes\n",
      "Mode: ONLY assembling (without read error correction)\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Isolate mode\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PA14wt0253G_R1.fastq.gz']\n",
      "      right reads: ['/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PA14wt0253G_R2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "  Assembly graph output will use GFA v1.2 format\n",
      "Other parameters:\n",
      "  Dir for temp files: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 10\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/spades.log\n",
      "\n",
      "/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PA14wt0253G_R1.fastq.gz: max reads length: 151\n",
      "/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PA14wt0253G_R2.fastq.gz: max reads length: 151\n",
      "\n",
      "Reads length: 151\n",
      "\n",
      "Default k-mer sizes were set to [21, 33, 55, 77] because estimated read length (151) is equal to or greater than 150\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Assembling started. \n",
      "\n",
      "\n",
      "===== K21 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K21/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=21\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.004     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.004     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.006     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (read_converter.cpp        :  78)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (read_converter.cpp        :  99)   Converting paired reads\n",
      "  0:00:00.151    34M / 40M   INFO    General                 (binary_converter.cpp      : 127)   16384 reads processed\n",
      "  0:00:00.268    39M / 45M   INFO    General                 (binary_converter.cpp      : 127)   32768 reads processed\n",
      "  0:00:00.696    49M / 54M   INFO    General                 (binary_converter.cpp      : 127)   65536 reads processed\n",
      "  0:00:01.370    59M / 64M   INFO    General                 (binary_converter.cpp      : 127)   131072 reads processed\n",
      "  0:00:02.955    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   262144 reads processed\n",
      "  0:00:07.518    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   524288 reads processed\n",
      "  0:00:13.700    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   1048576 reads processed\n",
      "  0:00:26.443    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   2097152 reads processed\n",
      "  0:00:54.224    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   4194304 reads processed\n",
      "  0:01:20.447    59M / 65M   INFO    General                 (binary_converter.cpp      : 143)   6236138 reads written\n",
      "  0:01:20.453     2M / 65M   INFO    General                 (read_converter.cpp        : 113)   Converting single reads\n",
      "  0:01:20.457    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:01:20.459     2M / 65M   INFO    General                 (read_converter.cpp        : 119)   Converting merged reads\n",
      "  0:01:20.594    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:01:20.611     1M / 65M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:01:20.624     1M / 65M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:01:20.625     1M / 65M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:01:20.625     1M / 65M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:01:20.626     1M / 65M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:01:20.628     1M / 65M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:01:20.628     1M / 65M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:01:20.628     1M / 65M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:01:30.230   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1097460 reads\n",
      "  0:01:39.575   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2195287 reads\n",
      "  0:01:48.980   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3293976 reads\n",
      "  0:01:59.097   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4393151 reads\n",
      "  0:02:08.412   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5493475 reads\n",
      "  0:02:17.674   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6592500 reads\n",
      "  0:02:27.388   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 7691770 reads\n",
      "  0:02:36.832   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8791536 reads\n",
      "  0:02:46.202   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9891207 reads\n",
      "  0:03:52.479   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 17586170 reads\n",
      "  0:04:56.276     1M / 627M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24944552 reads\n",
      "  0:04:56.276     1M / 627M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:05:37.926     1M / 627M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 62772672 kmers in total.\n",
      "  0:05:37.927     1M / 627M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:05:37.931     1M / 627M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:05:37.931     1M / 627M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:05:37.932     1M / 627M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:05:37.932     1M / 627M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:05:37.932     1M / 627M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:05:49.708   601M / 807M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 30214005 kmers\n",
      "  0:06:02.179   601M / 1088M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 60420700 kmers\n",
      "  0:06:03.309   601M / 1088M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 62772679 kmers\n",
      "  0:06:03.310   601M / 1088M INFO    General                 (kmer_splitters.hpp        : 202)   Used 62772679 kmers.\n",
      "  0:06:03.355     1M / 1088M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:06:09.282     1M / 1088M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 61399365 kmers in total.\n",
      "  0:06:09.282     1M / 1088M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:06:20.627    44M / 1088M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 61399365 kmers, 44353808 bytes occupied (5.77906 bits per kmer).\n",
      "  0:06:20.628    44M / 1088M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:06:23.762   104M / 1088M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:07:17.496   104M / 1088M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:07:17.523   104M / 1088M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:07:17.524   104M / 1088M INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:07:17.524   104M / 1088M INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:08:23.597   184M / 1088M INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 811377\n",
      "  0:08:24.415   184M / 1088M INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 823635\n",
      "  0:08:24.438   104M / 1088M INFO   Early tip clipping       (early_simplification.hpp  :  50)   15802391 22-mers were removed by early tip clipper\n",
      "  0:08:24.441   104M / 1088M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:08:24.597   104M / 1088M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:09:27.927   246M / 1088M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 3871487 sequences extracted\n",
      "  0:09:46.383   246M / 1088M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:09:57.602   246M / 1088M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:09:57.687   246M / 1088M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:09:59.730   246M / 1088M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:09:59.731   246M / 1088M INFO    General                 (debruijn_graph_constructor: 516)   Total 7742974 edges to create\n",
      "  0:09:59.733   547M / 1088M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:10:03.102   667M / 1088M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:10:05.742   667M / 1088M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:10:05.786   699M / 1088M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:10:06.643   699M / 1088M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:10:06.643   699M / 1088M INFO    General                 (debruijn_graph_constructor: 542)   Total 2498180 vertices to create\n",
      "  0:10:06.644   931M / 1088M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:10:08.290   720M / 1088M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:10:08.292   720M / 1088M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:10:18.165   763M / 1088M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 62772672 kmers, 45345752 bytes occupied (5.77904 bits per kmer).\n",
      "  0:10:18.271  1003M / 1088M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:17:53.715  1003M / 1088M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:17:59.145  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:18:04.509  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:18:10.026  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:18:15.534  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4000000 edges\n",
      "  0:18:21.722  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5000000 edges\n",
      "  0:18:28.706  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5963249 edges\n",
      "  0:18:33.514  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 6863249 edges\n",
      "  0:18:37.987  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 7704787 edges\n",
      "  0:18:38.174  1003M / 1088M INFO    General                 (coverage_filling.hpp      :  83)   Processed 7742700 edges\n",
      "  0:18:38.440   618M / 1088M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:18:38.446   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 18\n",
      "  0:18:38.447   618M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 255\n",
      "  0:18:38.447   618M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 255. Coverage mad: 65.2344\n",
      "  0:18:38.447   618M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:18:38.803   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:18:39.616   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:18:41.235   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:18:46.988   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:18:50.052   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 221.584. Fitted coverage std. dev: 76.4291\n",
      "  0:18:50.064   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0693853\n",
      "  0:18:50.065   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 21\n",
      "  0:18:50.065   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 21\n",
      "  0:18:50.065   619M / 1088M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6401196\n",
      "  0:18:50.068   618M / 1088M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 221.584\n",
      "  0:18:50.068   618M / 1088M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 21\n",
      "  0:18:50.068   618M / 1088M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:18:50.069   618M / 1088M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:18:50.070   618M / 1088M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:18:50.072   618M / 1088M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:18:50.072   618M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:18:50.644   618M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:18:50.645   618M / 1088M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:18:50.645   618M / 1088M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:18:50.646   618M / 1088M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:18:50.646   618M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:18:50.646   618M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:18:52.397   618M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 16397 times\n",
      "  0:18:52.398   618M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:19:59.142   673M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1106293 times\n",
      "  0:19:59.157   673M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:02.668   668M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 255861 times\n",
      "  0:20:02.691   668M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:20:02.692   668M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:02.978   662M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 13813 times\n",
      "  0:20:02.982   662M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.163   655M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 27156 times\n",
      "  0:20:04.169   655M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.296   655M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8758 times\n",
      "  0:20:04.299   655M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:20:04.299   655M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.315   655M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 244 times\n",
      "  0:20:04.316   655M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.461   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2005 times\n",
      "  0:20:04.462   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.486   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2068 times\n",
      "  0:20:04.487   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:20:04.487   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.491   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 43 times\n",
      "  0:20:04.492   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.531   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 393 times\n",
      "  0:20:04.532   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.544   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 913 times\n",
      "  0:20:04.544   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:20:04.544   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.546   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 9 times\n",
      "  0:20:04.546   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.569   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 105 times\n",
      "  0:20:04.570   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.576   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 528 times\n",
      "  0:20:04.577   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:20:04.577   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.578   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 6 times\n",
      "  0:20:04.578   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.592   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 36 times\n",
      "  0:20:04.592   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.596   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 284 times\n",
      "  0:20:04.596   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:20:04.596   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.597   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:20:04.597   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.608   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 18 times\n",
      "  0:20:04.608   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.611   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 207 times\n",
      "  0:20:04.612   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:20:04.612   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.612   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:20:04.612   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.619   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 12 times\n",
      "  0:20:04.619   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.621   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 127 times\n",
      "  0:20:04.621   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:20:04.622   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.622   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:20:04.622   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.629   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 8 times\n",
      "  0:20:04.630   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.631   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 99 times\n",
      "  0:20:04.631   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:20:04.631   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.632   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:04.632   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.637   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:20:04.637   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.639   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 94 times\n",
      "  0:20:04.639   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:20:04.639   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.667   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 37 times\n",
      "  0:20:04.668   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.731   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 15 times\n",
      "  0:20:04.732   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.758   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:20:04.758   653M / 1088M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:20:04.758   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.760   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:04.760   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.760   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:04.760   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:04.761   653M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:20:04.781   539M / 1088M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:20:04.781   539M / 1088M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:20:04.782   539M / 1088M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:20:04.782   539M / 1088M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:20:04.782   539M / 1088M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:20:04.782   539M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.806   539M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:04.806   539M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.840   540M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:04.841   540M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:04.864   540M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:04.864   540M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:04.922   540M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:04.923   540M / 1088M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:20:04.948   540M / 1088M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:20:05.009   539M / 1088M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 35758 times\n",
      "  0:20:05.022   538M / 1088M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:20:05.022   538M / 1088M INFO    General                 (simplification.cpp        : 509)     Average coverage = 229.856\n",
      "  0:20:05.034   538M / 1088M INFO    General                 (simplification.cpp        : 510)     Total length = 6488573\n",
      "  0:20:05.048   538M / 1088M INFO    General                 (simplification.cpp        : 511)     Median edge length: 2384\n",
      "  0:20:05.048   538M / 1088M INFO    General                 (simplification.cpp        : 512)     Edges: 22936\n",
      "  0:20:05.048   538M / 1088M INFO    General                 (simplification.cpp        : 513)     Vertices: 14822\n",
      "  0:20:05.049   538M / 1088M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:20:05.430   538M / 1088M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K21/simplified_contigs\"\n",
      "  0:20:05.484   542M / 1088M INFO    General                 (binary_converter.cpp      : 143)   11491 reads written\n",
      "  0:20:05.490   538M / 1088M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:20:05.552     1M / 1088M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 20 minutes 5 seconds\n",
      "\n",
      "===== K21 finished. \n",
      "\n",
      "\n",
      "===== K33 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/config.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K33/configs/isolate_mode.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=33\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.005     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.006     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.008     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.011     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.012     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K21/simplified_contigs\n",
      "  0:00:00.033     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.034     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:00:00.034     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.034     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.035     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.035     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.035     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:05.818   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 608983 reads\n",
      "  0:00:11.515   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1216623 reads\n",
      "  0:00:17.746   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1825881 reads\n",
      "  0:00:23.436   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2434386 reads\n",
      "  0:00:29.743   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3043283 reads\n",
      "  0:00:35.421   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3653144 reads\n",
      "  0:00:41.264   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4262379 reads\n",
      "  0:00:46.972   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4871742 reads\n",
      "  0:01:23.190   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8528968 reads\n",
      "  0:02:47.654   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 17057563 reads\n",
      "  0:04:11.842     1M / 571M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24967534 reads\n",
      "  0:04:11.842     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:08:49.121     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 74281939 kmers in total.\n",
      "  0:08:49.124     1M / 571M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:08:49.181     1M / 571M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:08:49.182     1M / 571M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:08:49.185     1M / 571M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:08:49.187     1M / 571M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:08:49.188     1M / 571M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:08:56.255   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 10150977 kmers\n",
      "  0:09:02.555   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 21181589 kmers\n",
      "  0:09:08.940   601M / 691M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 31250866 kmers\n",
      "  0:09:17.461   601M / 788M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 42935127 kmers\n",
      "  0:09:32.406   601M / 870M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 52996970 kmers\n",
      "  0:09:42.569   601M / 933M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 64378399 kmers\n",
      "  0:09:55.790   601M / 1030M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 74281972 kmers\n",
      "  0:09:55.795   601M / 1030M INFO    General                 (kmer_splitters.hpp        : 202)   Used 74281972 kmers.\n",
      "  0:09:55.844     1M / 1030M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:11:11.958     1M / 1030M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 73597170 kmers in total.\n",
      "  0:11:11.959     1M / 1030M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:11:25.905    54M / 1030M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 73597170 kmers, 53163784 bytes occupied (5.77889 bits per kmer).\n",
      "  0:11:25.905    54M / 1030M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:11:52.381   126M / 1030M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:13:27.340   126M / 1030M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:13:27.363   126M / 1030M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:13:27.364   126M / 1030M INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:13:27.364   126M / 1030M INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:14:50.306   206M / 1030M INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 977386\n",
      "  0:14:51.372   206M / 1030M INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 997454\n",
      "  0:14:51.394   126M / 1030M INFO   Early tip clipping       (early_simplification.hpp  :  50)   26010238 34-mers were removed by early tip clipper\n",
      "  0:14:51.396   126M / 1030M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:14:51.399   126M / 1030M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:16:07.229   236M / 1030M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2346402 sequences extracted\n",
      "  0:16:27.527   236M / 1030M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:16:43.843   236M / 1030M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:16:43.896   236M / 1030M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:16:45.158   236M / 1030M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:16:45.159   236M / 1030M INFO    General                 (debruijn_graph_constructor: 516)   Total 4692804 edges to create\n",
      "  0:16:45.162   417M / 1030M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:16:47.476   489M / 1030M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:16:49.016   489M / 1030M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:16:49.040   505M / 1030M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:16:49.569   505M / 1030M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:16:49.570   505M / 1030M INFO    General                 (debruijn_graph_constructor: 542)   Total 1661633 vertices to create\n",
      "  0:16:49.570   661M / 1030M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:16:50.582   537M / 1030M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:16:50.583   537M / 1030M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:17:03.922   591M / 1030M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 74281939 kmers, 53658448 bytes occupied (5.7789 bits per kmer).\n",
      "  0:17:04.090   875M / 1030M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:24:43.077   875M / 1030M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:24:53.872   875M / 1030M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:25:03.751   875M / 1030M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1966165 edges\n",
      "  0:25:11.729   875M / 1030M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2866165 edges\n",
      "  0:25:20.273   875M / 1030M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3758667 edges\n",
      "  0:25:26.866   875M / 1030M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4558667 edges\n",
      "  0:25:27.977   875M / 1030M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4692773 edges\n",
      "  0:25:28.427   413M / 1030M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:25:28.433   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 17\n",
      "  0:25:28.434   413M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 224\n",
      "  0:25:28.435   413M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 224. Coverage mad: 60.7866\n",
      "  0:25:28.435   413M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:25:28.652   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:25:29.565   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:25:31.013   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:25:35.693   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:25:37.696   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 197.76. Fitted coverage std. dev: 69.5032\n",
      "  0:25:37.706   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0364839\n",
      "  0:25:37.706   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 17\n",
      "  0:25:37.707   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 17\n",
      "  0:25:37.707   414M / 1030M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6411108\n",
      "  0:25:37.708   413M / 1030M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 197.76\n",
      "  0:25:37.708   413M / 1030M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 17\n",
      "  0:25:37.708   413M / 1030M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:25:37.708   413M / 1030M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:25:37.713   413M / 1030M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:25:37.718   413M / 1030M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:25:37.718   413M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:25:38.085   413M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 2 times\n",
      "  0:25:38.086   413M / 1030M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:25:38.086   413M / 1030M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:25:38.087   413M / 1030M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:25:38.087   413M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:25:38.087   413M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:25:39.211   413M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 33015 times\n",
      "  0:25:39.211   413M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:11.678   430M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 684216 times\n",
      "  0:26:11.687   430M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:12.741   428M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 68206 times\n",
      "  0:26:12.753   428M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:26:12.753   428M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:12.996   424M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 14463 times\n",
      "  0:26:13.005   424M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:13.655   420M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 14390 times\n",
      "  0:26:13.659   420M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:13.756   420M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 4216 times\n",
      "  0:26:13.758   420M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:26:13.758   420M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:13.777   420M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 531 times\n",
      "  0:26:13.778   420M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:13.912   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1264 times\n",
      "  0:26:13.913   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:13.934   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 547 times\n",
      "  0:26:13.934   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:26:13.935   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:13.937   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 47 times\n",
      "  0:26:13.938   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:13.967   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 171 times\n",
      "  0:26:13.968   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:13.975   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 157 times\n",
      "  0:26:13.975   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:26:13.976   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:13.976   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 11 times\n",
      "  0:26:13.977   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:13.984   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 24 times\n",
      "  0:26:13.985   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:13.988   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 73 times\n",
      "  0:26:13.988   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:26:13.988   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:13.989   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 6 times\n",
      "  0:26:13.989   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:13.996   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 10 times\n",
      "  0:26:13.997   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:14.000   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 61 times\n",
      "  0:26:14.000   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:26:14.000   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.001   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 4 times\n",
      "  0:26:14.002   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.005   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 4 times\n",
      "  0:26:14.005   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:14.007   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 37 times\n",
      "  0:26:14.007   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:26:14.007   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.007   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:26:14.007   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.011   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:26:14.012   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:14.013   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 36 times\n",
      "  0:26:14.013   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:26:14.014   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.014   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:26:14.014   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.017   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:26:14.017   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:14.019   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 24 times\n",
      "  0:26:14.019   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:26:14.019   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.019   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:26:14.020   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.022   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:26:14.022   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:14.025   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 13 times\n",
      "  0:26:14.025   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:26:14.025   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.050   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 12 times\n",
      "  0:26:14.051   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.077   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:26:14.078   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:14.100   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:26:14.100   419M / 1030M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:26:14.100   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.100   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:26:14.101   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.101   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:26:14.101   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:26:14.101   419M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:26:14.136   345M / 1030M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:26:14.137   345M / 1030M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:26:14.137   345M / 1030M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:26:14.137   345M / 1030M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:26:14.138   345M / 1030M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:26:14.138   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.177   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:26:14.177   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.202   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:26:14.202   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:14.225   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:26:14.225   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:14.251   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:26:14.251   345M / 1030M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:26:14.279   345M / 1030M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:26:14.451   343M / 1030M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 89158 times\n",
      "  0:26:14.460   341M / 1030M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:26:14.460   341M / 1030M INFO    General                 (simplification.cpp        : 509)     Average coverage = 203.895\n",
      "  0:26:14.468   341M / 1030M INFO    General                 (simplification.cpp        : 510)     Total length = 6513203\n",
      "  0:26:14.475   341M / 1030M INFO    General                 (simplification.cpp        : 511)     Median edge length: 22424\n",
      "  0:26:14.476   341M / 1030M INFO    General                 (simplification.cpp        : 512)     Edges: 4982\n",
      "  0:26:14.476   341M / 1030M INFO    General                 (simplification.cpp        : 513)     Vertices: 4046\n",
      "  0:26:14.476   341M / 1030M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:26:14.480   341M / 1030M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K33/simplified_contigs\"\n",
      "  0:26:14.505   345M / 1030M INFO    General                 (binary_converter.cpp      : 143)   2493 reads written\n",
      "  0:26:14.509   341M / 1030M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:26:14.549     1M / 1030M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 26 minutes 14 seconds\n",
      "\n",
      "===== K33 finished. \n",
      "\n",
      "\n",
      "===== K55 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/config.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K55/configs/isolate_mode.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=55\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.005     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.006     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.008     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.010     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K33/simplified_contigs\n",
      "  0:00:00.014     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.016     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.016     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.016     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:06.109   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 758726 reads\n",
      "  0:00:12.115   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1516834 reads\n",
      "  0:00:18.411   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2274976 reads\n",
      "  0:00:24.405   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3033866 reads\n",
      "  0:00:30.313   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3793509 reads\n",
      "  0:00:36.840   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4552653 reads\n",
      "  0:00:42.867   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5313017 reads\n",
      "  0:00:48.703   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6073528 reads\n",
      "  0:01:12.716   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9111813 reads\n",
      "  0:02:23.312   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 17466388 reads\n",
      "  0:03:26.182     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24949538 reads\n",
      "  0:03:26.183     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:06:16.168     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 80768556 kmers in total.\n",
      "  0:06:16.172     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:06:16.175     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:06:16.176     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:06:16.179     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:06:16.179     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:06:16.180     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:06:23.817   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 9634626 kmers\n",
      "  0:06:31.083   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 21363502 kmers\n",
      "  0:06:37.698   601M / 710M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 31698974 kmers\n",
      "  0:06:46.875   601M / 812M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 41312786 kmers\n",
      "  0:06:58.047   601M / 922M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 52181187 kmers\n",
      "  0:07:07.114   601M / 980M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 63344733 kmers\n",
      "  0:07:16.750   601M / 1070M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 72950204 kmers\n",
      "  0:07:25.649   601M / 1119M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 80768593 kmers\n",
      "  0:07:25.649   601M / 1119M INFO    General                 (kmer_splitters.hpp        : 202)   Used 80768593 kmers.\n",
      "  0:07:25.700     1M / 1119M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:08:33.838     1M / 1119M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 80765942 kmers in total.\n",
      "  0:08:33.839     1M / 1119M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:08:49.195    58M / 1119M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 80765942 kmers, 58341664 bytes occupied (5.77884 bits per kmer).\n",
      "  0:08:49.196    58M / 1119M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:08:58.130   138M / 1119M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:11:08.550   138M / 1119M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:11:08.564   138M / 1119M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:11:08.566   138M / 1119M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:12:58.099   294M / 1119M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 3123409 sequences extracted\n",
      "  0:13:32.655   294M / 1119M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:13:51.681   294M / 1119M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:13:51.774   294M / 1119M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:13:53.493   294M / 1119M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:13:53.494   294M / 1119M INFO    General                 (debruijn_graph_constructor: 516)   Total 6246818 edges to create\n",
      "  0:13:53.494   535M / 1119M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:13:56.963   631M / 1119M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:13:59.051   631M / 1119M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:13:59.103   663M / 1119M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:14:00.122   663M / 1119M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:14:00.122   663M / 1119M INFO    General                 (debruijn_graph_constructor: 542)   Total 3120795 vertices to create\n",
      "  0:14:00.123   952M / 1119M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:14:01.615   776M / 1119M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:14:01.617   776M / 1119M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:14:16.604   834M / 1119M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 80768556 kmers, 58343544 bytes occupied (5.77884 bits per kmer).\n",
      "  0:14:16.797  1146M / 1156M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:20:39.786  1146M / 1156M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:20:53.906  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:21:08.951  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:21:23.488  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:21:30.662  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3682742 edges\n",
      "  0:21:35.897  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4282742 edges\n",
      "  0:21:40.461  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4814020 edges\n",
      "  0:21:44.719  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5314020 edges\n",
      "  0:21:48.738  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5788719 edges\n",
      "  0:21:51.762  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 6151002 edges\n",
      "  0:21:52.579  1146M / 1156M INFO    General                 (coverage_filling.hpp      :  83)   Processed 6246815 edges\n",
      "  0:21:53.005   639M / 1156M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:21:53.008   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 14\n",
      "  0:21:53.008   639M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 174\n",
      "  0:21:53.009   639M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 174. Coverage mad: 50.4084\n",
      "  0:21:53.009   639M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:21:53.160   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:21:53.957   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:21:55.226   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:21:59.525   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:22:01.570   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 156.217. Fitted coverage std. dev: 56.7705\n",
      "  0:22:01.578   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.034713\n",
      "  0:22:01.579   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 15\n",
      "  0:22:01.579   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 15\n",
      "  0:22:01.579   640M / 1156M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6378969\n",
      "  0:22:01.579   639M / 1156M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 156.217\n",
      "  0:22:01.580   639M / 1156M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 15\n",
      "  0:22:01.580   639M / 1156M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:22:01.580   639M / 1156M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:22:01.587   653M / 1156M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:22:02.795   689M / 1156M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1663524 out of 6246815, length: 82189699\n",
      "  0:22:03.011   711M / 1156M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 6278055)\n",
      "  0:22:05.865   711M / 1156M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:22:58.478   753M / 1156M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 60230801 kmers, 43503272 bytes occupied (5.77821 bits per kmer).\n",
      "  0:22:58.748  1213M / 1328M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:23:33.465  1191M / 1328M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:23:56.837  1191M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:24:20.912  1192M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:24:44.631  1192M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:25:08.352  1193M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:25:31.887  1193M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:25:55.325  1194M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:27:51.439  1197M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:31:45.444  1202M / 1328M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:35:42.514  1224M / 1328M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:35:42.895   735M / 1328M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:35:42.897   735M / 1328M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:35:44.590   735M / 1328M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 6778 tips collected, total 7282 connection candidates\n",
      "  0:35:44.616   735M / 1328M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 118 gaps after checking 7153 candidates\n",
      "  0:35:44.759   735M / 1328M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:35:44.835   639M / 1328M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:35:44.846   639M / 1328M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:35:44.848   639M / 1328M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:35:44.848   639M / 1328M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:35:45.330   639M / 1328M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:35:45.331   639M / 1328M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:35:45.331   639M / 1328M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:35:45.332   639M / 1328M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:35:45.332   639M / 1328M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:35:45.332   639M / 1328M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:35:53.710   604M / 1328M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1161706 times\n",
      "  0:35:53.712   604M / 1328M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:50.629  1836M / 1883M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 226261 times\n",
      "  0:36:50.630  1836M / 1883M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:50.842  1837M / 1885M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8793 times\n",
      "  0:36:50.853  1837M / 1885M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:36:50.854  1837M / 1885M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:51.031  1835M / 1885M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 7943 times\n",
      "  0:36:51.034  1835M / 1885M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:51.864  1848M / 1896M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2524 times\n",
      "  0:36:51.870  1848M / 1896M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:51.923  1848M / 1897M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 756 times\n",
      "  0:36:51.925  1848M / 1897M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:36:51.926  1848M / 1897M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:51.947  1848M / 1897M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 228 times\n",
      "  0:36:51.948  1848M / 1897M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.029  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 164 times\n",
      "  0:36:52.031  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.043  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 75 times\n",
      "  0:36:52.044  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:36:52.045  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.047  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 12 times\n",
      "  0:36:52.048  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.062  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 17 times\n",
      "  0:36:52.063  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.067  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 18 times\n",
      "  0:36:52.068  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:36:52.069  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.072  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 4 times\n",
      "  0:36:52.074  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.081  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:36:52.082  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.089  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 12 times\n",
      "  0:36:52.090  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:36:52.091  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.093  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:36:52.094  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.102  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:36:52.107  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.110  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6 times\n",
      "  0:36:52.111  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:36:52.115  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.116  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:36:52.117  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.119  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:36:52.119  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.121  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:36:52.123  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:36:52.126  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.128  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:36:52.128  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.131  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:36:52.143  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.145  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:36:52.146  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:36:52.147  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.148  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:36:52.148  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.151  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:36:52.154  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.155  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:36:52.156  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:36:52.157  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.158  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:36:52.159  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.163  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:36:52.165  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.166  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:36:52.167  1849M / 1898M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:36:52.168  1849M / 1898M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.232  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 35 times\n",
      "  0:36:52.234  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.310  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:36:52.315  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.376  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:36:52.378  1849M / 1899M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:36:52.379  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:36:52.380  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:36:52.381  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:36:52.382  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:36:52.382  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:36:52.383  1849M / 1899M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:36:52.416  1795M / 1899M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:36:52.421  1809M / 1899M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:36:52.573  1825M / 1899M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 472438 out of 474291, length: 48768889\n",
      "  0:36:52.611  1829M / 1902M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 6278055)\n",
      "  0:36:53.623  1829M / 1902M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:37:09.596  1848M / 1940M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 25649236 kmers, 18526256 bytes occupied (5.77834 bits per kmer).\n",
      "  0:37:09.697  2044M / 2126M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:37:20.257  2040M / 2126M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:37:42.321  2040M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:38:03.909  2040M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:38:25.588  2041M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:38:47.626  2041M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:39:09.170  2041M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:39:30.744  2041M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:41:19.030  2043M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:44:56.726  2045M / 2126M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:48:41.958  2048M / 2126M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:48:42.084  1842M / 2131M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:48:42.091  1842M / 2131M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:48:42.230  1842M / 2131M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 449 tips collected, total 788 connection candidates\n",
      "  0:48:42.234  1842M / 2131M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 788 candidates\n",
      "  0:48:42.267  1842M / 2131M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:48:42.282  1795M / 2131M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:48:42.285  1795M / 2131M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:48:42.285  1795M / 2131M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:48:42.286  1795M / 2131M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:48:42.287  1795M / 2131M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:48:42.288  1795M / 2131M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:48:42.342  1795M / 2131M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:48:42.344  1795M / 2131M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:48:42.396  1796M / 2131M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:48:42.398  1796M / 2131M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:48:42.463  1796M / 2131M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:48:42.472  1796M / 2131M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:48:42.539  1796M / 2131M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:48:42.540  1796M / 2131M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:48:42.625  1796M / 2131M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:48:43.030  1792M / 2131M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 235840 times\n",
      "  0:48:43.048  1785M / 2131M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:48:43.049  1785M / 2131M INFO    General                 (simplification.cpp        : 509)     Average coverage = 158.432\n",
      "  0:48:43.058  1785M / 2131M INFO    General                 (simplification.cpp        : 510)     Total length = 6530004\n",
      "  0:48:43.071  1785M / 2131M INFO    General                 (simplification.cpp        : 511)     Median edge length: 52635\n",
      "  0:48:43.072  1785M / 2131M INFO    General                 (simplification.cpp        : 512)     Edges: 2612\n",
      "  0:48:43.073  1785M / 2131M INFO    General                 (simplification.cpp        : 513)     Vertices: 2742\n",
      "  0:48:43.074  1785M / 2131M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:48:45.981  1785M / 2131M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K55/simplified_contigs\"\n",
      "  0:48:46.013  1789M / 2131M INFO    General                 (binary_converter.cpp      : 143)   1306 reads written\n",
      "  0:48:46.019  1785M / 2131M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:48:50.742     1M / 2131M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 48 minutes 50 seconds\n",
      "\n",
      "===== K55 finished. \n",
      "\n",
      "\n",
      "===== K77 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/configs/isolate_mode.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/dataset.info\") with K=77\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.005     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.005     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.007     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.008     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.009     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K55/simplified_contigs\n",
      "  0:00:00.013     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.013     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 142.888\n",
      "  0:00:00.013     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.014     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.014     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:00:05.213   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 670453 reads\n",
      "  0:00:10.202   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1340357 reads\n",
      "  0:00:15.888   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2010788 reads\n",
      "  0:00:21.244   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2681456 reads\n",
      "  0:00:26.918   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3352592 reads\n",
      "  0:00:31.904   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4024031 reads\n",
      "  0:00:36.890   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4695551 reads\n",
      "  0:00:42.104   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5368155 reads\n",
      "  0:01:10.208   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8728117 reads\n",
      "  0:02:14.390   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 16788912 reads\n",
      "  0:03:23.319     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 24947164 reads\n",
      "  0:03:23.322     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:09:17.812     1M / 680M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 75620543 kmers in total.\n",
      "  0:09:17.814     1M / 680M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:09:17.837     1M / 680M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:09:17.838     1M / 680M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:09:17.854     1M / 680M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:09:17.854     1M / 680M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:09:17.855     1M / 680M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:09:26.658   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 5792426 kmers\n",
      "  0:09:37.526   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 12550071 kmers\n",
      "  0:09:43.947   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 19751620 kmers\n",
      "  0:09:54.279   601M / 680M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 27144535 kmers\n",
      "  0:10:00.151   601M / 712M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 34631252 kmers\n",
      "  0:10:11.139   601M / 777M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 42161625 kmers\n",
      "  0:10:20.770   601M / 801M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 49708820 kmers\n",
      "  0:10:32.900   601M / 854M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 57273014 kmers\n",
      "  0:10:43.955   601M / 885M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 64831267 kmers\n",
      "  0:10:50.473   601M / 930M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 72399278 kmers\n",
      "  0:10:53.116   601M / 930M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 75620598 kmers\n",
      "  0:10:53.117   601M / 930M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 75620598 kmers.\n",
      "  0:10:53.163     1M / 930M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:13:35.090     1M / 930M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 76046457 kmers in total.\n",
      "  0:13:35.093     1M / 930M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:13:53.638    55M / 930M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 76046457 kmers, 54932952 bytes occupied (5.77888 bits per kmer).\n",
      "  0:13:53.638    55M / 930M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:14:07.863   131M / 930M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:15:52.962   131M / 930M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:15:53.010   131M / 930M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:15:53.012   131M / 930M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:17:43.641   268M / 1031M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2575578 sequences extracted\n",
      "  0:18:16.176   268M / 1070M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:18:33.451   268M / 1070M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:18:33.549   268M / 1070M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:18:34.923   268M / 1070M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:18:34.924   268M / 1070M INFO    General                 (debruijn_graph_constructor: 516)   Total 5151156 edges to create\n",
      "  0:18:34.925   469M / 1070M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:18:38.054   549M / 1070M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:18:39.795   549M / 1070M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:18:39.840   581M / 1070M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:18:40.828   581M / 1070M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:18:40.829   581M / 1070M INFO    General                 (debruijn_graph_constructor: 542)   Total 3001492 vertices to create\n",
      "  0:18:40.829   862M / 1070M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:18:42.124   710M / 1070M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:18:42.126   710M / 1070M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:19:32.595   763M / 1070M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 75620543 kmers, 54625336 bytes occupied (5.77889 bits per kmer).\n",
      "  0:19:32.782  1055M / 1070M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:24:44.230  1055M / 1070M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:24:59.279  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:25:14.290  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:25:29.916  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:25:34.696  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3512737 edges\n",
      "  0:25:38.549  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3938006 edges\n",
      "  0:25:41.871  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4338006 edges\n",
      "  0:25:45.196  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4736484 edges\n",
      "  0:25:47.410  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5020558 edges\n",
      "  0:25:48.437  1055M / 1070M INFO    General                 (coverage_filling.hpp      :  83)   Processed 5151156 edges\n",
      "  0:25:49.023   580M / 1070M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:25:49.025   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 13\n",
      "  0:25:49.026   580M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 127\n",
      "  0:25:49.026   580M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 127. Coverage mad: 41.5128\n",
      "  0:25:49.026   580M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:25:49.149   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:25:49.587   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:25:50.718   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:25:54.287   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:25:57.528   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 116.71. Fitted coverage std. dev: 44.0833\n",
      "  0:25:57.535   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0263696\n",
      "  0:25:57.535   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 13\n",
      "  0:25:57.535   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 13\n",
      "  0:25:57.535   581M / 1070M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6327560\n",
      "  0:25:57.535   580M / 1070M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 116.71\n",
      "  0:25:57.535   580M / 1070M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 13\n",
      "  0:25:57.535   580M / 1070M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:25:57.535   580M / 1070M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:25:57.543   593M / 1070M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:25:58.700   661M / 1070M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1993397 out of 5151156, length: 98634764\n",
      "  0:25:58.978   685M / 1070M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 5176914)\n",
      "  0:26:02.310   685M / 1070M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:27:03.960   732M / 1070M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 67023893 kmers, 48409688 bytes occupied (5.7782 bits per kmer).\n",
      "  0:27:04.264  1244M / 1330M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:27:45.119  1220M / 1330M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:28:03.799  1221M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:28:22.623  1221M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:28:41.664  1222M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:29:00.581  1223M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:29:19.327  1223M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:29:38.578  1224M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:31:11.729  1227M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:34:17.824  1251M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:37:26.192  1258M / 1330M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:37:26.686   718M / 1330M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:37:26.687   718M / 1330M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:37:28.282   718M / 1330M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 7358 tips collected, total 7822 connection candidates\n",
      "  0:37:28.315   718M / 1330M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 243 gaps after checking 7552 candidates\n",
      "  0:37:28.463   718M / 1330M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:37:28.578   580M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:37:28.593   580M / 1330M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:37:28.595   580M / 1330M INFO    General                 (simplification.cpp        :  63)   Most init cleaning disabled on main iteration\n",
      "  0:37:28.596   580M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:37:29.028   580M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:37:29.029   580M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:37:29.029   580M / 1330M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:37:29.029   580M / 1330M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:37:29.030   580M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:37:29.030   580M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:37.122   543M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1077894 times\n",
      "  0:37:37.124   543M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.454   744M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 18331 times\n",
      "  0:37:41.459   744M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.565   744M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 858 times\n",
      "  0:37:41.566   744M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:37:41.567   744M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.628   744M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1059 times\n",
      "  0:37:41.629   744M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.745   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 158 times\n",
      "  0:37:41.750   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.790   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 158 times\n",
      "  0:37:41.791   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:37:41.791   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.799   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 52 times\n",
      "  0:37:41.799   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.813   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 20 times\n",
      "  0:37:41.814   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.820   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 25 times\n",
      "  0:37:41.820   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:37:41.821   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.823   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 7 times\n",
      "  0:37:41.823   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.824   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:37:41.825   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.826   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8 times\n",
      "  0:37:41.826   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:37:41.827   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.827   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:37:41.827   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.827   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:41.828   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.830   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 5 times\n",
      "  0:37:41.830   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:37:41.830   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.831   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 3 times\n",
      "  0:37:41.831   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.831   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:41.832   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.832   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:37:41.832   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:37:41.833   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.833   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:41.833   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.833   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:41.834   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.834   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:37:41.835   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:37:41.835   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.836   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:41.838   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.838   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:41.838   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.838   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:37:41.839   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:37:41.839   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.839   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:41.839   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.840   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:41.840   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.840   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:37:41.841   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:37:41.841   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.841   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:41.841   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:41.842   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:41.842   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:41.842   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:37:41.842   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:37:41.843   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:41.941   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 49 times\n",
      "  0:37:41.942   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:42.066   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:42.067   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:42.154   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:37:42.155   745M / 1330M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:37:42.155   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:37:42.156   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:37:42.156   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:37:42.156   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:37:42.156   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:37:42.157   745M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:37:42.208   708M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:37:42.216   721M / 1330M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:37:42.440   738M / 1330M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 890872 out of 892036, length: 69164747\n",
      "  0:37:42.518   745M / 1330M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 5176914)\n",
      "  0:37:44.126   745M / 1330M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:38:12.696   771M / 1330M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 35984151 kmers, 25992262 bytes occupied (5.7786 bits per kmer).\n",
      "  0:38:12.849  1047M / 1330M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:38:31.592  1040M / 1330M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:38:43.437  1041M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:38:55.013  1042M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:39:06.602  1042M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:39:18.369  1043M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:39:30.069  1043M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:39:41.785  1044M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:40:39.798  1047M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:42:34.923  1053M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:44:34.151  1076M / 1330M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:44:34.499   784M / 1330M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:44:34.500   784M / 1330M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:44:34.714   784M / 1330M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 1135 tips collected, total 2039 connection candidates\n",
      "  0:44:34.731   784M / 1330M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 2039 candidates\n",
      "  0:44:34.819   784M / 1330M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:44:34.851   708M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:44:34.858   708M / 1330M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:44:34.858   708M / 1330M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:44:34.858   708M / 1330M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:44:34.859   708M / 1330M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:44:34.859   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:44:34.959   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:44:34.963   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:44:35.074   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:44:35.075   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:44:35.155   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:44:35.156   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:44:35.282   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:44:35.283   708M / 1330M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:44:35.410   708M / 1330M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:44:36.133   697M / 1330M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 445215 times\n",
      "  0:44:36.143   688M / 1330M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:44:36.143   688M / 1330M INFO    General                 (simplification.cpp        : 509)     Average coverage = 116.51\n",
      "  0:44:36.151   688M / 1330M INFO    General                 (simplification.cpp        : 510)     Total length = 6505227\n",
      "  0:44:36.161   688M / 1330M INFO    General                 (simplification.cpp        : 511)     Median edge length: 67843\n",
      "  0:44:36.161   688M / 1330M INFO    General                 (simplification.cpp        : 512)     Edges: 1606\n",
      "  0:44:36.161   688M / 1330M INFO    General                 (simplification.cpp        : 513)     Vertices: 1652\n",
      "  0:44:36.162   688M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Mismatch Correction (id: mismatch_correction)\n",
      "  0:44:36.162   688M / 1330M INFO    General                 (graph_pack_helpers.cpp    :  44)   Index refill\n",
      "  0:44:36.163   688M / 1330M INFO    General                 (edge_index.hpp            : 175)   Using small index (max_id = 5176914)\n",
      "  0:44:36.544   688M / 1330M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:44:38.884   693M / 1330M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 6505227 kmers, 4701962 bytes occupied (5.78238 bits per kmer).\n",
      "  0:44:38.909   745M / 1330M INFO    General                 (edge_index_builders.hpp   : 253)   Collecting edge information from graph, this takes a while.\n",
      "  0:44:40.840   745M / 1330M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 2901750 kmers to process\n",
      "  0:44:42.155   748M / 1330M INFO    General                 (kmer_mapper.hpp           :  87)   Total 9902 kmers with non-trivial roots\n",
      "  0:44:42.170   745M / 1330M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:44:42.171   745M / 1330M INFO    General                 (mismatch_correction.cpp   : 392)   Collect potential mismatches\n",
      "  0:44:47.151   745M / 1330M INFO    General                 (mismatch_correction.cpp   : 193)   Total 908 edges (out of 1606) with 51629 potential mismatch positions (56.8601 positions per edge)\n",
      "  0:44:47.152   745M / 1330M INFO    General                 (mismatch_correction.cpp   : 394)   Potential mismatches collected\n",
      "  0:44:47.153   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:44:49.051   748M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:44:50.979   751M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:44:52.873   752M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:44:54.701   752M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:44:56.605   752M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:44:58.419   753M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:45:07.143   753M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:45:24.272   753M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:46:01.432   753M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 8400000 reads\n",
      "  0:47:13.230   753M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 16800000 reads\n",
      "  0:48:27.107   753M / 1330M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 24944552 reads processed\n",
      "  0:48:27.181   749M / 1330M INFO    General                 (mismatch_correction.cpp   : 387)   All edges processed\n",
      "  0:48:27.183   745M / 1330M INFO    General                 (mismatch_correction.cpp   : 442)   Corrected 0 nucleotides\n",
      "  0:48:27.183   745M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:48:27.183   745M / 1330M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/before_rr.fasta\"\n",
      "  0:48:27.611   745M / 1330M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_after_simplification.gfa\"\n",
      "  0:48:27.689   745M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Paired Information Counting (id: late_pair_info_count)\n",
      "  0:48:27.707   745M / 1330M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 2901750 kmers to process\n",
      "  0:48:27.707   745M / 1330M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:48:27.715   745M / 1330M INFO    General                 (pair_info_count.cpp       : 157)   Min edge length for estimation: 67843\n",
      "  0:48:27.715   745M / 1330M INFO    General                 (pair_info_count.cpp       : 168)   Estimating insert size for library #0\n",
      "  0:48:27.716   745M / 1330M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:48:27.717   745M / 1330M INFO    General                 (paired_info_utils.cpp     :  87)   Estimating insert size (takes a while)\n",
      "  0:48:27.727   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:48:31.220   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:48:34.622   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:48:38.084   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:48:41.553   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:48:44.919   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:48:48.212   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:49:05.225   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:49:37.860   777M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:50:13.581   777M / 1330M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:50:13.919   777M / 1330M INFO    General                 (paired_info_utils.cpp     : 104)   Edge pairs: 8496\n",
      "  0:50:13.920   777M / 1330M INFO    General                 (paired_info_utils.cpp     : 106)   2940158 paired reads (47.1471% of all) aligned to long edges\n",
      "  0:50:13.921   745M / 1330M INFO    General                 (pair_info_count.cpp       : 188)     Insert size = 213.149, deviation = 72.7554, left quantile = 128, right quantile = 310, read length = 151\n",
      "  0:50:13.922   745M / 1330M INFO    General                 (pair_info_count.cpp       : 203)   Filtering data for library #0\n",
      "  0:50:13.922   745M / 1330M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:50:13.923   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:50:17.385   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:50:20.816   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:50:24.133   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:50:27.373   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:50:30.731   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:50:33.980   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:50:50.964   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:51:23.026   745M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:51:58.086   745M / 1330M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:51:58.087   745M / 1330M INFO    General                 (pair_info_count.cpp       : 207)   Mapping library #0\n",
      "  0:51:58.088   745M / 1330M INFO    General                 (pair_info_count.cpp       : 209)   Mapping paired reads (takes a while)\n",
      "  0:51:58.088   745M / 1330M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:51:58.088   745M / 1330M INFO    General                 (paired_info_utils.cpp     : 138)   Left insert size quantile 128, right insert size quantile 310, filtering threshold 2, rounding threshold 0\n",
      "  0:51:58.092   758M / 1330M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:52:01.727   759M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:52:05.270   759M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:52:08.786   759M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:52:12.205   759M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:52:15.638   760M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:52:19.059   760M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:52:36.423   760M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:53:11.333   761M / 1330M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:53:48.253   762M / 1330M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6236138 reads processed\n",
      "  0:53:48.259   748M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Distance Estimation (id: distance_estimation)\n",
      "  0:53:48.260   748M / 1330M INFO    General                 (distance_estimation.cpp   :  42)   Processing library #0\n",
      "  0:53:48.260   748M / 1330M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:53:48.263   748M / 1330M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SIMPLE distance estimator\n",
      "  0:53:48.301   749M / 1330M INFO    General                 (distance_estimation_utils.: 131)   Refining clustered pair information\n",
      "  0:53:48.302   749M / 1330M INFO    General                 (distance_estimation_utils.: 133)   The refining of clustered pair information has been finished\n",
      "  0:53:48.302   749M / 1330M INFO    General                 (distance_estimation_utils.: 135)   Improving paired information\n",
      "  0:53:48.354   749M / 1330M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 34; contradictional = 54\n",
      "  0:53:48.405   749M / 1330M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 2; contradictional = 4\n",
      "  0:53:48.406   749M / 1330M INFO    General                 (distance_estimation_utils.:  80)   Filling scaffolding index\n",
      "  0:53:48.406   749M / 1330M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:53:48.406   749M / 1330M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SMOOTHING distance estimator\n",
      "  0:53:48.543   749M / 1330M INFO    General                 (distance_estimation.cpp   :  51)   Clearing raw paired index\n",
      "  0:53:48.546   745M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Repeat Resolving (id: repeat_resolving)\n",
      "  0:53:48.552   745M / 1330M INFO    General                 (repeat_resolving.cpp      :  88)   Using Path-Extend repeat resolving\n",
      "  0:53:48.554   745M / 1330M INFO    General                 (launcher.cpp              : 603)   ExSPAnder repeat resolving tool started\n",
      "  0:53:48.563   746M / 1330M INFO    General                 (launcher.cpp              : 420)   Creating main extenders, unique edge length = 2000\n",
      "  0:53:48.564   746M / 1330M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 116.51\n",
      "  0:53:48.565   746M / 1330M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5104\n",
      "  0:53:48.565   746M / 1330M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 116.51\n",
      "  0:53:48.565   746M / 1330M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5104\n",
      "  0:53:48.573   746M / 1330M INFO    General                 (extenders_logic.cpp       : 550)   Using 1 paired-end library\n",
      "  0:53:48.574   746M / 1330M INFO    General                 (extenders_logic.cpp       : 551)   Using 1 paired-end scaffolding library\n",
      "  0:53:48.574   746M / 1330M INFO    General                 (extenders_logic.cpp       : 552)   Using 0 single read libraries\n",
      "  0:53:48.574   746M / 1330M INFO    General                 (launcher.cpp              : 449)   Total number of extenders is 3\n",
      "  0:53:48.575   746M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 0 paths from 780 (0%)\n",
      "  0:53:48.593   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 79 paths from 780 (10%)\n",
      "  0:53:48.595   747M / 1330M INFO    General                 (path_extenders.cpp        :  34)   Processed 128 paths from 780 (16%)\n",
      "  0:53:48.597   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 158 paths from 780 (20%)\n",
      "  0:53:48.600   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 237 paths from 780 (30%)\n",
      "  0:53:48.601   747M / 1330M INFO    General                 (path_extenders.cpp        :  34)   Processed 256 paths from 780 (32%)\n",
      "  0:53:48.601   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 316 paths from 780 (40%)\n",
      "  0:53:48.602   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 395 paths from 780 (50%)\n",
      "  0:53:48.604   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 474 paths from 780 (60%)\n",
      "  0:53:48.604   747M / 1330M INFO    General                 (path_extenders.cpp        :  34)   Processed 512 paths from 780 (65%)\n",
      "  0:53:48.605   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 553 paths from 780 (70%)\n",
      "  0:53:48.606   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 632 paths from 780 (81%)\n",
      "  0:53:48.607   747M / 1330M INFO    General                 (path_extenders.cpp        :  36)   Processed 711 paths from 780 (91%)\n",
      "  0:53:48.607   747M / 1330M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:53:48.608   747M / 1330M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:53:48.608   747M / 1330M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:53:48.609   747M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:53:48.609   747M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:53:48.610   747M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:53:48.610   747M / 1330M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:53:48.611   747M / 1330M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:53:48.612   747M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:53:48.612   747M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:53:48.613   747M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:53:48.621   747M / 1330M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:53:48.621   747M / 1330M INFO    General                 (launcher.cpp              : 456)   Closing gaps in paths\n",
      "  0:53:48.622   747M / 1330M INFO    General                 (launcher.cpp              : 486)   Gap closing completed\n",
      "  0:53:48.623   745M / 1330M INFO    General                 (launcher.cpp              : 304)   Traversing tandem repeats\n",
      "  0:53:48.638   745M / 1330M INFO    General                 (launcher.cpp              : 314)   Traversed 3 loops\n",
      "  0:53:48.639   745M / 1330M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:53:48.639   745M / 1330M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:53:48.640   745M / 1330M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:53:48.640   745M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:53:48.640   745M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:53:48.640   745M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:53:48.641   745M / 1330M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:53:48.641   745M / 1330M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:53:48.642   745M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:53:48.642   745M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:53:48.643   745M / 1330M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:53:48.651   745M / 1330M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:53:48.651   745M / 1330M INFO    General                 (launcher.cpp              : 666)   ExSPAnder repeat resolving tool finished\n",
      "  0:53:48.651   745M / 1330M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:53:48.651   745M / 1330M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/before_rr.fasta\"\n",
      "  0:53:48.802   745M / 1330M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_with_scaffolds.gfa\"\n",
      "  0:53:48.920   745M / 1330M INFO    General                 (contig_output_stage.cpp   : 169)   Outputting FastG graph to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph.fastg\"\n",
      "  0:53:49.193   745M / 1330M INFO    General                 (contig_output_stage.cpp   : 200)   Breaking scaffolds\n",
      "  0:53:49.211   752M / 1330M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.fasta\n",
      "  0:53:49.302   752M / 1330M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.paths\n",
      "  0:53:49.320   752M / 1330M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.fasta\n",
      "  0:53:49.424   752M / 1330M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.paths\n",
      "  0:53:49.430   752M / 1330M INFO    General                 (contig_output_stage.cpp   : 114)   Populating GFA with scaffold paths\n",
      "  0:53:49.435   745M / 1330M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:53:49.858     1M / 1330M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 53 minutes 50 seconds\n",
      "\n",
      "===== K77 finished. \n",
      "\n",
      "\n",
      "===== Copy files started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/copy_files.py /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/before_rr.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/before_rr.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/strain_graph.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/strain_graph.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/scaffolds.paths /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.paths /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/assembly_graph.fastg /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph.fastg /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/K77/final_contigs.paths /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.paths\n",
      "\n",
      "\n",
      "===== Copy files finished. \n",
      "\n",
      "\n",
      "===== Assembling finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.fasta --misc_dir /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Assembled contigs are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.fasta\n",
      " * Assembled scaffolds are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.fasta\n",
      " * Paths in the assembly graph corresponding to the contigs are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/scaffolds.paths\n",
      " * Assembly graph is in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/assembly_graph_with_scaffolds.gfa\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PA14wt0253G_SPAdes/spades.log\n",
      "\n",
      "Thank you for using SPAdes! If you use it in your research, please cite:\n",
      "\n",
      "  Prjibelski, A., Antipov, D., Meleshko, D., Lapidus, A. and Korobeynikov, A., 2020. Using SPAdes de novo assembler. Current protocols in bioinformatics, 70(1), p.e102.\n",
      "  doi.org/10.1002/cpbi.102\n",
      "\n",
      "Command line: /home/student/miniconda3/envs/VariantCalling/bin/spades.py\t-o\t/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes\t-1\t/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PAO1wt12P_R1.fastq.gz\t-2\t/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PAO1wt12P_R2.fastq.gz\t--isolate\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 4.1.0\n",
      "  Python version: 3.11.0\n",
      "  OS: Linux-5.15.0-139-generic-x86_64-with-glibc2.31\n",
      "\n",
      "Output dir: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes\n",
      "Mode: ONLY assembling (without read error correction)\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Isolate mode\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PAO1wt12P_R1.fastq.gz']\n",
      "      right reads: ['/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PAO1wt12P_R2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "  Assembly graph output will use GFA v1.2 format\n",
      "Other parameters:\n",
      "  Dir for temp files: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 10\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/spades.log\n",
      "\n",
      "/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PAO1wt12P_R1.fastq.gz: max reads length: 151\n",
      "/media/student/EXTERNAL_USB/Prova/Analysis/fastq/PAO1wt12P_R2.fastq.gz: max reads length: 151\n",
      "\n",
      "Reads length: 151\n",
      "\n",
      "Default k-mer sizes were set to [21, 33, 55, 77] because estimated read length (151) is equal to or greater than 150\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Assembling started. \n",
      "\n",
      "\n",
      "===== K21 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K21/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=21\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.002     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  78)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  99)   Converting paired reads\n",
      "  0:00:00.190    34M / 40M   INFO    General                 (binary_converter.cpp      : 127)   16384 reads processed\n",
      "  0:00:00.321    39M / 45M   INFO    General                 (binary_converter.cpp      : 127)   32768 reads processed\n",
      "  0:00:00.813    49M / 55M   INFO    General                 (binary_converter.cpp      : 127)   65536 reads processed\n",
      "  0:00:01.542    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   131072 reads processed\n",
      "  0:00:03.165    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   262144 reads processed\n",
      "  0:00:06.180    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   524288 reads processed\n",
      "  0:00:12.186    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   1048576 reads processed\n",
      "  0:00:24.433    59M / 65M   INFO    General                 (binary_converter.cpp      : 127)   2097152 reads processed\n",
      "  0:00:41.495    59M / 65M   INFO    General                 (binary_converter.cpp      : 143)   3281626 reads written\n",
      "  0:00:41.502     2M / 65M   INFO    General                 (read_converter.cpp        : 113)   Converting single reads\n",
      "  0:00:41.505    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:00:41.506     2M / 65M   INFO    General                 (read_converter.cpp        : 119)   Converting merged reads\n",
      "  0:00:41.509    15M / 65M   INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:00:41.510     1M / 65M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:41.511     1M / 65M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:41.512     1M / 65M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:41.512     1M / 65M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:41.512     1M / 65M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:41.512     1M / 65M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:41.513     1M / 65M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:41.513     1M / 65M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:00:51.299   601M / 627M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1070676 reads\n",
      "  0:01:00.558   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2142854 reads\n",
      "  0:01:10.290   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3215498 reads\n",
      "  0:01:19.529   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4287837 reads\n",
      "  0:01:28.970   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5360694 reads\n",
      "  0:01:38.765   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6432611 reads\n",
      "  0:01:48.180   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 7505758 reads\n",
      "  0:01:57.590   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8576863 reads\n",
      "  0:02:06.957   601M / 628M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9649276 reads\n",
      "  0:02:37.732     1M / 628M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13126504 reads\n",
      "  0:02:37.732     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:02:44.438     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 48436930 kmers in total.\n",
      "  0:02:44.439     1M / 628M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:02:44.441     1M / 628M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:02:44.441     1M / 628M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:02:44.442     1M / 628M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:02:44.443     1M / 628M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:02:44.443     1M / 628M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 6710886\n",
      "  0:02:57.922   601M / 851M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 32076319 kmers\n",
      "  0:03:04.069   601M / 928M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 48436934 kmers\n",
      "  0:03:04.069   601M / 928M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 48436934 kmers.\n",
      "  0:03:04.108     1M / 928M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:08.056     1M / 928M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 47505179 kmers in total.\n",
      "  0:03:08.058     1M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:03:16.018    37M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 47505179 kmers, 34318528 bytes occupied (5.77933 bits per kmer).\n",
      "  0:03:16.019    37M / 928M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:03:19.055    85M / 928M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:03:58.984    85M / 928M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:03:59.014    85M / 928M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:03:59.014    85M / 928M  INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:03:59.015    85M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:04:46.534   125M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 589429\n",
      "  0:04:47.298   125M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 596916\n",
      "  0:04:47.314    85M / 928M  INFO   Early tip clipping       (early_simplification.hpp  :  50)   12821123 22-mers were removed by early tip clipper\n",
      "  0:04:47.316    85M / 928M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:04:47.318    85M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:05:33.540   187M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2712000 sequences extracted\n",
      "  0:05:47.164   187M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:05:54.225   187M / 928M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 1 loops collected\n",
      "  0:05:54.309   187M / 928M  INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:05:55.684   187M / 928M  INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:05:55.685   187M / 928M  INFO    General                 (debruijn_graph_constructor: 516)   Total 5424002 edges to create\n",
      "  0:05:55.685   395M / 928M  INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:05:58.170   479M / 928M  INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:06:00.041   479M / 928M  INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:06:00.072   495M / 928M  INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:06:00.635   495M / 928M  INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:06:00.636   495M / 928M  INFO    General                 (debruijn_graph_constructor: 542)   Total 1780250 vertices to create\n",
      "  0:06:00.636   660M / 928M  INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:06:01.759   516M / 928M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:06:01.759   516M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:06:08.980   552M / 928M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 48436930 kmers, 34991576 bytes occupied (5.77932 bits per kmer).\n",
      "  0:06:09.055   740M / 928M  INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:10:18.352   740M / 928M  INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:10:23.690   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:10:29.117   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:10:34.679   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 3000000 edges\n",
      "  0:10:41.825   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 3989215 edges\n",
      "  0:10:46.648   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 4889215 edges\n",
      "  0:10:49.683   740M / 928M  INFO    General                 (coverage_filling.hpp      :  83)   Processed 5423789 edges\n",
      "  0:10:49.827   433M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:10:49.830   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 15\n",
      "  0:10:49.831   433M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 119\n",
      "  0:10:49.831   433M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 124. Coverage mad: 31.1346\n",
      "  0:10:49.831   433M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:10:50.047   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:10:50.814   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:10:52.085   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:10:55.683   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:10:56.222   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 123.373. Fitted coverage std. dev: 32.3346\n",
      "  0:10:56.228   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0888995\n",
      "  0:10:56.229   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 18\n",
      "  0:10:56.229   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 18\n",
      "  0:10:56.229   434M / 928M  INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6171935\n",
      "  0:10:56.230   433M / 928M  INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 123.373\n",
      "  0:10:56.230   433M / 928M  INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 18\n",
      "  0:10:56.230   433M / 928M  INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:10:56.230   433M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:10:56.236   433M / 928M  INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:10:56.237   433M / 928M  INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:10:56.238   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:10:56.673   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:10:56.673   433M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:10:56.674   433M / 928M  INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:10:56.674   433M / 928M  INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:10:56.675   433M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:10:56.675   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:10:57.746   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 9936 times\n",
      "  0:10:57.747   433M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:33.263   466M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 810733 times\n",
      "  0:11:33.272   466M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:34.965   465M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 120365 times\n",
      "  0:11:34.990   465M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:11:34.991   465M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:35.132   463M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 8015 times\n",
      "  0:11:35.135   463M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:35.810   457M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 14404 times\n",
      "  0:11:35.814   457M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:35.961   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 9153 times\n",
      "  0:11:35.964   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:11:35.964   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:35.973   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 183 times\n",
      "  0:11:35.974   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.080   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1081 times\n",
      "  0:11:36.084   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.114   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1138 times\n",
      "  0:11:36.124   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:11:36.125   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.127   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 23 times\n",
      "  0:11:36.127   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.149   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 160 times\n",
      "  0:11:36.149   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.154   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 418 times\n",
      "  0:11:36.155   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:11:36.155   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.156   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 8 times\n",
      "  0:11:36.156   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.174   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 58 times\n",
      "  0:11:36.175   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.178   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 230 times\n",
      "  0:11:36.178   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:11:36.178   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.179   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:11:36.179   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.186   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 20 times\n",
      "  0:11:36.187   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.190   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 111 times\n",
      "  0:11:36.190   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:11:36.190   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.191   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:11:36.191   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.198   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 14 times\n",
      "  0:11:36.199   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.200   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 64 times\n",
      "  0:11:36.201   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:11:36.201   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.201   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:11:36.201   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.203   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:11:36.203   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.204   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 55 times\n",
      "  0:11:36.205   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:11:36.205   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.205   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:11:36.205   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.211   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 7 times\n",
      "  0:11:36.212   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.212   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 31 times\n",
      "  0:11:36.213   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:11:36.213   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.213   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:11:36.213   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.215   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:11:36.215   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.216   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:11:36.216   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:11:36.216   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.234   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 17 times\n",
      "  0:11:36.235   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.268   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 12 times\n",
      "  0:11:36.268   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.296   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:11:36.296   455M / 928M  INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:11:36.297   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.298   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:11:36.302   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.302   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:11:36.302   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:11:36.302   455M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:11:36.328   378M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:11:36.329   378M / 928M  INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:11:36.329   378M / 928M  INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:11:36.329   378M / 928M  INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:11:36.329   378M / 928M  INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:11:36.330   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.364   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:11:36.365   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.411   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:11:36.412   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:11:36.429   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:11:36.429   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:11:36.455   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:11:36.456   378M / 928M  INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:11:36.471   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:11:36.551   378M / 928M  INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 30034 times\n",
      "  0:11:36.571   377M / 928M  INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:11:36.572   377M / 928M  INFO    General                 (simplification.cpp        : 509)     Average coverage = 128.483\n",
      "  0:11:36.586   377M / 928M  INFO    General                 (simplification.cpp        : 510)     Total length = 6215872\n",
      "  0:11:36.605   377M / 928M  INFO    General                 (simplification.cpp        : 511)     Median edge length: 2364\n",
      "  0:11:36.612   377M / 928M  INFO    General                 (simplification.cpp        : 512)     Edges: 20281\n",
      "  0:11:36.613   377M / 928M  INFO    General                 (simplification.cpp        : 513)     Vertices: 12924\n",
      "  0:11:36.613   377M / 928M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:11:36.614   377M / 928M  INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K21/simplified_contigs\"\n",
      "  0:11:36.658   380M / 928M  INFO    General                 (binary_converter.cpp      : 143)   10159 reads written\n",
      "  0:11:36.661   377M / 928M  INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:11:36.705     1M / 928M  INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 11 minutes 36 seconds\n",
      "\n",
      "===== K21 finished. \n",
      "\n",
      "\n",
      "===== K33 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/config.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K33/configs/isolate_mode.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=33\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.002     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.003     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.003     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.004     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K21/simplified_contigs\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:06.000   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 592528 reads\n",
      "  0:00:11.890   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1184368 reads\n",
      "  0:00:17.860   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1777339 reads\n",
      "  0:00:24.152   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2370360 reads\n",
      "  0:00:29.963   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2964046 reads\n",
      "  0:00:36.175   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3556781 reads\n",
      "  0:00:41.816   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4149044 reads\n",
      "  0:00:47.550   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4742126 reads\n",
      "  0:01:29.723   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8892369 reads\n",
      "  0:02:14.289     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13146822 reads\n",
      "  0:02:14.290     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:03:52.615     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 56063755 kmers in total.\n",
      "  0:03:52.615     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:03:52.632     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:03:52.632     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:03:52.641     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:03:52.641     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:03:52.641     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:03:59.383   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 12497616 kmers\n",
      "  0:04:07.226   601M / 616M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 24991680 kmers\n",
      "  0:04:15.913   601M / 707M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 37496564 kmers\n",
      "  0:04:26.076   601M / 735M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 49952961 kmers\n",
      "  0:04:30.016   601M / 752M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 56063775 kmers\n",
      "  0:04:30.017   601M / 752M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 56063775 kmers.\n",
      "  0:04:30.064     1M / 752M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:05:20.763     1M / 752M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 55622538 kmers in total.\n",
      "  0:05:20.764     1M / 752M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:05:30.255    41M / 752M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 55622538 kmers, 40181504 bytes occupied (5.77917 bits per kmer).\n",
      "  0:05:30.256    41M / 752M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:05:36.041    97M / 752M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:06:32.509    97M / 752M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:06:32.521    97M / 752M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:06:32.522    97M / 752M  INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:06:32.522    97M / 752M  INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:07:25.300   177M / 752M  INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 686919\n",
      "  0:07:26.011   177M / 752M  INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 698361\n",
      "  0:07:26.021    97M / 752M  INFO   Early tip clipping       (early_simplification.hpp  :  50)   19565956 34-mers were removed by early tip clipper\n",
      "  0:07:26.024    97M / 752M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:07:26.027    97M / 752M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:08:17.606   172M / 1056M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1608188 sequences extracted\n",
      "  0:08:32.014   172M / 1081M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:08:41.505   172M / 1081M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:08:41.569   172M / 1081M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:08:42.412   172M / 1081M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:08:42.413   172M / 1081M INFO    General                 (debruijn_graph_constructor: 516)   Total 3216376 edges to create\n",
      "  0:08:42.414   297M / 1081M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:08:44.013   349M / 1081M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:08:45.113   349M / 1081M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:08:45.123   365M / 1081M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:08:45.509   365M / 1081M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:08:45.509   365M / 1081M INFO    General                 (debruijn_graph_constructor: 542)   Total 1166971 vertices to create\n",
      "  0:08:45.510   473M / 1081M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:08:46.187   380M / 1081M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:08:46.187   380M / 1081M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:08:55.257   420M / 1081M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 56063755 kmers, 40500264 bytes occupied (5.77917 bits per kmer).\n",
      "  0:08:55.412   636M / 1081M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:12:56.989   636M / 1081M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:13:07.273   636M / 1081M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:13:15.912   636M / 1081M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1916697 edges\n",
      "  0:13:23.241   636M / 1081M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2748032 edges\n",
      "  0:13:27.060   636M / 1081M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3216358 edges\n",
      "  0:13:27.628   285M / 1081M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:13:27.631   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 13\n",
      "  0:13:27.631   285M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 105\n",
      "  0:13:27.631   285M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 110. Coverage mad: 29.652\n",
      "  0:13:27.632   285M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:13:27.771   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:13:28.644   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:13:29.985   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:13:33.045   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:13:33.912   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 109.759. Fitted coverage std. dev: 29.6043\n",
      "  0:13:33.917   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0689098\n",
      "  0:13:33.918   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 15\n",
      "  0:13:33.918   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 15\n",
      "  0:13:33.918   286M / 1081M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6183950\n",
      "  0:13:33.918   285M / 1081M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 109.759\n",
      "  0:13:33.919   285M / 1081M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 15\n",
      "  0:13:33.919   285M / 1081M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:13:33.919   285M / 1081M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:13:33.922   285M / 1081M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:13:33.923   285M / 1081M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:13:33.924   285M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:13:34.193   285M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 2 times\n",
      "  0:13:34.198   285M / 1081M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:13:34.199   285M / 1081M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:13:34.199   285M / 1081M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:13:34.199   285M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:13:34.199   285M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:34.944   285M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 18250 times\n",
      "  0:13:34.945   285M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:52.009   298M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 479712 times\n",
      "  0:13:52.010   298M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:52.538   297M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 28417 times\n",
      "  0:13:52.545   297M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:13:52.545   297M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:52.677   296M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 7377 times\n",
      "  0:13:52.679   296M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:52.991   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 5931 times\n",
      "  0:13:52.993   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.156   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3084 times\n",
      "  0:13:53.158   294M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:13:53.158   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.170   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 189 times\n",
      "  0:13:53.170   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.229   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 349 times\n",
      "  0:13:53.230   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.284   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 296 times\n",
      "  0:13:53.288   294M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:13:53.288   294M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.294   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 17 times\n",
      "  0:13:53.294   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.311   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 45 times\n",
      "  0:13:53.318   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.323   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 95 times\n",
      "  0:13:53.324   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:13:53.324   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.325   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:13:53.325   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.330   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 16 times\n",
      "  0:13:53.330   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.333   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:13:53.333   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:13:53.333   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.334   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:13:53.334   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.335   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:13:53.335   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.337   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:13:53.337   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:13:53.337   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.338   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:13:53.338   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.338   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:13:53.338   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.341   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:13:53.341   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:13:53.342   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.342   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:13:53.342   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.342   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:13:53.343   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.343   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6 times\n",
      "  0:13:53.343   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:13:53.344   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.344   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:13:53.344   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.344   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:13:53.345   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.345   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 10 times\n",
      "  0:13:53.346   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:13:53.346   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.346   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:13:53.346   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.347   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:13:53.347   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.347   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 8 times\n",
      "  0:13:53.348   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:13:53.348   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.368   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 8 times\n",
      "  0:13:53.368   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.390   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:13:53.390   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.406   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:13:53.407   293M / 1081M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:13:53.407   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.407   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:13:53.408   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.408   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:13:53.408   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:13:53.408   293M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:13:53.421   239M / 1081M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:13:53.425   239M / 1081M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:13:53.425   239M / 1081M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:13:53.425   239M / 1081M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:13:53.426   239M / 1081M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:13:53.426   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.462   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:13:53.465   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.503   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:13:53.505   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:13:53.538   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:13:53.538   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:13:53.579   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:13:53.580   239M / 1081M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:13:53.604   239M / 1081M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:13:53.771   238M / 1081M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 77310 times\n",
      "  0:13:53.792   236M / 1081M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:13:53.793   236M / 1081M INFO    General                 (simplification.cpp        : 509)     Average coverage = 113.738\n",
      "  0:13:53.798   236M / 1081M INFO    General                 (simplification.cpp        : 510)     Total length = 6235779\n",
      "  0:13:53.811   236M / 1081M INFO    General                 (simplification.cpp        : 511)     Median edge length: 29936\n",
      "  0:13:53.812   236M / 1081M INFO    General                 (simplification.cpp        : 512)     Edges: 4221\n",
      "  0:13:53.812   236M / 1081M INFO    General                 (simplification.cpp        : 513)     Vertices: 3240\n",
      "  0:13:53.812   236M / 1081M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:13:53.844   236M / 1081M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K33/simplified_contigs\"\n",
      "  0:13:53.867   240M / 1081M INFO    General                 (binary_converter.cpp      : 143)   2112 reads written\n",
      "  0:13:53.871   236M / 1081M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:13:53.896     1M / 1081M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 13 minutes 53 seconds\n",
      "\n",
      "===== K33 finished. \n",
      "\n",
      "\n",
      "===== K55 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/config.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K55/configs/isolate_mode.info\"\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=55\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.000     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.000     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.003     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K33/simplified_contigs\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:00:06.490   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 732839 reads\n",
      "  0:00:12.742   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1466116 reads\n",
      "  0:00:19.009   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2200621 reads\n",
      "  0:00:25.022   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2934328 reads\n",
      "  0:00:31.106   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3667680 reads\n",
      "  0:00:38.216   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4401944 reads\n",
      "  0:00:44.397   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5137066 reads\n",
      "  0:00:50.525   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5870399 reads\n",
      "  0:01:15.416   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 8806720 reads\n",
      "  0:01:53.997     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13130728 reads\n",
      "  0:01:53.998     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:02:47.382     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 59773652 kmers in total.\n",
      "  0:02:47.383     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:02:47.385     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:02:47.385     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:02:47.385     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:02:47.385     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:02:47.386     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355443\n",
      "  0:02:53.970   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 12211803 kmers\n",
      "  0:03:01.057   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 24434436 kmers\n",
      "  0:03:09.516   601M / 692M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 36636966 kmers\n",
      "  0:03:17.702   601M / 783M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 48837641 kmers\n",
      "  0:03:25.423   601M / 850M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 59773672 kmers\n",
      "  0:03:25.424   601M / 850M  INFO    General                 (kmer_splitters.hpp        : 202)   Used 59773672 kmers.\n",
      "  0:03:25.461     1M / 850M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:04:21.028     1M / 850M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 59807304 kmers in total.\n",
      "  0:04:21.031     1M / 850M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:04:31.661    44M / 850M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 59807304 kmers, 43203896 bytes occupied (5.77908 bits per kmer).\n",
      "  0:04:31.662    44M / 850M  INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:04:37.665   104M / 850M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:05:43.034   104M / 850M  INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:05:43.043   104M / 850M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:05:43.047   104M / 850M  INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:06:56.992   215M / 1167M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 2141534 sequences extracted\n",
      "  0:07:20.864   215M / 1200M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:07:31.878   215M / 1200M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:07:31.986   215M / 1200M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:07:33.123   215M / 1200M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:07:33.124   215M / 1200M INFO    General                 (debruijn_graph_constructor: 516)   Total 4283068 edges to create\n",
      "  0:07:33.125   383M / 1200M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:07:35.444   451M / 1200M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:07:36.890   451M / 1200M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:07:36.932   483M / 1200M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:07:37.622   483M / 1200M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:07:37.623   483M / 1200M INFO    General                 (debruijn_graph_constructor: 542)   Total 2175186 vertices to create\n",
      "  0:07:37.623   688M / 1200M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:07:38.650   552M / 1200M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:07:38.653   552M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:07:48.520   596M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 59773652 kmers, 43179632 bytes occupied (5.77909 bits per kmer).\n",
      "  0:07:48.671   828M / 1200M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:11:19.252   828M / 1200M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:11:33.601   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:11:47.912   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:11:55.090   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2682776 edges\n",
      "  0:12:00.000   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3248828 edges\n",
      "  0:12:04.224   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3748828 edges\n",
      "  0:12:07.896   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4190744 edges\n",
      "  0:12:08.745   828M / 1200M INFO    General                 (coverage_filling.hpp      :  83)   Processed 4283068 edges\n",
      "  0:12:09.316   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:12:09.319   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 10\n",
      "  0:12:09.320   449M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 84\n",
      "  0:12:09.321   449M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 87. Coverage mad: 25.2042\n",
      "  0:12:09.321   449M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:12:09.411   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:12:09.792   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:12:11.053   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:12:13.434   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:12:13.799   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 86.4573. Fitted coverage std. dev: 24.6508\n",
      "  0:12:13.803   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.0955892\n",
      "  0:12:13.803   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 13\n",
      "  0:12:13.803   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 13\n",
      "  0:12:13.804   450M / 1200M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6179273\n",
      "  0:12:13.804   449M / 1200M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 86.4573\n",
      "  0:12:13.804   449M / 1200M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 13\n",
      "  0:12:13.804   449M / 1200M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:12:13.805   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:12:13.808   462M / 1200M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:12:14.663   498M / 1200M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1173697 out of 4283068, length: 62361050\n",
      "  0:12:14.801   513M / 1200M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 4304486)\n",
      "  0:12:16.902   513M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:12:53.139   545M / 1200M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 44394808 kmers, 32065504 bytes occupied (5.77824 bits per kmer).\n",
      "  0:12:53.317   885M / 1200M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:13:15.748   870M / 1200M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:13:39.306   871M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:14:02.691   871M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:14:26.103   872M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:14:49.347   873M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:15:12.330   873M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:15:35.915   874M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:17:31.811   877M / 1200M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:19:35.676   881M / 1200M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:19:35.873   525M / 1200M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:19:35.874   525M / 1200M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:19:36.921   525M / 1200M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 3555 tips collected, total 3760 connection candidates\n",
      "  0:19:36.937   525M / 1200M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 84 gaps after checking 3668 candidates\n",
      "  0:19:37.067   525M / 1200M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:19:37.104   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:19:37.106   449M / 1200M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:19:37.106   449M / 1200M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:19:37.107   449M / 1200M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:19:37.470   449M / 1200M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:19:37.472   449M / 1200M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:19:37.472   449M / 1200M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:19:37.473   449M / 1200M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:19:37.473   449M / 1200M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:19:37.473   449M / 1200M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:19:42.987   433M / 1200M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 779125 times\n",
      "  0:19:42.988   433M / 1200M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.036  1231M / 1263M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 149716 times\n",
      "  0:20:15.038  1231M / 1263M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.188  1231M / 1264M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 4463 times\n",
      "  0:20:15.190  1231M / 1264M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:20:15.190  1231M / 1264M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.309  1230M / 1264M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 3604 times\n",
      "  0:20:15.313  1230M / 1264M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.647  1234M / 1268M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 757 times\n",
      "  0:20:15.653  1234M / 1268M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.771  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 965 times\n",
      "  0:20:15.772  1234M / 1269M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:20:15.773  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.784  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 70 times\n",
      "  0:20:15.785  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.804  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 40 times\n",
      "  0:20:15.805  1234M / 1269M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.825  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 67 times\n",
      "  0:20:15.826  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:20:15.827  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.828  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:15.828  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.832  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 4 times\n",
      "  0:20:15.833  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.845  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 29 times\n",
      "  0:20:15.845  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:20:15.846  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.847  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:15.847  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.849  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:20:15.849  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.857  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:20:15.862  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:20:15.869  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.870  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:20:15.870  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.873  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:20:15.873  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.874  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:20:15.875  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:20:15.875  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.876  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:15.879  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.880  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:15.880  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.881  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:20:15.881  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:20:15.882  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.883  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:15.887  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.888  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:15.888  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.890  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:20:15.891  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:20:15.891  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.892  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:15.893  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.893  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:15.894  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.895  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:20:15.896  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:20:15.900  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.901  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:15.901  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:15.902  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:15.902  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:15.903  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:20:15.904  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:20:15.904  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:15.989  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 10 times\n",
      "  0:20:15.996  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:16.036  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:16.037  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:16.093  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:20:16.097  1234M / 1270M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:20:16.102  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:16.103  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:20:16.103  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:20:16.104  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:20:16.105  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:20:16.105  1234M / 1270M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:20:16.152  1208M / 1270M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:20:16.161  1221M / 1270M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:20:16.284  1230M / 1282M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 379843 out of 381554, length: 38553946\n",
      "  0:20:16.323  1233M / 1284M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 4304486)\n",
      "  0:20:17.234  1233M / 1284M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:20:28.503  1247M / 1317M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 20240601 kmers, 14619792 bytes occupied (5.7784 bits per kmer).\n",
      "  0:20:28.582  1403M / 1471M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:20:35.897  1400M / 1471M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:21:05.805  1401M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:21:35.475  1401M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:22:05.093  1401M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:22:34.706  1402M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:23:04.599  1402M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:23:34.164  1402M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:26:02.098  1404M / 1471M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:28:42.979  1405M / 1471M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:28:43.053  1240M / 1471M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:28:43.055  1240M / 1471M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:28:43.120  1240M / 1471M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 168 tips collected, total 313 connection candidates\n",
      "  0:28:43.122  1240M / 1471M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 313 candidates\n",
      "  0:28:43.146  1240M / 1471M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:28:43.156  1208M / 1471M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:28:43.157  1208M / 1471M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:28:43.158  1208M / 1471M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:28:43.158  1208M / 1471M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:28:43.159  1208M / 1471M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:28:43.159  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:28:43.231  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:28:43.232  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:28:43.309  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:28:43.316  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:28:43.365  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:28:43.367  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:28:43.406  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:28:43.407  1208M / 1471M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:28:43.493  1208M / 1471M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:28:43.907  1204M / 1471M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 189682 times\n",
      "  0:28:43.922  1199M / 1471M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:28:43.923  1199M / 1471M INFO    General                 (simplification.cpp        : 509)     Average coverage = 88.2127\n",
      "  0:28:43.930  1199M / 1471M INFO    General                 (simplification.cpp        : 510)     Total length = 6251864\n",
      "  0:28:43.939  1199M / 1471M INFO    General                 (simplification.cpp        : 511)     Median edge length: 76224\n",
      "  0:28:43.941  1199M / 1471M INFO    General                 (simplification.cpp        : 512)     Edges: 2190\n",
      "  0:28:43.941  1199M / 1471M INFO    General                 (simplification.cpp        : 513)     Vertices: 2090\n",
      "  0:28:43.942  1199M / 1471M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:28:43.944  1199M / 1471M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K55/simplified_contigs\"\n",
      "  0:28:43.965  1203M / 1471M INFO    General                 (binary_converter.cpp      : 143)   1095 reads written\n",
      "  0:28:43.968  1199M / 1471M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:28:46.111     1M / 1471M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 28 minutes 46 seconds\n",
      "\n",
      "===== K55 finished. \n",
      "\n",
      "\n",
      "===== K77 started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/spades-core /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/config.info /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/isolate_mode.info\n",
      "\n",
      "  0:00:00.000     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/config.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/configs/isolate_mode.info\"\n",
      "  0:00:00.001     1M / 18M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 10 Gb\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.003     1M / 18M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/dataset.info\") with K=77\n",
      "  0:00:00.004     1M / 18M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.005     1M / 18M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.006     1M / 18M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.007     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.007     1M / 18M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.007     1M / 18M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.008     1M / 18M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.008     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.010     1M / 18M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.010     1M / 18M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.012     1M / 18M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K55/simplified_contigs\n",
      "  0:00:00.014     1M / 18M   INFO    General                 (construction.cpp          : 150)   Max read length 151\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (construction.cpp          : 156)   Average read length 145.891\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.015     1M / 18M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:00:00.016     1M / 18M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:00:00.016     1M / 18M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33328 Gb\n",
      "  0:00:00.017     1M / 18M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:00:05.120   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 641445 reads\n",
      "  0:00:10.109   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1283786 reads\n",
      "  0:00:15.462   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 1927256 reads\n",
      "  0:00:20.669   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2570681 reads\n",
      "  0:00:26.729   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3215168 reads\n",
      "  0:00:33.112   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3857538 reads\n",
      "  0:00:38.923   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4501434 reads\n",
      "  0:00:44.841   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 5144971 reads\n",
      "  0:01:19.039   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 128)   Processed 9003256 reads\n",
      "  0:01:56.780     1M / 570M  INFO    General                 (kmer_splitters.hpp        : 134)   Used 13128694 reads\n",
      "  0:01:56.780     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:04:25.411     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 55473231 kmers in total.\n",
      "  0:04:25.412     1M / 570M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:04:25.442     1M / 570M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:04:25.442     1M / 570M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 10 files using 1 threads. This might take a while.\n",
      "  0:04:25.442     1M / 570M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 4096\n",
      "  0:04:25.442     1M / 570M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 3.33327 Gb\n",
      "  0:04:25.443     1M / 570M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 2236962\n",
      "  0:04:30.557   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 6508312 kmers\n",
      "  0:04:37.165   601M / 601M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 14004744 kmers\n",
      "  0:04:50.349   601M / 705M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 21305028 kmers\n",
      "  0:05:00.209   601M / 799M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 27792127 kmers\n",
      "  0:05:07.936   601M / 867M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 34360547 kmers\n",
      "  0:05:15.289   601M / 948M  INFO    General                 (kmer_splitters.hpp        : 197)   Processed 41990532 kmers\n",
      "  0:05:21.462   601M / 1045M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 49175349 kmers\n",
      "  0:05:28.930   601M / 1094M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 55473269 kmers\n",
      "  0:05:28.931   601M / 1094M INFO    General                 (kmer_splitters.hpp        : 202)   Used 55473269 kmers.\n",
      "  0:05:28.977     1M / 1094M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:07:22.675     1M / 1094M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 55798725 kmers in total.\n",
      "  0:07:22.678     1M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:07:33.685    41M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 55798725 kmers, 40308768 bytes occupied (5.77917 bits per kmer).\n",
      "  0:07:33.685    41M / 1094M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:07:43.524    97M / 1094M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:08:58.976    97M / 1094M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:08:59.031    97M / 1094M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:08:59.034    97M / 1094M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:10:09.583   191M / 1094M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 1748112 sequences extracted\n",
      "  0:10:31.828   191M / 1094M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:10:42.055   191M / 1094M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:10:42.112   191M / 1094M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:10:43.059   191M / 1094M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:10:43.060   191M / 1094M INFO    General                 (debruijn_graph_constructor: 516)   Total 3496224 edges to create\n",
      "  0:10:43.060   327M / 1094M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:10:45.133   383M / 1094M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:10:46.340   383M / 1094M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:10:46.357   399M / 1094M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:10:47.024   399M / 1094M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:10:47.025   399M / 1094M INFO    General                 (debruijn_graph_constructor: 542)   Total 2073606 vertices to create\n",
      "  0:10:47.026   592M / 1094M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:10:47.961   493M / 1094M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:10:47.965   493M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:10:57.899   533M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 55473231 kmers, 40073624 bytes occupied (5.77917 bits per kmer).\n",
      "  0:10:58.019   745M / 1094M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:13:48.206   745M / 1094M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:14:03.209   745M / 1094M INFO    General                 (coverage_filling.hpp      :  83)   Processed 1000000 edges\n",
      "  0:14:18.700   745M / 1094M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2000000 edges\n",
      "  0:14:24.154   745M / 1094M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2528789 edges\n",
      "  0:14:27.587   745M / 1094M INFO    General                 (coverage_filling.hpp      :  83)   Processed 2928789 edges\n",
      "  0:14:30.880   745M / 1094M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3306614 edges\n",
      "  0:14:32.463   745M / 1094M INFO    General                 (coverage_filling.hpp      :  83)   Processed 3496224 edges\n",
      "  0:14:33.275   397M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:14:33.293   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 7\n",
      "  0:14:33.293   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 61\n",
      "  0:14:33.294   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 64. Coverage mad: 20.7564\n",
      "  0:14:33.294   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:14:33.659   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:14:34.075   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:14:35.206   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:14:37.314   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:14:37.990   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 64.4987. Fitted coverage std. dev: 19.6675\n",
      "  0:14:37.993   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.253612\n",
      "  0:14:37.993   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 11\n",
      "  0:14:37.994   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 11\n",
      "  0:14:37.994   397M / 1094M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 6167580\n",
      "  0:14:37.994   397M / 1094M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 64.4987\n",
      "  0:14:37.994   397M / 1094M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 11\n",
      "  0:14:37.995   397M / 1094M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:14:37.995   397M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:14:38.000   410M / 1094M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:14:38.756   446M / 1094M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 1393903 out of 3496224, length: 72584154\n",
      "  0:14:38.922   463M / 1094M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 3513708)\n",
      "  0:14:41.346   463M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:15:22.788   498M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 48309169 kmers, 34892696 bytes occupied (5.77823 bits per kmer).\n",
      "  0:15:22.995   870M / 1094M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:15:49.597   853M / 1094M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:16:08.278   854M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:16:26.941   855M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:16:45.783   856M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:17:04.310   856M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:17:22.652   857M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:17:41.109   858M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:19:13.509   862M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:20:52.511   884M / 1094M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:20:52.884   487M / 1094M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:20:52.887   487M / 1094M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:20:53.860   487M / 1094M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 3527 tips collected, total 3718 connection candidates\n",
      "  0:20:53.876   487M / 1094M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 166 gaps after checking 3526 candidates\n",
      "  0:20:53.971   487M / 1094M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:20:54.015   397M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:20:54.024   397M / 1094M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:20:54.026   397M / 1094M INFO    General                 (simplification.cpp        :  63)   Most init cleaning disabled on main iteration\n",
      "  0:20:54.027   397M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:20:54.352   397M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:20:54.353   397M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:20:54.353   397M / 1094M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:20:54.354   397M / 1094M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:20:54.354   397M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:20:54.354   397M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:20:59.716   365M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 713838 times\n",
      "  0:20:59.718   365M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.350   448M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 7702 times\n",
      "  0:21:01.352   448M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.437   448M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 442 times\n",
      "  0:21:01.438   448M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:21:01.438   448M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.463   448M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 351 times\n",
      "  0:21:01.466   448M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.483   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 28 times\n",
      "  0:21:01.484   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.604   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 341 times\n",
      "  0:21:01.609   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:21:01.609   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.611   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 12 times\n",
      "  0:21:01.612   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.624   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 7 times\n",
      "  0:21:01.624   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.642   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 22 times\n",
      "  0:21:01.643   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:21:01.643   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.643   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:21:01.644   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.644   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.644   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.646   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 5 times\n",
      "  0:21:01.646   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:21:01.650   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.650   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:01.651   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.651   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:21:01.652   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.653   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:21:01.653   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:21:01.653   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.653   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:01.654   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.654   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.654   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.655   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:21:01.655   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:21:01.656   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.656   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:01.656   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.656   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.657   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.657   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:21:01.657   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:21:01.657   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.658   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:01.658   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.658   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.659   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.659   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:21:01.659   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:21:01.659   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.664   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:01.664   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.664   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.665   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.665   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:21:01.665   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:21:01.665   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.666   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:01.666   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.666   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.666   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.667   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:21:01.667   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:21:01.667   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.763   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 19 times\n",
      "  0:21:01.764   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.837   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.842   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.954   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:21:01.955   449M / 1094M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:21:01.955   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:21:01.955   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:21:01.956   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:21:01.956   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:21:01.956   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:21:01.956   449M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:21:02.013   431M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:21:02.021   444M / 1094M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:21:02.253   461M / 1094M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 668151 out of 669302, length: 51836680\n",
      "  0:21:02.317   467M / 1094M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 3513708)\n",
      "  0:21:03.591   467M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:21:22.972   486M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 27152757 kmers, 19613528 bytes occupied (5.77872 bits per kmer).\n",
      "  0:21:23.084   694M / 1094M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:21:36.224   688M / 1094M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:21:52.634   689M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:22:09.358   689M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:22:25.718   690M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:22:42.298   691M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:22:58.915   691M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:23:15.339   692M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:24:37.489   695M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:26:05.941   698M / 1094M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:26:06.090   482M / 1094M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:26:06.091   482M / 1094M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:26:06.288   482M / 1094M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 374 tips collected, total 699 connection candidates\n",
      "  0:26:06.295   482M / 1094M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 0 gaps after checking 699 candidates\n",
      "  0:26:06.339   482M / 1094M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:26:06.358   431M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:26:06.361   431M / 1094M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:26:06.361   431M / 1094M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:26:06.361   431M / 1094M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:26:06.362   431M / 1094M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:26:06.362   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:06.420   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:26:06.421   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:06.478   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:26:06.479   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:26:06.590   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:26:06.595   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:26:06.695   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:26:06.696   431M / 1094M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:26:06.786   431M / 1094M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:26:07.359   425M / 1094M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 333943 times\n",
      "  0:26:07.366   416M / 1094M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:26:07.367   416M / 1094M INFO    General                 (simplification.cpp        : 509)     Average coverage = 64.978\n",
      "  0:26:07.371   416M / 1094M INFO    General                 (simplification.cpp        : 510)     Total length = 6237203\n",
      "  0:26:07.380   416M / 1094M INFO    General                 (simplification.cpp        : 511)     Median edge length: 94276\n",
      "  0:26:07.380   416M / 1094M INFO    General                 (simplification.cpp        : 512)     Edges: 1416\n",
      "  0:26:07.380   416M / 1094M INFO    General                 (simplification.cpp        : 513)     Vertices: 1292\n",
      "  0:26:07.381   416M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Mismatch Correction (id: mismatch_correction)\n",
      "  0:26:07.381   416M / 1094M INFO    General                 (graph_pack_helpers.cpp    :  44)   Index refill\n",
      "  0:26:07.381   416M / 1094M INFO    General                 (edge_index.hpp            : 175)   Using small index (max_id = 3513708)\n",
      "  0:26:07.767   416M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:26:09.748   420M / 1094M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 6237203 kmers, 4508168 bytes occupied (5.78229 bits per kmer).\n",
      "  0:26:09.773   468M / 1094M INFO    General                 (edge_index_builders.hpp   : 253)   Collecting edge information from graph, this takes a while.\n",
      "  0:26:11.598   468M / 1094M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 1212192 kmers to process\n",
      "  0:26:12.156   470M / 1094M INFO    General                 (kmer_mapper.hpp           :  87)   Total 3086 kmers with non-trivial roots\n",
      "  0:26:12.161   468M / 1094M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:26:12.161   468M / 1094M INFO    General                 (mismatch_correction.cpp   : 392)   Collect potential mismatches\n",
      "  0:26:14.270   468M / 1094M INFO    General                 (mismatch_correction.cpp   : 193)   Total 572 edges (out of 1416) with 21750 potential mismatch positions (38.0245 positions per edge)\n",
      "  0:26:14.271   468M / 1094M INFO    General                 (mismatch_correction.cpp   : 394)   Potential mismatches collected\n",
      "  0:26:14.272   468M / 1094M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:26:16.256   470M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:26:18.237   471M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:26:20.141   472M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:26:22.017   472M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:26:23.944   472M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:26:25.880   472M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:26:35.109   472M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:26:53.295   472M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:27:30.509   472M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 8400000 reads\n",
      "  0:28:14.335   472M / 1094M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 13126504 reads processed\n",
      "  0:28:14.384   470M / 1094M INFO    General                 (mismatch_correction.cpp   : 387)   All edges processed\n",
      "  0:28:14.385   468M / 1094M INFO    General                 (mismatch_correction.cpp   : 442)   Corrected 0 nucleotides\n",
      "  0:28:14.385   468M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:28:14.386   468M / 1094M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/before_rr.fasta\"\n",
      "  0:28:14.560   468M / 1094M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_after_simplification.gfa\"\n",
      "  0:28:14.663   468M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Paired Information Counting (id: late_pair_info_count)\n",
      "  0:28:14.684   468M / 1094M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 1212192 kmers to process\n",
      "  0:28:14.685   468M / 1094M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:28:14.695   468M / 1094M INFO    General                 (pair_info_count.cpp       : 157)   Min edge length for estimation: 94276\n",
      "  0:28:14.699   468M / 1094M INFO    General                 (pair_info_count.cpp       : 168)   Estimating insert size for library #0\n",
      "  0:28:14.699   468M / 1094M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:28:14.700   468M / 1094M INFO    General                 (paired_info_utils.cpp     :  87)   Estimating insert size (takes a while)\n",
      "  0:28:14.723   500M / 1094M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:28:18.451   500M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:28:22.028   501M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:28:25.629   501M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:28:29.226   501M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:28:33.012   501M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:28:36.831   501M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:28:54.333   501M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:29:14.518   500M / 1094M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:29:14.866   500M / 1094M INFO    General                 (paired_info_utils.cpp     : 104)   Edge pairs: 9274\n",
      "  0:29:14.874   500M / 1094M INFO    General                 (paired_info_utils.cpp     : 106)   1470073 paired reads (44.7971% of all) aligned to long edges\n",
      "  0:29:14.875   468M / 1094M INFO    General                 (pair_info_count.cpp       : 188)     Insert size = 262.599, deviation = 93.6092, left quantile = 153, right quantile = 381, read length = 151\n",
      "  0:29:14.875   468M / 1094M INFO    General                 (pair_info_count.cpp       : 203)   Filtering data for library #0\n",
      "  0:29:14.876   468M / 1094M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:29:14.876   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:29:18.577   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:29:22.108   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:29:25.681   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:29:29.245   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:29:32.990   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:29:36.797   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:29:54.090   469M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:30:14.020   469M / 1094M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:30:14.021   468M / 1094M INFO    General                 (pair_info_count.cpp       : 207)   Mapping library #0\n",
      "  0:30:14.022   468M / 1094M INFO    General                 (pair_info_count.cpp       : 209)   Mapping paired reads (takes a while)\n",
      "  0:30:14.022   468M / 1094M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:30:14.022   468M / 1094M INFO    General                 (paired_info_utils.cpp     : 138)   Left insert size quantile 153, right insert size quantile 381, filtering threshold 2, rounding threshold 0\n",
      "  0:30:14.027   482M / 1094M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:30:17.918   482M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:30:21.604   483M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:30:25.310   483M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:30:29.058   483M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:30:32.977   483M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:30:36.921   483M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:30:55.135   484M / 1094M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:31:16.045   484M / 1094M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 3281626 reads processed\n",
      "  0:31:16.050   471M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Distance Estimation (id: distance_estimation)\n",
      "  0:31:16.051   471M / 1094M INFO    General                 (distance_estimation.cpp   :  42)   Processing library #0\n",
      "  0:31:16.051   471M / 1094M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:31:16.055   471M / 1094M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SIMPLE distance estimator\n",
      "  0:31:16.085   471M / 1094M INFO    General                 (distance_estimation_utils.: 131)   Refining clustered pair information\n",
      "  0:31:16.086   471M / 1094M INFO    General                 (distance_estimation_utils.: 133)   The refining of clustered pair information has been finished\n",
      "  0:31:16.087   471M / 1094M INFO    General                 (distance_estimation_utils.: 135)   Improving paired information\n",
      "  0:31:16.127   471M / 1094M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 156; contradictional = 96\n",
      "  0:31:16.165   471M / 1094M INFO   PairInfoImprover         (pair_info_improver.hpp    :  78)   Paired info stats: missing = 0; contradictional = 0\n",
      "  0:31:16.166   471M / 1094M INFO    General                 (distance_estimation_utils.:  80)   Filling scaffolding index\n",
      "  0:31:16.167   471M / 1094M INFO    General                 (distance_estimation_utils.:  23)   Estimating distances\n",
      "  0:31:16.167   471M / 1094M INFO   DistanceEstimator        (distance_estimation.hpp   : 132)   Using SMOOTHING distance estimator\n",
      "  0:31:16.228   471M / 1094M INFO    General                 (distance_estimation.cpp   :  51)   Clearing raw paired index\n",
      "  0:31:16.229   469M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Repeat Resolving (id: repeat_resolving)\n",
      "  0:31:16.229   469M / 1094M INFO    General                 (repeat_resolving.cpp      :  88)   Using Path-Extend repeat resolving\n",
      "  0:31:16.231   469M / 1094M INFO    General                 (launcher.cpp              : 603)   ExSPAnder repeat resolving tool started\n",
      "  0:31:16.237   469M / 1094M INFO    General                 (launcher.cpp              : 420)   Creating main extenders, unique edge length = 2000\n",
      "  0:31:16.238   469M / 1094M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 64.978\n",
      "  0:31:16.239   469M / 1094M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5820\n",
      "  0:31:16.240   469M / 1094M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 64.978\n",
      "  0:31:16.240   469M / 1094M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 5820\n",
      "  0:31:16.245   469M / 1094M INFO    General                 (extenders_logic.cpp       : 550)   Using 1 paired-end library\n",
      "  0:31:16.246   469M / 1094M INFO    General                 (extenders_logic.cpp       : 551)   Using 1 paired-end scaffolding library\n",
      "  0:31:16.252   469M / 1094M INFO    General                 (extenders_logic.cpp       : 552)   Using 0 single read libraries\n",
      "  0:31:16.252   469M / 1094M INFO    General                 (launcher.cpp              : 449)   Total number of extenders is 3\n",
      "  0:31:16.255   469M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 0 paths from 681 (0%)\n",
      "  0:31:16.337   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 69 paths from 681 (10%)\n",
      "  0:31:16.346   471M / 1094M INFO    General                 (path_extenders.cpp        :  34)   Processed 128 paths from 681 (18%)\n",
      "  0:31:16.347   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 138 paths from 681 (20%)\n",
      "  0:31:16.351   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 207 paths from 681 (30%)\n",
      "  0:31:16.353   471M / 1094M INFO    General                 (path_extenders.cpp        :  34)   Processed 256 paths from 681 (37%)\n",
      "  0:31:16.354   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 276 paths from 681 (40%)\n",
      "  0:31:16.362   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 345 paths from 681 (50%)\n",
      "  0:31:16.366   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 414 paths from 681 (60%)\n",
      "  0:31:16.370   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 483 paths from 681 (70%)\n",
      "  0:31:16.370   471M / 1094M INFO    General                 (path_extenders.cpp        :  34)   Processed 512 paths from 681 (75%)\n",
      "  0:31:16.375   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 552 paths from 681 (81%)\n",
      "  0:31:16.375   471M / 1094M INFO    General                 (path_extenders.cpp        :  36)   Processed 621 paths from 681 (91%)\n",
      "  0:31:16.376   471M / 1094M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:31:16.376   471M / 1094M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:31:16.377   471M / 1094M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:31:16.377   471M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:31:16.377   471M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:31:16.377   471M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:31:16.378   471M / 1094M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:31:16.379   471M / 1094M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:31:16.380   471M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:31:16.380   471M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:31:16.383   471M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:31:16.393   471M / 1094M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:31:16.394   471M / 1094M INFO    General                 (launcher.cpp              : 456)   Closing gaps in paths\n",
      "  0:31:16.394   471M / 1094M INFO    General                 (launcher.cpp              : 486)   Gap closing completed\n",
      "  0:31:16.396   469M / 1094M INFO    General                 (launcher.cpp              : 304)   Traversing tandem repeats\n",
      "  0:31:16.419   469M / 1094M INFO    General                 (launcher.cpp              : 314)   Traversed 2 loops\n",
      "  0:31:16.419   469M / 1094M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:31:16.420   469M / 1094M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:31:16.420   469M / 1094M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:31:16.420   469M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:31:16.421   469M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:31:16.421   469M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:31:16.421   469M / 1094M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:31:16.422   469M / 1094M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:31:16.422   469M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:31:16.423   469M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:31:16.423   469M / 1094M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:31:16.428   469M / 1094M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:31:16.436   469M / 1094M INFO    General                 (launcher.cpp              : 666)   ExSPAnder repeat resolving tool finished\n",
      "  0:31:16.436   469M / 1094M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:31:16.437   469M / 1094M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/before_rr.fasta\"\n",
      "  0:31:16.621   469M / 1094M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_with_scaffolds.gfa\"\n",
      "  0:31:16.730   469M / 1094M INFO    General                 (contig_output_stage.cpp   : 169)   Outputting FastG graph to \"/media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph.fastg\"\n",
      "  0:31:17.003   469M / 1094M INFO    General                 (contig_output_stage.cpp   : 200)   Breaking scaffolds\n",
      "  0:31:17.036   475M / 1094M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.fasta\n",
      "  0:31:17.135   475M / 1094M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.paths\n",
      "  0:31:17.152   475M / 1094M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.fasta\n",
      "  0:31:17.260   475M / 1094M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.paths\n",
      "  0:31:17.267   475M / 1094M INFO    General                 (contig_output_stage.cpp   : 114)   Populating GFA with scaffold paths\n",
      "  0:31:17.270   469M / 1094M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:31:17.456     1M / 1094M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 31 minutes 17 seconds\n",
      "\n",
      "===== K77 finished. \n",
      "\n",
      "\n",
      "===== Copy files started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/copy_files.py /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/before_rr.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/before_rr.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph_after_simplification.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/first_pe_contigs.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/strain_graph.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/strain_graph.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.fasta /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/scaffolds.paths /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.paths /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph_with_scaffolds.gfa /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/assembly_graph.fastg /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph.fastg /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/K77/final_contigs.paths /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.paths\n",
      "\n",
      "\n",
      "===== Copy files finished. \n",
      "\n",
      "\n",
      "===== Assembling finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /home/student/miniconda3/envs/VariantCalling/bin/python /home/student/miniconda3/envs/VariantCalling/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.fasta --misc_dir /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Assembled contigs are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.fasta\n",
      " * Assembled scaffolds are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.fasta\n",
      " * Paths in the assembly graph corresponding to the contigs are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/scaffolds.paths\n",
      " * Assembly graph is in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/assembly_graph_with_scaffolds.gfa\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /media/student/EXTERNAL_USB/Prova/Analysis/assemblies/PAO1wt12P_SPAdes/spades.log\n",
      "\n",
      "Thank you for using SPAdes! If you use it in your research, please cite:\n",
      "\n",
      "  Prjibelski, A., Antipov, D., Meleshko, D., Lapidus, A. and Korobeynikov, A., 2020. Using SPAdes de novo assembler. Current protocols in bioinformatics, 70(1), p.e102.\n",
      "  doi.org/10.1002/cpbi.102\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if configuration[\"denovo\"]:\n",
    "\n",
    "    samples = {file.split(\"_\")[0] for file in os.listdir(fastq_directory_path)}\n",
    "\n",
    "    for sample in samples:\n",
    "        files = os.listdir(fastq_directory_path)\n",
    "\n",
    "        R1_file = sample + \"_R1.fastq.gz\"\n",
    "        R1_file_path = os.path.join(fastq_directory_path, R1_file)\n",
    "        R2_file = sample + \"_R2.fastq.gz\"\n",
    "        R2_file_path = os.path.join(fastq_directory_path, R2_file)\n",
    "\n",
    "        denovo_directory = sample + \"_SPAdes\"\n",
    "        denovo_directory_path = os.path.join(assemblies_directory_path, denovo_directory)\n",
    "\n",
    "        if os.path.exists(denovo_directory_path):\n",
    "            shutil.rmtree(denovo_directory_path)\n",
    "\n",
    "        command = [\n",
    "            \"python\", spades, \"-o\", denovo_directory_path,\n",
    "            \"-1\", R1_file_path, \"-2\", R2_file_path,\n",
    "            \"--isolate\"\n",
    "        ]\n",
    "\n",
    "        subprocess.run(command, check=True)\n",
    "\n",
    "        contigs_file_path = os.path.join(denovo_directory_path, \"contigs.fasta\")\n",
    "\n",
    "        denovo_file = sample + \".SPAdes.denovoassembly.fasta\"\n",
    "        denovo_file_path = os.path.join(assemblies_directory_path, denovo_file)\n",
    "        \n",
    "        os.rename(contigs_file_path, denovo_file_path)\n",
    "\n",
    "        shutil.rmtree(denovo_directory_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
